{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## 分析同源基因bias表达\n",
    "rm(list=ls())\n",
    "#setwd('G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads')\n",
    "setwd('G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/homo_gene2/gene_bias/DESeq')\n",
    "library(DESeq2)\n",
    "library(dplyr)\n",
    "#tissues <- list('anthers', 'cotyledon',  'hypocotyl', 'leaf', 'ovules', 'petals', 'radicle', 'root', 'stem', 'stigma')\n",
    "tissues <- list('fiber_5DPA', 'fiber_10DPA', 'fiber_20DPA')\n",
    "#tissues <- list('anthers')\n",
    "for (t in tissues){\n",
    "    print(t)\n",
    "    sample = paste(t, \"/\", t, '_hg_reads.bed', sep=\"\")\n",
    "    out_f = paste(t, \"/\", t, '_DEG.bed', sep=\"\")\n",
    "    countData <- read.table(sample, sep='\\t', header=T, row.names=1)\n",
    "    #countData <- select(countData, reads1_at, reads2_at, reads3_at, reads1_dt, reads2_dt, reads3_dt)\n",
    "    countData <- select(countData, reads1_at, reads2_at, reads1_dt, reads2_dt)\n",
    "    # print(countData)\n",
    "    #condition <- factor(c(\"at\", \"at\", \"at\", \"dt\", \"dt\", \"dt\"))\n",
    "    condition <- factor(c(\"at\", \"at\", \"dt\", \"dt\"))\n",
    "    colData <- data.frame(row.names=colnames(countData), condition)\n",
    "    all(rownames(colData)) == colnames(countData)\n",
    "    dds <- DESeqDataSetFromMatrix(countData, DataFrame(colData), design=~condition)\n",
    "    #print(dds)\n",
    "    # #View(dds)\n",
    "    dds <- DESeq(dds)\n",
    "    resultsNames(dds)\n",
    "    res <- results(dds)\n",
    "    # table(res$padj==0)\n",
    "    # mcols(res, use.names=TRUE)\n",
    "    res <- res[order(res$padj), ]\n",
    "    res <- subset(res, padj != 0)\n",
    "    GeneId <- row.names(res)\n",
    "    logFC <- res$log2FoldChange\n",
    "    padj <- res$padj\n",
    "    data <- data.frame(GeneId=GeneId, logFC=logFC, padj=padj)\n",
    "    data$sig[(data$padj>0.01)|((data$logFC < 1)&(data$logFC> -1))] <- \"Not\"\n",
    "    data$sig[data$padj <= 0.01 & data$logFC >= 1] <- \"Up\"\n",
    "    data$sig[data$padj <= 0.01 & data$logFC <= -1] <- \"Down\"\n",
    "    data_out = subset(data, sig != 'Not')\n",
    "    write.table(data_out, file=out_f, sep='\\t', row.names=F)\n",
    "} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## 分析同源基因bias表达\n",
    "rm(list=ls())\n",
    "setwd('G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data')\n",
    "library(DESeq2)\n",
    "library(dplyr)\n",
    "tissues <- list('anther', 'cotyledon',  'hypocotyl', 'leaf', 'ovule', 'petal', 'radicle', 'root', 'stem', 'stigma', 'fiber_10DPA', 'fiber_20DPA')\n",
    "batch1_sample <- c('anther', 'hypocotyl')\n",
    "batch2_sample <- c('cotyledon', 'leaf', 'ovule', 'petal', 'root', 'stem', 'stigma')\n",
    "batch3_sample <- c('radicle', 'fiber_10DPA', 'fiber_20DPA')\n",
    "df <- read.csv(\"J668_all_Tissue_hg_reads.csv\",header=TRUE, row.names=1)\n",
    "#print(df)\n",
    "for (t in tissues){\n",
    "    at_rep1 = paste(t, \"_rep1_at\", sep=\"\")\n",
    "    at_rep2 = paste(t, \"_rep2_at\", sep=\"\")\n",
    "    at_rep3 = paste(t, \"_rep3_at\", sep=\"\")\n",
    "    dt_rep1 = paste(t, \"_rep1_dt\", sep=\"\")\n",
    "    dt_rep2 = paste(t, \"_rep2_dt\", sep=\"\")\n",
    "    dt_rep3 = paste(t, \"_rep3_dt\", sep=\"\")\n",
    "    out_f = paste(t, '_DEG.bed', sep=\"\")\n",
    "    #print(at_rep1)\n",
    "    countData <- select(df, at_rep1, at_rep2, at_rep3, dt_rep1, dt_rep2, dt_rep3)\n",
    "    #rename\n",
    "    names(countData)[names(countData) == at_rep1] <- \"Control1\"\n",
    "    names(countData)[names(countData) == at_rep2] <- \"Control2\"\n",
    "    names(countData)[names(countData) == at_rep3] <- \"Control3\"\n",
    "    names(countData)[names(countData) == dt_rep1] <- \"Treatment1\"\n",
    "    names(countData)[names(countData) == dt_rep2] <- \"Treatment2\"\n",
    "    names(countData)[names(countData) == dt_rep3] <- \"Treatment3\"\n",
    "    # 批次效应\n",
    "    if (t %in% batch1_sample) {\n",
    "        at_batch_sample <- c(1, 2, 2)\n",
    "        dt_batch_sample <- c(1, 2, 2)\n",
    "    } else if (t %in% batch2_sample) {\n",
    "        at_batch_sample <- c(1, 1, 2)\n",
    "        dt_batch_sample <- c(1, 1, 2)\n",
    "    } else {\n",
    "        at_batch_sample <- c(1, 1, 3)\n",
    "        dt_batch_sample <- c(1, 1, 3)\n",
    "    }\n",
    "    batch <- factor(c(at_batch_sample, dt_batch_sample))\n",
    "    # # # print(countData)\n",
    "    condition <- factor(c(\"Control\", \"Control\", \"Control\", \"Treatment\", \"Treatment\", \"Treatment\"))\n",
    "    #colData <- data.frame(row.names=colnames(countData), condition)\n",
    "    colData <- data.frame(condition=condition, batch=batch)\n",
    "    all(rownames(colData)) == colnames(countData)\n",
    "    dds <- DESeqDataSetFromMatrix(countData, DataFrame(colData), design=~batch+condition)\n",
    "    #print(dds)\n",
    "    # #View(dds)\n",
    "    dds <- DESeq(dds)\n",
    "    resultsNames(dds)\n",
    "    res <- results(dds)\n",
    "    # table(res$padj==0)\n",
    "    # mcols(res, use.names=TRUE)\n",
    "    res <- res[order(res$padj), ]\n",
    "    res <- subset(res, padj != 0)\n",
    "    GeneId <- row.names(res)\n",
    "    logFC <- res$log2FoldChange\n",
    "    padj <- res$padj\n",
    "    data <- data.frame(GeneId=GeneId, logFC=logFC, padj=padj)\n",
    "    data$sig[(data$padj>0.01)|((data$logFC < 1)&(data$logFC> -1))] <- \"Not\"\n",
    "    data$sig[data$padj <= 0.01 & data$logFC >= 1] <- \"Up\"\n",
    "    data$sig[data$padj <= 0.01 & data$logFC <= -1] <- \"Down\"\n",
    "    data_out = subset(data, sig != 'Not')\n",
    "    write.table(data_out, file=out_f, sep='\\t', row.names=F)\n",
    "} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## 分析fiber_5DPA 和fiber_10DPA之间差异表达的基因，\n",
    "rm(list=ls())\n",
    "#setwd('G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads')\n",
    "setwd('G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/5DPA_vs_10DPA')\n",
    "library(DESeq2)\n",
    "library(dplyr)\n",
    "countData <- read.table(\"5DPA_10DPA_reads.bed\", sep='\\t', header=T, row.names=1)\n",
    "countData <- select(countData, control1, control2, KD1, KD2)\n",
    "condition <- factor(c(\"control\", \"control\", \"KD\", \"KD\"))\n",
    "colData <- data.frame(row.names=colnames(countData), condition)\n",
    "all(rownames(colData)) == colnames(countData)\n",
    "dds <- DESeqDataSetFromMatrix(countData, DataFrame(colData), design=~condition)\n",
    "dds <- DESeq(dds)\n",
    "resultsNames(dds)\n",
    "res <- results(dds)\n",
    "res <- res[order(res$padj), ]\n",
    "res <- subset(res, padj != 0)\n",
    "GeneId <- row.names(res)\n",
    "logFC <- res$log2FoldChange\n",
    "padj <- res$padj\n",
    "data <- data.frame(GeneId=GeneId, logFC=logFC, padj=padj)\n",
    "data$sig[(data$padj>0.01)|((data$logFC < 1)&(data$logFC> -1))] <- \"Not\"\n",
    "data$sig[data$padj <= 0.01 & data$logFC >= 1] <- \"Up\"\n",
    "data$sig[data$padj <= 0.01 & data$logFC <= -1] <- \"Down\"\n",
    "data_out = subset(data, sig != 'Not')\n",
    "write.table(data_out, file=\"5DPA_10DPA_DEG.bed\", sep='\\t', row.names=F)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## 分析ovules 和fiber_5DPA之间差异表达的基因，\n",
    "rm(list=ls())\n",
    "setwd('G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/10DPA_vs_20DPA')\n",
    "library(DESeq2)\n",
    "library(dplyr)\n",
    "countData <- read.table(\"10DPA_20DPA_reads.bed\", sep='\\t', header=T, row.names=1)\n",
    "countData <- select(countData, CT1, CT2, KD1, KD2)\n",
    "condition <- factor(c(\"CT\", \"CT\", \"KD\", \"KD\"))\n",
    "colData <- data.frame(row.names=colnames(countData), condition)\n",
    "all(rownames(colData)) == colnames(countData)\n",
    "dds <- DESeqDataSetFromMatrix(countData, DataFrame(colData), design=~condition)\n",
    "dds <- DESeq(dds)\n",
    "resultsNames(dds)\n",
    "res <- results(dds)\n",
    "res <- res[order(res$padj), ]\n",
    "res <- subset(res, padj != 0)\n",
    "GeneId <- row.names(res)\n",
    "logFC <- res$log2FoldChange\n",
    "padj <- res$padj\n",
    "data <- data.frame(GeneId=GeneId, logFC=logFC, padj=padj)\n",
    "data$sig[(data$padj>0.01)|((data$logFC < 1)&(data$logFC> -1))] <- \"Not\"\n",
    "data$sig[data$padj <= 0.01 & data$logFC >= 1] <- \"Up\"\n",
    "data$sig[data$padj <= 0.01 & data$logFC <= -1] <- \"Down\"\n",
    "data_out = subset(data, sig != 'Not')\n",
    "write.table(data_out, file=\"10DPA_20DPA_DEG.bed\", sep='\\t', row.names=F)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## 分析fiber_5DPA 和fiber_20DPA之间差异表达的基因，\n",
    "rm(list=ls())\n",
    "#setwd('G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads')\n",
    "setwd('G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/5DPA_vs_20DPA')\n",
    "library(DESeq2)\n",
    "library(dplyr)\n",
    "countData <- read.table(\"5DPA_20DPA_reads.bed\", sep='\\t', header=T, row.names=1)\n",
    "countData <- select(countData, DPA5_rep1, DPA5_rep2, DPA20_rep1, DPA20_rep2)\n",
    "condition <- factor(c(\"DPA5\", \"DPA5\", \"DPA20\", \"DPA20\"))\n",
    "colData <- data.frame(row.names=colnames(countData), condition)\n",
    "all(rownames(colData)) == colnames(countData)\n",
    "dds <- DESeqDataSetFromMatrix(countData, DataFrame(colData), design=~condition)\n",
    "dds <- DESeq(dds)\n",
    "resultsNames(dds)\n",
    "res <- results(dds)\n",
    "res <- res[order(res$padj), ]\n",
    "res <- subset(res, padj != 0)\n",
    "GeneId <- row.names(res)\n",
    "logFC <- res$log2FoldChange\n",
    "padj <- res$padj\n",
    "data <- data.frame(GeneId=GeneId, logFC=logFC, padj=padj)\n",
    "data$sig[(data$padj>0.01)|((data$logFC < 1)&(data$logFC> -1))] <- \"Not\"\n",
    "data$sig[data$padj <= 0.01 & data$logFC >= 1] <- \"Up\"\n",
    "data$sig[data$padj <= 0.01 & data$logFC <= -1] <- \"Down\"\n",
    "data_out = subset(data, sig != 'Not')\n",
    "write.table(data_out, file=\"5DPA_20DPA_DEG.bed\", sep='\\t', row.names=F)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## 比较组织间两两差异\n",
    "library(DESeq2)\n",
    "library(dplyr)\n",
    "setwd(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data\")\n",
    "tissues = c('anther', 'ovule', 'root',  'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma')\n",
    "df = read.csv(\"all_tissues_reads.bed\",  header=T, row.names=1)\n",
    "#print(df)\n",
    "for (t1 in seq(1, 9)) {\n",
    "    one <- tissues[t1]\n",
    "    if (one == \"Anther\" | one== \"Hypocotyl\") {\n",
    "        ac <- c(1,2,2)\n",
    "    } else {\n",
    "        ac <- c(1,1,2)\n",
    "    }\n",
    "    for (t2 in seq(t1+1,10)) {\n",
    "        two <- tissues[t2]\n",
    "            if (two == \"Anther\" | two == \"Hypocotyl\") {\n",
    "            bc <- c(1,2,2)\n",
    "        } else {\n",
    "            bc <- c(1,1,2)\n",
    "        }\n",
    "        out_f <- paste(one, \"_\", two, \"_DEG.bed\", sep=\"\")\n",
    "        c1 = paste(one, 1, sep=\"\")\n",
    "        c2 = paste(one, 2, sep=\"\")\n",
    "        c3 = paste(one, 3, sep=\"\")\n",
    "        t1 = paste(two, 1, sep=\"\")\n",
    "        t2 = paste(two, 2, sep=\"\")\n",
    "        t3 = paste(two, 3, sep=\"\")\n",
    "        control_treatment = c(c1,c2,c3,t1,t2,t3)\n",
    "        print(control_treatment)\n",
    "        #print(control_treatment)\n",
    "        countData = select(df, control_treatment)\n",
    "        condition <- factor(c(\"control\", \"control\", \"control\", \"treat\", \"treat\", \"treat\"))\n",
    "        batch <- factor(c(ac, bc))\n",
    "        #colData <- data.frame(row.names=colnames(countData), condition)\n",
    "        colData <- data.frame(condition=condition,batch=batch)\n",
    "        all(rownames(colData)) == colnames(countData)\n",
    "        dds <- DESeqDataSetFromMatrix(countData, DataFrame(colData), design=~batch+condition)\n",
    "        #View(dds)\n",
    "        dds <- DESeq(dds)\n",
    "        resultsNames(dds)\n",
    "        res <- results(dds)\n",
    "        table(res$padj==0)\n",
    "        mcols(res, use.names=TRUE)\n",
    "        res <- res[order(res$padj), ]\n",
    "        res <- subset(res, padj != 0)\n",
    "        GeneId <- row.names(res)\n",
    "        logFC <- res$log2FoldChange\n",
    "        padj <- res$padj\n",
    "        data <- data.frame(GeneId=GeneId, logFC=logFC, padj=padj)\n",
    "        data$sig[(data$padj>0.05)|((data$logFC < 1)&(data$logFC> -1))] <- \"Not\"\n",
    "        data$sig[data$padj <= 0.05 & data$logFC >= 1] <- \"Up\"\n",
    "        data$sig[data$padj <= 0.05 & data$logFC <= -1] <- \"Down\"\n",
    "        #data_out = subset(data, sig != 'Not')\n",
    "        write.table(data, file=out_f, sep='\\t', row.names=F)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## 比较两两组织间两两差异\n",
    "library(DESeq2)\n",
    "library(dplyr)\n",
    "setwd(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data\")\n",
    "tissues = c('anther', 'ovule', 'root',  'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma')\n",
    "df = read.csv(\"J668_all_Tissue_reads.csv\",  header=T, row.names=1)\n",
    "batch1_sample <- c('anther', 'hypocotyl')\n",
    "batch2_sample <- c('cotyledon', 'leaf', 'ovule', 'petal', 'root', 'stem', 'stigma')\n",
    "batch3_sample <- c('radicle', 'fiber_10DPA', 'fiber_20DPA')\n",
    "#print(df)\n",
    "for (t1 in seq(1, 11)) {\n",
    "    one <- tissues[t1]\n",
    "    if (t1 %in% batch1_sample) {\n",
    "        control_batch_sample <- c(1, 2, 2)\n",
    "    } else if (t1 %in% batch2_sample) {\n",
    "        control_batch_sample <- c(1, 1, 2)\n",
    "    } else {\n",
    "        control_batch_sample <- c(1, 1, 3)\n",
    "    }\n",
    "\n",
    "    for (t2 in seq(t1+1,12)) {\n",
    "        if (t2 %in% batch1_sample) {\n",
    "            treatment_batch_sample <- c(1, 2, 2)\n",
    "        } else if (t2 %in% batch2_sample) {\n",
    "            treatment_batch_sample <- c(1, 1, 2)\n",
    "        } else {\n",
    "            treatment_batch_sample <- c(1, 1, 3)\n",
    "        }\n",
    "        two <- tissues[t2]\n",
    "        out_f <- paste(one, \"_\", two, \"_DEG.csv\", sep=\"\")\n",
    "        print(out_f)\n",
    "        c1 = paste(one, \"_rep1\", sep=\"\")\n",
    "        c2 = paste(one, \"_rep2\", sep=\"\")\n",
    "        c3 = paste(one, \"_rep3\", sep=\"\")\n",
    "        t1 = paste(two, \"_rep1\", sep=\"\")\n",
    "        t2 = paste(two, \"_rep2\", sep=\"\")\n",
    "        t3 = paste(two, \"_rep3\", sep=\"\")\n",
    "        control_treatment = c(c1,c2,c3,t1,t2,t3)\n",
    "        print(control_treatment)\n",
    "        #print(control_treatment)\n",
    "        countData = select(df, control_treatment)\n",
    "        print(head(countData))\n",
    "            #rename\n",
    "        names(countData)[names(countData) == c1] <- \"Control1\"\n",
    "        names(countData)[names(countData) == c2] <- \"Control2\"\n",
    "        names(countData)[names(countData) == c3] <- \"Control3\"\n",
    "        names(countData)[names(countData) == t1] <- \"Treatment1\"\n",
    "        names(countData)[names(countData) == t2] <- \"Treatment2\"\n",
    "        names(countData)[names(countData) == t3] <- \"Treatment3\"        \n",
    "        condition <- factor(c(\"Control\", \"Control\", \"Control\", \"Treat\", \"Treat\", \"Treat\"))\n",
    "        batch <- factor(c(control_batch_sample, treatment_batch_sample))\n",
    "        #colData <- data.frame(row.names=colnames(countData), condition)\n",
    "        colData <- data.frame(condition=condition,batch=batch)\n",
    "        all(rownames(colData)) == colnames(countData)\n",
    "        dds <- DESeqDataSetFromMatrix(countData, DataFrame(colData), design=~batch+condition)\n",
    "        #View(dds)\n",
    "        dds <- DESeq(dds)\n",
    "        resultsNames(dds)\n",
    "        res <- results(dds)\n",
    "        table(res$padj==0)\n",
    "        mcols(res, use.names=TRUE)\n",
    "        res <- res[order(res$padj), ]\n",
    "        res <- subset(res, padj != 0)\n",
    "        GeneId <- row.names(res)\n",
    "        logFC <- res$log2FoldChange\n",
    "        padj <- res$padj\n",
    "        data <- data.frame(GeneId=GeneId, logFC=logFC, padj=padj)\n",
    "        data$sig[(data$padj>0.01)|((data$logFC < 1)&(data$logFC> -1))] <- \"Not\"\n",
    "        data$sig[data$padj <= 0.01 & data$logFC >= 1] <- \"Up\"\n",
    "        data$sig[data$padj <= 0.01 & data$logFC <= -1] <- \"Down\"\n",
    "        #data_out = subset(data, sig != 'Not')\n",
    "        write.csv(data, file=out_f,row.names=F)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Warning message in all(rownames(colData)):\n",
      "\"coercing argument of type 'character' to logical\"\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## 分析同源差异表达基因\n",
    "library(DESeq2)\n",
    "library(dplyr)\n",
    "#setwd(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AA\")\n",
    "#setwd(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AD\")\n",
    "#setwd(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/DD\")\n",
    "setwd(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/DA\")\n",
    "\n",
    "tissues = c('anther', 'ovule', 'root',  'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma')\n",
    "batch1_sample <- c('anther', 'hypocotyl')\n",
    "batch2_sample <- c('cotyledon', 'leaf', 'ovule', 'petal', 'root', 'stem', 'stigma')\n",
    "batch3_sample <- c('radicle', 'fiber_10DPA', 'fiber_20DPA')\n",
    "for (t1 in seq(1, 11)) {\n",
    "    one <- tissues[t1]\n",
    "    if (t1 %in% batch1_sample) {\n",
    "        control_batch_sample <- c(1, 2, 2)\n",
    "    } else if (t1 %in% batch2_sample) {\n",
    "        control_batch_sample <- c(1, 1, 2)\n",
    "    } else {\n",
    "        control_batch_sample <- c(1, 1, 3)\n",
    "    }\n",
    "    for (t2 in seq(t1+1,12)) {\n",
    "        two <- tissues[t2]\n",
    "        if (t2 %in% batch1_sample) {\n",
    "            treatment_batch_sample <- c(1, 2, 2)\n",
    "        } else if (t2 %in% batch2_sample) {\n",
    "            treatment_batch_sample <- c(1, 1, 2)\n",
    "        } else {\n",
    "            treatment_batch_sample <- c(1, 1, 3)\n",
    "        }\n",
    "        input_f <- paste(one, \"_\", two, \".bed\", sep=\"\")\n",
    "        out_f <- paste(one, \"_\", two, \"_DEG.bed\", sep=\"\")\n",
    "        df = read.table(input_f, sep=\"\\t\", header=T, row.names=1)\n",
    "        countData <- as.matrix(df)\n",
    "        condition <- factor(c(\"control\", \"control\", \"control\", \"treat\", \"treat\", \"treat\"))\n",
    "        batch <- factor(c(control_batch_sample, treatment_batch_sample))\n",
    "        #colData <- data.frame(row.names=colnames(countData), condition)\n",
    "        colData <- data.frame(condition=condition,batch=batch)\n",
    "        all(rownames(colData)) == colnames(countData)\n",
    "        # colData <- data.frame(row.names=colnames(countData), condition)\n",
    "        # all(rownames(colData)) == colnames(countData)\n",
    "        dds <- DESeqDataSetFromMatrix(countData, DataFrame(colData), design=~batch+condition)\n",
    "        #View(dds)\n",
    "        dds <- DESeq(dds)\n",
    "        resultsNames(dds)\n",
    "        res <- results(dds)\n",
    "        table(res$padj==0)\n",
    "        mcols(res, use.names=TRUE)\n",
    "        res <- res[order(res$padj),]\n",
    "        res <- subset(res, padj != 0)\n",
    "        GeneId <- row.names(res)\n",
    "        logFC <- res$log2FoldChange\n",
    "        padj <- res$padj\n",
    "        data <- data.frame(GeneId=GeneId, logFC=logFC, padj=padj)\n",
    "        data$sig[(data$padj>0.01)|((data$logFC < 1)&(data$logFC> -1))] <- \"Not\"\n",
    "        data$sig[data$padj <= 0.01 & data$logFC >= 1] <- \"Up\"\n",
    "        data$sig[data$padj <= 0.01 & data$logFC <= -1] <- \"Down\"\n",
    "        #data_out = subset(data, sig != 'Not')\n",
    "        write.csv(data, file=out_f, row.names=F)\n",
    "    }\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "starting httpd help server ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "browseVignettes(\"HTSFilter\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"d\"\n"
     ]
    }
   ],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "a0f669ea42a4dcea88d385f8f9b4c3d9b966e24234758831d1dbfdac925619a2"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
