{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 组织间两两比较统计多少基因能发生差异表达\n",
    "import pandas as pd\n",
    "tissues = ['anthers', 'cotyledon', 'ovules',  'hypocotyl', 'leaf', 'petals', 'radicle', 'root', 'stem', 'stigma']\n",
    "df_result = pd.DataFrame()\n",
    "h = 1\n",
    "for i in range(9):\n",
    "    first = tissues[i]\n",
    "    for j in range(i+1, 10):\n",
    "        second = tissues[j]\n",
    "        inputf = \"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/T_VS_T/{0}_{1}_DEG.bed\".format(first, second)\n",
    "        #print(inputf)\n",
    "        df = pd.read_csv(inputf, sep=\"\\t\")\n",
    "        df['number'] = [1]*df.shape[0]\n",
    "        df_tmp = df.groupby(by=[\"sig\"], as_index=False)['number'].sum()\n",
    "        df_tmp['start'] = first\n",
    "        df_tmp['end'] = second\n",
    "        df_tmp['group'] = h\n",
    "        h += 1\n",
    "        df_result = pd.concat([df_result, df_tmp], axis=0)\n",
    "for t in tissues:\n",
    "    add_line1 = ['Up', 1, t, t, h]\n",
    "    add_line2 = ['Down', 1, t, t, h]\n",
    "    df_result.loc[df_result.shape[0]] = add_line1\n",
    "    df_result.loc[df_result.shape[0]] = add_line2\n",
    "    h+=1\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/T_VS_T/all_DEG_statistics.bed\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Subgenome   sig  number    start        end  group\n",
      "0         At  Down    7912  anthers  cotyledon      1\n",
      "1         At    Up    8102  anthers  cotyledon      1\n",
      "2         Dt  Down    8514  anthers  cotyledon      1\n",
      "3         Dt    Up    8394  anthers  cotyledon      1\n",
      "0         At  Down    8242  anthers     ovules      2\n",
      "..       ...   ...     ...      ...        ...    ...\n",
      "3         Dt    Up    7533     root     stigma     44\n",
      "0         At  Down    7521     stem     stigma     45\n",
      "1         At    Up    7061     stem     stigma     45\n",
      "2         Dt  Down    7906     stem     stigma     45\n",
      "3         Dt    Up    7341     stem     stigma     45\n",
      "\n",
      "[180 rows x 6 columns]\n"
     ]
    }
   ],
   "source": [
    "## 组织间两两比较统计多少基因能发生差异表达, 区分两个亚基因组\n",
    "import pandas as pd\n",
    "tissues = ['anthers', 'cotyledon', 'ovules',  'hypocotyl', 'leaf', 'petals', 'radicle', 'root', 'stem', 'stigma']\n",
    "df_result = pd.DataFrame()\n",
    "h = 1\n",
    "for i in range(9):\n",
    "    first = tissues[i]\n",
    "    for j in range(i+1, 10):\n",
    "        second = tissues[j]\n",
    "        inputf = \"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/T_VS_T/result/{0}_{1}_DEG.bed\".format(first, second)\n",
    "        #print(inputf)\n",
    "        df = pd.read_csv(inputf, sep=\"\\t\")\n",
    "        df['number'] = [1]*df.shape[0]\n",
    "        df['Subgenome'] = df['GeneId'].apply(lambda x: 'At' if \"Ghir_A\" in str(x) else \"Dt\")\n",
    "        df_tmp = df.groupby(by=[\"Subgenome\", \"sig\"], as_index=False)['number'].sum()\n",
    "        df_tmp['start'] = first\n",
    "        df_tmp['end'] = second\n",
    "        df_tmp['group'] = h\n",
    "        h += 1\n",
    "        df_result = pd.concat([df_result, df_tmp], axis=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tkinter import _flatten\n",
    "df_at = df_result.query(\"Subgenome=='At'\").copy()\n",
    "df_dt = df_result.query(\"Subgenome=='Dt'\").copy()\n",
    "other_list = list(_flatten([[i, i] for i in range(46, 91)]))\n",
    "df_dt['group'] = other_list\n",
    "df_dt.columns = ['Subgenome', 'sig', 'number', 'end', 'start', 'group']\n",
    "df_at.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/T_VS_T/all_DEG_At_statistics.bed\", sep=\"\\t\", index=False)\n",
    "df_dt.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/T_VS_T/all_DEG_Dt_statistics.bed\", sep=\"\\t\", index=False)\n",
    "\n",
    "\n",
    "## 添加对称轴\n",
    "\n",
    "# for t in tissues:\n",
    "#     add_line1 = ['Up', 1, t, t, h]\n",
    "#     add_line2 = ['Down', 1, t, t, h]\n",
    "#     df_result.loc[df_result.shape[0]] = add_line1\n",
    "#     df_result.loc[df_result.shape[0]] = add_line2\n",
    "#     h+=1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将组织差异表达基因，添加各组织TPM 值\n",
    "import pandas as pd\n",
    "tissues = ['Anther', 'Cotyledon', 'Ovule',  'Hypocotyl', 'Leaf', 'Petal', 'Radicle', 'Root', 'Stem', 'Stigma']\n",
    "df_result = pd.DataFrame()\n",
    "df_TPM = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/expression_gene/J668_all_tissue_average_TPM.txt\", sep=\"\\t\")\n",
    "h = 1\n",
    "for i in range(9):\n",
    "    first = tissues[i]\n",
    "    for j in range(i+1, 10):\n",
    "        second = tissues[j]\n",
    "        inputf = \"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/T_VS_T/result/{0}_{1}_DEG.bed\".format(first, second)\n",
    "        outputf = \"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/T_VS_T/result_TPM/{0}_{1}_DEG.bed\".format(first, second)\n",
    "        df = pd.read_csv(inputf, sep=\"\\t\")\n",
    "        df_TPM_tissues = df_TPM[['GeneId', first, second]].copy()\n",
    "        df_result = pd.merge(df, df_TPM_tissues, how='inner', on=['GeneId'])\n",
    "        df_result.columns = ['GeneId', 'logFC', 'padj', 'sig', 'control', 'treatment']\n",
    "        df_result.to_csv(outputf, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将组织差异表达基因，添加各组织TPM 值, 新RNA-Seq数据\n",
    "import pandas as pd\n",
    "tissues = ['anther', 'ovule', 'root',  'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df_result = pd.DataFrame()\n",
    "df_TPM = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data/J668_all_Tissue_mean_TPM.csv\")\n",
    "h = 1\n",
    "for i in range(11):\n",
    "    first = tissues[i]\n",
    "    for j in range(i+1, 12):\n",
    "        second = tissues[j]\n",
    "        inputf = \"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/{0}_{1}_DEG.csv\".format(first, second)\n",
    "        outputf = \"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/point_figure/{0}_{1}_DEG.bed\".format(first, second)\n",
    "        df = pd.read_csv(inputf)\n",
    "        df_TPM_tissues = df_TPM[['GeneId', first, second]].copy()\n",
    "        df_result = pd.merge(df, df_TPM_tissues, how='inner', on=['GeneId'])\n",
    "        df_result.columns = ['GeneId', 'logFC', 'padj', 'sig', 'control', 'treatment']\n",
    "        df_result.to_csv(outputf, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 整合组织差异\n",
    "import pandas as pd\n",
    "tissues = ['Anther', 'Cotyledon', 'Ovule',  'Hypocotyl', 'Leaf', 'Petal', 'Radicle', 'Root', 'Stem', 'Stigma']\n",
    "df_result = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    inputf = \"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/{0}/{0}_reads.bed\".format(t)\n",
    "    df_tmp = pd.read_csv(inputf, sep=\"\\t\")\n",
    "    t1, t2, t3 = \"{}1\".format(t), \"{}2\".format(t), \"{}3\".format(t)\n",
    "    df_tmp.columns = ['GeneId', t1, t2, t3]\n",
    "    if df_result.shape[0]==0:\n",
    "        df_result = df_tmp.copy()\n",
    "    else:\n",
    "        df_result = pd.merge(df_result, df_tmp, how='inner', on=['GeneId'])\n",
    "    \n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/all_tissues_reads.bed\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 分析各组织亚基因组reads\n",
    "import pandas as pd\n",
    "tissues = ['Anther', 'Cotyledon', 'Ovule',  'Hypocotyl', 'Leaf', 'Petal', 'Radicle', 'Root', 'Stem', 'Stigma']\n",
    "df_reads = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/all_tissues_reads.bed\", sep=\"\\t\")\n",
    "df_hg = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/homo_gene2/Ghir_homo_gene_pairs.txt\", sep=\"\\t\")\n",
    "df_At , df_Dt = df_hg[['At']].copy(), df_hg[['Dt']].copy()\n",
    "df_At.columns = ['GeneId']\n",
    "df_Dt.columns = ['GeneId']\n",
    "for i in range(9):\n",
    "    A1, A2, A3 = \"{}1\".format(tissues[i]), \"{}2\".format(tissues[i]), \"{}3\".format(tissues[i])\n",
    "    df_A = df_reads[['GeneId', A1, A2, A3]].copy()\n",
    "    df_A_At = pd.merge(df_At, df_A, how='left', on=['GeneId'])\n",
    "    df_A_Dt = pd.merge(df_Dt, df_A, how='left', on=['GeneId'])\n",
    "    for j in range(i+1, 10):\n",
    "        B1, B2, B3 = \"{}1\".format(tissues[j]), \"{}2\".format(tissues[j]), \"{}3\".format(tissues[j])\n",
    "        df_B = df_reads[['GeneId', B1, B2, B3]].copy()\n",
    "        df_B_At = pd.merge(df_At, df_B, how='left', on=['GeneId'])[[B1, B2, B3]]\n",
    "        df_B_Dt = pd.merge(df_Dt, df_B, how='left', on=['GeneId'])[[B1, B2, B3]]\n",
    "        AA_f = \"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/ST_VS_ST/AA/{0}_{1}.bed\".format(tissues[i], tissues[j])\n",
    "        AD_f = \"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/ST_VS_ST/AD/{0}_{1}.bed\".format(tissues[i], tissues[j])\n",
    "        DA_f = \"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/ST_VS_ST/DA/{0}_{1}.bed\".format(tissues[i], tissues[j])\n",
    "        DD_f = \"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/ST_VS_ST/DD/{0}_{1}.bed\".format(tissues[i], tissues[j])\n",
    "        df_AA = pd.concat([df_A_At, df_B_At], axis=1)\n",
    "        df_AD = pd.concat([df_A_At, df_B_Dt], axis=1)\n",
    "        df_DD = pd.concat([df_A_Dt, df_B_Dt], axis=1)\n",
    "        df_DA = pd.concat([df_A_Dt, df_B_At], axis=1)\n",
    "        df_AA.columns = ['GeneId', 'control1', 'control2', 'control3', 'treatment1', 'treatment2', 'treatment3']\n",
    "        df_AD.columns = ['GeneId', 'control1', 'control2', 'control3', 'treatment1', 'treatment2', 'treatment3']\n",
    "        df_DD.columns = ['GeneId', 'control1', 'control2', 'control3', 'treatment1', 'treatment2', 'treatment3']\n",
    "        df_DA.columns = ['GeneId', 'control1', 'control2', 'control3', 'treatment1', 'treatment2', 'treatment3']\n",
    "        df_AA.to_csv(AA_f, sep=\"\\t\", index=False)\n",
    "        df_AD.to_csv(AD_f, sep='\\t', index=False)\n",
    "        df_DA.to_csv(DA_f, sep=\"\\t\", index=False)\n",
    "        df_DD.to_csv(DD_f, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 分析各组织亚基因组reads\n",
    "import pandas as pd\n",
    "tissues = ['anther', 'ovule', 'root',  'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df_reads = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data/J668_all_Tissue_reads.csv\")\n",
    "df_hg = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/homo_gene2/Ghir_homo_gene_pairs.txt\", sep=\"\\t\")\n",
    "df_At , df_Dt = df_hg[['GeneId_At']].copy(), df_hg[['GeneId_Dt']].copy()\n",
    "df_At.columns = ['GeneId']\n",
    "df_Dt.columns = ['GeneId']\n",
    "for i in range(11):\n",
    "    A1, A2, A3 = \"{}_rep1\".format(tissues[i]), \"{}_rep2\".format(tissues[i]), \"{}_rep3\".format(tissues[i])\n",
    "    df_A = df_reads[['GeneId', A1, A2, A3]].copy()\n",
    "    df_A_At = pd.merge(df_At, df_A, how='left', on=['GeneId'])\n",
    "    df_A_Dt = pd.merge(df_Dt, df_A, how='left', on=['GeneId'])\n",
    "    for j in range(i+1, 12):\n",
    "        B1, B2, B3 = \"{}_rep1\".format(tissues[j]), \"{}_rep2\".format(tissues[j]), \"{}_rep3\".format(tissues[j])\n",
    "        df_B = df_reads[['GeneId', B1, B2, B3]].copy()\n",
    "        df_B_At = pd.merge(df_At, df_B, how='left', on=['GeneId'])[[B1, B2, B3]]\n",
    "        df_B_Dt = pd.merge(df_Dt, df_B, how='left', on=['GeneId'])[[B1, B2, B3]]\n",
    "        AA_f = \"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AA/{0}_{1}.bed\".format(tissues[i], tissues[j])\n",
    "        AD_f = \"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AD/{0}_{1}.bed\".format(tissues[i], tissues[j])\n",
    "        DA_f = \"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/DA/{0}_{1}.bed\".format(tissues[i], tissues[j])\n",
    "        DD_f = \"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/DD/{0}_{1}.bed\".format(tissues[i], tissues[j])\n",
    "        df_AA = pd.concat([df_A_At, df_B_At], axis=1)\n",
    "        df_AD = pd.concat([df_A_At, df_B_Dt], axis=1)\n",
    "        df_DD = pd.concat([df_A_Dt, df_B_Dt], axis=1)\n",
    "        df_DA = pd.concat([df_A_Dt, df_B_At], axis=1)\n",
    "        df_AA.columns = ['GeneId', 'control1', 'control2', 'control3', 'treatment1', 'treatment2', 'treatment3']\n",
    "        df_AD.columns = ['GeneId', 'control1', 'control2', 'control3', 'treatment1', 'treatment2', 'treatment3']\n",
    "        df_DD.columns = ['GeneId', 'control1', 'control2', 'control3', 'treatment1', 'treatment2', 'treatment3']\n",
    "        df_DA.columns = ['GeneId', 'control1', 'control2', 'control3', 'treatment1', 'treatment2', 'treatment3']\n",
    "        df_AA.to_csv(AA_f, sep=\"\\t\", index=False)\n",
    "        df_AD.to_csv(AD_f, sep='\\t', index=False)\n",
    "        df_DA.to_csv(DA_f, sep=\"\\t\", index=False)\n",
    "        df_DD.to_csv(DD_f, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将所有同源基因对FPKM 整理好\n",
    "import pandas as pd\n",
    "tissues = ['anther', 'cotyledon', 'ovule',  'hypocotyl', 'leaf', 'petal', 'radicle', 'root', 'stem', 'stigma', 'fiber_10DPA', 'fiber_20DPA']\n",
    "df_result = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_seq/homo_gene2/gene_bias/TPM/{0}/{0}_pairs_tpm.txt\".format(t)\n",
    "    df = pd.read_csv(input_f, sep=\"\\t\")\n",
    "    df['At_TPM'] = (df['at_tpm1'] + df['at_tpm2'] + df['at_tpm3'])/3\n",
    "    df['Dt_TPM'] = (df['dt_tpm1'] + df['dt_tpm2'] + df['dt_tpm3'])/3\n",
    "    df_tmp = df[['GeneId_At', 'GeneId_Dt', 'At_TPM', 'Dt_TPM']].copy()\n",
    "    #df_tmp.columns = ['GeneId_at', 'GeneId_dt', 'At_TPM', 'Dt_TPM']\n",
    "    df_tmp['tags'] = [t]*df_tmp.shape[0]\n",
    "    df_result = pd.concat([df_result, df_tmp], axis=0)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_seq/homo_gene2/all_HG_TPM.bed\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将亚组间差异表达基因添加表达量\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "tissues = ['Anther', 'Cotyledon', 'Ovule',  'Hypocotyl', 'Leaf', 'Petal', 'Radicle', 'Root', 'Stem', 'Stigma']\n",
    "df_hg = pd.read_csv('G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/homo_gene2/Ghir_homo_gene_pairs.txt', sep=\"\\t\")\n",
    "df_hg.columns = ['GeneId_at', 'GeneId_dt']\n",
    "df_hg_tpm = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/homo_gene2/all_HG_TPM.bed\", sep=\"\\t\")\n",
    "df_hg_tpm['At_log2TPM_1'] = df_hg_tpm['At_TPM'].apply(lambda x: np.log2(x+1.5))\n",
    "df_hg_tpm['Dt_log2TPM_1'] = df_hg_tpm['Dt_TPM'].apply(lambda x: -(np.log2(x+1.5)))\n",
    "h = 1\n",
    "def hg_tpm(F_path, df_h, tag1, tag2, df1, df2, tags, sigs):\n",
    "    df1.columns = [tag1, 'control']\n",
    "    df2.columns = [tag2, 'treatment']\n",
    "    df = pd.read_csv(F_path, sep=\"\\t\")\n",
    "    df.rename(columns={'GeneId' : tag1}, inplace=True)\n",
    "    df = pd.merge(df, df_hg, how='inner', on=[tag1])\n",
    "    dfA = pd.merge(df, df1, how='inner', on=[tag1])\n",
    "    df_AA = pd.merge(dfA, df2, how='inner', on=[tag2])\n",
    "    df_AA['tags'] = [tags]*df_AA.shape[0]\n",
    "    df_AA['sigs'] = [sigs]*df_AA.shape[0]\n",
    "    return df_AA\n",
    "for i in range(9):\n",
    "    first = tissues[i]\n",
    "    df_A_At = df_hg_tpm.query(\"tags==@first\")[['GeneId_at', 'At_log2TPM_1']].copy()\n",
    "    df_A_Dt = df_hg_tpm.query(\"tags==@first\")[['GeneId_dt', 'Dt_log2TPM_1']].copy()\n",
    "    for j in range(i+1, 10):\n",
    "        second = tissues[j]\n",
    "        output_f = \"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/ST_VS_ST/{0}_{1}_subg_TPM.bed\".format(first, second)\n",
    "        df_D_At = df_hg_tpm.query(\"tags==@second\")[['GeneId_at', 'At_log2TPM_1']].copy()\n",
    "        df_D_Dt = df_hg_tpm.query(\"tags==@second\")[['GeneId_dt', 'Dt_log2TPM_1']].copy()\n",
    "        ## AA\n",
    "        input_AA = \"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/ST_VS_ST/AA/{0}_{1}_DEG.bed\".format(first, second)\n",
    "        df_AA = hg_tpm(input_AA, df_hg, 'GeneId_at', 'GeneId_at', df_A_At, df_D_At, 'Same', 'AA')\n",
    "        ## AD\n",
    "        input_AD = \"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/ST_VS_ST/AD/{0}_{1}_DEG.bed\".format(first, second)\n",
    "        df_AD = hg_tpm(input_AD, df_hg, \"GeneId_at\", \"GeneId_dt\", df_A_At, df_D_Dt, 'Subg', 'AD')\n",
    "        ## DD\n",
    "        input_DD = \"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/ST_VS_ST/DD/{0}_{1}_DEG.bed\".format(first, second)\n",
    "        df_DD = hg_tpm(input_DD, df_hg, \"GeneId_dt\", \"GeneId_dt\", df_A_Dt, df_D_Dt, 'Same', \"DD\")\n",
    "        ## DA\n",
    "        input_DA = \"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/ST_VS_ST/DA/{0}_{1}_DEG.bed\".format(first, second)\n",
    "        df_DA = hg_tpm(input_DA, df_hg, \"GeneId_dt\", \"GeneId_at\", df_A_Dt, df_D_At, 'Subg', \"DA\")\n",
    "        df_result = pd.concat([df_AA, df_AD, df_DD, df_DA], axis=0)\n",
    "        df_result['sig_tags'] = df_result['sig'].map(str) + \"-\" + df_result['tags'].map(str)\n",
    "        df_result.to_csv(output_f, sep=\"\\t\", index=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将所有同源基因对的TPM 整理好\n",
    "tissues = ['anther',  'cotyledon', 'ovule',  'hypocotyl', 'leaf', 'petal', 'radicle', 'root', 'stem', 'stigma', 'fiber_10DPA', 'fiber_20DPA']\n",
    "df_result = pd.DataFrame()\n",
    "\n",
    "def HG_TPM(tags):\n",
    "    df_HG_pairs = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/homo_gene2/Ghir_homo_gene_pairs.txt\", sep=\"\\t\")\n",
    "    df_TPM = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data/J668_all_Tissue_mean_TPM.csv\")\n",
    "    df_TPM_tag = df_TPM[['GeneId', tags]].copy()\n",
    "    df_TPM_tag.columns = ['GeneId_At', 'At_TPM']\n",
    "    df_At = pd.merge(df_HG_pairs, df_TPM_tag, how='left', on=['GeneId_At'])\n",
    "    df_TPM_tag.columns = ['GeneId_Dt', 'Dt_TPM']\n",
    "    df_At_Dt = pd.merge(df_At, df_TPM_tag, how='left', on=['GeneId_Dt'])\n",
    "    df_At_Dt['tags'] = [tags]*df_At_Dt.shape[0]\n",
    "    return df_At_Dt\n",
    "for t in tissues:\n",
    "    df_tmp = HG_TPM(t)\n",
    "    df_result = pd.concat([df_result, df_tmp], axis=0)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data/J668_all_Tissue_HG_pair_TPM.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将亚组间差异表达基因添加表达量\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "tissues = ['Anther', 'Cotyledon', 'Ovule',  'Hypocotyl', 'Leaf', 'Petal', 'Radicle', 'Root', 'Stem', 'Stigma']\n",
    "df_hg = pd.read_csv('G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/homo_gene2/Ghir_homo_gene_pairs.txt', sep=\"\\t\")\n",
    "df_hg.columns = ['GeneId_at', 'GeneId_dt']\n",
    "df_hg_tpm = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/homo_gene2/all_HG_TPM.bed\", sep=\"\\t\")\n",
    "df_hg_tpm['At_log2TPM_1'] = df_hg_tpm['At_TPM'].apply(lambda x: np.log2(x+1.5))\n",
    "df_hg_tpm['Dt_log2TPM_1'] = df_hg_tpm['Dt_TPM'].apply(lambda x: -(np.log2(x+1.5)))\n",
    "h = 1\n",
    "def hg_tpm(F_path, df_h, tag1, tag2, df1, df2, tags, sigs):\n",
    "    df1.columns = [tag1, 'control']\n",
    "    df2.columns = [tag2, 'treatment']\n",
    "    df = pd.read_csv(F_path, sep=\"\\t\")\n",
    "    df.rename(columns={'GeneId' : tag1}, inplace=True)\n",
    "    df = pd.merge(df, df_hg, how='inner', on=[tag1])\n",
    "    dfA = pd.merge(df, df1, how='inner', on=[tag1])\n",
    "    df_AA = pd.merge(dfA, df2, how='inner', on=[tag2])\n",
    "    df_AA['tags'] = [tags]*df_AA.shape[0]\n",
    "    df_AA['sigs'] = [sigs]*df_AA.shape[0]\n",
    "    return df_AA\n",
    "for i in range(9):\n",
    "    first = tissues[i]\n",
    "    df_A_At = df_hg_tpm.query(\"tags==@first\")[['GeneId_at', 'At_log2TPM_1']].copy()\n",
    "    df_A_Dt = df_hg_tpm.query(\"tags==@first\")[['GeneId_dt', 'Dt_log2TPM_1']].copy()\n",
    "    for j in range(i+1, 10):\n",
    "        second = tissues[j]\n",
    "        output_f = \"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/ST_VS_ST/{0}_{1}_subg_TPM.bed\".format(first, second)\n",
    "        df_D_At = df_hg_tpm.query(\"tags==@second\")[['GeneId_at', 'At_log2TPM_1']].copy()\n",
    "        df_D_Dt = df_hg_tpm.query(\"tags==@second\")[['GeneId_dt', 'Dt_log2TPM_1']].copy()\n",
    "        ## AA\n",
    "        input_AA = \"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/ST_VS_ST/AA/{0}_{1}_DEG.bed\".format(first, second)\n",
    "        df_AA = hg_tpm(input_AA, df_hg, 'GeneId_at', 'GeneId_at', df_A_At, df_D_At, 'Same', 'AA')\n",
    "        ## AD\n",
    "        input_AD = \"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/ST_VS_ST/AD/{0}_{1}_DEG.bed\".format(first, second)\n",
    "        df_AD = hg_tpm(input_AD, df_hg, \"GeneId_at\", \"GeneId_dt\", df_A_At, df_D_Dt, 'Subg', 'AD')\n",
    "        ## DD\n",
    "        input_DD = \"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/ST_VS_ST/DD/{0}_{1}_DEG.bed\".format(first, second)\n",
    "        df_DD = hg_tpm(input_DD, df_hg, \"GeneId_dt\", \"GeneId_dt\", df_A_Dt, df_D_Dt, 'Same', \"DD\")\n",
    "        ## DA\n",
    "        input_DA = \"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/ST_VS_ST/DA/{0}_{1}_DEG.bed\".format(first, second)\n",
    "        df_DA = hg_tpm(input_DA, df_hg, \"GeneId_dt\", \"GeneId_at\", df_A_Dt, df_D_At, 'Subg', \"DA\")\n",
    "        df_result = pd.concat([df_AA, df_AD, df_DD, df_DA], axis=0)\n",
    "        df_result['sig_tags'] = df_result['sig'].map(str) + \"-\" + df_result['tags'].map(str)\n",
    "        df_result.to_csv(output_f, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将亚组间差异表达基因添加表达量, 新的RNA_Seq数据\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "tissues = ['anther', 'ovule', 'root',  'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df_hg = pd.read_csv('G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_seq/homo_gene2/Ghir_homo_gene_pairs.txt', sep=\"\\t\")\n",
    "df_hg.columns = ['GeneId_At', 'GeneId_Dt']\n",
    "df_hg_tpm = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data/J668_all_Tissue_HG_pair_TPM.csv\")\n",
    "df_hg_tpm['At_log2TPM_1'] = df_hg_tpm['At_TPM'].apply(lambda x: np.log2(x+1.5))\n",
    "df_hg_tpm['Dt_log2TPM_1'] = df_hg_tpm['Dt_TPM'].apply(lambda x: -(np.log2(x+1.5)))\n",
    "h = 1\n",
    "def hg_tpm(F_path, df_h, tag1, tag2, df1, df2, tags, sigs):\n",
    "    df1.columns = [tag1, 'control']\n",
    "    df2.columns = [tag2, 'treatment']\n",
    "    #print(df1)\n",
    "    df = pd.read_csv(F_path)\n",
    "    #print(df.query(\"sig=='Up'\").shape[0])\n",
    "    df.rename(columns={'GeneId' : tag1}, inplace=True)\n",
    "    df = pd.merge(df, df_hg, how='inner', on=[tag1])\n",
    "    #print(df.query(\"sig=='Up'\").shape[0])\n",
    "    dfA = pd.merge(df, df1, how='inner', on=[tag1])\n",
    "    #print(dfA.query(\"sig=='Up'\").shape[0])\n",
    "    #rint(dfA.head())\n",
    "    #print(df2.head())\n",
    "    df_AA = pd.merge(dfA, df2, how='inner', on=[tag2])\n",
    "    #print(df_AA.query(\"sig=='Up'\").shape[0])\n",
    "    df_AA['tags'] = [tags]*df_AA.shape[0]\n",
    "    df_AA['sigs'] = [sigs]*df_AA.shape[0]\n",
    "    return df_AA\n",
    "for i in range(11):\n",
    "    first = tissues[i]\n",
    "    df_A_At = df_hg_tpm.query(\"tags==@first\")[['GeneId_At', 'At_log2TPM_1']].copy()\n",
    "    df_A_Dt = df_hg_tpm.query(\"tags==@first\")[['GeneId_Dt', 'Dt_log2TPM_1']].copy()\n",
    "    for j in range(i+1, 12):\n",
    "        second = tissues[j]\n",
    "        #print(first, second)\n",
    "        output_f = \"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/point_figure/{0}_{1}_subg_TPM.bed\".format(first, second)\n",
    "        df_D_At = df_hg_tpm.query(\"tags==@second\")[['GeneId_At', 'At_log2TPM_1']].copy()\n",
    "        df_D_Dt = df_hg_tpm.query(\"tags==@second\")[['GeneId_Dt', 'Dt_log2TPM_1']].copy()\n",
    "        ## AA\n",
    "        input_AA = \"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AA/{0}_{1}_DEG.bed\".format(first, second)\n",
    "        df_AA = hg_tpm(input_AA, df_hg, 'GeneId_At', 'GeneId_At', df_A_At, df_D_At, 'Same', 'AA')\n",
    "        ## AD\n",
    "        input_AD = \"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AD/{0}_{1}_DEG.bed\".format(first, second)\n",
    "        df_AD = hg_tpm(input_AD, df_hg, \"GeneId_At\", \"GeneId_Dt\", df_A_At, df_D_Dt, 'Subg', 'AD')\n",
    "        ## DD\n",
    "        input_DD = \"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/DD/{0}_{1}_DEG.bed\".format(first, second)\n",
    "        df_DD = hg_tpm(input_DD, df_hg, \"GeneId_Dt\", \"GeneId_Dt\", df_A_Dt, df_D_Dt, 'Same', \"DD\")\n",
    "        #print(df_DD.head())\n",
    "        ## DA\n",
    "        input_DA = \"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/DA/{0}_{1}_DEG.bed\".format(first, second)\n",
    "        df_DA = hg_tpm(input_DA, df_hg, \"GeneId_Dt\", \"GeneId_At\", df_A_Dt, df_D_At, 'Subg', \"DA\")\n",
    "        df_result = pd.concat([df_AA, df_AD, df_DD, df_DA], axis=0)\n",
    "        df_result['sig_tags'] = df_result['sig'].map(str) + \"-\" + df_result['tags'].map(str)\n",
    "        #print('a', df_result.query(\"sigs=='AA' & sig=='Up'\").shape[0])\n",
    "        df_result.to_csv(output_f, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6315\n",
      "6325\n",
      "24803 24914\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/anther_ovule_DEG.csv\")\n",
    "df_hg = pd.read_csv('G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_seq/homo_gene2/Ghir_homo_gene_pairs.txt', sep=\"\\t\")\n",
    "df.columns = ['GeneId_At', 'logFC', 'padj', 'sig']\n",
    "#print(df.shape[0], df_hg.shape[0])\n",
    "df_At = pd.merge(df, df_hg, how='inner', on=['GeneId_At'])\n",
    "print(df_At.query(\"sig=='Up'\").shape[0])\n",
    "df.columns = ['GeneId_Dt', 'logFC', 'padj', 'sig']\n",
    "df_Dt = pd.merge(df, df_hg, how='inner', on=['GeneId_Dt'])\n",
    "print(df_Dt.query(\"sig=='Up'\").shape[0])\n",
    "print(df_At.shape[0], df_Dt.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 分析同源基因在组织间差异表达，同时表达模式并不一致，统计两两组织间所有的这种情况\n",
    "import pandas as pd\n",
    "## 整理同源基因在组织间差异表达模式\n",
    "def Hg_tissues_DEG_mode(df_AA, df_DD, df_hg):\n",
    "    df_AA, df_DD, df_hg = df_AA[['GeneId', 'sig']].copy(), df_DD[['GeneId', 'sig', 'logFC']].copy(), df_hg.copy()\n",
    "    df_AA.columns = ['GeneId_At', 'sig_At', 'logFC_At']\n",
    "    df_DD.columns = ['GeneId_Dt', 'sig_Dt', 'logFC_Dt']\n",
    "    df_A = pd.merge(df_hg, df_AA, how='left', on=['GeneId_At'])\n",
    "    df_AD = pd.merge(df_A, df_DD, how='left', on=['GeneId_Dt'])\n",
    "    df_AD.fillna(\"Not\", inplace=True)\n",
    "    return df_AD\n",
    "\n",
    "## 整理所有组织间的同源基因差异表达模式\n",
    "def All_Tissue_DEM():\n",
    "    tissues = ['anther', 'ovule', 'root',  'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "    df_hg = pd.read_csv('G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_seq/homo_gene2/Ghir_homo_gene_pairs.txt', sep=\"\\t\")\n",
    "    df_result = pd.DataFrame()\n",
    "    for i in range(11):\n",
    "        for j in range(i+1, 12):\n",
    "            T1 = tissues[i]\n",
    "            T2 = tissues[j]\n",
    "            AA_input = \"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AA/{0}_{1}_DEG.bed\".format(T1, T2)\n",
    "            DD_input = \"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/DD/{0}_{1}_DEG.bed\".format(T1, T2)\n",
    "            df_AA = pd.read_csv(AA_input)\n",
    "            df_DD = pd.read_csv(DD_input)\n",
    "            df_tmp = Hg_tissues_DEG_mode(df_AA, df_DD, df_hg)\n",
    "            df_tmp['tissues'] = [\"{0}_{1}\".format(T1, T2)]*df_tmp.shape[0]\n",
    "            df_result = pd.concat([df_result, df_tmp], axis=0)\n",
    "    return df_result\n",
    "## 同源基因组织差异表达模式\n",
    "df_Tissues_DEM = All_Tissue_DEM()\n",
    "df_Tissues_DEM.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AA_DD_HDEM/J668_allTissue_DEM.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              tissues sig_At sig_Dt  number\n",
      "0    anther_cotyledon   Down   Down    4514\n",
      "1    anther_cotyledon   Down    Not    1756\n",
      "2    anther_cotyledon   Down     Up     223\n",
      "3    anther_cotyledon    Not   Down    1714\n",
      "4    anther_cotyledon    Not    Not    8983\n",
      "5    anther_cotyledon    Not     Up    1761\n",
      "6    anther_cotyledon     Up   Down     238\n",
      "7    anther_cotyledon     Up    Not    1652\n",
      "8    anther_cotyledon     Up     Up    4452\n",
      "9  anther_fiber_10DPA   Down   Down    4373\n"
     ]
    }
   ],
   "source": [
    "### 统计不同模式下同源基因对的数目\n",
    "import pandas\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AA_DD_HDEM/J668_allTissue_DEM.csv\")\n",
    "df['number'] = [1]*df.shape[0]\n",
    "df_sta = df.groupby(by=['tissues', 'sig_At', 'sig_Dt'], as_index=False)['number'].sum()\n",
    "print(df_sta.head(10))\n",
    "df_sta.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AA_DD_HDEM/J668_allTissue_DEM_sta.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将九种不同的表达模式转换为funkyheatmap能识别的格式\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AA_DD_HDEM/J668_allTissue_DEM_sta.csv\")\n",
    "df['sig'] = df['sig_At'].map(str) + \"_\" + df['sig_Dt'].map(str)\n",
    "df_wide = df.pivot(columns='sig', index='tissues', values='number')\n",
    "df_wide.reset_index(inplace=True)\n",
    "df_wide.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AA_DD_HDEM/J668_allTissue_DEM_funkyheatmap.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 绘制下标，图形格式\n",
    "tissues = ['anther', 'ovule', 'root',  'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df_result = pd.DataFrame(columns = ['index', 'number', 'tags1', 'tags2'])\n",
    "n = 1\n",
    "for i in range(11):\n",
    "    for j in range(i+1, 12):\n",
    "        T1 = tissues[i]\n",
    "        T2 = tissues[j]\n",
    "        df_result.loc[df_result.shape[0], ] = [n, 1, T1, T2]\n",
    "        n = n+1\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AA_DD_HDEM/J668_tissues_tags.csv\", index=False)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 组织差异表达基因数目，在两个组织间差异表达的数目，在三个组织间差异表达的数目，多组织差异表达基因数目\n",
    "import pandas as pd\n",
    "tissues = ['anther', 'ovule', 'root',  'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/all_gene_sorted.bed6\", sep=\"\\t\", names=['chr', 'start', 'end', 'GeneId', 'a', 'b'])\n",
    "df_GeneId = df[['GeneId']].copy()\n",
    "for i in range(11):\n",
    "    for j in range(i+1, 12):\n",
    "        T1, T2 = tissues[i], tissues[j]\n",
    "        input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/{0}_{1}_DEG.csv\".format(T1,T2)\n",
    "        df_tmp = pd.read_csv(input_f)\n",
    "        df_tmp = df_tmp[['GeneId', 'sig']].copy()\n",
    "        df_tmp.columns = ['GeneId', \"{0}-{1}\".format(T1,T2)]\n",
    "        df_GeneId = pd.merge(df_GeneId, df_tmp, how='left', on=['GeneId'])\n",
    "df_GeneId.set_index('GeneId', inplace=True)\n",
    "df_GeneId.dropna(inplace=True, how='all')\n",
    "df_GeneId.fillna(0, inplace=True)\n",
    "old_tags = ['Up', 'Down', 'Not']\n",
    "new_tags = [1,1,0]\n",
    "\n",
    "df_GeneId.replace(old_tags, new_tags, inplace=True)\n",
    "\n",
    "df_GeneId['DEG_T_N'] = df_GeneId.sum(axis=1)\n",
    "df_GeneId['number'] = [1]*df_GeneId.shape[0]\n",
    "df_GeneId.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/statistics/J668_DEG_N_tissues.csv\")\n",
    "\n",
    "df_result = df_GeneId.groupby(by=['DEG_T_N'], as_index=False)['number'].sum()\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/statistics/J668_DEG_N_tissues_sta.csv\", index=False)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 筛选反向表达\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AA_DD_HDEM/J668_allTissue_DEM.csv\")\n",
    "tissues = ['anther', 'ovule', 'root',  'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df_result = pd.DataFrame()\n",
    "\n",
    "for i in range(11):\n",
    "    for j in range(i+1, 12):\n",
    "        T1, T2 = tissues[i], tissues[j]\n",
    "        T1_T2 = \"{0}_{1}\".format(T1,T2)\n",
    "        df_tmp = df.query(\"(tissues==@T1_T2) & ((sig_At == 'Up' & sig_Dt == 'Down') | (sig_Dt == 'Up' & sig_At == 'Down'))\").copy()\n",
    "        #print(df_tmp.shape[0])\n",
    "        df_tmp['sig'] = df_tmp['sig_At'].map(str) + \"-\" + df_tmp['sig_Dt'].map(str)\n",
    "        df_tmp = df_tmp[['GeneId_At', 'GeneId_Dt', 'sig']]\n",
    "        df_tmp.columns = ['GeneId_At', 'GeneId_Dt', T1_T2]\n",
    "        if df_result.shape[0] == 0:\n",
    "            df_result = df_tmp.copy()\n",
    "        else:\n",
    "            df_result = pd.merge(df_result, df_tmp, how='outer', on=['GeneId_At', 'GeneId_Dt'])\n",
    "# 转换为complexheatmap 可识别的数据格式\n",
    "df_result.fillna(\"Not\", inplace=True)\n",
    "old_tags = ['Up-Down', 'Down-Up', 'Not']\n",
    "new_tags = [1, -1, 0]\n",
    "df_result.replace(old_tags, new_tags, inplace=True)\n",
    "df_result.drop(['GeneId_Dt'], axis=1, inplace=True)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AA_DD_HDEM/opposite/J668_HG_tissue_pairs_opposite_complaxheatmap.csv\", index=False)\n",
    "\n",
    "# ## 将宽数据转换为长数据\n",
    "# df_result['index'] = range(df_result.shape[0])\n",
    "# df_result_long = df_result.melt(id_vars=['GeneId_At', 'GeneId_Dt', 'index'])\n",
    "# df_result_long.fillna(\"Not\", inplace=True)\n",
    "# df_result_long.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AA_DD_HDEM/opposite/J668_HG_tissue_pairs_opposite.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析在11对组织间差异表达的基因\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/statistics/J668_DEG_N_tissues.csv\")\n",
    "## 找到11对能发生差异表达的同源基因\n",
    "df_11 = df.query(\"DEG_T_N==11\").copy()\n",
    "df_11.drop(['DEG_T_N', 'number'], axis=1, inplace=True)\n",
    "df_11.set_index('GeneId', inplace=True)\n",
    "df_11_T = df_11.T\n",
    "\n",
    "## 分析组织间差异表达，选最大的比较数目\n",
    "def sta(df):\n",
    "    df_index = df.index.to_list()\n",
    "    tissues1 = [i.split('-')[0] for i in df_index]\n",
    "    tissues2 = [i.split('-')[1] for i in df_index]\n",
    "    tissues = tissues1 + tissues2\n",
    "    df_tissues = pd.DataFrame({'tissues':tissues})\n",
    "    max_tissues = df_tissues['tissues'].value_counts().idxmax()\n",
    "    max_tissues_number = df_tissues['tissues'].value_counts().max()\n",
    "    return max_tissues, max_tissues_number\n",
    "## \n",
    "df_result = pd.DataFrame(columns=['GeneId', 'tissues', 'times'])\n",
    "for c in df_11_T.columns:\n",
    "    df_tmp = df_11_T[df_11_T[c]>0][[c]].copy()\n",
    "    max_T, max_T_N = sta(df_tmp)\n",
    "    df_result.loc[df_result.shape[0]] = [c, max_T, max_T_N]\n",
    "\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/statistics/J668_11_diffTissPair_maxTissue_number.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 统计分析一个基因在11个组织对中差异表达，其中出现次数最多的组织的次数\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/statistics/J668_11_diffTissPair_maxTissue_number.csv\")\n",
    "df['number'] = [1]*df.shape[0]\n",
    "df_sta = df.groupby(by=['tissues', 'times'], as_index=False)['number'].sum()\n",
    "df_sta.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/statistics/J668_11_diffTissPair_maxTissue_sta.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析在20, 21对组织间差异表达的基因, \n",
    "# 存在两种情况，基因A 在组织A中表达量差异高于所有组织，同时在组织B中也高于所有组织，此外其它组织表达水平没明显差异\n",
    "# 第二：基因A 在组织A中高于除组织B以外所有组织(10),同时组织B也高于所有组织 \n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/statistics/J668_DEG_N_tissues.csv\")\n",
    "## 找到11对能发生差异表达的同源基因\n",
    "df_11 = df.query(\"DEG_T_N==20 | DEG_T_N==21\").copy()\n",
    "df_11.drop(['DEG_T_N', 'number'], axis=1, inplace=True)\n",
    "df_11.set_index('GeneId', inplace=True)\n",
    "df_11_T = df_11.T\n",
    "\n",
    "## 分析组织间差异表达，选最大的比较数目\n",
    "def sta(df):\n",
    "    df_index = df.index.to_list()\n",
    "    tissues1 = [i.split('-')[0] for i in df_index]\n",
    "    tissues2 = [i.split('-')[1] for i in df_index]\n",
    "    tissues = tissues1 + tissues2\n",
    "    df_tissues = pd.DataFrame({'tissues':tissues})\n",
    "    value_counts = df_tissues['tissues'].value_counts().sort_values(ascending=False)\n",
    "    max_tissues1 = value_counts.index[0]\n",
    "    max_tissues_number1 = value_counts.iloc[0]\n",
    "    max_tissues2 = value_counts.index[1]\n",
    "    max_tissues_number2 = value_counts.iloc[1]\n",
    "    return max_tissues1, max_tissues_number1, max_tissues2, max_tissues_number2,\n",
    "## \n",
    "df_result = pd.DataFrame(columns=['GeneId', 'tissues', 'times'])\n",
    "for c in df_11_T.columns:\n",
    "    df_tmp = df_11_T[df_11_T[c]>0][[c]].copy()\n",
    "    max_T1, max_T_N1, max_T2, max_T_N2  = sta(df_tmp)\n",
    "    df_result.loc[df_result.shape[0]] = [c, max_T1, max_T_N1]\n",
    "    df_result.loc[df_result.shape[0]] = [c, max_T2, max_T_N2]\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/statistics/J668_20_21_diffTissPair_maxTissue_number.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 统计分析一个基因在11个组织对中差异表达，其中出现次数最多的组织的次数\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/statistics/J668_20_21_diffTissPair_maxTissue_number.csv\")\n",
    "df['number'] = [1]*df.shape[0]\n",
    "df_sta = df.groupby(by=['tissues', 'times'], as_index=False)['number'].sum()\n",
    "df_sta.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/statistics/J668_20_21_diffTissPair_maxTissue_sta.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6935\n"
     ]
    }
   ],
   "source": [
    "## 分析同源基因TPM 在组织中的差异倍数\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "tissues = ['anther', 'ovule', 'root',  'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df_oppsite_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AA_DD_HDEM/opposite/J668_HG_tissue_pairs_opposite_complaxheatmap.csv\")\n",
    "df_oppsite_gene = df_oppsite_gene[['GeneId_At']].copy()\n",
    "df_hg = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/J668_HG_pairs.csv\")\n",
    "df_oppsite_HG = pd.merge(df_hg, df_oppsite_gene, how='inner', on=['GeneId_At'])\n",
    "print(df_oppsite_HG.shape[0])\n",
    "df_TPM = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data/J668_all_Tissue_mean_TPM_long.csv\")\n",
    "\n",
    "## 整理同源基因对TPM\n",
    "def HG_pair_TPM(df_hg, df_TMP, tissue):\n",
    "    df_hg = df_hg.copy()\n",
    "    df_TPM = df_TMP[['GeneId', 'TPM']].copy()\n",
    "    df_TPM.columns = ['GeneId_At', '{}_TPM_At'.format(tissue)]\n",
    "    df_At = pd.merge(df_hg, df_TPM, how='left', on=['GeneId_At'])\n",
    "    df_TPM.columns = ['GeneId_Dt', \"{}_TPM_Dt\".format(tissue)]\n",
    "    df_At_Dt = pd.merge(df_At, df_TPM, how='left', on=['GeneId_Dt'])\n",
    "    return df_At_Dt\n",
    "\n",
    "## 分析同源基因TPM比例\n",
    "def HG_TPM_ratio(df, T1, T2):\n",
    "    T1_At, T1_Dt, T2_At, T2_Dt = \"{}_TPM_At\".format(T1), \"{}_TPM_Dt\".format(T1), \"{}_TPM_At\".format(T2), \"{}_TPM_Dt\".format(T2)\n",
    "    df['At'] = (df[T1_At]+0.001)/(df[T2_At]+0.001)\n",
    "    df['Dt'] = (df[T1_Dt]+0.001)/(df[T2_Dt]+0.001)\n",
    "    df['AA_DD'] = np.log2(df['At']/df['Dt'])\n",
    "    df_return = df[['GeneId_At', 'GeneId_Dt', 'AA_DD']].copy()\n",
    "    df_return.columns = ['GeneId_At', 'GeneId_Dt', \"{0}_{1}\".format(T1,T2)]\n",
    "    return df_return\n",
    "## 计算两个组织的TPM 比例\n",
    "def tissues_TPM_ratio(df_hg):\n",
    "    df_return = pd.DataFrame()\n",
    "    for i in range(11):\n",
    "        for j in range(i+1, 12):\n",
    "            T1, T2 = tissues[i], tissues[j]\n",
    "            df_T1 = df_TPM.query(\"tissues==@T1\").copy()\n",
    "            df_T2 = df_TPM.query(\"tissues==@T2\").copy()\n",
    "            df_T1_TPM = HG_pair_TPM(df_hg, df_T1, T1)\n",
    "            df_T2_TPM = HG_pair_TPM(df_hg, df_T2, T2)\n",
    "            #print(df_T1_TPM)\n",
    "            df_T1T2_TPM = pd.merge(df_T1_TPM, df_T2_TPM, how='inner', on=['GeneId_At', 'GeneId_Dt'])\n",
    "            df_tmp = HG_TPM_ratio(df_T1T2_TPM, T1, T2)\n",
    "            if df_return.shape[0] == 0:\n",
    "                df_return = df_tmp.copy()\n",
    "            else:\n",
    "                df_return = pd.merge(df_return, df_tmp, how='inner', on=['GeneId_At', 'GeneId_Dt'])\n",
    "    return df_return\n",
    "df_result = tissues_TPM_ratio(df_oppsite_HG)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AA_DD_HDEM/opposite/J668_HG_tissue_pairs_TPM_opposite_complaxheatmap.csv\", index=False)\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AA_DD_HDEM/opposite/J668_opposite_mode.csv\")\n",
    "df = df.query(\"sig!='Not-Not'\")\n",
    "df.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AA_DD_HDEM/opposite/J668_opposite_mode.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 统计所有相反的差异表达基因，其它的表达模式\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AA_DD_HDEM/opposite/J668_HG_tissue_pairs_TPM_opposite_complaxheatmap.csv\")\n",
    "df_At = df[['GeneId_At']].copy()\n",
    "df_all_tissues = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AA_DD_HDEM/J668_allTissue_DEM.csv\")\n",
    "df_result = pd.merge(df_At, df_all_tissues, how='inner', on=['GeneId_At'])\n",
    "df_result['sig'] = df_result['sig_At'].map(str) + \"-\" + df_result['sig_Dt'].map(str)\n",
    "df_result['number'] = [1]*df_result.shape[0]\n",
    "df_result = df_result.groupby(by=['sig', 'tissues'], as_index=False)['number'].sum()\n",
    "\n",
    "## 重新排序\n",
    "tissues = ['anther', 'ovule', 'root',  'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df_result2 = pd.DataFrame(columns=['sig', 'tissues', 'number'])\n",
    "for i in range(11):\n",
    "        for j in range(i+1, 12):\n",
    "            T1, T2 = tissues[i], tissues[j]\n",
    "            T1_T2 = \"{0}_{1}\".format(T1, T2)\n",
    "            df_tmp = df_result.query(\"tissues == @T1_T2\").copy()\n",
    "            df_result2 = pd.concat([df_result2, df_tmp], axis=0)\n",
    "df_result2.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/AA_DD_HDEM/opposite/J668_opposite_mode.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 统计分析差异表达基因，分析AA, DD, AD, DA /Up 和 Down 的占比\n",
    "import pandas as pd\n",
    "tissues = ['Anther', 'Cotyledon', 'Ovule',  'Hypocotyl', 'Leaf', 'Petal', 'Radicle', 'Root', 'Stem', 'Stigma']\n",
    "df_result = pd.DataFrame()\n",
    "df_hg = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/homo_gene2/Ghir_homo_gene_pairs.txt\", sep=\"\\t\")\n",
    "for i in range(9):\n",
    "    first = tissues[i]\n",
    "    for j in range(i+1, 10):\n",
    "        second = tissues[j]\n",
    "        ab = \"{0}_{1}\".format(first, second)\n",
    "        df_tmp = pd.DataFrame()\n",
    "        for k in ['AA', 'AD', 'DA', 'DD']:\n",
    "            df_DEG = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/ST_VS_ST/{0}/{1}_DEG.bed\".format(k, ab), sep=\"\\t\")\n",
    "            if k =='AA' or k=='AD':\n",
    "                df_DEG.rename(columns={'GeneId':'GeneId_At'}, inplace=True)\n",
    "                df_hg_DEG = pd.merge(df_hg, df_DEG, how='inner', on=['GeneId_At'])\n",
    "            else:\n",
    "                df_DEG.rename(columns={'GeneId':'GeneId_Dt'}, inplace=True)\n",
    "                df_hg_DEG = pd.merge(df_hg, df_DEG, how='inner', on=['GeneId_Dt'])\n",
    "            df_hg_DEG = df_hg_DEG.query(\"sig!='Not'\")[['GeneId_At', 'GeneId_Dt', 'sig']].copy()\n",
    "            df_hg_DEG.rename(columns={'sig':k}, inplace=True)\n",
    "            if df_tmp.shape[0] == 0:\n",
    "                df_tmp = df_hg_DEG\n",
    "            else:\n",
    "                df_tmp = pd.merge(df_tmp, df_hg_DEG, how='outer', on=['GeneId_At', 'GeneId_Dt'])\n",
    "        df_tmp.fillna(\"Not\", inplace=True)\n",
    "        df_tmp['tags'] = [\"{0}_vs_{1}\".format(first, second)]*df_tmp.shape[0]\n",
    "        df_result = pd.concat([df_result, df_tmp], axis=0)\n",
    "#print(df_result.head())\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/ST_VS_ST/bar_figure/Multi_Subg_DEG.bed\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 统计各组织，亚基因差异表达基因比例\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/ST_VS_ST/bar_figure/Multi_Subg_DEG.bed\", sep=\"\\t\")\n",
    "df['number'] = [1]*df.shape[0]\n",
    "df_sta = df.groupby(by=['AA', 'AD', 'DA', 'DD', 'tags'], as_index=False)['number'].sum()\n",
    "df_sta['sig'] = df_sta['AA'].map(str) + '-' + df_sta['AD'].map(str) + '-' + df_sta['DA'].map(str) + '-' + df_sta['DD'].map(str)\n",
    "df_sta.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/ST_VS_ST/bar_figure/Multi_Subg_DEG_sta.bed\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty DataFrame\n",
      "Columns: [GeneId_At, GeneId_Dt, AA, AD, DA, DD, tags]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "## 统计各组织，亚基因差异表达基因比例 2\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/ST_VS_ST/bar_figure/Multi_Subg_DEG.bed\", sep=\"\\t\")\n",
    "df_Keep = df.query(\"(AA=='Up' & AD=='Up' & DA=='Up' & DD=='Up') | (AA=='Down' & AD=='Down' & DA=='Down' & DD=='Down')\").copy()\n",
    "df_Hg = df.query(\"((AA=='Up' & DD=='Up') & (AD!='Up' | DA!='Up')) | ((AA=='Down' & DD=='Down') & (AD!='Down' | DA!='Down'))\").copy()\n",
    "df_Subg = df.query(\"((AD=='Up' & DA=='Up') & (AA!='Up' |  DD!='Up')) | ((AD=='Down' & DA=='Down') & (AA!='Down' | DD!='Down'))\").copy()\n",
    "df_Not = df.query(\"AA=='Not' & AD=='Not' & DA=='Not' & DD=='Not'\").copy()\n",
    "\n",
    "df_dyname = pd.concat([df, df_Keep, df_Hg, df_Subg, df_Not], axis=0)\n",
    "df_dyname.drop_duplicates(keep=False, inplace=True)\n",
    "df_Keep['sig'] =['Stable']*df_Keep.shape[0]\n",
    "df_Hg['sig'] = ['Hg_Stable']*df_Hg.shape[0]\n",
    "df_Subg['sig'] = ['Subg_Stable']*df_Subg.shape[0]\n",
    "df_dyname['sig'] = ['Dynamic']*df_dyname.shape[0]\n",
    "df_result = pd.concat([df_Keep, df_Hg, df_Subg, df_dyname], axis=0)\n",
    "df_result['number'] = [1]*df_result.shape[0]\n",
    "df_sta = df_result.groupby(by=['tags', 'sig'], as_index=False)['number'].sum()\n",
    "df_sta.to_csv('G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/ST_VS_ST/bar_figure/Multi_Subg_DEG_sta2.bed', sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 确定所有组织间差异表达的基因数目\n",
    "import pandas as pd\n",
    "tissues = ['anther', 'ovule', 'root', 'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "# 统计上调和下调的基因数目\n",
    "def up_down_number(df):\n",
    "    df_tmp = df.query(\"sig=='Up' | sig=='Down'\").copy()\n",
    "    df_tmp['number'] = [1]*df_tmp.shape[0]\n",
    "    df_return = df_tmp.groupby(by=['sig'], as_index=False)['number'].sum()\n",
    "    return df_return\n",
    "df_result = pd.DataFrame()\n",
    "for i in range(11):\n",
    "    for j in range(i+1, 12):\n",
    "        input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/{0}_{1}_DEG.csv\".format(tissues[i], tissues[j])\n",
    "        tags = \"{0}_{1}\".format(tissues[i], tissues[j])\n",
    "        df = pd.read_csv(input_f)\n",
    "        df_sta_tmp = up_down_number(df)\n",
    "        df_sta_tmp['tags'] = [tags]*df_sta_tmp.shape[0]\n",
    "        df_result = pd.concat([df_result, df_sta_tmp], axis=0)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/J668_DEG_number_statistics.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析确定两个亚基因组间比较的差异\n",
    "import pandas as pd\n",
    "tissues = ['anther', 'ovule', 'root', 'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "\n",
    "def statis_number(tags):\n",
    "    df_return = pd.DataFrame(columns = ['number', 'tags', 'tissues'])\n",
    "    for i in range(11):\n",
    "        for j in range(i+1, 12):\n",
    "            first = tissues[i]\n",
    "            last = tissues[j]\n",
    "            input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/{0}/{1}_{2}_DEG.bed\".format(tags, first, last)\n",
    "            df_tmp = pd.read_csv(input_f)\n",
    "            Up_number = df_tmp.query(\"sig=='Up'\").shape[0]\n",
    "            df_return.loc[df_return.shape[0], ] = [Up_number, tags, \"{0}_{1}\".format(first, last)]\n",
    "    return df_return\n",
    "df_AA_up = statis_number(\"AA\")\n",
    "df_DD_up = statis_number(\"DD\")\n",
    "df_result = pd.concat([df_AA_up, df_DD_up], axis=0)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue_hg/J668_AA_DD_up_number.csv\", index=False)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.3 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "a0f669ea42a4dcea88d385f8f9b4c3d9b966e24234758831d1dbfdac925619a2"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
