{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ -0.31864038   3.830054  ]\n",
      " [  0.16738613   0.7980106 ]\n",
      " [  9.989476     8.636723  ]\n",
      " ...\n",
      " [ -2.98316    -15.160893  ]\n",
      " [ -3.3057296  -11.269796  ]\n",
      " [ -1.8772815    3.5438497 ]]\n"
     ]
    }
   ],
   "source": [
    "import umap\n",
    "import numpy as np\n",
    "from sklearn.datasets import load_digits\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "%matplotlib inline\n",
    "chromHMM_3Kb = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/ChromHMM/cotyledon/cotyledon_3K_chromHMM.bed\", sep=\"\\t\", names=['chr', 'start', 'end', 'E1', 'E2', 'E3', 'E4'])\n",
    "reducer = umap.UMAP()\n",
    "chromHMM_data = chromHMM_3Kb[\n",
    "    [\n",
    "        \"E1\",\n",
    "        \"E2\",\n",
    "        \"E3\",\n",
    "        \"E4\",\n",
    "    ]\n",
    "].values\n",
    "scaled_penguin_data = StandardScaler().fit_transform(chromHMM_data)\n",
    "embedding = reducer.fit_transform(scaled_penguin_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/ChromHMM/cotyledon/cotyledon_3K_chromHMM.bed\", sep=\"\\t\", names=['chr', 'start', 'end', 'E1', 'E2', 'E3', 'E4'])\n",
    "df['idx'] = df['chr'].astype(str) + '-' + df['start'].astype(str) + '-' + df['end'].astype(str)\n",
    "tags_list = []\n",
    "for index, row in df.iterrows():\n",
    "    if row['E1']==1:\n",
    "        tags = 'E1'\n",
    "    elif row['E2']==1:\n",
    "        tags = 'E2'\n",
    "    elif row['E3']==1:\n",
    "        tags = 'E3'\n",
    "    elif row['E4']==1:\n",
    "        tags = 'E4'\n",
    "    elif row['E1']>0 and row['E2']>0 and row['E3']==0 and row['E4']==0:\n",
    "        tags = 'E1_E2'\n",
    "    elif row['E1']>0 and row['E2']==0 and row['E3']>0 and row['E4']==0:\n",
    "        tags = 'E1_E3'\n",
    "    elif row['E1']>0 and row['E2']==0 and row['E3']==0 and row['E4']>0:\n",
    "        tags = 'E1_E4'    \n",
    "    elif row['E1']==0 and row['E2']>0 and row['E3']>0 and row['E4']==0:\n",
    "        tags = 'E2_E3'\n",
    "    elif row['E1']==0 and row['E2']>0 and row['E3']==0 and row['E4']>0:\n",
    "        tags = 'E2_E4'\n",
    "    elif row['E1']==0 and row['E2']==0 and row['E3']>0 and row['E4']>0:\n",
    "        tags = 'E3_E4'\n",
    "    elif row['E1']>0 and row['E2']>0 and row['E3']>0 and row['E4']==0:\n",
    "        tags = 'E1_E2_E3'\n",
    "    elif row['E1']>0 and row['E2']>0 and row['E3']==0 and row['E4']>0:\n",
    "        tags = 'E1_E2_E4'\n",
    "    elif row['E1']>0 and row['E2']==0 and row['E3']>0 and row['E4']>0:\n",
    "        tags = 'E1_E3_E4'\n",
    "    elif row['E1']==0 and row['E2']>0 and row['E3']>0 and row['E4']>0:\n",
    "        tags = 'E2_E3_E4'\n",
    "    elif row['E1']>0 and row['E2']>0 and row['E3']>0 and row['E4']>0:\n",
    "        tags = 'E1_E2_E3_E4'\n",
    "    tags_list.append(tags)\n",
    "df['tags'] = tags_list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_umap = pd.DataFrame(embedding, columns =['umap1', 'umap2'])\n",
    "df_result = pd.concat([df, df_umap], axis=1)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/ChromHMM/cotyledon/cotyledon_3K_chromHMM_umap.bed\", sep=\"\\t\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tags\n",
      "E1             528461\n",
      "E1_E2           27656\n",
      "E1_E2_E3        21132\n",
      "E1_E2_E3_E4     17293\n",
      "E1_E2_E4         8421\n",
      "E1_E3           16663\n",
      "E1_E3_E4        35397\n",
      "E1_E4           57651\n",
      "E2               6469\n",
      "E2_E3            5644\n",
      "E2_E3_E4         9490\n",
      "E2_E4            2420\n",
      "E3                 17\n",
      "E3_E4            5630\n",
      "E4               2334\n",
      "Name: number, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/ChromHMM/cotyledon/cotyledon_3K_chromHMM_umap.bed\", sep=\"\\t\")\n",
    "df['number'] = [1]*df.shape[0]\n",
    "df_result = df.groupby(['tags'])['number'].sum()\n",
    "print(df_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 计算TSS score\n",
    "import pandas as pd\n",
    "import sys\n",
    "input_f = sys.argv[1]\n",
    "df = pd.read_csv(in)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 对于ChIP-Seq peaks 位点位置，进行高斯展宽分析\n",
    "import pandas as pd\n",
    "import multiprocessing as mp\n",
    "def guass_line(df):\n",
    "    for "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>a</th>\n",
       "      <th>b</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   a  b\n",
       "0  4  5\n",
       "1  6  6\n",
       "2  7  7"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "def myfun(x):\n",
    "    return x + 3\n",
    "df = pd.DataFrame({'a':[1,3,4], 'b':[2,3,4]})\n",
    "df.apply(myfun)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将computematrix 结果z-score 标准化，之后再归一化\n",
    "import pandas as pd\n",
    "import sys\n",
    "import scipy.stats as stats\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "\n",
    "input_f = sys.argv[1]\n",
    "output_f = sys.argv[2]\n",
    "df = pd.read_csv(input_f, sep=\"\\t\", header=None)\n",
    "df_tags = df.loc[:, 0:5].copy()\n",
    "df_number = df.loc[:, 5:].copy()\n",
    "\n",
    "#! Z-score\n",
    "df_number.astype('float')\n",
    "df_array = df_number.to_numpy()\n",
    "df_zscore = stats.zscore(df_array, axis=1, ddof=1)\n",
    "\n",
    "#! normal\n",
    "tool = MinMaxScaler(feature_range=(-1, 1))\n",
    "df_zscore_normal = tool.fit_transform(df_zscore)\n",
    "df_result = pd.concat([df_tags, df_zscore_normal], axis=1)\n",
    "df_result.to_csv(output_f, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/expression_gene/J668_all_tissue_average_TPM_long.txt\", sep=\"\\t\")\n",
    "df2 = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/Gh_gene_position.bed\", sep=\"\\t\")\n",
    "df_result = pd.merge(df, df2, how='inner', on=['GeneId'])\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/expression_gene/J668_all_tissue_average_TPM_long.txt\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1668\n",
      "1612\n"
     ]
    }
   ],
   "source": [
    "# diffbind 附近(左右各500 bp)所包含的基因表达水平上的差异\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "def diffpeaks_TAU(df_diff_peaks, df_TAU, region):\n",
    "    df_diff_peaks['start'] = df_diff_peaks['start'] - region\n",
    "    df_diff_peaks['end'] = df_diff_peaks['end'] + region\n",
    "    df_return = pd.DataFrame()\n",
    "    for index, row in df_diff_peaks.iterrows():\n",
    "        df_tmp = pd.DataFrame()\n",
    "        C, S, E, T = row['chrs'], row['start'], row['end'], row['log2FC']\n",
    "        df_tmp = df_TAU.query(\"chrs==@C and start>=@S and end<=@E\").copy()\n",
    "        if df_tmp.shape[0] > 0:\n",
    "            if T >0:\n",
    "                df_tmp['sigs'] = ['Hypocotyl']*df_tmp.shape[0]\n",
    "            else:\n",
    "                df_tmp['sigs'] = ['Cotyledon']*df_tmp.shape[0]\n",
    "            df_return = pd.concat([df_return, df_tmp], axis=0)\n",
    "    df_return.columns = ['index', 'treatment','value', 'chrs', 'start', 'end', 'class']\n",
    "    return df_return[['value', 'class', 'treatment', 'index']]\n",
    "\n",
    "df_diff_peaks = pd.read_csv('G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/hypocotyl_cotyledon/result_edgeR_sig.bed', sep=\"\\t\", names=['chrs', 'start', 'end', 'a', 'log2FC'])\n",
    "df_TAU = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/expression_gene/J668_all_tissue_average_TPM_long.txt\", sep=\"\\t\")\n",
    "df_TAU_tissue = df_TAU.query(\"tags=='cotyledon'|tags=='hypocotyl'\").copy()\n",
    "df_result = diffpeaks_TAU(df_diff_peaks, df_TAU_tissue, 5000)\n",
    "df_result['value'] = np.log2(df_result['value']+0.001)\n",
    "df_result['treatment'].replace(['hypocotyl', 'cotyledon'], ['Hypocotyl', \"Cotyledon\"], inplace=True)\n",
    "print(df_result.shape[0])\n",
    "df_result.drop_duplicates(keep='first', inplace=True)\n",
    "print(df_result.shape[0])\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/hypocotyl_cotyledon/diff_peaks_TAU_10k.bed\", sep=\"\\t\", index=False)\n",
    "# print(df_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 统计分析差异peaks位置所包含的差异表达基因（H3K4me3）\n",
    "import pandas as pd\n",
    "def creat_tags():\n",
    "    tiss1 = ['anther', 'ovule', 'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'petal', 'stigma']\n",
    "    tiss2 = ['anther', 'cotyledon', 'hypocotyl', 'leaf', 'ovule', 'petal', 'radicle', 'stem', 'stigma']\n",
    "    old_tags = []\n",
    "    n = []\n",
    "    for t1 in range(9):\n",
    "        for t2 in range(t1+1, 9):\n",
    "            new = \"{0}_{1}\".format(tiss1[t1], tiss1[t2])\n",
    "            old = \"{0}_{1}\".format(tiss2[t1], tiss2[t2])\n",
    "            old_tags.append(old)\n",
    "            n.append(new)\n",
    "            #print(old)\n",
    "    new_tags = [n[4], n[3], n[2], n[0], n[6], n[1], n[5], n[7], n[26], n[22], n[11], n[31], n[17], n[30], \n",
    "               n[32], n[21], n[10], n[28], n[16], n[27], n[29], n[9], n[24], n[15], n[23], n[25], n[13],\n",
    "               n[8], n[12], n[14], n[19], n[33], n[35], n[18], n[20], n[34]]\n",
    "    return old_tags, new_tags\n",
    "old_tags, new_tags = creat_tags()       \n",
    "df_diffpeaks_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K4me3/rawdata/J668_H3K4me3_diffpeaks_gene.csv\", names=['GeneId', \"sig\", 'tags'])\n",
    "df_diffpeaks_gene['tags'].replace(old_tags, new_tags, inplace=True)\n",
    "df_diffpeaks_gene[['tissueA', 'tissueB']] = df_diffpeaks_gene['tags'].str.split(\"_\", expand=True)\n",
    "df_DEG = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/J668_mutipTissues_DEG.csv\")\n",
    "df_DEG = df_DEG[['GeneId', 'tags']]\n",
    "df_DEG['table'] = ['DEG']*df_DEG.shape[0]\n",
    "df_result = pd.merge(df_diffpeaks_gene, df_DEG, how='left', on=['GeneId', 'tags'])\n",
    "df_result.fillna('NDEG', inplace=True)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K4me3/DEG/J668_diffpeak_DEG.csv\", index=False)\n",
    "df_result['number'] = [1]*df_result.shape[0]\n",
    "df_statistics = df_result.groupby(by=['tissueA', 'tissueB', 'table'], as_index=False)['number'].sum()\n",
    "df_statistics.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K4me3/DEG/J668_diffpeak_DEG_statistics.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "## H3K4me3 替换名称\n",
    "import pandas as pd\n",
    "#df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K4me3/J668_H3K4me3_diffpeaks_statistics.csv\")\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K4me3/rawdata/J668_gene2Kb_H3K4me3_diffpeaks.csv\", names=['GeneId', 'tissue'])\n",
    "def creat_tags():\n",
    "    tiss1 = ['anther', 'ovule', 'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'petal', 'stigma']\n",
    "    tiss2 = ['anther', 'cotyledon', 'hypocotyl', 'leaf', 'ovule', 'petal', 'radicle', 'stem', 'stigma']\n",
    "    old_tags = []\n",
    "    n = []\n",
    "    for t1 in range(9):\n",
    "        for t2 in range(t1+1, 9):\n",
    "            new = \"{0}_{1}\".format(tiss1[t1], tiss1[t2])\n",
    "            old = \"{0}_{1}\".format(tiss2[t1], tiss2[t2])\n",
    "            old_tags.append(old)\n",
    "            n.append(new)\n",
    "            #print(old)\n",
    "    new_tags = [n[4], n[3], n[2], n[0], n[6], n[1], n[5], n[7], n[26], n[22], n[11], n[31], n[17], n[30], \n",
    "               n[32], n[21], n[10], n[28], n[16], n[27], n[29], n[9], n[24], n[15], n[23], n[25], n[13],\n",
    "               n[8], n[12], n[14], n[19], n[33], n[35], n[18], n[20], n[34]]\n",
    "    return old_tags, new_tags\n",
    "old_tags, new_tags = creat_tags()\n",
    "df['tissue'].replace(old_tags, new_tags, inplace=True)\n",
    "\n",
    "#df.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K4me3/J668_H3K4me3_diffpeaks_statistics.csv\", index=False)\n",
    "df.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K4me3/rawdata/J668_gene2Kb_H3K4me3_diffpeaks.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "# H3K27ac 替换名称\n",
    "import pandas as pd\n",
    "#df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K27ac/J668_H3K27ac_diffpeaks_statistics.csv\")\n",
    "#df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K27ac/rawdata/J668_H3K27ac_diffpeaks_gene.csv\", names=['GeneId', 'sig', 'tissue'])\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K27ac/rawdata/J668_gene2Kb_H3K27ac_diffpeaks.csv\", names=['GeneId', 'tissue'])\n",
    "def creat_tags():\n",
    "    tiss1 = ['ovule', 'root', 'radicle', 'hypocotyl', 'cotyledon', 'stem', 'stigma']\n",
    "    tiss2 = ['cotyledon', 'hypocotyl', 'ovule', 'radicle', 'root', 'stem', 'stigma']\n",
    "    old_tags = []\n",
    "    n = []\n",
    "    for t1 in range(7):\n",
    "        for t2 in range(t1+1, 7):\n",
    "            new = \"{0}_{1}\".format(tiss1[t1], tiss1[t2])\n",
    "            old = \"{0}_{1}\".format(tiss2[t1], tiss2[t2])\n",
    "            old_tags.append(old)\n",
    "            n.append(new)\n",
    "    new_tags = [n[15], n[3], n[12], n[8], n[18], n[19], n[2], n[11], n[7], n[16], n[17], \n",
    "                n[1], n[0], n[4], n[5], n[6], n[13], n[14], n[9], n[10], n[20]]\n",
    "    return old_tags, new_tags\n",
    "old_tags, new_tags = creat_tags()\n",
    "df['tissue'].replace(old_tags, new_tags, inplace=True)\n",
    "df.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K27ac/rawdata/J668_gene2Kb_H3K27ac_diffpeaks.csv\", index=False)\n",
    "#df.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K27ac/rawdata/J668_H3K27ac_diffpeaks_gene.csv\", index=False, header=False)\n",
    "# df['tissue'].replace(old_tags, new_tags, inplace=True)\n",
    "# df.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K27ac/J668_H3K27ac_diffpeaks_statistics.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 统计分析差异peaks位置所包含的差异表达基因（H3K27ac）\n",
    "import pandas as pd \n",
    "df_diffpeaks_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K27ac/rawdata/J668_H3K27ac_diffpeaks_gene.csv\", names=['GeneId', \"sig\", 'tags'])\n",
    "df_diffpeaks_gene[['tissueA', 'tissueB']] = df_diffpeaks_gene['tags'].str.split(\"_\", expand=True)\n",
    "df_DEG = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/J668_mutipTissues_DEG.csv\")\n",
    "df_DEG = df_DEG[['GeneId', 'tags']]\n",
    "df_DEG['table'] = ['DEG']*df_DEG.shape[0]\n",
    "df_result = pd.merge(df_diffpeaks_gene, df_DEG, how='left', on=['GeneId', 'tags'])\n",
    "df_result.fillna('NDEG', inplace=True)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K27ac/DEG/J668_diffpeak_DEG.csv\", index=False)\n",
    "df_result['number'] = [1]*df_result.shape[0]\n",
    "df_statistics = df_result.groupby(by=['tissueA', 'tissueB', 'table'], as_index=False)['number'].sum()\n",
    "df_statistics.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K27ac/DEG/J668_diffpeak_DEG_statistics.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "110\n",
      "90\n"
     ]
    }
   ],
   "source": [
    "## 整理H3K27ac与H3K4me3 差异peaks 所造成的基因表达差异\n",
    "import pandas as pd\n",
    "tissues = ['anther', 'ovule', 'root', 'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'petal', 'stigma']\n",
    "df_up = pd.DataFrame(columns=['tissueA', 'tissueB', 'table'])\n",
    "df_down = pd.DataFrame(columns=['tissueA', 'tissueB', 'table'])\n",
    "for i in range(10):\n",
    "    for j in range(i, 10):\n",
    "        for k in ['DEG', 'NDEG']:\n",
    "            df_up.loc[df_up.shape[0], ] = [tissues[i], tissues[j], k]\n",
    "for i in range(10):\n",
    "    for j in range(i+1, 10):\n",
    "        for k in ['DEG', 'NDEG']:\n",
    "            df_down.loc[df_down.shape[0], ] = [tissues[j], tissues[i], k]\n",
    "#print(df_up.shape[0])\n",
    "#print(df_down.shape[0])\n",
    "df_H3K27ac = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K27ac/DEG/J668_diffpeak_DEG_statistics.csv\")\n",
    "df_H3K4me3 = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K4me3/DEG/J668_diffpeak_DEG_statistics.csv\")\n",
    "#df_H3K27ac['tags'] = ['H3K27ac']*df_H3K27ac.shape[0]\n",
    "#df_H3K4me3['tags'] = ['H3K4me3']*df_H3K4me3.shape[0]\n",
    "df_H3K27ac.columns = ['tissueB', 'tissueA', 'table', 'number']\n",
    "df_H3K27ac = pd.merge(df_down, df_H3K27ac, how='left', on=['tissueA', 'tissueB', 'table'])\n",
    "df_H3K4me3 = pd.merge(df_up, df_H3K4me3, how='left', on=['tissueA', 'tissueB', 'table'])\n",
    "df_H3K4me3.fillna(1, inplace=True)\n",
    "df_H3K27ac.fillna(1, inplace=True)\n",
    "df_H3K27ac['tags'] = ['H3K27ac']*df_H3K27ac.shape[0]\n",
    "df_H3K4me3['tags'] = ['H3K4me3']*df_H3K4me3.shape[0]\n",
    "df_result = pd.concat([df_H3K4me3, df_H3K27ac], axis=0)\n",
    "#df_result.drop_duplicates(keep='last', inplace=True, subset=['tissueA', 'tissueB', 'table'])\n",
    "# df_result.fillna(1, inplace=True)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/J668_diffChIP_DEG.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       class  number\n",
      "0  Cotyledon    1120\n",
      "1  Hypocotyl     758\n"
     ]
    }
   ],
   "source": [
    "## 统计差异ChIP 修饰位点附近基因所包含的差异表达基因\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/hypocotyl_cotyledon/diff_peaks_TAU_10k.bed\", sep=\"\\t\")\n",
    "df_DEG = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/reads/T_VS_T/cotyledon_hypocotyl_DEG.bed\", sep=\"\\t\")\n",
    "df.columns = ['value', 'class', 'treatment', 'GeneId']\n",
    "df_result = pd.merge(df, df_DEG, how='inner', on=['GeneId'])\n",
    "df_result['number'] = [1]*df_result.shape[0]\n",
    "dz = df_result.groupby(by=['class'], as_index=False)['number'].sum()\n",
    "print(dz) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析各种chromHMM的占比\n",
    "import pandas as pd\n",
    "samples = ['hypocotyl', 'leaf', 'radicle', 'stigma']\n",
    "df_result = pd.DataFrame()\n",
    "for s in samples:\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/ChromHMM/statistics/{}_4_segments.bed\".format(s)\n",
    "    df = pd.read_csv(input_f, sep=\"\\t\")\n",
    "    df['size'] = df['end'] - df['start']\n",
    "    df['tags'] = [s]*df.shape[0]\n",
    "    df_result = pd.concat([df_result, df], axis=0)\n",
    "df = df_result.groupby(by=['tags', 'sigs'], as_index=False)['size'].sum()\n",
    "df.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/ChromHMM/statistics/type_ratio.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             GeneId_At        GeneId_Dt table_At table_Dt stem_bias  \\\n",
      "3      Ghir_A01G000100  Ghir_D01G000120      DEG      DEG        no   \n",
      "4      Ghir_A01G000110  Ghir_D01G000130      DEG      DEG        no   \n",
      "5      Ghir_A01G000130  Ghir_D01G000150       no      DEG        no   \n",
      "8      Ghir_A01G000170  Ghir_D01G000180      DEG       no     BiasA   \n",
      "10     Ghir_A01G000190  Ghir_D01G000200       no      DEG        no   \n",
      "...                ...              ...      ...      ...       ...   \n",
      "25463  Ghir_A13G024820  Ghir_D13G025590      DEG       no        no   \n",
      "25464  Ghir_A13G024830  Ghir_D13G025600       no      DEG        no   \n",
      "25468  Ghir_A13G024870  Ghir_D13G025640      DEG      DEG        no   \n",
      "25469  Ghir_A13G024880  Ghir_D13G025650      DEG      DEG        no   \n",
      "25471  Ghir_A13G024900  Ghir_D13G025670      DEG       no        no   \n",
      "\n",
      "      stigma_bias  \n",
      "3              no  \n",
      "4              no  \n",
      "5              no  \n",
      "8              no  \n",
      "10             no  \n",
      "...           ...  \n",
      "25463       BiasD  \n",
      "25464       BiasA  \n",
      "25468       BiasA  \n",
      "25469          no  \n",
      "25471          no  \n",
      "\n",
      "[6710 rows x 6 columns]\n",
      "7271\n"
     ]
    }
   ],
   "source": [
    "## 统计分析同源基因受ChIP-Seq差异表达信号的影响\n",
    "import pandas as pd\n",
    "df_hg = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/J668_HG_pairs.csv\")\n",
    "df_hg_bias = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/all_tissue_bias_gene_pair.bed\", sep=\"\\t\")\n",
    "df_bias_stem = df_hg_bias.query(\"tags=='stem'\")[['GeneId_At', 'GeneId_Dt', 'sig']].copy()\n",
    "df_bias_stigma = df_hg_bias.query(\"tags=='stigma'\")[['GeneId_At', 'GeneId_Dt', 'sig']].copy()\n",
    "df_bias_stem.columns = ['GeneId_At', 'GeneId_Dt', 'stem_bias']\n",
    "df_bias_stigma.columns = ['GeneId_At', 'GeneId_Dt', 'stigma_bias']\n",
    "df_H3K4me3 = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K4me3/DEG/J668_diffpeak_DEG.csv\")\n",
    "df_H3K4me3_tissue = df_H3K4me3.query(\"tags == 'stem_stigma'\")[['GeneId', 'table']].copy()\n",
    "df_H3K4me3_tissue.columns = ['GeneId_At', 'table_At']\n",
    "df_a = pd.merge(df_hg, df_H3K4me3_tissue, how='left', on=['GeneId_At'])\n",
    "df_H3K4me3_tissue.columns = ['GeneId_Dt', 'table_Dt']\n",
    "df_ab = pd.merge(df_a, df_H3K4me3_tissue, how='left', on=['GeneId_Dt'])\n",
    "df_ab_stem = pd.merge(df_ab, df_bias_stem, how='left', on=['GeneId_At', 'GeneId_Dt'])\n",
    "df_ab_stem_stigma = pd.merge(df_ab_stem, df_bias_stigma, how='left', on=['GeneId_At', 'GeneId_Dt'])\n",
    "df_diffpeak = df_ab_stem_stigma.query(\"table_At=='DEG'| table_Dt=='DEG'\").copy()\n",
    "df_diffpeak.fillna(\"no\", inplace=True)\n",
    "print(df_diffpeak)\n",
    "df_true = df_diffpeak.query(\"(table_At!=table_Dt) & ((stem_bias != stigma_bias))\")\n",
    "df_ab = df_diffpeak.query(\"(table_At=='DEG' | table_Dt=='DEG') & ((stem_bias != stigma_bias))\")\n",
    "df_stem_stigma = df_hg_bias.query(\"tags=='stem'|tags=='stigma'\").copy()\n",
    "df_stem_stigma.drop_duplicates(keep=False, subset=['GeneId_At', 'GeneId_Dt', 'sig'], inplace=True)\n",
    "print(df_stem_stigma.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析各组织间差异同源bias基因，有多少是受到表观修饰差异的影响\n",
    "import pandas as pd\n",
    "# 选择H3K27ac和H3K4me3信号\n",
    "df_H3K27ac = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K27ac/rawdata/J668_gene2Kb_H3K27ac_diffpeaks.csv\")\n",
    "df_H3K4me3 = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K4me3/rawdata/J668_gene2Kb_H3K4me3_diffpeaks.csv\")\n",
    "#df_H3K27ac_tissue = df_H3K27ac.query(\"tissue=='ovule_radicle' | tissue=='ovule_hypocotyl' | tissue=='ovule_cotyledon' | tissue=='ovule_stem' | tissue=='ovule_stigma' | tissue=='radicle_hypocotyl' | tissue=='radicle_cotyledon' | tissue=='radicle_stem' | tissue=='radicle_stigma' | tissue=='hypocotyl_cotyledon' | tissue=='hypocotyl_stem' | tissue=='hypocotyl_stigma' | tissue=='cotyledon_stem' | tissue=='cotyledon_stigma' | tissue=='stem_stigma'\")\n",
    "#df_H3K4me3_tissue = df_H3K4me3.query(\"tissue=='ovule_radicle' | tissue=='ovule_hypocotyl' | tissue=='ovule_cotyledon' | tissue=='ovule_stem' | tissue=='ovule_stigma' | tissue=='radicle_hypocotyl' | tissue=='radicle_cotyledon' | tissue=='radicle_stem' | tissue=='radicle_stigma' | tissue=='hypocotyl_cotyledon' | tissue=='hypocotyl_stem' | tissue=='hypocotyl_stigma' | tissue=='cotyledon_stem' | tissue=='cotyledon_stigma' | tissue=='stem_stigma'\")\n",
    "#print(df_H3K27ac.shape[0], df_H3K4me3.shape[0], df_H3K27ac_tissue.shape[0], df_H3K4me3_tissue.shape[0])\n",
    "tissue = ['ovule', 'radicle', 'hypocotyl', 'cotyledon', 'stem', 'stigma']\n",
    "df_hg = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/J668_HG_pairs.csv\")\n",
    "df_hg_bias = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/all_tissue_bias_gene_pair.bed\", sep=\"\\t\")\n",
    "# 调整两个组织差异peaks所包含的基因\n",
    "def tissue_diff(df_hg, df_diff, tags):\n",
    "    if tags == 'H3K27ac':\n",
    "        df_diff.columns = ['GeneId_At', 'diff_At_H3K27ac']\n",
    "        df_a = pd.merge(df_hg, df_diff, how='left', on=['GeneId_At'])\n",
    "        df_diff.columns = ['GeneId_Dt', 'diff_Dt_H3K27ac']\n",
    "        df_ab = pd.merge(df_a, df_diff, how='left', on=['GeneId_Dt'])\n",
    "    else:\n",
    "        df_diff.columns = ['GeneId_At', 'diff_At_H3K4me3']\n",
    "        df_a = pd.merge(df_hg, df_diff, how='left', on=['GeneId_At'])\n",
    "        df_diff.columns = ['GeneId_Dt', 'diff_Dt_H3K4me3']\n",
    "        df_ab = pd.merge(df_a, df_diff, how='left', on=['GeneId_Dt'])\n",
    "    return df_ab\n",
    "\n",
    "# 确定两个组织间同源基因差异\n",
    "def tissue_hg_bias(df_hg, df_A_bias, df_B_bias, df_AB_H3K27ac_diff, df_AB_H3K4me3_diff):\n",
    "    df_A_bias.columns = ['GeneId_At', 'GeneId_Dt', 'tissueA_bias']\n",
    "    df_B_bias.columns = ['GeneId_At', 'GeneId_Dt', 'tissueB_bias']\n",
    "    df_hg_A_bias = pd.merge(df_hg, df_A_bias, how='left', on=['GeneId_At', 'GeneId_Dt'])\n",
    "    df_hg_AB_bias = pd.merge(df_hg_A_bias, df_B_bias, how='left', on=['GeneId_At', 'GeneId_Dt'])\n",
    "    df_hg_H3K27ac_diff = tissue_diff(df_hg, df_AB_H3K27ac_diff, 'H3K27ac')\n",
    "    df_hg_H3K4me3_diff = tissue_diff(df_hg, df_AB_H3K4me3_diff, 'H3K4me3')\n",
    "    df_H3K27ac_return = pd.merge(df_hg_AB_bias, df_hg_H3K27ac_diff, how='left', on=['GeneId_At', 'GeneId_Dt'])\n",
    "    df_return = pd.merge(df_H3K27ac_return,df_hg_H3K4me3_diff, how='left', on=['GeneId_At', 'GeneId_Dt'] )\n",
    "    return df_return\n",
    "tissue = ['ovule', 'radicle', 'hypocotyl', 'cotyledon', 'stem', 'stigma']\n",
    "df_result = pd.DataFrame()\n",
    "for i in range(6):\n",
    "    tags1 = tissue[i]\n",
    "    df_tissueA = df_hg_bias.query(\"tags==@tags1\")[['GeneId_At', 'GeneId_Dt', 'sig']].copy()\n",
    "    for j in range(i+1, 6):\n",
    "        tags2 = tissue[j]\n",
    "        df_tissueB = df_hg_bias.query(\"tags==@tags2\")[['GeneId_At', 'GeneId_Dt', 'sig']].copy()\n",
    "        tags = \"{0}_{1}\".format(tissue[i], tissue[j])\n",
    "        df_tissue_H3K27ac_diff = df_H3K27ac.query(\"tissue==@tags\")[['GeneId']].copy()\n",
    "        df_tissue_H3K4me3_diff = df_H3K4me3.query(\"tissue==@tags\")[['GeneId']].copy()\n",
    "        df_tissue_H3K27ac_diff['tags'] = ['diff']*df_tissue_H3K27ac_diff.shape[0]\n",
    "        df_tissue_H3K4me3_diff['tags'] = ['diff']*df_tissue_H3K4me3_diff.shape[0]\n",
    "        df_tmp = tissue_hg_bias(df_hg, df_tissueA, df_tissueB, df_tissue_H3K27ac_diff, df_tissue_H3K4me3_diff)\n",
    "        df_tmp['tags'] = [tags]*df_tmp.shape[0]\n",
    "        df_tmp.fillna(\"stable\", inplace=True)\n",
    "        df_tmp = df_tmp.query(\"~(tissueA_bias=='stable' & tissueB_bias=='stable' & diff_At_H3K27ac=='stable' & diff_Dt_H3K27ac=='stable' & diff_At_H3K4me3=='stable' & diff_Dt_H3K4me3=='stable')\")\n",
    "        df_result = pd.concat([df_result, df_tmp], axis=0)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K27ac_H3K4me3/J668_hg_bias_diff.csv\", index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 统计分析上一个cell的结果\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K27ac_H3K4me3/J668_hg_bias_diff.csv\")\n",
    "df['number'] = [1]*df.shape[0]\n",
    "#df_a = df.query(\"~(tissueA_bias=='stable' & tissueB_bias=='stable')\")\n",
    "df_a = df.query(\"(tissueA_bias != tissueB_bias)\")\n",
    "df_b = df_a.query(\"diff_At_H3K27ac=='diff' | diff_Dt_H3K27ac=='diff' | diff_At_H3K4me3=='diff' | diff_Dt_H3K4me3=='diff'\").copy()\n",
    "df_a = df_a.groupby('tags', as_index=False)['number'].sum()\n",
    "df_b = df_b.groupby('tags', as_index=False)['number'].sum()\n",
    "df_a.columns = ['tags', 'All']\n",
    "df_b.columns = ['tags', 'Need']\n",
    "df_result = pd.merge(df_a, df_b, how='inner', on=['tags'])\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/J668_tissue_hg_bias_diffpeak.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ovule_radicle\n",
      "ovule_hypocotyl\n",
      "ovule_cotyledon\n",
      "ovule_stem\n",
      "ovule_stigma\n",
      "radicle_hypocotyl\n",
      "radicle_cotyledon\n",
      "radicle_stem\n",
      "radicle_stigma\n",
      "hypocotyl_cotyledon\n",
      "hypocotyl_stem\n",
      "hypocotyl_stigma\n",
      "cotyledon_stem\n",
      "cotyledon_stigma\n",
      "stem_stigma\n"
     ]
    }
   ],
   "source": [
    "tissue = ['ovule', 'radicle', 'hypocotyl', 'cotyledon', 'stem', 'stigma']\n",
    "for i in range(6):\n",
    "    for j in range(i+1, 6):\n",
    "        print(\"{0}_{1}\".format(tissue[i], tissue[j]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 选择一些基因来作图用于举例说明（cotyledon vs stigma）\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K27ac_H3K4me3/J668_hg_bias_diff.csv\")\n",
    "df_cotyledon_stigma = df.query(\"tags==('cotyledon_stigma') & (tissueA_bias!=tissueB_bias) & (diff_At_H3K27ac=='diff' | diff_Dt_H3K27ac=='diff' | diff_At_H3K4me3=='diff' | diff_Dt_H3K4me3=='diff')\")[['GeneId_At', 'GeneId_Dt']].copy()\n",
    "#df_cotyledon_stigma.to_csv('G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/cotyledon_stigma/cotyledon_stigma.csv', index=False)\n",
    "df_hg_position = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/all_homo_gene_pairs_position.txt\", sep=\"\\t\")\n",
    "df_hg_position.columns = ['GeneId_At', 'GeneId_Dt', 'chr_at', 'start_at', 'end_at', 'chr_dt', 'start_dt', 'end_dt']\n",
    "df = pd.merge(df_hg_position, df_cotyledon_stigma, how='inner', on=['GeneId_At', 'GeneId_Dt'])\n",
    "fp = open(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K27ac_H3K4me3/plot.sh\", 'w')\n",
    "for index, row in df.iterrows():\n",
    "    GeneIdAt, GeneIdDt, chrAt, startAt, endAt, chrDt, startDt, endDt = row['GeneId_At'], row['GeneId_Dt'], row['chr_at'], row['start_at']-2000, row['end_at']+2000, row['chr_dt'], row['start_dt']-2000, row['end_dt']+2000\n",
    "    if startAt<0:\n",
    "        startAt=0\n",
    "    if startDt<0:\n",
    "        startDt=0\n",
    "    order_line1 = \"bsub -q q2680v2 -e %J.err -o %J.out 'pyGenomeTracks --tracks cotyledon_stigma.ini --region {0}:{1}-{2} -o figure/{3}_{4}.pdf'\".format(chrAt, startAt, endAt, index, GeneIdAt)\n",
    "    order_line2 = \"bsub -q q2680v2 -e %J.err -o %J.out 'pyGenomeTracks --tracks cotyledon_stigma.ini --region {0}:{1}-{2} -o figure/{3}_{4}.pdf'\".format(chrDt, startDt, endDt, index, GeneIdDt)\n",
    "    print(order_line1, file=fp)\n",
    "    print(order_line2, file=fp)\n",
    "fp.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 统计各组织各染色质状态占比\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/ChromHMM/J668_chromHMM.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'tags', 'tissue'])\n",
    "old_tags = ['E1', 'E2', 'E3', 'E4']\n",
    "new_tags = ['None', 'H3K4me3', 'H3K27ac', 'H3K4me3_H3K27ac']\n",
    "df['tags'].replace(old_tags, new_tags, inplace=True)\n",
    "df['length'] = df['end'] - df['start']\n",
    "df_statis = df.groupby(by=['tissue', 'tags'], as_index=False)['length'].sum()\n",
    "df_statis.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/ChromHMM/J668_chromHMM_statis.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "334\n",
      "1337\n"
     ]
    }
   ],
   "source": [
    "## 拉取stigma、ovule组织特异性同源基因，并且打上标签，还给出坐标\n",
    "import pandas as pd\n",
    "## 打标签，加坐标\n",
    "def add_position(df):\n",
    "    df_hg_position = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/all_homo_gene_pairs_position.txt\", sep=\"\\t\")\n",
    "    df_position = pd.merge(df, df_hg_position, how='inner', on=['GeneId_At', 'GeneId_Dt'])\n",
    "    df_At = df_position[['GeneId_At', 'chr_at', 'start_at', 'end_at']].copy()\n",
    "    df_Dt = df_position[['GeneId_Dt', 'chr_dt', 'start_dt', 'end_dt']].copy()\n",
    "    df_At.columns = ['GeneId', 'Chr', 'Start', 'End']\n",
    "    df_Dt.columns = ['GeneId', 'Chr', 'Start', 'End']\n",
    "    df_At['tags'] = range(df_At.shape[0])\n",
    "    df_Dt['tags'] = range(df_Dt.shape[0])\n",
    "    df_AtDt = pd.concat([df_At, df_Dt], axis=0)\n",
    "    return df_AtDt\n",
    "def main(df, tags):\n",
    "    #df_tags = df.query(\"TAU_Tags_At == @tags\")[['GeneId_At', 'GeneId_Dt']].copy()\n",
    "    df_tags = df.query(\"TAU_Tags_At == @tags\")[['GeneId_At', 'GeneId_Dt']].copy()\n",
    "    print(df_tags.shape[0])\n",
    "    df_result = add_position(df_tags)\n",
    "    return df_result\n",
    "df_TSG_HG = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/TAU/HG/TSG_HG_diff_type.bed\", sep=\"\\t\")\n",
    "\n",
    "df_stigma = main(df_TSG_HG, 'stigma')\n",
    "df_stigma.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/TAU/example/stigma_TSG_Hbias.bed\", sep=\"\\t\", index=False, header=False, columns=['Chr', 'Start', 'End', 'GeneId', 'tags'])\n",
    "\n",
    "df_ovule = main(df_TSG_HG, 'ovule')\n",
    "df_ovule.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/TAU/example/ovule_TSG_Hbias.bed\", sep=\"\\t\", index=False, header=False, columns=['Chr', 'Start', 'End', 'GeneId', 'tags'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析有多少组织差异表达基因在promoter区域存在差异, 统计有多少基因符合差异表达与ChIP-Seq peaks 信号存在与否有关\n",
    "import pandas as pd\n",
    "tissues = ['anther', 'ovule', 'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'petal', 'stigma']\n",
    "\n",
    "#df 表示差异表达基因，df_T1表示组织1在promoter区域有信号注释的基因，df_T2表示组织2在promoter区域有信号注释的基因\n",
    "def add_chip_signal(df, df_T1, df_T2):\n",
    "    df, df_T1, df_T2 = df.copy(), df_T1.copy(), df_T2.copy()\n",
    "    df_a = pd.merge(df, df_T1, how='left', on=['GeneId'])\n",
    "    df_ab = pd.merge(df_a, df_T2, how='left', on=['GeneId'])\n",
    "    return df_ab\n",
    "## 分析有peaks注释促进基因上调表达，没有peaks注释导致基因下调表达\n",
    "def DEG_peaks(df):\n",
    "    df = df.copy()\n",
    "    df.fillna(0, inplace=True)\n",
    "    df_all = df.query(\"sig!='Not' & control != treatment\")\n",
    "    df_result = df.query(\"(sig=='Down' & control==1 & treatment==0) | (sig=='Up' & control==0 & treatment==1)\").copy()\n",
    "    return df_result, df_all\n",
    "\n",
    "df_H3K4me3 = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/peaks/annotation/J668_H3K4me3_promoter_peaks.csv\")\n",
    "df_result = pd.DataFrame() ##记录chip peaks 的存在和缺失正向调控基因的上调和下调\n",
    "df_result2 = pd.DataFrame(columns=['control', 'treatment', 'all', 'positive']) ##记录chip peaks 的差异所能够影响的所有差异表达基因，其中包含了正向调控和负向调控\n",
    "for i in range(9):\n",
    "    for j in range(i+1, 10):\n",
    "        T1,T2 = tissues[i], tissues[j]\n",
    "        input_f = 'G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/{0}_{1}_DEG.csv'.format(T1,T2)\n",
    "        #print(input_f)\n",
    "        df = pd.read_csv(input_f)\n",
    "        df_T1 = df_H3K4me3.query(\"tissues ==@T1\")[['GeneId']].copy()\n",
    "        df_T1.drop_duplicates(keep='first', inplace=True)\n",
    "        df_T2 = df_H3K4me3.query(\"tissues==@T2\")[['GeneId']].copy()\n",
    "        df_T2.drop_duplicates(keep='first', inplace=True)\n",
    "        df_T1['control'] = [1]*df_T1.shape[0]\n",
    "        df_T2['treatment'] = [1]*df_T2.shape[0]\n",
    "        df_tmp = add_chip_signal(df, df_T1, df_T2)\n",
    "        df_tmp.dropna(subset=['control', 'treatment'], inplace=True, how='all')\n",
    "\n",
    "        #print(df_tmp)\n",
    "        #print(df_tmp.query(\"control != treatment & sig =='Not'\").shape[0])\n",
    "        #print(df_tmp.query(\"control != treatment & sig =='Not'\").head())\n",
    "        df_tmp, df_tmp2 = DEG_peaks(df_tmp)\n",
    "        df_result2.loc[df_result2.shape[0], ] = [T1, T2, df_tmp2.shape[0],  df_tmp.shape[0]]        \n",
    "        #print(df_tmp.shape[0])\n",
    "        df_tmp['C_tissue'] = [T1]*df_tmp.shape[0]\n",
    "        df_tmp['T_tissue'] = [T2]*df_tmp.shape[0]\n",
    "        df_result = pd.concat([df_result, df_tmp], axis=0)\n",
    "print(df_result2)\n",
    "df_result2.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/epigenome/J668_DEG_H3K4me3_sta.csv\", index=False)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/epigenome/J668_DEG_H3K4me3.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析有多少组织差异表达基因在promoter区域存在差异, 统计有多少基因符合差异表达与ChIP-Seq peaks 信号存在与否有关\n",
    "import pandas as pd\n",
    "tissues = ['ovule', 'root', 'radicle', 'hypocotyl', 'cotyledon', 'stem', 'stigma']\n",
    "\n",
    "#df 表示差异表达基因，df_T1表示组织1在promoter区域有信号注释的基因，df_T2表示组织2在promoter区域有信号注释的基因\n",
    "def add_chip_signal(df, df_T1, df_T2):\n",
    "    df, df_T1, df_T2 = df.copy(), df_T1.copy(), df_T2.copy()\n",
    "    df_a = pd.merge(df, df_T1, how='left', on=['GeneId'])\n",
    "    df_ab = pd.merge(df_a, df_T2, how='left', on=['GeneId'])\n",
    "    return df_ab\n",
    "## 分析有peaks注释促进基因上调表达，没有peaks注释导致基因下调表达\n",
    "def DEG_peaks(df):\n",
    "    df = df.copy()\n",
    "    df.fillna(0, inplace=True)\n",
    "    df_all = df.query(\"sig!='Not' & control != treatment\")\n",
    "    df_result = df.query(\"(sig=='Down' & control==1 & treatment==0) | (sig=='Up' & control==0 & treatment==1)\").copy()\n",
    "    print(df_all.shape[0], df_result.shape[0])\n",
    "    return df_result, df_all\n",
    "\n",
    "df_H3K4me3 = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/peaks/annotation/J668_H3K27ac_promoter_peaks.csv\")\n",
    "df_result = pd.DataFrame()\n",
    "df_result2 = pd.DataFrame(columns=['control', 'treatment', 'all', 'positive']) ##记录chip peaks 的差异所能够影响的所有差异表达基因，其中包含了正向调控和负向调控\n",
    "for i in range(6):\n",
    "    for j in range(i+1, 7):\n",
    "        T1,T2 = tissues[i], tissues[j]\n",
    "        input_f = 'G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/{0}_{1}_DEG.csv'.format(T1,T2)\n",
    "        df = pd.read_csv(input_f)\n",
    "        df_T1 = df_H3K4me3.query(\"tissues ==@T1\")[['GeneId']].drop_duplicates().copy()\n",
    "        df_T2 = df_H3K4me3.query(\"tissues==@T2\")[['GeneId']].drop_duplicates().copy()\n",
    "        df_T1['control'] = [1]*df_T1.shape[0]\n",
    "        df_T2['treatment'] = [1]*df_T2.shape[0]\n",
    "        df_tmp = add_chip_signal(df, df_T1, df_T2)\n",
    "        #print(T1, T2)\n",
    "        df_tmp, df_tmp2  = DEG_peaks(df_tmp)\n",
    "        df_result2.loc[df_result2.shape[0], ] = [T1, T2, df_tmp2.shape[0],  df_tmp.shape[0]]   \n",
    "        df_tmp['C_tissue'] = [T1]*df_tmp.shape[0]\n",
    "        df_tmp['T_tissue'] = [T2]*df_tmp.shape[0]\n",
    "        df_result = pd.concat([df_result, df_tmp], axis=0)\n",
    "df_result2.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/epigenome/J668_DEG_H3K27ac_sta.csv\", index=False)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/epigenome/J668_DEG_H3K27ac.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 统计组织间差异表达基因有多少是与ChIP-Seq修饰信号有关\n",
    "import pandas as pd\n",
    "def sta_number(df):\n",
    "    df['number'] = [1]*df.shape[0]\n",
    "    df_result = df.groupby(by=['sig', 'C_tissue', 'T_tissue'], as_index=False)['number'].sum()\n",
    "    df_result['C_T'] = df_result['C_tissue'].map(str) + \"_\" + df_result['T_tissue'].map(str)\n",
    "    return df_result\n",
    "\n",
    "df_H3K27ac = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/epigenome/J668_DEG_H3K27ac.csv\")\n",
    "df_H3K27ac = sta_number(df_H3K27ac)\n",
    "df_H3K27ac.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/epigenome/J668_DEG_H3K27ac_number.csv\", index=False)\n",
    "\n",
    "df_H3K4me3 = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/epigenome/J668_DEG_H3K4me3.csv\")\n",
    "df_H3K4me3 = sta_number(df_H3K4me3)\n",
    "df_H3K4me3.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/epigenome/J668_DEG_H3K4me3_number.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 整理为饼图可识别的格式\n",
    "import pandas as pd\n",
    "df_pie = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/J668_diffChIP_DEG.csv\")\n",
    "old_tags = ['DEG', 'NDEG']\n",
    "new_tags = ['Up', 'Down']\n",
    "df_pie['table'].replace(old_tags, new_tags, inplace=True)\n",
    "# H3K4me3\n",
    "df_H3K4me3 = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/epigenome/J668_DEG_H3K4me3_number.csv\")\n",
    "df_H3K4me3 = df_H3K4me3[['sig', 'C_tissue', 'T_tissue', 'number']].copy()\n",
    "df_H3K4me3['tags'] = ['H3K4me3']*df_H3K4me3.shape[0]\n",
    "df_H3K4me3.columns = ['table', 'tissueA', 'tissueB', 'number', 'tags']\n",
    "# H3K27ac\n",
    "df_H3K27ac = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/epigenome/J668_DEG_H3K27ac_number.csv\")\n",
    "df_H3K27ac = df_H3K27ac[['sig', 'C_tissue', 'T_tissue', 'number']].copy()\n",
    "df_H3K27ac['tags'] = ['H3K27ac']*df_H3K27ac.shape[0]\n",
    "old_tags = ['Up', 'Down']\n",
    "new_tags = ['Down', 'Up']\n",
    "df_H3K27ac['sig'].replace(old_tags, new_tags, inplace=True)\n",
    "\n",
    "df_H3K27ac.columns = ['table', 'tissueB', 'tissueA', 'number', 'tags']\n",
    "df_H3K27ac_H3K4me3 = pd.concat([df_H3K4me3, df_H3K27ac], axis=0)\n",
    "\n",
    "\n",
    "## data concat\n",
    "df_tmp = pd.merge(df_pie, df_H3K27ac_H3K4me3, how='left', on=['table', 'tissueA', 'tissueB', 'tags'])\n",
    "df_tmp['number_x'] = df_tmp['number_y'].fillna(df_tmp['number_x'])\n",
    "df_tmp.drop('number_y', inplace=True, axis=1)\n",
    "df_tmp.rename(columns={'number_x':'number'}, inplace=True)\n",
    "\n",
    "df_tissue = pd.DataFrame({'tissueA':['ovule', 'radicle', 'hypocotyl', 'cotyledon', 'stem', 'stigma']})\n",
    "\n",
    "df_tmp_a = pd.merge(df_tmp, df_tissue, how='inner', on=['tissueA'])\n",
    "df_tissue.columns = ['tissueB']\n",
    "df_tmp_ab = pd.merge(df_tmp_a, df_tissue, how='inner', on=['tissueB'])\n",
    "\n",
    "df_tmp_ab.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/epigenome/J668_tiss_piar_DEG_chir_ab.csv\", index=False)\n",
    "df_tmp.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/epigenome/J668_tiss_piar_DEG_chir.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析有多少同源基因bias表示受到了ChiP修饰的影响 (H3K4me3)\n",
    "import pandas as pd\n",
    "df_H3K4me3 = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/peaks/annotation/J668_H3K4me3_promoter_peaks.csv\")\n",
    "df_H3K4me3.drop_duplicates(keep='first', subset=['GeneId', 'tissues'], inplace=True)\n",
    "df_H3K4me3 = df_H3K4me3[['GeneId', 'tissues']].copy()\n",
    "df_H3K4me3['tags'] = [1]*df_H3K4me3.shape[0]\n",
    "df_H3K4me3.columns = ['GeneId_At', 'tissues', 'sig_At']\n",
    "df_hg = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/all_tissue_bias_gene_pair.bed\", sep=\"\\t\")\n",
    "df_a = pd.merge(df_hg, df_H3K4me3, how='left', on=['GeneId_At', 'tissues'])\n",
    "df_H3K4me3.columns = ['GeneId_Dt', 'tissues', 'sig_Dt']\n",
    "df_ab = pd.merge(df_a, df_H3K4me3, how='left', on=['GeneId_Dt', 'tissues'])\n",
    "df_ab.dropna(subset=['sig_At', 'sig_Dt'], how='all', inplace=True)\n",
    "df_ab.fillna(0, inplace=True)\n",
    "df_result = df_ab.query(\"(sig=='BiasA' & sig_At==1 & sig_Dt==0) | (sig=='BiasD' & sig_At==0 & sig_Dt==1)\").copy()\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/epigenome/J668_HG_bias_H3K4me3.csv\", index=False)\n",
    "df_result['number'] = [1]*df_result.shape[0]\n",
    "df_sta = df_result.groupby(by=['tissues', 'sig'], as_index=False)['number'].sum()\n",
    "df_sta.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/epigenome/J668_HG_bias_H3K4me3_sta.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析有多少同源基因bias表达受到chip修饰的影响 (H3K27ac)\n",
    "import pandas as pd\n",
    "df_H3K27ac = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/peaks/annotation/J668_H3K27ac_promoter_peaks.csv\")\n",
    "df_H3K27ac.drop_duplicates(keep='first', subset=['GeneId', 'tissues'], inplace=True)\n",
    "df_H3K27ac = df_H3K27ac[['GeneId', 'tissues']].copy()\n",
    "df_H3K27ac['tags'] = [1]*df_H3K27ac.shape[0]\n",
    "df_H3K27ac.columns = ['GeneId_At', 'tissues', 'sig_At']\n",
    "df_hg = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/all_tissue_bias_gene_pair.bed\", sep=\"\\t\")\n",
    "df_a = pd.merge(df_hg, df_H3K27ac, how='left', on=['GeneId_At', 'tissues'])\n",
    "df_H3K27ac.columns = ['GeneId_Dt', 'tissues', 'sig_Dt']\n",
    "df_ab = pd.merge(df_a, df_H3K27ac, how='left', on=['GeneId_Dt', 'tissues'])\n",
    "df_ab.dropna(subset=['sig_At', 'sig_Dt'], how='all', inplace=True)\n",
    "df_ab.fillna(0, inplace=True)\n",
    "df_result = df_ab.query(\"(sig=='BiasA' & sig_At==1 & sig_Dt==0) | (sig=='BiasD' & sig_At==0 & sig_Dt==1)\").copy()\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/epigenome/J668_HG_bias_H3K27ac.csv\", index=False)\n",
    "df_result['number'] = [1]*df_result.shape[0]\n",
    "df_sta = df_result.groupby(by=['tissues', 'sig'], as_index=False)['number'].sum()\n",
    "df_sta.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/epigenome/J668_HG_bias_H3K27ac_sta.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析有多少同源基因bias表达受到ATAC\n",
    "import pandas as pd\n",
    "df_ATAC = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/rawdata/J668_ATAC_promoter_peaks.csv\")\n",
    "df_ATAC.drop_duplicates(keep='first', subset=['GeneId', 'tissues'], inplace=True)\n",
    "df_ATAC = df_ATAC[['GeneId', 'tissues']].copy()\n",
    "df_ATAC['tags'] = [1]*df_ATAC.shape[0]\n",
    "df_ATAC.columns = ['GeneId_At', 'tissues', 'sig_At']\n",
    "df_hg = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/all_tissue_bias_gene_pair.bed\", sep=\"\\t\")\n",
    "df_a = pd.merge(df_hg, df_ATAC, how='left', on=['GeneId_At', 'tissues'])\n",
    "df_ATAC.columns = ['GeneId_Dt', 'tissues', 'sig_Dt']\n",
    "df_ab = pd.merge(df_a, df_ATAC, how='left', on=['GeneId_Dt', 'tissues'])\n",
    "df_ab.dropna(subset=['sig_At', 'sig_Dt'], how='all', inplace=True)\n",
    "df_ab.fillna(0, inplace=True)\n",
    "df_result = df_ab.query(\"(sig=='BiasA' & sig_At==1 & sig_Dt==0) | (sig=='BiasD' & sig_At==0 & sig_Dt==1)\").copy()\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/epigenome/J668_HG_bias_ATAC.csv\", index=False)\n",
    "df_result['number'] = [1]*df_result.shape[0]\n",
    "df_sta = df_result.groupby(by=['tissues', 'sig'], as_index=False)['number'].sum()\n",
    "df_sta.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/epigenome/J668_HG_bias_ATAC_sta.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "## hypocotyl 所包含的bias整理出来，满足H3K27ac，H3K4me3、ATAC均在TSS区域存在或是缺失造成同源基因bias表达\n",
    "from functools import reduce \n",
    "import pandas as pd\n",
    "df_ATAC = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/epigenome/J668_HG_bias_ATAC.csv\")\n",
    "df_H3K27ac = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/epigenome/J668_HG_bias_H3K27ac.csv\")\n",
    "df_H3K4me3 = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/epigenome/J668_HG_bias_H3K4me3.csv\")\n",
    "# 找到H3K27ac，H3K4me3和ATAC均位于启动子区域有差异并且基因表达bias有差异的同源基因\n",
    "df_ATAC_hypocotyl = df_ATAC.query(\"tissues=='hypocotyl'\")[['GeneId_At', 'GeneId_Dt']].copy()\n",
    "df_H3K27ac_hypocotyl = df_H3K27ac.query(\"tissues=='hypocotyl'\")[['GeneId_At', 'GeneId_Dt']].copy()\n",
    "df_H3K4me3_hypocotyl = df_H3K4me3.query(\"tissues=='hypocotyl'\")[['GeneId_At', 'GeneId_Dt']].copy()\n",
    "df_result = reduce(lambda left,right: pd.merge(left, right, on=['GeneId_At','GeneId_Dt'], how='inner'), [df_ATAC_hypocotyl, df_H3K27ac_hypocotyl, df_H3K4me3_hypocotyl])\n",
    "# 分析这些同源基因对的位置\n",
    "df_hg = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/all_homo_gene_pairs_position.txt\", sep=\"\\t\")\n",
    "df_result = pd.merge(df_hg, df_result, how='inner', on=['GeneId_At', 'GeneId_Dt'])\n",
    "df_result['tags'] = range(df_result.shape[0])\n",
    "df_at = df_result[['chr_at', 'start_at', 'end_at', 'GeneId_At', 'tags']].copy()\n",
    "df_dt = df_result[['chr_dt', 'start_dt', 'end_dt', 'GeneId_Dt', 'tags']].copy()\n",
    "df_at.columns = ['chr', 'start', 'end', 'GeneId', 'tags']\n",
    "df_dt.columns = ['chr', 'start', 'end', 'GeneId', 'tags']\n",
    "df_result = pd.concat([df_at, df_dt], axis=0)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/epigenome/hypocotyl_bias_effected_epignome.bed\", sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析hypocotyl和stigma 间受H3K4me3和 H3K27ac以及ATAC影响的差异表达基因, 20230515\n",
    "import pandas as pd\n",
    "from functools import reduce \n",
    "df_H3K4me3 = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/epigenome/J668_DEG_H3K4me3.csv\")\n",
    "df_H3K27ac = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/epigenome/J668_DEG_H3K27ac.csv\")\n",
    "df_ATAC = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/diffpeaks/gene/hypocotyl_stigma_annotation_gene.bed\", sep=\" \")\n",
    "df_ATAC_gene = df_ATAC[['geneId']].drop_duplicates(keep='first').copy()\n",
    "df_ATAC_gene.columns = ['GeneId']\n",
    "df_H3K27ac_gene = df_H3K27ac.query(\"C_tissue=='hypocotyl' & T_tissue=='stigma'\")[['GeneId']].copy()\n",
    "df_H3K4me3_gene = df_H3K4me3.query(\"C_tissue=='hypocotyl' & T_tissue=='stigma'\")[['GeneId']].copy()\n",
    "df_DEG_id = reduce(lambda leaf, right: pd.merge(leaf, right, on=['GeneId'], how='inner'), [df_ATAC_gene, df_H3K27ac_gene, df_H3K4me3_gene])\n",
    "## 找到对应的基因id\n",
    "df_GeneId = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/all_gene_sorted.bed6\", sep=\"\\t\", names=['Chr', 'Start', 'End', 'GeneId', 'a', 'b'])\n",
    "df_GeneId = df_GeneId[['Chr', 'Start', 'End', 'GeneId']].copy()\n",
    "df_result = pd.merge(df_DEG_id, df_GeneId, how='inner', on=['GeneId'])\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/epigenome/hypocotyl_stigma_DEG_epigenome.bed\", sep=\"\\t\", index=False, header=False, columns=['Chr', 'Start', 'End', 'GeneId'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 抽取组织特异表达基因, hypocotyl 和stigma， 2023.05.15\n",
    "import pandas as pd\n",
    "def gene_pst(df):\n",
    "    df = df.copy()\n",
    "    df_position = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/all_gene_sorted.bed6\", sep=\"\\t\", names=['chr', 'start', 'end', 'GeneId', 'a', 'b'])\n",
    "    df_return = pd.merge(df, df_position, how='inner', on=['GeneId'])\n",
    "    return df_return\n",
    "for t in ['hypocotyl', 'stigma']:\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/TAU/{}_TAU_GeneId.bed\".format(t)\n",
    "    output_f = \"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/TSG/hypocotyl_stigma/{}_TAU_Gene_position.bed\".format(t)\n",
    "    df = pd.read_csv(input_f, sep=\"\\t\", names=['GeneId'])\n",
    "    df_return = gene_pst(df)\n",
    "    df_return.to_csv(output_f, sep=\"\\t\", index=False, header=False, columns=['chr', 'start', 'end', 'GeneId', 'a', 'b'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.3 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "a0f669ea42a4dcea88d385f8f9b4c3d9b966e24234758831d1dbfdac925619a2"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
