{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# analysis micore environment in special chromosome region\n",
    "import pandas as pd\n",
    "import os\n",
    "import sys\n",
    "import argparse\n",
    "\n",
    "def main(argv):\n",
    "    #! init\n",
    "    parser = argparse.ArgumentParser(description=\"\")\n",
    "    parser.add_argument(\"-g\", \"--gene\", required=True, help=\"The gene file\")\n",
    "    parser.add_argument(\"-c\", \"--chip\", required=True, help=\"The chip file\")\n",
    "    parser.add_argument(\"-o\", \"--output\", required=True, help='The output file')\n",
    "    parser.add_argument(\"-l\", \"--length\", required='True', help=\"The length of region to be extend\")\n",
    "    parser.add_argument(\"-n\", \"--number\", required='True', help=\"The number of splits for the region\")\n",
    "    parser.add_argument(\"-cn\", \"--chrom_name\", required='True', help=\"The number of chromsome\")\n",
    "    args = parser.parse_args()\n",
    "    genef = args.gene\n",
    "    chipf = args.chip\n",
    "    L = int(args.length)\n",
    "    N = int(args.number)\n",
    "    cn = args.chrom_name\n",
    "    outputf = args.output\n",
    "    #! analysis\n",
    "    #* 1、region extend\n",
    "    df_gene = pd.read_csv(genef, sep=\"\\t\")\n",
    "    df_gene = region_extend(df_gene, L)\n",
    "    df_chip = pd.read_csv(chipf, sep=\"\\t\")\n",
    "    df_result = pd.DataFrame()\n",
    "    #* single chrom \n",
    "    #for c in ['Ghir_A01', 'Ghir_A02', 'Ghir_A03', 'Ghir_A04', 'Ghir_A05', 'Ghir_A06', 'Ghir_A07', 'Ghir_A08', 'Ghir_A09', 'Ghir_A10', 'Ghir_A11', 'Ghir_A12', 'Ghir_A13', 'Ghir_D01', 'Ghir_D02', 'Ghir_D03', 'Ghir_D04', 'Ghir_D05', 'Ghir_D06', 'Ghir_D07', 'Ghir_D08', 'Ghir_D09', 'Ghir_D10', 'Ghir_D11', 'Ghir_D12', 'Ghir_D13']:\n",
    "    for c in [cn]:\n",
    "        df_gene_chr = df_gene.query(\"chrom==@c\").copy()\n",
    "        df_chip_chr = df_chip.query(\"chrom==@c\").copy()\n",
    "        print(df_gene_chr)\n",
    "        print(df_chip_chr)\n",
    "        #* 2、region chip modified \n",
    "        df_region_chip = region_chip(df_gene_chr, df_chip_chr, L, N)\n",
    "        #* 3、statistics chip modified \n",
    "        df_return = data_format(df_gene_chr, df_region_chip, N)\n",
    "        df_result = pd.concat([df_result, df_return], axis=0)\n",
    "    df_result.to_csv(outputf)\n",
    "    \n",
    "# extend the region\n",
    "def region_extend(df, length):\n",
    "    df['s'] = df['start'] - length\n",
    "    df['e'] = df['end'] + length\n",
    "    df_return = df[['chrom', 's', 'e', 'GeneId']]\n",
    "    df_return.columns = ['chrom', 'start', 'end', 'GeneId']\n",
    "    return df_return\n",
    "# find chip modified on the region\n",
    "def region_chip(df_gene, df_chip, length, number):\n",
    "    df_return = pd.DataFrame()\n",
    "    for index, row in df_gene.iterrows():\n",
    "        ## add geneid region\n",
    "        C, S, E, G = row['chrom'], row['start'], row['end'], row['GeneId']\n",
    "        #print(C, S, E, G)\n",
    "        df_tmp = df_chip.query(\"chrom==@C & start<=@E & end>=@S\").copy()\n",
    "        df_tmp.reset_index(inplace=True, drop=True)\n",
    "        df_tmp.loc[0, 'start'] , df_tmp.loc[df_tmp.shape[0]-1, 'end']= S, E\n",
    "        df_tmp['GeneId'] = [G]*df_tmp.shape[0]\n",
    "        #print(df_tmp)\n",
    "        #df_tmp['Total_Length'] = [E-S + 2*length]*df_tmp.shape[0]\n",
    "        ## split region and add tags\n",
    "        average_length = int((E-S)/number)\n",
    "        gene_region_length = E-S\n",
    "        #print(average_length)\n",
    "        for i in range(number):\n",
    "            S_tmp = S\n",
    "            E_tmp = S + average_length\n",
    "            df_tmp2 = df_tmp.query(\"chrom==@C & start <= @E_tmp & end>= @S_tmp\").copy()\n",
    "            df_tmp2.reset_index(drop=True, inplace=True)\n",
    "            df_tmp2.loc[0, 'start'] = S_tmp\n",
    "            df_tmp2.loc[df_tmp2.shape[0]-1, 'end'] = E_tmp\n",
    "            tags = \"R{}\".format(i)\n",
    "            df_tmp2['tags'] = [tags]*df_tmp2.shape[0]\n",
    "            df_tmp2['gene_region_length'] = [gene_region_length]*df_tmp2.shape[0]\n",
    "            df_tmp2['sig_total_length'] = [E_tmp-S_tmp]*df_tmp2.shape[0]\n",
    "            df_tmp2['length'] = df_tmp2['end'] - df_tmp2['start'] \n",
    "            df_return = pd.concat([df_return , df_tmp2], axis=0)\n",
    "            S = E_tmp+1\n",
    "    return df_return\n",
    "# adjust data format\n",
    "def data_format(df_gene, df_chip, number):\n",
    "    df_tmp = df_chip.groupby(by=['GeneId','gene_region_length', 'sig_total_length', 'tags', 'sigs'], as_index=False)['length'].sum()\n",
    "    df_tmp['ratio'] = df_tmp['length']/df_tmp['sig_total_length']\n",
    "    #print(df_tmp.head(30))\n",
    "    df_tags = df_gene[['GeneId']].copy()\n",
    "    # tags\n",
    "    for i in range(number):\n",
    "        t = 'R{}'.format(i)\n",
    "        df_tags[t] = [t]*df_tags.shape[0]\n",
    "    df_tags = pd.melt(df_tags, id_vars=['GeneId'])\n",
    "    df_tags = df_tags[['GeneId', 'variable']].copy()\n",
    "    df_tags.columns = ['GeneId', 'tags']\n",
    "    # sigs\n",
    "    df_sigs = df_gene[['GeneId']].copy()\n",
    "    for i in range(1,5):\n",
    "        s = \"E{}\".format(i)\n",
    "        df_sigs[s] = [s]*df_sigs.shape[0]\n",
    "    df_sigs = pd.melt(df_sigs, id_vars=['GeneId'])\n",
    "    df_sigs = df_sigs[['GeneId', 'variable']].copy()\n",
    "    df_sigs.columns = ['GeneId', 'sigs']\n",
    "    df_gene_tags = pd.merge(df_tags, df_sigs, how='outer', on=['GeneId'])\n",
    "    df_gene_tags['ratio'] = [0]*df_gene_tags.shape[0]\n",
    "    df_tmp = df_tmp[['GeneId', 'tags', 'sigs', 'ratio']]\n",
    "    df_tmp = pd.concat([df_tmp, df_gene_tags], axis=0)\n",
    "    df_tmp.drop_duplicates(keep='first', inplace=True, subset=['GeneId', 'tags', 'sigs'])\n",
    "    df_tmp.sort_values(by=['GeneId', 'tags', 'sigs'], inplace=True)\n",
    "    df_tmp['tags_sigs'] = df_tmp['tags'].map(str) + \"-\" + df_tmp['sigs'].map(str)\n",
    "    df_tmp = df_tmp[['GeneId', 'tags_sigs', 'ratio']]\n",
    "    df_return = df_tmp.pivot(index='GeneId', columns='tags_sigs', values='ratio')\n",
    "    return df_return\n",
    "if __name__ == \"__main__\":\n",
    "    main(sys.argv[1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\pc\\AppData\\Local\\Temp\\ipykernel_22872\\2494101199.py:19: FutureWarning: Passing 'suffixes' which cause duplicate columns {'R1-E1_x', 'R4-E4_x', 'R2-E1_x', 'R1-E3_x', 'R2-E2_x', 'R2-E4_x', 'R4-E2_x', 'R4-E1_x', 'R1-E4_x', 'R3-E2_x', 'R1-E2_x', 'R0-E4_x', 'R0-E3_x', 'R3-E4_x', 'R0-E2_x', 'R0-E1_x', 'R3-E3_x', 'R3-E1_x', 'R2-E3_x', 'R4-E3_x'} in the result is deprecated and will raise a MergeError in a future version.\n",
      "  df_micro_state = pd.merge(df_micro_state, df_tmp, how='inner', on=['GeneId'])\n"
     ]
    }
   ],
   "source": [
    "## 将多个数据文件整理为一个文件，并且z-score 标准化，再用umap 来聚类分析\n",
    "\n",
    "# 1、数据合并\n",
    "samples = ['hypocotyl', 'leaf', 'radicle', 'stigma']\n",
    "#samples = ['hypocotyl']\n",
    "chroms = ['Ghir_A01', 'Ghir_A02', 'Ghir_A03', 'Ghir_A04', 'Ghir_A05', 'Ghir_A06', 'Ghir_A07', 'Ghir_A08', 'Ghir_A09', 'Ghir_A10', 'Ghir_A11', 'Ghir_A12', 'Ghir_A13', \n",
    "          'Ghir_D01', 'Ghir_D02', 'Ghir_D03', 'Ghir_D04', 'Ghir_D05', 'Ghir_D06', 'Ghir_D07', 'Ghir_D08', 'Ghir_D09', 'Ghir_D10', 'Ghir_D11', 'Ghir_D12', 'Ghir_D13']\n",
    "df_micro_state = pd.DataFrame()\n",
    "for s in samples:\n",
    "    df_tmp = pd.DataFrame()\n",
    "    for c in chroms:\n",
    "        input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/micro_state/chromHMM_ratio/5/{0}_micro_state_{1}.csv\".format(s, c)\n",
    "        df = pd.read_csv(input_f)\n",
    "        df_tmp = pd.concat([df_tmp, df], axis=0)\n",
    "        df_tmp.reset_index(drop=True, inplace=True)\n",
    "    if df_micro_state.shape[0] == 0:\n",
    "        df_micro_state = df_tmp.copy()\n",
    "    else:\n",
    "        df_micro_state = pd.merge(df_micro_state, df_tmp, how='inner', on=['GeneId'])\n",
    "        #df_micro_state = pd.merge(df_micro_state, df_tmp, how='inner', on=['GeneId'])\n",
    "#print(df_micro_state)\n",
    "\n",
    "# z-score 标准化， umap 聚类\n",
    "#print(df_micro_state)\n",
    "col_list = []\n",
    "for s in samples:\n",
    "    for k in range(1,5):\n",
    "        for i in range(5):\n",
    "            col = \"{0}_R{1}-E{2}\".format(s, k, i)\n",
    "            col_list.append(col)\n",
    "\n",
    "def zscore_umap(df):\n",
    "    import scipy.stats as stats\n",
    "    from sklearn.preprocessing import StandardScaler\n",
    "    import umap.umap_ as umap\n",
    "    #z-score\n",
    "    id_list = df['GeneId'].tolist()\n",
    "    df.set_index('GeneId', inplace=True)\n",
    "    df_numpy = df.to_numpy()\n",
    "    # df_zscore = stats.zscore(df_numpy, axis=1)\n",
    "    # df_zscore = pd.DataFrame(df_zscore, columns=col_list)\n",
    "    df_zscore = pd.DataFrame(df_numpy, columns=col_list)\n",
    "    # umap\n",
    "    reducer = umap.UMAP()\n",
    "    scaled_penguin_data = StandardScaler().fit_transform(df_zscore)\n",
    "    embedding = reducer.fit_transform(scaled_penguin_data)\n",
    "    df_umap = pd.DataFrame(embedding, columns =['umap1', 'umap2'], index=id_list)\n",
    "    return df_umap\n",
    "df_umap = zscore_umap(df_micro_state)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_umap.reset_index(inplace=True)\n",
    "df_umap.columns = ['GeneId', 'umap1', 'umap2']\n",
    "df_gene = pd.read_csv('G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/expression_gene/J668_all_tissue_average_TPM.txt', sep=\"\\t\")\n",
    "df_gene = df_gene[['GeneId', 'Hypocotyl']]\n",
    "df_result = pd.merge(df_umap, df_gene, how='inner', on=['GeneId'])\n",
    "# df_gene = df_result['Hypocotyl'].to_numpy()\n",
    "# df_gene_zscore = stats.zscore(df_gene, axis=0)\n",
    "# df_result['Hypocotyl'] = df_gene_zscore\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/micro_state/gene/hypocotyl_gene_micro_state.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                GeneId      umap1     umap2\n",
      "0      Ghir_A01G000010   6.050754  7.385537\n",
      "1      Ghir_A01G000020   6.954064  1.978401\n",
      "2      Ghir_A01G000030   6.013783  5.801857\n",
      "3      Ghir_A01G000040   9.074519  9.409721\n",
      "4      Ghir_A01G000050  13.280522  6.789733\n",
      "...                ...        ...       ...\n",
      "68895  Ghir_D13G025640  13.688496  6.244218\n",
      "68896  Ghir_D13G025650  11.114699  6.351100\n",
      "68897  Ghir_D13G025660   9.548344  9.074274\n",
      "68898  Ghir_D13G025670  10.023056  1.930667\n",
      "68899  Ghir_D13G025680   6.016584  7.462289\n",
      "\n",
      "[68900 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "### 将TSG 选出来，umap 聚类的结果\n",
    "df_umap.reset_index(inplace=True)\n",
    "df_umap.columns = ['GeneId', 'umap1', 'umap2']\n",
    "print(df_umap)\n",
    "df_TSG = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/tau_gini/J668_tissue_special_gene.txt\", sep=\"\\t\")\n",
    "df_TSG_id = df_TSG.query(\"Tags=='hypocotyl' | Tags=='leaf' | Tags=='radicle' | Tags=='stigma'\")[['GeneId', 'Tags']].copy()\n",
    "df_TSG_umap = pd.merge(df_TSG_id, df_umap, how='inner', on=['GeneId'])\n",
    "df_TSG_umap.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/micro_state/TSG/result.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 分析bam文件中的信号在基因处的分布\n",
    "# 分析基因所处于的bam区间\n",
    "#Nutures\n",
    "#20221101\n",
    "#Analysis the gene located in the bins\n",
    "\n",
    "import pandas as pd\n",
    "import sys\n",
    "\n",
    "gene_f = sys.argv[1]\n",
    "bin_f = sys.argv[2]\n",
    "output_f = sys.argv[3]\n",
    "def gene_bin(gene, bins, output):\n",
    "    df_gene = pd.read_csv(gene, sep=\"\\t\", names=['chrom', 'start', 'end', 'GeneId'])\n",
    "    df_bin = pd.read_csv(bins, sep=\"\\t\", names=['chrom', 'start', 'end', 'signal'])\n",
    "    chrom = ['Ghir_A01', 'Ghir_A02', 'Ghir_A03', 'Ghir_A04', 'Ghir_A05', 'Ghir_A06', 'Ghir_A07', 'Ghir_A08', 'Ghir_A09', 'Ghir_A10', 'Ghir_A11', 'Ghir_A12', 'Ghir_A13',\n",
    "             'Ghir_D01', 'Ghir_D02', 'Ghir_D03', 'Ghir_D04', 'Ghir_D05', 'Ghir_D06', 'Ghir_D07', 'Ghir_D08', 'Ghir_D09', 'Ghir_D10', 'Ghir_D11', 'Ghir_D12', 'Ghir_D13']\n",
    "    df_return = pd.DataFrame()\n",
    "    for c in chrom:\n",
    "        df_gene_chr = df_gene.query(\"chrom==@c\").copy()\n",
    "        df_bin_chr = df_bin.query(\"chrom==@c\").copy()\n",
    "        for index, row in df_gene_chr.iterrows():\n",
    "            S, E , G = row['start'], row['end'], row['GeneId']\n",
    "            df_tmp = df_bin_chr.query(\"start<=@S & end>=@E\").copy()\n",
    "            df_tmp['GeneId'] = [G]*df_tmp.shape[0]\n",
    "            df_return = pd.concat([df_return, df_tmp], axis=0)\n",
    "    df_return['Start'] = df_return['start'] - 3000\n",
    "    df_return['End'] = df_return['end'] + 3000\n",
    "    df_return.to_csv(output, sep=\"\\t\", columns=['chrom', 'Start', 'End'], index=False)\n",
    "gene_bin(gene_f, bin_f, output_f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析基因周围的chip信号丰度\n",
    "import pandas as pd\n",
    "import sys\n",
    "gene_f = sys.argv[1]\n",
    "single_f = sys.argv[2]\n",
    "output_f = sys.argv[3]\n",
    "\n",
    "def epigenome_single(gene, single, output):\n",
    "    df_gene = pd.read_csv(gene_f, sep=\"\\t\", names=['chrom', 'start', 'end', 'sigs'])\n",
    "    df_single = pd.read_csv(single_f, sep=\"\\t\", names=['chrom', 'start', 'end', 'single'])\n",
    "    df_gene[['GeneId', 'Tags']] = df_gene['sigs'].split(\"-\", expand = True)\n",
    "    df_gene = df_gene[['chrom', 'start', 'end', 'GeneId', 'Tags']].copy()\n",
    "    df_tmp = pd.merge(df_gene, df_single, how='inner', on=['chrom', 'start', 'end', 'GeneId'])\n",
    "    df_return = df_tmp.pivot(columns='Tags', index='GeneId', values='single')\n",
    "    df_return.to_csv(output, index=False)\n",
    "epigenome_single(gene_f, single_f, output_f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 对结果zscore 和umap 标准化\n",
    "import pandas as pd\n",
    "import scipy.stats as stats\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import umap.umap_ as umap\n",
    "def zscore_umap(df):\n",
    "    #z-score\n",
    "    id_list = df['GeneId'].tolist()\n",
    "    df.set_index('GeneId', inplace=True)\n",
    "    df[df<=0] = 0\n",
    "    df.fillna(0, inplace=True)\n",
    "    df_numpy = df.to_numpy()\n",
    "    #print(df_numpy)\n",
    "    # df_zscore = stats.zscore(df_numpy, axis=1)\n",
    "    # df_zscore = pd.DataFrame(df_zscore)\n",
    "    # df_zscore.fillna(0, inplace=True)\n",
    "    df_zscore = pd.DataFrame(df_numpy)\n",
    "    # # umap\n",
    "    reducer = umap.UMAP()\n",
    "    scaled_penguin_data = StandardScaler().fit_transform(df_zscore)\n",
    "    embedding = reducer.fit_transform(scaled_penguin_data)\n",
    "    df_umap = pd.DataFrame(embedding, columns =['umap1', 'umap2'], index=id_list)\n",
    "    return df_umap\n",
    "# df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/micro_state/epigenome_single/gene_cotyledon_H3K27ac_3K_100bp.csv\")\n",
    "# #print(df.head())\n",
    "# df_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/expression_gene/J668_all_tissue_average_TPM.txt\", sep=\"\\t\")\n",
    "# df_umap = zscore_umap(df)\n",
    "# df_umap.reset_index(inplace=True)\n",
    "# df_umap.columns = ['GeneId', 'umap1', 'umap2']\n",
    "# df_result = pd.merge(df_umap, df_gene, how='inner', on=['GeneId'])\n",
    "# df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/micro_state/epigenome_single/Gene_cotyledon_H3K27ac.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 整理，将H3K4me3 和 H3K27ac 都加进去\n",
    "import pandas as pd\n",
    "df_H3K4me3 = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/micro_state/epigenome_single/gene_cotyledon_H3K4me3_3K_100bp.csv\")\n",
    "df_H3K27ac = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/micro_state/epigenome_single/gene_cotyledon_H3K27ac_3K_100bp.csv\")\n",
    "df = pd.merge(df_H3K4me3, df_H3K27ac, how='inner', on=['GeneId'])\n",
    "df_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/expression_gene/J668_all_tissue_average_TPM.txt\", sep=\"\\t\")\n",
    "df_umap = zscore_umap(df)\n",
    "df_umap.reset_index(inplace=True)\n",
    "df_umap.columns = ['GeneId', 'umap1', 'umap2']\n",
    "df_result = pd.merge(df_umap, df_gene, how='inner', on=['GeneId'])\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/micro_state/epigenome_single/Gene_cotyledon_H3K4me3_H3K27ac.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 整理，将H3K4me3 \n",
    "import pandas as pd\n",
    "df= pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/micro_state/epigenome_single/gene_cotyledon_H3K4me3_3K_100bp.csv\")\n",
    "df_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/expression_gene/J668_all_tissue_average_TPM.txt\", sep=\"\\t\")\n",
    "# df_gene.reset_index(inplace=True)\n",
    "# df_gene.rename(columns={'index':'GeneId'}, inplace=True)\n",
    "df_umap = zscore_umap(df)\n",
    "df_umap.reset_index(inplace=True)\n",
    "df_umap.columns = ['GeneId', 'umap1', 'umap2']\n",
    "df_result = pd.merge(df_umap, df_gene, how='inner', on=['GeneId'])\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/micro_state/epigenome_single/Gene_cotyledon_H3K4me3.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 整理，将H3K27ac\n",
    "import pandas as pd\n",
    "df= pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/micro_state/epigenome_single/gene_cotyledon_H3K27ac_3K_100bp.csv\")\n",
    "df_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/expression_gene/J668_all_tissue_average_TPM.txt\", sep=\"\\t\")\n",
    "df_umap = zscore_umap(df)\n",
    "df_umap.reset_index(inplace=True)\n",
    "df_umap.columns = ['GeneId', 'umap1', 'umap2']\n",
    "df_result = pd.merge(df_umap, df_gene, how='inner', on=['GeneId'])\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/micro_state/epigenome_single/Gene_cotyledon_H3K27ac.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 对于非编码区的功能调控元件\n",
    "## 将多个数据文件整理为一个文件，并且z-score 标准化，再用umap 来聚类分析\n",
    "\n",
    "# 1、数据合并\n",
    "#samples = ['hypocotyl', 'leaf', 'radicle', 'stigma', 'cotyledon']\n",
    "samples = ['hypocotyl']\n",
    "\n",
    "#samples = ['hypocotyl']\n",
    "chroms = ['Ghir_A01', 'Ghir_A02', 'Ghir_A03', 'Ghir_A04', 'Ghir_A05', 'Ghir_A06', 'Ghir_A07', 'Ghir_A08', 'Ghir_A09', 'Ghir_A10', 'Ghir_A11', 'Ghir_A12', 'Ghir_A13', \n",
    "          'Ghir_D01', 'Ghir_D02', 'Ghir_D03', 'Ghir_D04', 'Ghir_D05', 'Ghir_D06', 'Ghir_D07', 'Ghir_D08', 'Ghir_D09', 'Ghir_D10', 'Ghir_D11', 'Ghir_D12', 'Ghir_D13']\n",
    "df_micro_state = pd.DataFrame()\n",
    "for s in samples:\n",
    "    df_tmp = pd.DataFrame()\n",
    "    for c in chroms:\n",
    "        input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/5/{0}_micro_state_{1}.csv\".format(s, c)\n",
    "        df = pd.read_csv(input_f)\n",
    "        df_tmp = pd.concat([df_tmp, df], axis=0)\n",
    "        df_tmp.reset_index(drop=True, inplace=True)\n",
    "    if df_micro_state.shape[0] == 0:\n",
    "        df_micro_state = df_tmp.copy()\n",
    "    else:\n",
    "        df_micro_state = pd.merge(df_micro_state, df_tmp, how='inner', on=['GeneId'])\n",
    "        #df_micro_state = pd.merge(df_micro_state, df_tmp, how='inner', on=['GeneId'])\n",
    "#print(df_micro_state)\n",
    "df_micro_state.fillna(0, inplace=True)\n",
    "# z-score 标准化， umap 聚类\n",
    "#print(df_micro_state)\n",
    "col_list = []\n",
    "for s in samples:\n",
    "    for k in range(1,5):\n",
    "        for i in range(5):\n",
    "            col = \"{0}_R{1}-E{2}\".format(s, k, i)\n",
    "            col_list.append(col)\n",
    "\n",
    "def zscore_umap(df):\n",
    "    import scipy.stats as stats\n",
    "    from sklearn.preprocessing import StandardScaler\n",
    "    import umap.umap_ as umap\n",
    "    #z-score\n",
    "    id_list = df['GeneId'].tolist()\n",
    "    df.set_index('GeneId', inplace=True)\n",
    "    df_numpy = df.to_numpy()\n",
    "    # df_zscore = stats.zscore(df_numpy, axis=1)\n",
    "    # df_zscore = pd.DataFrame(df_zscore, columns=col_list)\n",
    "    df_zscore = pd.DataFrame(df_numpy, columns=col_list)\n",
    "    # umap\n",
    "    reducer = umap.UMAP()\n",
    "    scaled_penguin_data = StandardScaler().fit_transform(df_zscore)\n",
    "    embedding = reducer.fit_transform(scaled_penguin_data)\n",
    "    df_umap = pd.DataFrame(embedding, columns =['umap1', 'umap2'], index=id_list)\n",
    "    return df_umap\n",
    "df_umap = zscore_umap(df_micro_state)\n",
    "df_umap.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/PETRE_micro_state.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 将umap结果与平均信号结合起来\n",
    "import pandas as pd\n",
    "import scipy.stats as stats\n",
    "df_single = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/micro_state/epigenome_single/Gene_cotyledon_H3K4me3_Gene.csv\")\n",
    "df_umap = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/micro_state/epigenome_single/gene_cotyledon_H3K27ac_3K_100bp.csv\", index_col=0)\n",
    "df_umap['average'] = df_umap.median(axis=1)\n",
    "df_umap.reset_index(inplace=True)\n",
    "df_single = df_single[['GeneId', 'umap1', 'umap2']]\n",
    "df_umap = df_umap[['GeneId', 'average']].copy()\n",
    "df_result = pd.merge(df_single, df_umap, how='inner', on=['GeneId'])\n",
    "## 将小于0的和空值替换为0\n",
    "df_result_tmp = df_result.query(\"average<0\").copy()\n",
    "df_result_tmp['average'] = 0\n",
    "df_result = pd.concat([df_result, df_result_tmp], axis=0)\n",
    "df_result.fillna(0, inplace=True)\n",
    "df_result.drop_duplicates(keep='last', inplace=True, subset=['GeneId'])\n",
    "# 将数据z-score 标准化\n",
    "df_result['average_zscore'] = stats.zscore(df_result['average'], axis=0)\n",
    "# 将数据归一化到0-1之间\n",
    "max_value = df_result['average_zscore'].max()\n",
    "min_value = df_result['average_zscore'].min()\n",
    "df_result['normal'] = (df_result['average_zscore']-min_value)/(max_value-min_value)\n",
    "df_result['normal'].replace(0, 0.00001, inplace=True)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/micro_state/epigenome_single/gene_cotyledon_H3K4me3_3K_100bp_chip.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                GeneId      umap1     umap2\n",
      "0      Ghir_A01G000010   9.311453  5.451591\n",
      "1      Ghir_A01G000020  10.173335  8.722606\n",
      "2      Ghir_A01G000030   3.787272  7.875386\n",
      "3      Ghir_A01G000040  11.109644  3.293409\n",
      "4      Ghir_A01G000050   5.330428  2.857170\n",
      "...                ...        ...       ...\n",
      "68894  Ghir_D13G025640   0.587722  3.416919\n",
      "68895  Ghir_D13G025650   4.998712  2.426601\n",
      "68896  Ghir_D13G025660  10.901816  3.161812\n",
      "68897  Ghir_D13G025670  -1.615474  4.846691\n",
      "68898  Ghir_D13G025680  -0.376509  8.412956\n",
      "\n",
      "[68899 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "## 将hypocotyl 和stigma 结合起来， 分析TSG 是否具有H3K27ac 的偏好性\n",
    "import pandas as pd\n",
    "df_cotyledon = pd.read_csv('G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/micro_state/epigenome_single/cotyledon/gene_cotyledon_H3K4me3_3K_100bp.csv')\n",
    "df_stigma = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/micro_state/epigenome_single/stigma/gene_stigma_H3K4me3_3K_100bp.csv\")\n",
    "df = pd.merge(df_cotyledon, df_stigma, how='inner', on=['GeneId'])\n",
    "df.fillna(0, inplace=True)\n",
    "def zscore_umap(df):\n",
    "    import scipy.stats as stats\n",
    "    from sklearn.preprocessing import StandardScaler\n",
    "    import umap.umap_ as umap\n",
    "    #z-score\n",
    "    df.set_index('GeneId', inplace=True)\n",
    "    df = df.loc[~(df==0).all(axis=1), :]\n",
    "    id_list = df.index.to_list()\n",
    "    df_numpy = df.to_numpy()\n",
    "    df_zscore = stats.zscore(df_numpy, axis=1)\n",
    "    df_zscore = pd.DataFrame(df_zscore)\n",
    "    #df_zscore = pd.DataFrame(df_numpy)\n",
    "    # umap\n",
    "    reducer = umap.UMAP()\n",
    "    scaled_penguin_data = StandardScaler().fit_transform(df_zscore)\n",
    "    embedding = reducer.fit_transform(scaled_penguin_data)\n",
    "    df_umap = pd.DataFrame(embedding, columns =['umap1', 'umap2'], index=id_list)\n",
    "    return df_umap\n",
    "df_normal = zscore_umap(df)\n",
    "df_normal.reset_index(inplace=True)\n",
    "df_normal.rename(columns={'index': 'GeneId'}, inplace=True)\n",
    "print(df_normal)\n",
    "# ## 打上TSG 标签\n",
    "df_tsg = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/tau_gini/J668_tissue_special_geneid.txt\", sep=\"\\t\")\n",
    "df_tsg_cotyledon_stigam = df_tsg.query(\"tags=='cotyledon' | tags=='stigma'\")[['GeneId', 'tags']].copy()\n",
    "df_result = pd.merge(df_normal, df_tsg_cotyledon_stigam, how='inner', on=['GeneId'])\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/micro_state/TSG/cotyledon_hypocotyl_tsg.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.3 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "a0f669ea42a4dcea88d385f8f9b4c3d9b966e24234758831d1dbfdac925619a2"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
