{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "211724\n",
      "5477\n"
     ]
    }
   ],
   "source": [
    "## 分析 Fit-Hi-C, Mustache, Juicer, fanc 所鉴定的loop合集\n",
    "import pandas as pd\n",
    "# fithic_f=\"\"\n",
    "# mustache_f=\"\"\n",
    "# juicer_f=\"\"\n",
    "# fanc_f=\"\"\n",
    "# df_fithic = pd.read_csv(fithic_f, sep=\"\\t\", names=['chr1', 's1', 'e1', 'chr2', 's2', 'e2'])\n",
    "# df_mustache = pd.read_csv(mustache_f, sep=\"\\t\")\n",
    "# df_juicer = pd.read_csv(juicer_f, sep=\"\\t\")\n",
    "# df_fanc = pd.read_csv(fanc_f, sep=\"\\t\")\n",
    "# reaname columns\n",
    "def re_col(df, tags):\n",
    "    chr_id = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26]\n",
    "    chr_n = ['Ghir_A01', 'Ghir_A02', 'Ghir_A03', 'Ghir_A04', 'Ghir_A05', 'Ghir_A06', 'Ghir_A07', 'Ghir_A08', 'Ghir_A09', 'Ghir_A10', 'Ghir_A11', 'Ghir_A12', 'Ghir_A13', 'Ghir_D01', 'Ghir_D02', 'Ghir_D03', 'Ghir_D04', 'Ghir_D05', 'Ghir_D06', 'Ghir_D07', 'Ghir_D08', 'Ghir_D09', 'Ghir_D10', 'Ghir_D11', 'Ghir_D12', 'Ghir_D13']\n",
    "    if tags == 'mustache':\n",
    "        df_loops = df[['BIN1_CHR', 'BIN1_START', 'BIN1_END', 'BIN2_CHROMOSOME', 'BIN2_START', 'BIN2_END']].copy()\n",
    "        df_loops.columns = ['chr1', 's1', 'e1', 'chr2', 's2', 'e2']\n",
    "        df_loops['tags'] = [tags]*df_loops.shape[0]\n",
    "        df_loops['chr1'].replace(chr_id, chr_n , inplace=True)\n",
    "        df_loops['chr2'].replace(chr_id, chr_n , inplace=True)\n",
    "        #print(df_loops)\n",
    "        return df_loops\n",
    "    if tags == 'fithic':\n",
    "        df_loops = df.copy()\n",
    "        df_loops['tags'] =[tags]*df_loops.shape[0]\n",
    "        return df_loops\n",
    "    if tags == 'juicer':\n",
    "        df_loops = df[['#chr1', 'x1', 'x2', 'chr2', 'y1', 'y2']].copy()\n",
    "        df_loops.columns = ['chr1', 's1', 'e1', 'chr2', 's2', 'e2']\n",
    "        df_loops['tags'] =[tags]*df_loops.shape[0]\n",
    "        return df_loops\n",
    "    if tags == 'fanc':\n",
    "        df_loop = df.copy()\n",
    "        df_loops['tags'] =[tags]*df_loops.shape[0]\n",
    "        return df_loops\n",
    "# Data structure standardisation\n",
    "# df_fithic_loop = re_col(df_fithic, 'fithic')\n",
    "# df_mustache_loop = re_col(df_mustache, 'mustache')\n",
    "# df_juicer_loop = re_col(df_juicer, 'juicer')\n",
    "# df_fanc_loop = re_col(df_fanc, 'fanc')\n",
    "# df_all_loop = df_fithic_loop.concat([df_mustache_loop, df_fithic_loop, df_fanc_loop], axis=1)\n",
    "# choose_overlap_loops\n",
    "def choose_overlap_loops(df):\n",
    "    df_loop = df.copy()\n",
    "    df_uniq_loop = df_loop.drop_duplicates(keep=False, subset=['chr1','s1', 'e1','chr2','s2', 'e2'])\n",
    "    print(df_uniq_loop.shape[0])\n",
    "    df_all_loop = pd.concat([df_loop, df_uniq_loop], axis=0)\n",
    "    df_overlap_loop = df_all_loop.drop_duplicates(keep=False)\n",
    "    df_need_loop = df_overlap_loop[['chr1','s1', 'e1','chr2','s2', 'e2']].copy()\n",
    "    df_need_loop.drop_duplicates(keep='first', inplace=True)\n",
    "    return df_need_loop\n",
    "\n",
    "df_a = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/mustache/anthers.tsv\", sep=\"\\t\")\n",
    "df_b = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/fithic/anthers_15_00008_loop.bedpe\", sep=\"\\t\", names=['chr1', 's1', 'e1', 'chr2', 's2', 'e2'])\n",
    "df_c = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/juicer/enriched_pixels_3000.bedpe\", sep='\\t')\n",
    "df_a_loop = re_col(df_a, 'mustache')\n",
    "df_b_loop = re_col(df_b, 'fithic')\n",
    "df_c_loop = re_col(df_c, 'juicer')\n",
    "df_all_loop = pd.concat([df_a_loop, df_b_loop, df_c_loop], axis=0)\n",
    "df_true_loop = choose_overlap_loops(df_all_loop)\n",
    "print(df_true_loop.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "def choose_overlap_loops(df):\n",
    "    df_loop = df.copy()\n",
    "    #print(df_loop)\n",
    "    df_uniq_loop = df_loop.drop_duplicates(keep=False, subset=['chr1','start1', 'end1','chr2','start2', 'end2'])\n",
    "    #print(df_uniq_loop.shape[0])\n",
    "    df_all_loop = pd.concat([df_loop, df_uniq_loop], axis=0)\n",
    "    df_overlap_loop = df_all_loop.drop_duplicates(keep=False)\n",
    "    df_need_loop = df_overlap_loop[['chr1','start1', 'end1','chr2','start2', 'end2']].copy()\n",
    "    df_need_loop.drop_duplicates(keep='first', inplace=True)\n",
    "    return df_need_loop\n",
    "def overlap_loops(fithic, mustache,  chromosight, output_f):\n",
    "    chr_id = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26]\n",
    "    chr_idn = ['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10', 'chr11', 'chr12', 'chr13',\n",
    "               'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22', 'chr23', 'chr24', 'chr25', 'chr26']\n",
    "    chr_n = ['Ghir_A01', 'Ghir_A02', 'Ghir_A03', 'Ghir_A04', 'Ghir_A05', 'Ghir_A06', 'Ghir_A07', 'Ghir_A08',\n",
    "     'Ghir_A09', 'Ghir_A10', 'Ghir_A11', 'Ghir_A12', 'Ghir_A13', 'Ghir_D01', 'Ghir_D02', 'Ghir_D03', 'Ghir_D04',\n",
    "      'Ghir_D05', 'Ghir_D06', 'Ghir_D07', 'Ghir_D08', 'Ghir_D09', 'Ghir_D10', 'Ghir_D11', 'Ghir_D12', 'Ghir_D13']\n",
    "    # Fit-Hi-C\n",
    "    df_fithic = pd.read_csv(fithic, sep='\\t', names=['chr1', 'start1', 'end1', 'chr2', 'start2', 'end2'])\n",
    "    # Mustache\n",
    "    df_mustache = pd.read_csv(mustache, sep='\\t', names=['chr1', 'start1', 'end1', 'chr2', 'start2', 'end2'])\n",
    "    df_mustache['chr1'].replace(chr_id, chr_n, inplace=True)\n",
    "    df_mustache['chr2'].replace(chr_id, chr_n, inplace=True)\n",
    "    # Peakachu\n",
    "    df_peakachu = pd.read_csv()\n",
    "    \n",
    "    #df_juicer = pd.read_csv(juicer, sep='\\t', names=['chr1', 'start1', 'end1', 'chr2', 'start2', 'end2'])\n",
    "    df_chromosight = pd.read_csv(chromosight, sep='\\t', names=['chr1', 'start1', 'end1', 'chr2', 'start2', 'end2'])\n",
    "    #df_LASCA = pd.read_csv(LASCA, sep=\"\\t\", names=['chr1', 'start1', 'end1', 'chr2', 'start2', 'end2'])\n",
    "    df_fithic['tags'] = ['fithic']*df_fithic.shape[0]\n",
    "    #df_juicer['tags'] = ['juicer']*df_juicer.shape[0]\n",
    "    df_mustache['tags'] = ['mustache']*df_mustache.shape[0]\n",
    "    df_chromosight['tags'] = ['chromosight']*df_chromosight.shape[0]\n",
    "    #df_LASCA['tags'] = ['LASCA']*df_LASCA.shape[0]\n",
    "    df_all = pd.concat([df_fithic, df_mustache, df_chromosight], axis=0)\n",
    "   \n",
    "    df_overlap = choose_overlap_loops(df_all)\n",
    "    df_overlap.replace(chr_n, chr_id, inplace=True)\n",
    "    \n",
    "    ##将所有染色体名称替换\n",
    "    df_all.replace(chr_n, chr_id, inplace=True)\n",
    "    df_all.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/overlap_loop/{0}_all.csv\".format(t), index=False)\n",
    "    #print(df_overlap.shape[0])\n",
    "    df_overlap.to_csv(output_f, sep=\"\\t\", index=False, header=False)\n",
    "    return df_overlap\n",
    "\n",
    "tissues = ['root', 'leaf', 'anthers', 'cotyledon', 'hypocotyl', 'stem', 'stigma', 'petal', 'fiber_10DPA', 'fiber_20DPA', 'ovule', 'radicle']\n",
    "for t in tissues:\n",
    "    fithic_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/fithic/{0}_10_0001_loop.bedpe\".format(t)\n",
    "    mustache_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/mustache/{0}_mustache.bed\".format(t)\n",
    "    #juicer_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/juicer/{0}_juicer.bed\".format(t)\n",
    "    chromosight_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/chromosight/{0}_chromosight.bed\".format(t)\n",
    "    #LASCA_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/LASCA/{0}_LASCA.bed\".format(t)\n",
    "    output_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/overlap_loop/{0}_overlap.bed\".format(t)\n",
    "    df_overlap = overlap_loops(fithic_f, mustache_f, chromosight_f, output_f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "root 19977\n",
      "leaf 16314\n",
      "anthers 8215\n",
      "cotyledon 25229\n",
      "hypocotyl 9744\n",
      "stem 13236\n",
      "stigma 7842\n",
      "petal 16065\n",
      "fiber_10DPA 13992\n",
      "fiber_20DPA 12530\n",
      "ovule 19558\n",
      "radicle 26855\n"
     ]
    }
   ],
   "source": [
    "#统计各overlap的数量\n",
    "import pandas as pd\n",
    "tissues = ['root', 'leaf', 'anthers', 'cotyledon', 'hypocotyl', 'stem', 'stigma', 'petal', 'fiber_10DPA', 'fiber_20DPA', 'ovule', 'radicle']\n",
    "for t in tissues:\n",
    "    df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/overlap_loop/{0}_overlap.bed\".format(t), sep=\"\\t\")\n",
    "    print(t, df.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "118041\n",
      "163830\n",
      "139540\n",
      "180537\n",
      "154797\n",
      "165406\n",
      "175759\n",
      "157033\n",
      "160463\n",
      "156524\n",
      "142253\n",
      "115527\n"
     ]
    }
   ],
   "source": [
    "## 合并Fit-Hi-C，Mustache，Chromosight和Peakachu前25% 的loop，以及四种软件鉴定的重叠loops\n",
    "import pandas as pd\n",
    "def choose_overlap_loops(df):\n",
    "    df_loop = df.copy()\n",
    "    df_uniq_loop = df_loop.drop_duplicates(keep=False, subset=['chr1','start1', 'end1','chr2','start2', 'end2'])\n",
    "    df_all_loop = pd.concat([df_loop, df_uniq_loop], axis=0)\n",
    "    df_overlap_loop = df_all_loop.drop_duplicates(keep=False)\n",
    "    df_need_loop = df_overlap_loop[['chr1','start1', 'end1','chr2','start2', 'end2']].copy()\n",
    "    df_need_loop.drop_duplicates(keep='first', inplace=True)\n",
    "    df_need_loop.columns = ['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2']\n",
    "    return df_need_loop\n",
    "## 将overlap loop与前25% 的loops合并\n",
    "def concat_25precent_overlap_loops(df, tags):\n",
    "    chr_id = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26]\n",
    "    chr_idn = ['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10', 'chr11', 'chr12', 'chr13',\n",
    "               'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22', 'chr23', 'chr24', 'chr25', 'chr26']\n",
    "    chr_n = ['Ghir_A01', 'Ghir_A02', 'Ghir_A03', 'Ghir_A04', 'Ghir_A05', 'Ghir_A06', 'Ghir_A07', 'Ghir_A08',\n",
    "    'Ghir_A09', 'Ghir_A10', 'Ghir_A11', 'Ghir_A12', 'Ghir_A13', 'Ghir_D01', 'Ghir_D02', 'Ghir_D03', 'Ghir_D04',\n",
    "    'Ghir_D05', 'Ghir_D06', 'Ghir_D07', 'Ghir_D08', 'Ghir_D09', 'Ghir_D10', 'Ghir_D11', 'Ghir_D12', 'Ghir_D13']\n",
    "    input_25f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/final_loops/25_precent/{}_fithic_mustache_chromsight_peakchu_loop.bed\".format(tags)\n",
    "    df_25_precent = pd.read_csv(input_25f, sep=\"\\t\")\n",
    "    df_25_precent['chrom1'].replace(chr_id, chr_n , inplace=True)\n",
    "    df_25_precent['chrom2'].replace(chr_id, chr_n , inplace=True)\n",
    "    df_result = pd.concat([df, df_25_precent], axis=0)\n",
    "    df_result.drop_duplicates(keep='first', inplace=True, subset=['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2'])\n",
    "    return df_result\n",
    "def overlap_loops(fithic, mustache,  chromosight, peakachu, tags):\n",
    "    chr_id = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26]\n",
    "    chr_idn = ['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10', 'chr11', 'chr12', 'chr13',\n",
    "               'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22', 'chr23', 'chr24', 'chr25', 'chr26']\n",
    "    chr_n = ['Ghir_A01', 'Ghir_A02', 'Ghir_A03', 'Ghir_A04', 'Ghir_A05', 'Ghir_A06', 'Ghir_A07', 'Ghir_A08',\n",
    "     'Ghir_A09', 'Ghir_A10', 'Ghir_A11', 'Ghir_A12', 'Ghir_A13', 'Ghir_D01', 'Ghir_D02', 'Ghir_D03', 'Ghir_D04',\n",
    "      'Ghir_D05', 'Ghir_D06', 'Ghir_D07', 'Ghir_D08', 'Ghir_D09', 'Ghir_D10', 'Ghir_D11', 'Ghir_D12', 'Ghir_D13']\n",
    "    # Fit-Hi-C\n",
    "    df_fithic = pd.read_csv(fithic, sep='\\t', names=['chr1', 'start1', 'end1', 'chr2', 'start2', 'end2'])\n",
    "    # Mustache\n",
    "    df_mustache = pd.read_csv(mustache, sep='\\t', names=['chr1', 'start1', 'end1', 'chr2', 'start2', 'end2'])\n",
    "    df_mustache['chr1'].replace(chr_id, chr_n, inplace=True)\n",
    "    df_mustache['chr2'].replace(chr_id, chr_n, inplace=True)\n",
    "    # Peakachu\n",
    "    df_peakachu = pd.read_csv(peakachu, sep=\"\\t\", names=['chr1', 'start1', 'end1', 'chr2', 'start2', 'end2'])\n",
    "    df_peakachu.replace(chr_idn, chr_n, inplace=True)\n",
    "    # chromsight\n",
    "    df_chromosight = pd.read_csv(chromosight, sep='\\t', names=['chr1', 'start1', 'end1', 'chr2', 'start2', 'end2'])\n",
    "    # Add tags\n",
    "    df_fithic['tags'] = ['fithic']*df_fithic.shape[0]\n",
    "    df_peakachu['tags'] = ['peakachu']*df_peakachu.shape[0]\n",
    "    df_mustache['tags'] = ['mustache']*df_mustache.shape[0]\n",
    "    df_chromosight['tags'] = ['chromosight']*df_chromosight.shape[0]\n",
    "    # print(df_fithic.head())\n",
    "    # print(df_peakachu.head())\n",
    "    # print(df_mustache.head())\n",
    "    # print(df_chromosight.head())\n",
    "    #df_all = pd.concat([df_fithic, df_mustache, df_chromosight], axis=0)\n",
    "    df_all = pd.concat([df_fithic, df_mustache, df_chromosight, df_peakachu], axis=0)\n",
    "    ## find overlap loops\n",
    "    df_overlap = choose_overlap_loops(df_all)\n",
    "    #print(df_overlap.shape[0])\n",
    "    # concat loop\n",
    "    df_overlap = concat_25precent_overlap_loops(df_overlap, tags)\n",
    "    return df_overlap\n",
    "\n",
    "tissues = ['anthers', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'root', 'radicle', 'ovule', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "for t in tissues:\n",
    "    fithic_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/fithic/{0}_10_0001_loop.bedpe\".format(t)\n",
    "    mustache_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/mustache/{0}_mustache.bed\".format(t)\n",
    "    chromosight_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/chromosight/{0}_chromosight.bed\".format(t)\n",
    "    peakachu_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/peakachu/{0}_peakachu.bed\".format(t)\n",
    "    output_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/final_loops/{0}_final_loop.bed\".format(t)\n",
    "    df_overlap = overlap_loops(fithic_f, mustache_f, chromosight_f, peakachu_f, t)\n",
    "    print(df_overlap.shape[0])\n",
    "    df_overlap.to_csv(output_f, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将训练集合定义为3W条loops\n",
    "import pandas as pd\n",
    "tissues = ['root', 'leaf', 'anthers', 'cotyledon', 'hypocotyl', 'stem', 'stigma', 'petal', 'fiber_10DPA', 'fiber_20DPA', 'ovule', 'radicle']\n",
    "for t in tissues:\n",
    "    all_loop = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/overlap_loop/{0}_all.csv\".format(t)\n",
    "    over_loop = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/overlap_loop/{0}_overlap.bed\".format(t)\n",
    "    test_loop = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/overlap_loop/{0}_test.bed\".format(t)\n",
    "    df_all = pd.read_csv(all_loop)\n",
    "    df_over = pd.read_csv(over_loop, sep=\"\\t\", names=['chr1', 'start1', 'end1', 'chr2', 'start2', 'end2'])\n",
    "    #df_test = pd.read_csv(test_loop, sep=\"\\t\")\n",
    "    df_all = df_all[['chr1', 'start1', 'end1', 'chr2', 'start2', 'end2']].copy()\n",
    "    df_all_over = pd.concat([df_all, df_over], axis=0)\n",
    "    df_all_over.drop_duplicates(keep=False, inplace=True)\n",
    "    samples = 30000 - df_over.shape[0]\n",
    "    df_choose = df_all_over.sample(samples)\n",
    "    df_test = pd.concat([df_choose, df_over], axis=0)\n",
    "    df_test.sort_values(by=['chr1', 'start1', 'end1', 'chr2', 'start2', 'end2'], inplace=True)\n",
    "    df_test.to_csv(test_loop, sep=\"\\t\", header=False, index=False)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissues = ['root', 'anthers', 'cotyledon', 'hypocotyl']\n",
    "chr_id = ['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22', 'chr23', 'chr24', 'chr25', 'chr26']\n",
    "chr_n = ['Ghir_A01', 'Ghir_A02', 'Ghir_A03', 'Ghir_A04', 'Ghir_A05', 'Ghir_A06', 'Ghir_A07', 'Ghir_A08', 'Ghir_A09', 'Ghir_A10', 'Ghir_A11', 'Ghir_A12', 'Ghir_A13', 'Ghir_D01', 'Ghir_D02', 'Ghir_D03', 'Ghir_D04', 'Ghir_D05', 'Ghir_D06', 'Ghir_D07', 'Ghir_D08', 'Ghir_D09', 'Ghir_D10', 'Ghir_D11', 'Ghir_D12', 'Ghir_D13']\n",
    "for t in tissues:\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/peakachu/{}_peakechu.bed\".format(t)\n",
    "    output_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/peakachu/{}.bed\".format(t)\n",
    "    df = pd.read_csv(input_f, sep=\"\\t\", names=['chr1', 'start1', 'end1', 'chr2', 'start2', 'end2', 'a', 'b'])\n",
    "    df.replace(chr_id, chr_n , inplace=True)\n",
    "    df.to_csv(output_f, sep=\"\\t\", index=False, columns=['chr1', 'start1', 'end1', 'chr2', 'start2', 'end2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 给peakachu的结果加上互作reads数目\n",
    "import sys\n",
    "interaction_f = sys.argv[1]\n",
    "peakachu_f = sys.argv[2]\n",
    "output_f = sys.argv[3]\n",
    "\n",
    "df_interaction = pd.read_csv(interaction_f, sep=\"\\t\")\n",
    "df_peakachu = pd.read_csv(peakachu_f, sep=\"\\t\", names=['chr1', 'start1', 'end1', 'chr2', 'start2', 'end2', 'k', 'p-value'])\n",
    "chr_id = ['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22', 'chr23', 'chr24', 'chr25', 'chr26']\n",
    "chr_n = ['Ghir_A01', 'Ghir_A02', 'Ghir_A03', 'Ghir_A04', 'Ghir_A05', 'Ghir_A06', 'Ghir_A07', 'Ghir_A08', 'Ghir_A09', 'Ghir_A10', 'Ghir_A11', 'Ghir_A12', 'Ghir_A13', 'Ghir_D01', 'Ghir_D02', 'Ghir_D03', 'Ghir_D04', 'Ghir_D05', 'Ghir_D06', 'Ghir_D07', 'Ghir_D08', 'Ghir_D09', 'Ghir_D10', 'Ghir_D11', 'Ghir_D12', 'Ghir_D13']\n",
    "df_peakachu.replace(chr_id, chr_n, inplace=True)\n",
    "df_peakachu['fragmentMid1'] = df_peakachu['start1'] + 1500\n",
    "df_peakachu['fragmentMid2'] = df_peakachu['start2'] + 1500\n",
    "df_interaction['fragmentMid1'] = df_interaction['start1'] + 1500\n",
    "df_interaction['fragmentMid2'] = df_interaction['start2'] + 1500\n",
    "\n",
    "df_interaction_N = df_interaction[['chr1', 'fragmentMid1', 'chr2', 'fragmentMid2', 'contactCount']].copy()\n",
    "df_peakachu_N = df_peakachu[['chr1', 'fragmentMid1', 'chr2', 'fragmentMid2',  'p-value']].copy()\n",
    "df_result = pd.merge(df_peakachu_N, df_interaction_N, how='left', on=['chr1', 'fragmentMid1', 'chr2', 'fragmentMid2'])\n",
    "df_result.to_csv(output_f, sep='\\t', index=False)\n",
    "df2 = df_result.dropna()\n",
    "print(df_result.shape[0], df2.shape[0])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "tissues = ['root', 'anthers', 'cotyledon', 'hypocotyl', 'leaf', 'stem', 'stigma', 'petal']\n",
    "df = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/peakachu_hiACT/network/{}_final_network.txt\".format(t)\n",
    "    df_tmp = pd.read_csv(input_f, sep=\"\\t\", names=['source', 'target', 'weight'])\n",
    "    df = pd.concat([df, df_tmp], axis=0)\n",
    "df.drop_duplicates(keep='first', inplace=True)\n",
    "df_A01 = df.query(\"source.str.contains('Ghir_A01') and target.str.contains('Ghir_A01')\", engine='python').copy()\n",
    "df_A01.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/peakachu_hiACT/louvain_clusting/A01_loop.nx\", sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAV0AAADnCAYAAAC9roUQAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAgAElEQVR4nOydd3gUVdfAfzPb0nuDEAi9S5UqTUSKiEoRK4pgw977ay+v7bN3BSmKIEiTIgrSW2ihh56QhJC+yW62zdzvj4Ukm50NLdR3f8/jIzt35t67k50z5557iiSEwI8fP378nB/kCz0BP378+Plfwi90/fjx4+c84he6fvz48XMe8QtdP378+DmP+IWuHz9+/JxH9NU1xsTEiOTk5PM0FT9+/Pi5PNi4cWOeECJWq61aoZucnExKSsq5mZUfP378XKZIknTYV5vfvODHjx8/5xG/0PXjx4+f84hf6Prx48fPecQvdP348ePnPFLtRpofP5cSQgiOWDdSYD9MtKk+iUFtkSS/XuHn4sIvdP1cFuTbDjIz/SGcouz4EYkQfRzD631FkD6qRsYwl5Txy6wNLFm9h8AAA7fd2In+vVrUSN9+/neQqssy1rFjR+F3GfNzsaMIJz+kDUbB4dUWF9Cc4fW+Ousx8gst3P7IT5Ra7R7Hr2ieyFdv33rW/fu5vJAkaaMQoqNWm3/t5eei4KjFzBMr59JlxhdcM/t7pu3dinqKaUd3Fy/SFLgAubY9WF0FZz2/z35a4iVwAVJ3ZbJ4+c6z7t/P/w5+oesHoRZjK7gXS3Z9LNn1sGQ3xFb0AkJ4C5lzwba8bK764xv+OLiDo9ZS9pnzeXbtAoYtnISiqie9/ljZbp9tAoFdLT3rOa5K2e+zbcqsDWfdv5//HfxC9xLiWEY+K/5Yz9ZlO1FPQRidCkI4seZeh2L/CzjRpwul7BdsBaNqZIyT8dCK2biE9/fZkpfFn4d3nfT6mIBGPttkZMIMtc9qfiejzOY8p/37ubzwb6RdAjjsTt667XPWL9yCUN1Lbr1Bx+NfjaHfHT3Oqm/FthjUTM021bEBxbkDnaHlWY1RHfk2KxmlRZptApi4ZxND6lc/fpOwa1h97GtNE0ObyJvRSWf/M+/UJpnl6/dptl3To9lZ9+/nfwe/pnsJ8O1zU1i3YHO5wAVwORU+uv97tq30vbQ+FRRnChUablVUVMeWs+r/ZEjH//OFcgp2XZMuhJvqfYZJDkMq702idcRQusbdVxPT5PmH+2My6ryOhwSZuGVIxX5JdmEJL09dxD1fT2f80hRsTleNjO/n8sGv6V7kOB0uFo1f5lb7qiBUwY+vTOOTpf854/4lOQG32NMSbjKSLu6M+z4VogKCSAqN5HBJoWb7bY3bnlI/cQFNuafRLPLse3GpdmIDmqKXjV7n7czK4afVmxBCcHe39rROTDil/sNCApn944O899Ui1m46hAT07NKYh+/qRWhwAADT16Tyxox/yq/ZsP8IX/+1hrnPjSY+IuSUxvFz+eN3GbvIMeeXMLLew6iKtjYaGhnM71nfAHC0zMy0A5s4UJJP66jaDEtuS4QxkP0FBRTZbDSLiSHY6CmICm370BX0Qydp9C+FEhS/GUkynPJ8La4Clh/9jHTLBkCQGNSWnvGPEGas5fOaXQU5XD//Zy+7bvOIWOZcdzcG2VvDPBPG/TKbJXsOeBzr0age39859Kz7tjocdHnxS81XV9NaMfz+1J1nPYafS4fqXMb8mu5FTkhkMAHBJqzmMs32Ok3cwmxDbjpjVk7BoSioCOYf2cnH25YSWBRGYYnb1qmTZB7s1IknunYDIN+ez4s7PqCBoRX3x6R6/BgkOYTA6Omo6Fidu50dRYepGxxLn/g2BOpMmnMxO7KZcmC0h231kGUtGQc3MjL5W6JMyZrXNY+KZ/XQcbyzaQmrsg8TqDdwT/OO3N6kXY0J3Lmpu7wELsCKfYeZvnEbIzq0Pqv+f1mxRVPgAuzJzkMIgSRVZ0hxR9StWrSdX79ZQmFeKQ2b1+K+5weT1PDcrjb8nF/8QvciR5Zlxr49ks8eneBlAZD1Mne+MhRVCB5b+zs2xdN+6FQVHKGFiJIgQMIlVD5ft5bE0DBubtWKnw//jEN1sNsewzOZvWgbdIw6hlL22SLon/Q2iaIOdy1/DbPLWt7nJ7tn8kXHh2kSVsdrrstzPtfczFKEk2VHP+Omeh/7/J5xQSF8ctWQ07o3p8M3y9f7bPt+xYazFrqFFu2X4gmEgOpkrhCC956cyooFWzmx+EzJLWHjio955sNb6DP41Mwsfi5+/BtplwDXje3Lve/cit5QofUZAgw8/uUYOvRtzZ7iHIodGg/98V0qKdBTGH+wagUAu8wV7lgudKRYazGruDHb7bHMz1nEYylfeQhcAIdw8dimr9EyS6VbNvr8DlllqafyVWscp6JQYC3DXGbzeU6pXTuw4nQY3N63B0NYoAlZ1pa4hYUW8vJK2JOawcpFqVS9rULAx89Px2Y9+zn6uTjwa7qXCMMfH8TwxweRfTAXm9VGUpNa6A3uP58qhM9dfkkCyaQgyirssoU2twCSqvEbsCl20suKNdvKFDvbig5xRWT98mNCCFQF8GENkGvAbet0yC21cN+M2Ww/moPArV2oetC5vL0lujRIOuvxmteJp0ViHDszj3m1vTriGq9je3Zn8epLv5OfVwJAgMmAS5aQFe+/o6KobFq9l27XnDvXPT/nD7+me4lRq34s9VsmlQtcgGYR8ehk7T+lECAUz7ZAvfvaZqG+tbPmoVdUO49sW77H56nT1rFvdbyXpnaCJmF9q+2vJjHbbPT9bjzbjgtcOO4UZwCliuzXyzIvDexdI+NOfvQWbu7aGqNehwTEh4fw1ZgbuPaKJh7nZWUV8vD9E8jLLXH/fYQ7wEKJCEJoaMQSoDiVGpmjnwuPX9O9DNBJMo+16MWH25d4NwoQVs8/86g27QAYW38sT6U+hUt4mh8iDZFcm9CXr/au8Lk51DGqQpAoisr4SSsRuiYktMgjOMpWbr8UAoQ1lKsaP3jG3+90+WH9RqxOjSgxCWQjSIpby2+REMsnNw8mOiS4RsY16HS8MuwaXhnmrdlW5sP35mmaZ5DAFWzEUOIZfi3JEm27+Y6683Np4Re6lwn3NetOgd3ChL3rAAlZkkCAo1APxzVdCeiUWIenuncHIMwYxodXfMiEQxPYWbITGZku0V0YUWcEQfogusa0YHWedzKX5mF1iTaFlX/Ozy9FUVRUp4G/3upBUocs6nXOAgGH1yXCsZaYvj99P1UhBEdLZ3Co6FMcyjGMuliSIx4jIWR4tZ4Ai9O0I8dO3IQVz9xHTLCnoBVCsGzHAdakpVM/PorBHZoREqDtpXG27NmV7WNuElQJwJBkiTHPDCI0POiczMXP+ccvdC8jnm9zLWObdmNVzgEMso6eCY2wOpzM3b0bs93BNQ0b0DreMxgg3BjOY00e0+zvtdajeG3bRNbm7UIgAIm2EQ14r+1Yj/NCwwLK/y1UmfQNdUjfUOHd0KfXqQUgVCXD/CMHCz8C3Etrh3KMtPxXcCi51IsY5/O66OAg9uZrZxaTkAg1eQrT9Nwibv3kF0rKKjTM92Yu5b93DqJ/2yZVuzhrDAYZu49cQvUbxxNU28mxrCLq1I/h7icH0LJDco3Pwc+Fwy90LzNiAkK4oV6FPTbEYGJMB00f7ZNilPW80+YeihwWjtoKSAiIJMLorbEGBhi5uldz/l7qrRXLssSdt3U77bFV4eBQ4f9xQuBWauFQ0eckhY9FlrwjzgBGX9meDRmZmpuLN7Rshklf8bNXVcGoz3/zELju8QXPT55Pm+RaJESEnvb8q+PagVcwc7p2ZrJHnxpE6yvOfmPPz8WLX+j6qZZiRxm/7N3C0qx9xAWG8GirHjSN8HbWf/qJAeQVlLJla3r5MYNBxysvDKF+cuxpj1vmTEfgK3uXgtmWSkSg9svk6oYNGNm2NVO3bPPIydu+di3eHtDP49xNBzIpsmr72CqqYG7KTu69pvNpz7867h7Ti53719C09ypikoqwFAWwYW4LWtQd4he4/wP4ha4fnxwqKWDIop+wuCp8RBce2cPoJlfycntP4WUyGfi/92/l6NFitu04QnRUMO3a1jtpFJYvJKn6SLSiwgK+eHoi65ftRpIkegxozf0vDSE0IghJknjj2r7c3aEd83btwepwMrR1C5rExnj1k1ti0U47cZz8EqvvxjOkQPmH/o/+AUIFCYIjbAx+ZDUJwYEIMfCM75mfSwO/y5gfnzy5Zo6HwD3B+LQNbMo7onlNQkI4/fq2pH275LMSHoH6ukg+dQKJ54f/zaq/tuO0u3DYnPwzaxNjr/0Ae6XctuEhgrDEHZhj5jA7/zt2mXd49dSyTrzPSDFZkujcuO4ZfwctFNXG1rz/AGq5w7B0PIjlqHUpebZ1NTqen4sPv9C9SJlxIJXOf3xKo6nv0mraB7y18W/syvlLE1jmcpJa4GOXHfh8+8pzOr4k6agX8QjeP1GJvF3tKc71vsZcaOGP8e5ouwxLOi9vf4bV+SvIc+Syt3QPn6R9wJ9Zsz2uqRsbwVXN6nt3BiRGh9GjuXbbmVJg24wQvnxuVTLMc2t0PD8XH36hexHy3y1LeWbdPHJtFlQhsCpOfkpbz4jFE7X9O88Bbm8F32Pl2yznfA51wx+gYdQL6KRQQIdMIElh9zH9Nd8eBUtmu0ORv9z/CUoV4SYQzMn6A6vLc+4fjx7MbVe1xaBzPw46WaJ/myZMf+oO9LoafkSq0f5DJDuRrt/JyEwiI7MBefmPoCgVQSg2q4MVC1OZ98sa8nO0owX9XPz4bbo1xMGcAvbn5JMUHUHTxNPfODpBgd3KD7vXarbtLMphdc4huifUrPalRZDeSFJwBOkW7aoO19SpeVeqqkiSRJ2wu0gMvROXWoJeDkaS9EjSGz6vkWUZRSgUOPI12wWClML19IztU37MoNPx/NA+PD+0j+Y1NUlUQHskyeCl7UbKVurqLZVClF2U2X7HnrOUWgkrWbHgEB88Ow3F5U5/+eXrs2ndsT5vjx+Dweh/jC8l/JruWWK1Oxj12VRuen8iT06Yx4iPJtPj5a84kKP90J+MtTmHfVbBVYVg3bF0zbZzwefdbtLMzhCiN3Fn4w7nbR6SJGPQhSMdz9/Qc5DvEOX+IzphV3wntwF3XokLhU4y0j7uPSo/ehKCJL1FUwlWRQGZGV96CNwTbEs5yFuPTjnHM4YvNq2hw6QvaTH+E2778zfSi7UTzvs5NfxC9yx5+ddFbDmU7SEoi612bv5oCmWO0y9YGKg3VJuIJsIYeEbzPBNaRdfizwFjaR4Rh4yEDoleCQ1YMHAskaYLFyE16okBhIR534fYWhFcf0c3AnVBGKpJvN4pqsu5nB77c/P5aPFK3vjzHzane9efqxV8NX3qzCI+qDcmOYYoQwKyT28NQXHxbC+Be4INy3ZjLjxzU4+qClTVtxlp6OwpfLhxJfk2K1aXk9VZ6fSe/iO7C7wT+/g5NfzrkrPAaneyZJt2aW6HS2H66lRG9T49jbBbfDJGnc4rNy6ADonr67U4o7meKU0j4pg3YOzJTzyPBIcG8MPiZ/n544UsX7AVWZbpN6wjdzzSrzz95dDEm/ntiLcWeEV4WyKMEedkXqoqeHbmAv7cvqfcGv7LhlQaxUYzdcxIj7DiEGMynRI+A8Dp3MPRY77zNZSafb9AJCAns5CwSO/8EWVOJ3aXQniAycuTZG/aUd59fRYZh/ORJOhwZQMef3YQ8Qnh5edsyD7CpmNZ3t9TCB5dMo+/ht/jc15+fOMXumdBqc1+fMNJmw37Mk5b6Jp0er7vOYLR//7mUb5GAj7ocj2xgdXnMEi3pLLHvJJQfTQtI/oSavD2Tb0cCI8K5tG3hvHoW8M02/smXItJZ+L3I1OxKlb0kp6r4/oxtM7N52xOs1N3Mn9HmtcvYl9uPk/PXMA3t92oeZ1e3wSdLgFF0XLDk7GVDAdyfI4bVzvS43OexcLDM+ex6UgWAvfG4C1tW/Ofa69GliS2bj7M049MLj9fCEhZf4AHR//ApOkPERziDuuetMt3UdK9hWdmPvPjF7pnRXRoEHpZxumjflnDhDMTeN0T6rPupkf5OS2FLflZNA6LYVzLbtUu6e2KhfEHxmF2Viz7VuROpHf8GK6MPvsaYJciV8X2ontMT1zChV7Sn/Ogg/FrNvm0x6/Yd4gSm51QjSQ6kiQRGz2ZnNxBCFE5GEMiKGgEnXo8SGDwu5RZvG3R3fu3IjyqQst1qSo3jv+FnJLS8mOKKpiyKZVjpRa+HHo9b7w8Q3OOpaU2Fv65lWEj3RF4Jp3vABV/AMeZ47fpngU6WfapyeplmZHdq89JWx2RpiAeb92TCb1v4aX215zUhjr18AseAhdAoPJvzo/k2g6d8TwudSRJwiAbzouQKPYRTgzuRDu+wo0BDIam1E7YRkT4mwSY+hIUeDPxsX8THfkJQSEBfLfgSZIaVnjFnIjCe/r9kR79LNt/kNxSbRvv32n72ZJ2hJIS7Y1GIWBTyqHyzw+08R3+3DE+0Webn+rxa7pnyWPXXUVRaRkz120/nocLDHodH901mFqRYSe7vEYoceaSY9NOZyhQSS1aQN+E85fP9mwoKlvDwYJXsbkOI0km4oKHUS/yaXRyzeS8PZf0alyfaZu2a7YZ9ToSwn0nzilyZGJx5REXdDuhId429Jj4cL6b/xQ2q4Pc7CKi4sIIDg3wOi8tN9+ntg2Qmp2DLEkoPsxi8fEVv9mGEVGMbNKa39K2eZwToNPz9TXnrp7d5Y5f6NYAr47sxzM39iJl/xH0Oh0dGyZiPJ7JqrC0jMWpaZTaHHRuXJeWSfE1Pn6JMw8Jyad92eLU9rW92Mgpmcn+gmc4EZQhhJOjpT9TUPYP7Wr/hU72FjIXAkUopFszMEh6EgMTy7Xocb26MDt1F3aXpw+uLEk80bc7Bo3leqE9nT/SH8Wmngh2kGgadi1XJzynqZ0HBBmrrQ5cJ9z3i16SJBrVjSUo2EiJ2VvblSQYdoundvvfXgMY0qg5/7dxFUX2Mq6p14jH23cnQO8XHWeK/87VEEEmY3nI6ImH5Z/UvTw7eQEuRS3XPurHRTH1iVsJMmmnJTwTokzelXkr0yisa42Nda5QhYODhS+jFQXnUDLJs8wmPnSk94U1gNlZioxEiKF6bdqluvjp4ARW568tf8EZJSP3NRjLldEdSAgPZcHDd/P0zAVsyXC7EYYHBvD8tT25qZ13fTNFdTHt0FgUj2xqgj3mRQTrY+kSO+a0v8u1TRsRoNdjc3l7v4SaTHRJTuLZl4bwxsszcFYpAXTP/X1IrBPlccxcbGXTxG245h0kSFEpuEKm6LUrSKiyeefn1JGqCyvt2LGjSElJOY/TOT8IIcizWQnU6wkxnH11gBXph3h+yV9klZYgA72TG/BSl56MeH8KDpd3nH3dmAjmvXB3jdoZF2V9xtaiBV7HQ/Wx3Nf4J3TnuTDkyXA6tmG1/groCQq6DZuws+3oCEA7v0SYqSutEmo2EGC3eR8f7fkWs8tdHFIn6RiaOJBhda7z+tsIIXhj59scsBzU7Ovppk/QOrzVaY2fWjCDlblfaLbpJBP3N1l4Wv2dYM+xPG6eONWjZFGQQc+0UbfSNM69ubt/bw6/TVnNnl3Z1EmKYuyDV1O/igZtK3MwdugX5B71DDmWJIlPJ46laavqX/b/y0iStFEIoZl79OJ6Es8DP+3cwPublmNTXEhA+9hEvu1zEzGBZ2YzXLAvjXELK5KUqMCSQwfYcOQIkqqd2CQjr4iNBzLp2LDmfrT9aj2ESQ4mpWAWKi4kZBqEdGRw4vMXlcAVwkFB7nCczoqXeZnlB3TGroCKjEBGoCAhKgWJGOSa9a3NKsvhjR0fo1DheaIIhelH5iFLMkPrDPI4P610r0+BC/D9gZ/4rN3HpzWHozbvpO8VcznzqLmmcTFsffphlu7dz5aso7RMiKNPowYe5o2GjeN58bWbqu3n73lbycsxex0XQvDak1P59a+nmZKyhfHrNlHmdHF14wY827eHpoeGnwounqfxPPD9jvW8nbK0/LMANuZmMnDOT6y7+WF3XbHTQAjBs/8s0mwrcTkwBoGx1NtBRACbD2bVqNCVJR29E8bQK340TuHAIHk7xF8MmIvf9hC4J1Aca4nX6VAqFcm0C5kSYQB0JITdUaPzmJW5wEPgVmbmkfkMqX0terni8Ugt2qZ57gnMTjMWl5Vg/alH6sWaGrGvRKOYKG5N92TkFZQy/tdVrFy3D4NBx/X923Dn8M7IxytD92nckD6NG57yfKqyeukunwmWCvNLGfrdFLYfq/CY+W3zNmZv28myR8YSGeyv6eaLy9JlrNRlJqVgBZsL1+BQ3flghRB8sGm55vm5Niv/Zh447XHSzcVYnN75Zk+g+ojY1ckS4UHnZlNIkmSMcsBFKXAByiyTfLQIAnEhSZT/Z5JUImU7ccG3EB5Qs3bpfaWHfLapQiXf4Zlf4FRWC/qTJF6vSuvIoUhoX9M2qvogjpxcM7c/+ANz/0qlsNjKsbwSfpyyklvu/x6XD7/x00Ur1PoEZfF6D4F7AptL4Zk52oqIHzeXldBVhcrU9G95Zfv9TEn/komHP+X51LuYn/UbWZYSHD6W+wBrsw+f9ng6Sa5WuMUFa5ss9DrdOSl4eLHj1pp8J6OpeiclCfQSxJia1vhc4gzx+NrOEECo3jPyr1fsVdX21ySkMSbd6S2r9bKJoXW/QC95CrfGodfQOab6ENtvfl5Gmc07t8fRY2Ym/LrqtObhi2F3+H7RlTbxrTSsPZRRI+NfrlxWQnf5sQWsK1hK5R1wgWDxsVkcsG6qJo0MJIedfDfW4rKw27yDzLIMhBDUCQujVoi276UE/HbXrXRqVIfjhQHQ62QCjXo+u2fIOdN0L2YkSUKnS9ZsE8LXFhoUlnx5VuNuK8hi5JLxdJrzIcP/+ZFNeRmk7QvWFLpCQJKhIUF6T0EYbYrmqujumv0bJAP3N7z3jOYWH9iMexv/ybC6XzGw9luMbTSPfrVfOul1q1O0c34AzF2cekZzqUrTVnUYfPOVXsf1eh3JjX27Pp6fjM+XLpeVTffvnD98tAiW5c2kcXgb0orzvFr1ksSIxr6jx+yKjY/2vEN6mXsjRUIi2hjD402e59tBQxj2+6/YFU8t+s3e11AnPJwfx40gq8DMun0ZBJsM9Ghen0Cj7wQmlzthEe9RmH8baNhTfW0dKcJ7M+dUmXlwC89vrNjoLHKUccu/EyDPRFh+AvWbHvVIqVhqDmBXTjRo7Dvf2/AekoLq8EfmbGyqDT16Oka15576d5+2llsZSZKID2x+WtfoZN/6kkvDY8apKkzfl8pPuzdgU5z0T2rKM217EaCv/rf4yAuD6dO/NT99vpjco2aatkrk9vt6cxgr9/02S/OarsnuvYrdxSnMyfoOs7MAg2yka/R19Eu47aI1fZ0vLiuXsSe33OozQMAgGXms4RcMmPMTJc6Kx1snSXzdezDt4msTYwzXdBV6ZfvT5Dm87VcRhkjeaf0JhTYbU7ZtZU1mBvUjInisUzfig6tPTHOpYLPakWUJY0DN+RXbbCswFz6BqmYBMnp9C4qce3xW/w0OGETtmB9OexwhBK1mvoNTaNg4BagHgwCJsKhSDAaFkuJAHDYTCSEhrBxXveaqChVZunALxU+++5uZf27WbBvUtxXPPzqw/LOiqoz4axKb8zwzhgXqDKy4aRzRAWe26dXz3W846jqeK0KS3MsEFR6t05arhuqYnvGp1zUNgq9gTMPXzmi8S4n/GZexKGMs+RrCESDGlEBiSBgbRz7Cn4d3s/5oBpKpgO3WzXx08Ds4COGGYF5sfhdtIyvsrfstezUFLoDZWcze0t00DW3Bo5268igXfxDCqfLb+7OY+Pp0HGXujcLGHRrwn+lPkZDs6cu5qzibt7bOJ82cQ6g+gHsad+O2Bp2qFUgBAT0IqJWCzeVkc34mBllHg/DV5Bc/oXG2jpiIky+3tdhemK0tcMsnooJNj7nA00TUo369k/Z9IQUuwLjRvVm6cg+FxZ7VioMCjYy53dP+/M+RfWzJ807RWKY4eXDZDKb1v/O0xz94MBfdv8WEJchYEmWEThBQIAjfrzBL2kh+J+0XwgFLKvm2bKIDap32mJcLl5VNd2ji3ZrHJSSG1L4dAKNOx00NWtK3QRQbS9djVyu0q2KnhRe3fcthy9HyY4csvm1nKqrPsjCXMj88P5kfnp9SLnAB9m48wAPtnsFaUpG0ZWn2boYv/ZbNBRlYXA6O2sy8s20h966efNJabp+mLqfVtA+5459fuGXxZK5emEGJ9CJQoXUZdPWpG/c3Rv2ZlSeq1gVQ8t64Q4AOGN223RmNdz4xGvTMHP8gt9x0JVERwYSFBjCobysmfzmG2GjPl8iv+7b4tLOm5B6pNleDLxb9vR0EhGarJKS4qLXOReReBVkFIak4VN8bppuL/j3t8S4nLitNt0V4e25PeojpR37Acdy53CiZuDnpXpqFtSk/TwjBV/u07b+KUPg1/S+ebz4KgHBDhM+8BhISSYEn14rOF/m2g5QpxcQFNsUon1mFibLSMqZ/pF2R1mK2snjSMm4YNwBVqDyTMkPzYV5z7ABXPfIhumMySbWjePa+frRpUeGT/M6mf/hhd0WpcYHA7HRwyzInK2/cTEyAjCQFIJ2mC1ZVWkQkYJR1ml4reknm5xtv5am5C8gqLQEBpgKFuBQbj8z4mhfeGkrvfhURZvsK83lmyQJSc915bVvFxPN/1wyiQUSUV9/nC51OZtzdvRl3d+9qz5OrefcIQBFqNZUrtLFopJn06LQagnS+E//4otCRzzFbNjGmeKJNZ16D8GLgshK6AB2je9AxugdWVwkO1UmYIcJrKWh2WbBWU0drj7miDlmbiA4YZSN21ftHVisgkTpBdWtu8mfIEcsW/jzyEk5xQguVaB1xAz3jHzntTYv9W6txnROwY9Vubhg3gH3mXMoU3+WILLXtBB8N4NCRfB5+9Tc+fGkYndsmU2Qv48dKArcyLgQ/7F7Hi+19V1E4gYXjlbMAACAASURBVLnMxgdzl7EodS8uVaVZ7VheG9aPJrUqchhLksS7HYfw1HrvF+wb7QfRMSmRhDlmAqqUuxHAe6/8QbsrGxAeEcSBogIGTvsZp1phqtiae5R+U8ezaOTdNIqMPul8LyR3Ne3IkkztFVticDgG+fRfbl07N2LBom0oGj7BRr2BGFMieQ7vUkUSEldG9z/lcRyqg+/3/x97SraVKz4RhmiebPoakcYL98I7Gy4r80JlgvShRBijNG1vgTpTtXXI6gZXuMMYZSNPNnmJwCqaY1JgXZ5r9mrNTfgMsTjzmZXxVCWBCyDYVjSLrQXayaqrIyy6Gi1Er8dlCmTvjszjd8/3PZSUijZVCN772p1HYOXRg9UqQmtyvIW+KgRKJYFnLrMx6L/j+SNlJ1aHE4dLITX9KMM/mcymg54P+vV1WzH3mvvoHFuPWFMwHaPr8kffsQyv344De3MwV7GJlo+pCpYsdEeh/Xftcg+BewJFCJ771zMQQAiBcGxG2BYhFG876oWgZ+0GdIjxzn+rkyQ+7HrdGfXZuVMDmjet5VVMU5LgnbeGcU+D19BL3puvN9V5CIN86t47Uw59y+6SVI+VZpEzn7d3PoNSqaLyvsPHeOX/5nL/K7/y7dSV5BaUanV3UXDJarp2xzbMll9RXFkEBvQhNPjO8vDHk2GUDVwV04bleRXG/hDZQZTexjFnMKOSB3qcXy+4Ph+1/Ya9pXsocRbTKKQpEcaLI8vSurzx+FrPbcifSNvo4afVX91midSqH0fmvgq7NjodcnISksHA6nXprL75KyJigwl+ykCpoh2RZ9rv+WDlFliwlDlOGmpdufBmTkkpz85ZwLpDR1CBTnUTeW1gXxZt3kOR1XulogrBC1MXsugFz+xcTSPimdRrlNf5llJ7tbbnouMa8JpM387+W3KyK8Z3pUPBbaDmwfHsysLYEynyq/JKxheK3weM4tsda/lh1zpKnQ6aR8bxasd+tImpfUb96XQyH71/C3PmbWHmrBSsZQ5at6rDYw/1IybG/eL+T8vJrMydw37LNiINMVwTfxuhx5+bUqedIJ2h2mfWrtjYUrReu021sT5/JV2ie/LmlwtZuLwij8W2PVlMnbeRz/4zgtZNzuz7nUsuSaGbX/QOxaVfc8Kd3mpfRIH5dWrHLsBkPLXopSeajiTLlkuudR+v11lJkqnELbokGYMSiRCvIVXSkmVJpmno6flSng9ybHt8ttmUUn7esplRbdqelpnhvb9e4aErn8Oc79YW5Pr1kPQ6Kqs1RbkWYqcEYL1FQj0h9I//z7RTj65KzglJcify7lGrATpJRvHhVXB3U7czfm6phT5f/Iirkoa5Pj2ToT9MoZbkOzlRdlEJ5jIbYYEnDz5p3LwWsixrLpElCdp2dG/ghRqNmB3aNswTS3NVVSF/CHiU2xHg+BdR+ChS1Fcnnc+55v6WXbi/Zc1VQjYa9Qwf2pHhQzU9o9DJenrFD6UXFeWiXli1iKlpqeWbd10Skpjc/2bNXMMWV0m1NQjTSnZQsjuORSu8EwfZHS7+88k8Zn5570XnF3zJmRfsjm0Ul35D1fglIcrIyr0eIU6t7HmQPoAv2j/Oj41WkGQqQZLcGw4yKor1ZxwlH56D2dc80aYGPtvKHAZeX7aUO2b+flp9JiTHMe3oD7w17wV6j74GyaDHax0JqDvsfBx1AzfUuYJ6wdE0lGMIXRhI0GZvgde9QwMMBh2hBhNvXTlAc9x20bXok9gIgNG/zPAQuCewKQpFJ8nApfUAaxEYaOS2MT002+o1iKNtx2QAxrX3XbbmxibHX8RlE6oI3Eo4/kaoF+9y93zxzIoF/LJnq4e3xNqjGQycPUHz/HBjZLVmwHhTLSbNWu8znLugyEp6dqF24wXkkhO6JdYZ4MOJXggrVpt21iYtXNZpyJRqyRNclu8RwneuhouFzjGj0LKtCgEr9jYDYM2RDDZl+7YvulwKBQUW7PaK+6rT6eg8qD2tr/YdqSdUgchz8W7HoSzo9wh/DB5Hp1rJXufFRYfw0sMVJpuRjdoyq//dtIupTbDeSEJgKC+178vUfqOQJYniMht7c92ueJJOIapeATGN8jCFuk0KDln1+Si2S659WhF/d47txZMvX09YRBCSBAajjoE3tOOLCWPLNaRbW7ShT13vl1vdiGAe73RcIJf5TvIiBPyzcjy5hf+7gldVVX7fq52pbW9RPvuLvV0vdZKeztE9Na+Rkeka25viEh8vOtxeMUK9+IKSLz3zgvCd1QtUXEp2Ne2euOx/V9PqAFEKUvgp93chCDcmMijxDRZlvoHruJYvgPUHGrIjs8KdbcKWzbSv5Wnf2r8/h9ffnMWRTLc2IAGdOzfkpReGEBzsDmutXS8GWZZQNX68kiTRsHlFn3qdzGev3szmHRnMXLQFh8NF/54tuLpbU68l3hXRtZlx7d2a36mwzL0pGF67mOgGFZpKWEIp9lIDlt316dw0mZV7DnlcFxpg4v3bBmFzufhk3WoW7ksjyGDgvvZXcmOzFj7v4YAh7RgwxLdvrixJjL9uKNtyj/LD1hTMSiGlwWmUKJncveE14gOi+bK+THVxXYvWpfHiT99xbZemvHn/oItuyXuuySmz+Eik6Wb5kUM0DPf2Arm17ljMzmJ2mCv2X2R03NvgCcINkbRvWZe/Vu7W7DMkyES9xIvPw+GSE7rBgQMxW35GK3Yf9JgMrU+5L0lXXT5bPUin709YEyiqysHMfIwGPUnxESd9QBuEduf+pvPp8fNr6GUXGYXRVF3E6I73IYTg12VbmPR3CkfzStApCsEG0Dndwnrtuv08/Ngkxv/gLo7YrmtDIqJDKMgt8Ro3uXE8jZp7CnJJkmjfqi7tW525K13t8DBMQQrRDQq9ViGmECe2hpkszCmDhmBEJllEcGfLNozs2gaby0XHH746nnLTffETixcwbed2fhlafbrEk9E6NoFnu3Vi7IZ3UZWKl9BRWz6fHlF4vra2P4dLldiw1+098NfaPTRPjuf2Adp20MuVqIDq/cabRmr73kqSxAONnsbsKGaXeSsBukCah7fBKLs9I+4Z3pXl6/dhc3iaG2VJ4s3HB1+UL7dLTugGmK7CZOyM3bGmSouEyXgFJmP7U+7LGHw3ZdZJVBXgQkCZbgjBFyDU86vfVjD5zxRU1V03IToimP97diiN63r+KDcVHOLTXQvZXeKuxVUnMJKwkCT2ZGovYe/rcCWKqnLrf6eQlnk86Y9ewhWmozhUR1i6E73dLUgOH85j1+4smjerjU4n89lvD/LivRPYW5hHUTMDznCZBqZw3npudLXfxe0Z4ESq4jokhGDDgUwKLBauapxMSKBnshijTkenDjJHNFaGkgRhERayc9wajAOVfVIBxcEOjpQVcPuKrzHWdWAEEGAtCMReFMCazAz+PXyQ3vXOLLrtBB+nTa3YOKzEypIE9pZF0ySwYpksBKgC3v69N1ZHxT0YP3fd/5zQNen0tIqOZ3t+jldbiMFIt9p1KXHY+St9L8UOG53jk2gZXeG6GWYMp3OMt6mhbu0ofnj3dj7+aQlbdh1BkiSaJMfyykMDSa5zcfpPX5IJb4RQKDR/SHHpdwhhBfSEBI8kJvwNZPn0knc4LJNwmP8DlXbTNxxL5N4Vw/hv3wEMa+FdUPBc8c20lUyY4+0iYzTomP/lA4QEuYXT39nbeWnrdJwakVb23CDKijw3sgY0asxX113P7yu38dZUbZOKbFOJOFxh0711ZGfuHdun/POP21N4Z91SXJUEjgS81KkP97b2TP8nhILD/DGq9TvcG546ZNO1GCP+y9r9BTz082yPirnXt2vGuzcP8NBKntv6OanFPsrKC9i9O8njmMkkCI0vQhWee35CgN1sxJobQvc6dZl80wjNPk+Vwcufxim0k1DqUPm9XXsCbL9QVJLD7sxoPvuzKwdyPB9+SYJ14588q3mcDwqOmVk8fR1Zh/Pp0LMpVw1qc8pumVpYnA6u/eMnjpRWZI0L1huYO+Qu0oqO8fCyOSjH5ZEEdK9Vjwn9bkZ/FmNeKC67hDeSpCMq/Dmiwp87vtlVfTLx6jAG38nw2fm0Ct9EmNHOwoxGpJe6tajn/1nEgEaNCTbWXIYtX7hcCpP/1H7BOZwKC1bsYET/9qhC5b0dczUFLkBgrJXrE9uz8nAGYSYTD3fqwqAm7gQ+v/y7yef4qkk67lnqJiSkYjmYbi7infX/eghccJsj3l6/lGZRsfRITC4/bi94FOGYX+lMBdW+gLKc1Tw+6VbsLs/7OXfzburGRDCub0XCoK7RrTWFrhDgdHp7JwRGFbvnr+GsbwpzYCtUcFWX/OY4Nqudpb+vY+uKPcQmRjHwrh7Url85yY9vJUVFxmq6kaCwu/nkj/ksXKNta4yPvDBmq9NhzeJtvPXA+HIzyl/T1hEaEcT3S14gPOrMMugFG4ysuvkBUvOOsiLzEC2iYumT1JACm5Vx/872uLMCWJl9mC+2rubxdtUnkL/UuPReIVWQJN1Z2W3Mdjvb8xSm7m/Dd7s6lQtccEcc/X3Qd8KbmiTzWLFH1FVVdh10ZzrLtBZicfl2mVKBm9o2ZuWYe5l/x6hygQtU239VBg+uyFUx/9AenwJLAB+kVJRBUl05VQRu5ZOLuaWttuAfv2yj5/i1exAgmTxk3IlFWVZWlc0RSSDrq9+l1ge46BtfvWmhKNfMqDbP8+njk/h3xnqmf7aQ+7r8hxWzK16GHSJ9+2pHGkKJNro3Xh+5uadmzltJgidu713tPC40ZRY7bz84wcNuDVBSZOXp4Z+fdf9XxCTwUJsu9Ely12/7ePMKn6+yn3ZefCvts+WSF7onQwjBqiOHefLv+Tzy11xWZXiGmeZaqnfjKbb5ztFQk4SFBFTrk9i8vtu+FaQ3+QgsEOhlFwF6B69vn8yHO2dgdnq60wzp4sNUIgQ6q9sNS5Jg1B3dCa2k6doVXzUd3GSUVJToVmyzfZ4nSXBjyx2abWUOTzfAolwL0huhiAMyQj1eWcIlk5ERQ1mZp/nkVF65chn8+vlKMo8W+TznpeGflAeEnMDlVHh3zHfYrO4X3V11rgOX5OEbKgSgwoNJI8oVgNjIEKa8eQe1Yyu8XwJNBl4efS19OjQG4NCODCa8+hvzvl2MolwY90QhnJiL3yc7qxlZmXXJOdqTrWumori0X7JHDhyj4NiZJ5XXYndhrs82i6s6b6VLk0vSvHCqOBWFO+ZOZ33WkfI36dx9e2gSFc2c4XcQoDdQOzQMnSSV25KqMrBR43M2P6vTgSIEoUYTkWFBtGyUwLa93i5vep3M9b3dGa+iTSG0CE8ktahyaKrApHchSwJJcif0mZW1jllZ63ij9e1cHe/WWm/t1ZbfV6SSXejpiSBLEnXKDCS3imfUnT3o0D7Zo/3qpIZ8vMl33S3PTFvVCw+drH2fQ6tspk3+dgnWXCd8WrEUz+9kxNLe5CVlhZAI0wdSopShiYDIpQYcisqUP9Yz+o5uzN+Zhkmv48YrWhBgMFBabGX/Nu1wX1UV/DlhGcPGXcvn366mZH8CAd0L0NexgwRKvh7bsmim1dpNz5cr/JobJMYw64Mx2B0urDYH4SGBx93vVB7p+hJpKRXFUD9/5Cde/f1Jug3xLo9zLsnPuwVHpU1pRdlHUrMX6NCnKxuXanug5BwpICourMbm0Do6gZRj3slxAOICL49iAJW5rIXuLztTPQTuCdIK8nlu6SI+7TeYQIOBW1u1YfK2LV7X90muT+w5qACx+WgW4/6aQ/ZxLTspNIwvrx3C+0/cwJhXfyU7r7hckwow6fnmpZEEmCoc/t9rN5IRyz/Hori1L50kygVuVf6zbQqNQmpTNziWIJORGS+PYvKSTfyxejuKEPS5oiFP3NCDAJMBVVXZt/0IRw4co06DCjtm65gE2sYksCXvqPcAwOtd+pb/W2fqjavkfc3zBLD2sHYqzMf6u+uPKapKau5R/kzZgRImYSgRSMfvRdx6BwgobG8qz1dokGXe7nENjePCGLvue81BjUuDkBW3zfq3jF38+EmFtv3q/CU82qsr7UurDxveveEAFquddVsPIdBTtijO65wtRUcoMluJCPPczDUZ9ZiMFY/aR2O/9RC4AKqi8trQj5hdPIHA4PNTP8/hSMXhWOt1XJLgnhfXs3FpElXfcJIsUatejNc1Z8MDrTszcfcmTcXn5SuvrtGxLgYueaFrU0pwqGWE6mM8ciUA/Lxtk09b0Z/70vjgagWjTsdrva8mxGjgp80bcaoqellmZKvWvNarr4+rz5w9BbkM/eMXj3lllJgZ+scvLL11DDM+voetezLZfSiH5NpRdG6d7GWzTgiM4J9rXuD/di1gbuZmXFg1Be4JPk+bwwft3ElggkxG7hvYhfsGesbg//LJQqZ8vKA8CCIoNIA3Jz1Ai47uSKyZ19/BbfOnsjbnSPk1OknivasG0ComofyYbGiOpG+DcG31moeEjtbNPiFi9T/lCWtkSeK+qztxS5c2fLh2JV9uWuu+N/2NgBFUiNlgI2FvGY1GptO3Rx46oyDEmEDLsNvoGt+v3Hb6fed7eXLdJErEcZNQGRiWhqDPcr+wLAkSZQHey+bPlq3h5ZYdfN9AILFhPCVWu8/fkwhUEXoosdq9hG5VlvyyouKDLCMZ9AgBwuFg6nuzGP3mLdVeX1PY7cvwtTFoClSoVa+E7MOeGm3Xfq2IiK5ZRSQ+KJQf+w7j/qWzcBw3ZcmSxGNtunNd/WY1OtbFwCXpMgZQ7DjGjPQXKXC6hYBRCuLahEdpFtG7/JyuP39LtsXbqR/cAiNl9DgiKzltu1SVErudUJPpnLmpXP/7JLblVvVVVBncaAv966cRoFMINTalefTLhAe00uyjKl/vnc+Uw//6bI8zRTCzx4s+2xdPW8fHT07xOi7LEr9ufYewyIoEM5mlZqbt38Q282GCjAb61mrOoMRWmHQVmrgQNhyFj6PaK4XGSvHIIQ9hs/+FUMuw0Y8S5/U0iE9Ap5P5ZftWXly+WHN+YSYLd165Cp1XCgiJnnH30yG6IqFKYZGF68d9iZBAKvP8G+a2kX1m9G4YHYnllz0oCAL3F6GzV5hJJFniuzVvUKtBHL1GfVL+YlLjnahXlkBkRcKfuMAw3mhzMx2ifefE6Kcb6e43OBDJ4BmynNw4nm/XvuVxzOlSWLJlH1v2ZRIfGcp1XZoTG372gq+0dCLm4ud9tErM/fEx5kw4ihACSZbpPaQdj//3FgzGc6OrORSFdTnpOBWVTgl1CDGcebHPC011LmOXpNB1KGV8kTYCtUoOBgmZ4UnvUC/EHdL55qql/Lh1o1YXRAYEsmn0uPMesdL424+r5GYVvNRtNkmhBR4CRUJHx4QJRAX61sDKXE4sLjtmVymj1n7s87x2kQ34vMMDXsdtNifL/93FZ6/OxGEuQy61IVX5PQy5pxcPvjGs/PPHO/5i0oE15d4MElA7MIIZvR8k2OC5LFZVBaEeQcKIufA+FGeVullSBBFxy8i2u7hp5iRyzZWd1ioY2WE18aHaOTJkdDzUdDZ6ucIN7d0vFvLnku1e5+a21fnedVMhMI9yF4mAg0VErMkCSeKWJwZx98s3AfDJxCX8tmAzriZW6FTm7q9Kn0ZZz69XPUrdYO1l+E0x92CxOpECAzR/f+/OeJx2vdxhy9n5Zka8MRFr5bwYssTHDwyhxxW+BfupoKpWjmY3QTu6M4RatXfjcqoU5pYQFhVMQOC5d528XKhO6F4Q7wUhXJQ5dmF3VVOloBqWH/vJS+ACCFSW5nxT/vmRDl0IMXj/UHSSxCvdel+QEMGqWfrbxh/yErgAAoUdeS9r9pFtLeb6v7+i/dx36LngI0YsGU+rUN8P4MONB3sdy8osZOSwT3n/3XnYAoyosWG4kuNQTZ5azL5tFVU0dhVlMX7/Kg/3MQFklhUxZvXPXmPIsg6dvh5l1pneAhcQooiNh/szbuPbxNc9TKtW6dSufSIXrRudrPgUuAAqCnn2gx7HXnh4APeM7IbJWHGvr+rUCL3Ox89dgHRCsZXg5k4bmfXlZH7b9he/pS5m+P0zUVX3jv3jo67m2t5N4coy99OjMS+H6uKXgyu1xwLu/+BOpACTz9/f9/9xZ4VTVcEtb0/2ELgAiip44us52B2nllHPF7IcRHj4exotOiKjvkGSZAxGPXGJkTUicBWXctLaef8LnHebbqFlDkcKnkMVVtxvWJno4DtIjHoNSTp5diiXqrK/VLvcC0ChoyKbVkRAICvvvJcn/p7PiozDuIRKbFAwr111NTERMjcu+4CjZUXoZJm+8a15qvlgwo1nVo76VBnWrCWTtlds2vVJ2uXzXKvrsHtpV+nhtDjtDFj8eXndLwFYFQfLs44xsG4L1hVW5BaVkXiq2U00DfPMMSGE4JEHJ2ApreSOI0mAQKkdjXQwp1yW1G9WUXHgi92+M7jtKM4izZxDk7B45mYu5df0+ViUMnTo+CJ5HiYNeScBDUxHAbX8O0ZFWTAanRw6VKv8nOoQAnYcySOhsWce5XtGduOekd2wO1zodTI6nUzg/H/4ZWOqZj8GM0gIvrhzKq0Sj1YS8iqqcxmlOV0Ijd+AJAdz3bBmzN+wutp5bS/ynfh8wOg+fPrCNBSXoqn15GYWALAhLYMSq7ZPtioE05dv445r3GHvBTYr/2YewOJ00KN2Mslhp5boJTjkDoymrpiL38Dl2o/BeAVhoU+iNzQ6petPhS2r0/i/Z6ZyLLMQSZZo3bkhL389mtCImnnWDpXm8vrWOWwrchfZbB5ei+daDeKKyKSTX3wBOK9C12JPISP/CQSVfe9U8i0TcSrHqB+nsft8nFKnnYfXTmfNsUPcVL+UGB/5MwJ0nrauiIBAxg8e5nHszyObeCSlIsesqioszN7CpoKDzOj5pId9sqZ5oUtPVh9J50BRwfGk6Zqpao/jHWn3xe5/NQstCgSbcwtZ0u9tVuftwiDr6RLdlD05+by5aCkSMLB5E9on1ebzTxZRXKzhXnVc8KohAehKbUiSxJ3PDCpv3m3W9l44wbyMLYQGmJmXvaz8mIKCQao+/61JcmEThvIphIQ4MBqdOBwGXKoOm1NPgMGleZ+EgCcnpfDaDRHc1MHbD7my18DLA/qwOyePTUcqpbkUoC8BnQK9mqVVEbiVByrFXvIpAeEvntLvI8EYyfRVqegkiV6tGxId6hYwa1bs4f03ZqPUrw1CoLoUyMpFrpSwpW5T9wtn56Hq7/eejGMIIXhq5XxmHvA0p7SIjOO3AbcRajy5XdRgaEh0jPdKpSbYkXKAl+78FvV4onihClLX7OOu7m8wdeObGAPO7llbn3eA+9dO9PBd31aUyT2rf+K7LnfTPvriKRx7gvMqdHOKv6gicCsw2/7G7jyIyeAdNaQKwTULv6DA4Xb235qfyNWJaZoPR5eY6nd+XarCuzu0KwHn2s0sPrqNwYmnnjTndAkyGPlr5N38m36Q+QfSkBHAeM1zowO7eR37NzvNZ985ZSUYdQZ6x1+B3eXitonT2Xykwu934oYt1AkPQ5qbU60GKQIMGF0Kr/50r0fIZ6QhmGM27Y1JgBKnlRWFy72OO4RMgOQ7Gs4pvMN6w8Mt5OZGAPD3npZc39rbG0IIWLuxCU5F5c25S7iuTTOM+opKDrP+3MKGTQdJiAvn+oFtaJAcy6+jR7I/N581hzIINRnZsDWdeevdq42br9xUrReIs2wuAeEv0ioiCb2kw+Ur37KAxX9ks9jidvrXTVvCC8P70DoqmlefnVZxniSBXgd1E1APHEFW3ZuXY151l1iKjah+s6xFvTheX/+Pl8AF2Fl4jDsX/8as67zLFJ1PvvrPzHKBW5kyi53Jny7knueuP6V+ymxOduzNwmTU06JRLXQ6GSEEL2yaoRks5FAVPtixgF97eu9lXGjOq03X7qoupFbC6vB+sAB+O7ixXOAC7DfHkl4aiSrwiAyqH9yZdlE3eFyrqoJjBSWUlrm1rT3mrGpCWgWrc32Xv6kpdLJM3+SGfHT1QB5s9xj5hYneEU4iiDaxH3ldG2nyXaomSF9hd3vnr2UeAvcER4rN5LRzn6cCLp2EyyjhMsiox0PSbryzO7P3fUT7np7uOqMbd6/2e7WMitR0QFpc3FAzu78QsM8WgdYiWwi39NNLEo+0Gc2g2i/jOh4JJgQoCqxc35wde9yhpDani7SjbiF3OD2Pfjd+zKdf/83qdfuZOXcTYx6ewMRf3SaBhrHR3HFlW264ogVP3tgT3XGPhmBT9dFPmQeKGNXqGRZPWskrLW7SPkmAa3sIWCr0GUVVeWf6Et77cJ73+SekfEwEAcEmXvjhPlp0cn+nvu0aY9BrV8GQJRjYuRk/79beKAbYkpdNrvXCJk5P3+dbW1/xp/bzXpVf52yg/91f8Pibv/PgK1Ppf/cXbN6RwWFLPsVO30nMdxVfHIVBq3JeNd1AQ3McrkOabRIyBl28Ztu8jKqhoxKLMlpQK8hM4/AchiW3pW3kYGoFetr1Zi/bxv9N/heb3QkSNE+O5+EHulWr5cWazm8yklcn/8OCjQPo0CKN7h22YdApbN3dkH83dKD7O0YMVcwoYxt35+F1UzWF25jG3VBVwd70Y0zbop2lH8ARIeMySlBlY0nV61BVwZ339ta8bmBiKybuW8NOs/ePuV1EXcKcwbgNFJ7MLWxGm6Ackkxmj3tvVXV8neNdCkeS4O0rbyUhOIqWMXHoZZnsohLGT+1PgMmBoko4nd4bOwadDkVRGfPIz7iqhLGqqmD8lFX07N6U5LoVWb8CZB19guNZkp/FPzubMKbnWp9mjD9/TuRYRj6fPDKBJu2T+X76ffx352wOWXORkWkVlsTamUXg8H6sVCHY4yhB89clSdRqnsT46Q95ZPEKNBkY//TN3PPhNByuyi5s0KZTXXp9/QMiRUTftQAAIABJREFUhmoN36n5R+kbVHP22dMlMMhEiUNbMJpOwbSwKmU/X0zyXD2V2Zw8+vp0vv2s+lVtkP7idDk7r0I3LvxRissWU7W+GYBOjiLYpF2LKlTTX08i2xpOgS2aH7o95dW6eO1u3vmxkt+ngJ0Hc3jhzQWEDgukSOMNKSExrG7NFe47GU5FYUHKHlQhs2F7MzZs99Qs52/Yxc0923oc61OrCUPrtmNm+mYP4dYzvjFX2Oow6IGvKLHZURpTjbFYwhajI0ArDYFOYtLUtTx0n3ckkCzJ/Nb7fr7evZTx+1dRpjgJkA0k5kWzf0Yxr4nlRN8FcpU/lwsdb2T2ZXzba7AXTSLzSCb/Lk1iaVYDxAiLW2ZUmuotSQPpn+xZ6SGzsNg9b7vvB6lxfAwrVqfhdGov+1VV8Pe/Oxg7yp2X1elUGDH4YxwOhWhgcVoLbu2YQlCQp/1YCMg+FMiSGRUbkmmbDpH3dxa/3vIY1jIHq1MOsG5POmsdvjVLJdh37bbEpCjNtIktkhNY9dnDLE5JY8v+LBrUiuL9NatYdSgd1XjyBEb1T3FD7VxgV8wMetVI2v6dHEsNJXN5FJX/0DeN6XXSPr6YtEzzuCoEc+dsJ6FROOmWAq92CRiZ3OlMp35OOa9CN8jYivqx4zmc9wCqsBw/KqOX42gUP9UrouwEoxt1YWn2Xk3t7qZ62jW8Ppq0VPN4camN+wN686NrIc4qZoYnml7n07fyXJBvtnoU6avKgRzvonqSJPFWhyGMbtyVOempOISLEfU6IJkl7n5xIg7X8RTbqgS+nnEJJFVLJ3Uze95mTaHrdClYyhzc36Q3DzZz59q97/VfSU2r0HwL5yUQdUN2+diS5PaieLH5/ZjzE7h/zG6czgqbuXQ4CP1AM/o6TlrUSeaehjfSNMzbrl8vOsLHl3HTICYSWZZYu6H6rHA5lZK1PPXQRByOCgHtcuh5YNztPPv0X7RsmY0kgcsJ8ybUZerXLRBGI5gkcCngcDB/4nL+3n+U/2fvvMOkqNJw/zvVeaYn5xkmkXMGSQoSlaCYAFEMmN01x1V3dU1rFtecBSNgAFEQUAkCIjkPcQiTc+ocqu4fPamnq3sGRXDvve/z8DBTqU/1VH3nO1943/X1Lb0eA5AgBZ3sumUmUrRL3Shf//exQceskSTOG9yV8wZ35ekfVmNx+sIgwiWI1DiIMjgockTgaRYb7xudz/XttxFhW0iVK4Uw843ojadPSaHAuoWVBfci9/DQoTt0mFKG+85jLL+hF85KPd0HZDP2ktZ5JkrKg+cQcg4V89y0aVy74QPsXv/QUL/YTG7pfG6QM88sTnvJWKRpFD3b5WBz7cLpPoRem0G4YWCAwf3ys1/5Yt56rFYniclRDL02k1/tJ/wkmbPNcTza9/yWHwFAdV0Q8hPg2K5avrvuQRae2Mj2yqOkh8VxQ8exJBhPDYmHW3awv3Y9Ne4S4vWZdI4cgiT8LaCiKGyy7CVidDFCAuexMJxHzeBt+h4GdQouJ9QhMoG7eja1Kf9n/gpcnqaGBWO5giMRVQMgyaCvDW7sXS08RbfHy4ufrOLbtXvw1CdFhvbK5IJzevkZXABPqZGyDzMxdq1Dl+QkQRPPh1fegV6j558vLAjwQpUaLe4vYnED4x4ZS5d+6vSLCZFmhnfIZP0R9drud67xdaW1a0UT66z6tmZZVsjZG0iyYrMZeOzxKUgahb9d0Z837v8SwsJA21RJomg1KPowtmklvM04FDRO8CMlboF/X38+X8rrWNEilnnjbWNp31E9tNYcXlnmm7178RgVInUOXh/7FV3jfZSfMrC4oBsvHhzBmwO+pmtkGQ2xHo+7iNqqOzGF78Ac9c9WP+ePwiM7+bHwfuT6FW3DI6gLlzn/nYP0qZnDoFHd2kSIHmk2UlapPlGlJkbRPTqVZWPu4svjW9hYdoRYQziXZw9hwF+waqEBZ4R7QQhBuKEP4YY+qvsfe2A+G9Y2ZemL8qvgySrueeRs9sVVYPO4uTSrL+NSAwUPGyBpJLwqWVOAxFgzMXozN3YM7l38XpQ7TzDv6D24ZFvjBCEKJCal3kmv6KbPu2nzyxyyFKCtd+BMvWoxdrRQ82MyeCQiwwyM7tP2WFxufrnf7+EV4AkDj7neuNZ/TwKYmdyVnwke8w0z+cfa/vX2Mn7a5F818evu42zccUzVvigeCfueKOx7oE6A/mpf/HXrlqMqRzdh/S+HGDsuuMbdxQO7k1tRSVF1k/dj1Gn54NpLSI3xTZgXX9Cfdz4KrKAAX2XA6Prk4O6dJ1SPaYDsFYy8dCRvPPytjx+h2XMmhMCVGIG3RUuxAMLKZGyJzYxJ/Xm3TR5GVmIs9z5yAdf/bQw/LNmBwaBl0oUD0Btbfw1rnU6mfPIxFZIT9PDx5E9JDa9tNGga4KK0HM5OyCNKX6sy1zqwWz/AZL4OjSa15c5TigLbJrxKYOOGEKCNsNOpp6HNChSzLxvKs2+rt4dfN92X2I01hHNj55Hc2Ln1cMVfAWfE6JY6ann1wEp+Ls5BKySmtOvHbV3GYtDoKMir9DO4jVBg2Uub+HrlfW1aIo3s34GfNx9S3Tdr0p9Dn6coCvOP/wunbPXfjsx3hS+hEya6Rg3nq7y1HLL4e1lCACYZY+daYorT+eTey09qKdglK4k9LWghI/MU3CZwxAtMETouOasXt50zFKfVxfqlOTid6jy5M6c1xbXLqiysUvseFQVFAryhGxiaR09cTk+oomQizOrx2pJaC1PmzG1cVgPoNRJzrpjCyM7Zft+TyajnzlvGMOfNnwKu89LT0xtf9rzj5QH7m0Nv0BIVZ0YbYcLrDpy8PYlm1XvRusBcIOOKEHgNYNbrmPfPK8hKavLAo2PCmXFV6EqQlnho5Qryan2hke5xRX4GtwFCQLzBopIxaYCC27keTdgfkyxqDR7ZTiiFjTp3IbGGtjkUU8b0YvvePFau39/4LAkBt1xxDh0yTl8o8FTitBvdSqeFi1a/grVZDOaToxtYXribH0bfw9LFwSVlrFYntTV2otrQyfLPGyaw90gxJZVNXpEQcN9VY4gyh1Ym/b2ocOVR5wlOyPxTybt0jRrO13nqLaJCQHJPhUW3Xn/Snz1z0kC+XrnD71EXgN4O+jyFOQ9O5qzeWQCYDXreeuUqHvjnQkpbqPyOPqcrV0wfQrGjgEX5H7O/bg8Db/ViKQnnyE9ZOKvCmgarKKiWKzRDQ8lTaWEVwuZCCdMHNbyXTFNPpM5483M/gwvg8so8tHA5Gx65JeD4i6YMYEC/bN75aA35BVV065LCTdeOJDqq6bnp3Tf08rNrvcpxUIWfEPcsKWCsD99MHtfZz+D+Xiw/3CRd1C/+95ZCyQjx5/PTJpuCy9mDRLQ+q83XEkLw6B2TmD1tGOu2HEav0zLyrE7Ex/zv8uyedqP71sGf/QxuA8qcdbx3ZG1jcXswGFspM3F67Ti8Vsz6GL5+8TrWbD3Mqi2HSIwxM21cP5LjfctQRVFwOz3oDNpTllxwem1+MeeWqPWU4lU82L3BO7R+71hSE6P45y3n8/ibywL29e2WxuBe/kYmOzOeBfNuobLayuYtR3F7vAzol0VKUhQVzlJe2v8IznqKRI0OItOs9LtqLy6Ljt2fd8Nt83mlIUKYAEwd40t01lVb0VhdeIy6es4C//KAbl1TycwK9FxKai2U1KrH9KrtDg4Ul9ElOVC+O6NdLE8+EqSWFsjIiic+3kx5eeC1hYAZs3yeaExCBBXFNQHHaCusuNOiQnruGkkw69LQ1TCyLPPps9/y5ZxlOO0u9EYdF94yjmsfvaTRK1cUxY9r9mCVulx56xDojaN+57ltR7gugXZhw8i3BbZKJxq7E6VXJ0cPhfSUGC6f4luhbi4/zIO//kixo5oukSnc0fX0JsD/KE670f2xeF/QfYvytvHuxGv4Yt461WL6jMy4oLV9dq+VL44/S661KVaZGdadKwb8gzGDm3TCFEXh9Qe/4IdP1uP1ykiSYNTFg7jnv1f9IaVTgCRjh5D7tcKAhIYhcd1YVrxZ9ZizE9pG56iG88/uTt+uaTz/wY/sPVyEOdzAFZMHc9GY3kGNeWx0OBPG+n/mypLFjQa3AQ2n681u+l61l81vhfJm6o/VabjrynPZ+P1W/j39ZUhrh7bCiidcDyad76Kygsbm5IX/XqF6jcKq0NIwJyqqVY1uW/DWvBu5dsYb1NX63+ulM4YwaIjvb3ntQxfwwu0fB5yrK6rFnWgGnSZQfhjQajX894lppCRGBZxrszlZs2Y/5eUWti7eTM6KnY0Tl8vhZuHLS9n0ww7eWP84Gq1PAzDZbKbY4psgNpdlYHHrMetcgSEGooHAEiqAiJhXEMK3yvPKFiosH1NrX4pWSiAp6i5M+uDx9JPFuLRn2VDyAgdrv0PBJx6bGX4256Sokzi1FW8eXMFHuasbXZvSslrWlR3gqT6XMzbFf/xFBZW8+dwy9uw4jkYrcc7YHsy66Vyif6ew5qnCaad2HP/T85Q61F+kBEMEK8fez0v/WcLyJTv8DK9Wq+GD+beQnBoTcJ6iKLx26E5KnYHJkShdPHd3eauxeuCZm99nzaLALp7+o7rx1Be3nfT9KIrCYUseRfZyMsKTWVvyEgUO9YnlrLhLGZ00m3JnDVdseBpXCylvk8bA1yMew6jxJZ5qXCf4tfQ5Khw5aCUTPaNn0i1mekAlxO/B4hM7eGb3cmrcdiQEY1K68uzAizFqdDy1725KnYHdbE33DAeWdKQqN4bhvTOpqrKz/6g/R3C4Sc+8p2Zh9MrM6nQbHpcXIiKQEn2NCUKIRjKfy64/h9n3qFehOFwe+v87uBjihodvIdKopc65FSEEZkMfJNH2onhFUdix7TjrVucQEWnivAt6cUK7nnzrPuIN6fSNHc/Hj67iu3mBISHZ68GZHY833uxrEVMUhMODtrCaRx67lNGXBNaJfv75r7z7XrPaU0UBuwvd/hOIFonfe966nnEzfR73wl17eOCnFY370sMrmX/+JxikpmoQWcQQnbQGS81DuBzfNruSjojo1zCG+djm3N4yDhaNQlb8Pfh4882kxPwxo9gSsuLF4a1CL5nRSn9MEaPEUcOFq59DVllN6oSGVeMeRS/5/Mjd249z/00fBohrRkSaeHvB34hL+HOboP5SEuwXtOvLe4cDs8sGyc05yUW8mHMVxkvCuWn0SFa/a6Om2sagIR245qZRmCPUY7H59kOUOfNV99W6KzhQt5VukYNxOtyqBhdg2+ocaqssRJ5ErOi38t08s/8jXHJTpjbdFEv3iFissr+3kRnWl5GJvj74eEMUHw25nzkHvmJb5SGEkBgS1437u09vNLhl9n0szb+BhuChx2tnS8VrFNg3MT5tTpvHCFDntrC1ajtu2U3v6B78UlzAv3Y0vZQyCiuLcjj485ssGnkrpUUeaCUMGZFWR5LchWdvvxC9TsvRgnKWrN6D1e7i7P4dGNa3PZIkmP/cYp/BBairQ/a4kRLiULRa8Hrp3Ts1qMEFMOq1TOrdhe93BbZnn9s1gaOV43F5mzN6CTJjniQ5cmabvhshBP0GZNFvQBabyhfzfsnVTTvr4Jfyzxl151XMuOMJvp+3jsqSGgaN7cF7j35J8fEKTLnlkBuYlMvunhaw7edV+/wNrm8AYNLjSU9E14LgZtlHqxk3czg1dgf/+WoVOgHuCEDAhL7b+MGeQoLGQZTkplI2UCXr6Vf9LTZ3EjZ5LCm6aLIiL8RoPNuPwS+/8u4AgwtQbnmLGPPlGHWhV2wnA0loCNOemqX/hrIDqgYXwK142VqRy9CEziiKwlMPLggwuAB1tXY+/2Atf39g0ikZ0+/BaTe6N3cazTcntlHRrHMnXONgXNoBFMDqBau3morwzzjnX8O5NOOBVq9Z5sgLGktVUCiwHaZb5GAO7wzN37vn1yMMm6hextYSWytzeGLfewGfm2evRFY6c0/XsRyu+w1JaOkfO4kkoz/fbYopjmf73hj0+quLH0ItW1Nk20Sl4xCxRnXBzNzySn7NPYFeo2FCj07srtvOu7kfISP7xnoc7G4tWikej+z/5z9ureRv782noDCKTheVEaRXBQHcMPZ8JmRNaNyWnRbP7SrS4hVFLRo87A7kE02VG1URrcewn58+kagwIws27cYjy2iEYErfrlw06D5c3pYxWYXjVQ+jEWEkRExt9doNyLVsY2WJOsvd6rJ5pGZ0ZdrdIxtZ7KzVNl6+MzDsAJDZNYXsbv5GV1EU5sxZoXo8QqDERUILo6vUK1R8umE7DrcHDQJNBcRGV5EaXY0QgnLZRLnsc0YkvOyreq0xVFHmgN11mzgv/XMiDVmN17U41EvqAMpq55Ae98dl1v8MSMEeSHzPpFP2rRzzj1dQVxOck2H9zzn/bxldraThu9F38UrOCr4v2ImsKIxLO6aqorKvdj1ljjwSjKF5MWMNySGS6IJInc9tS2wXp3pEA9Latz02+PqhBUENfYGjHLO2I5PSAlnC1OBwH6Cg/G84PQcAgUk/BKeniGAtZbl1KwKMrqwoPLRoOYt35TSO6okfl9Fn5MGATJdR66FXYjH7yxOxe/w5DLZYjmPKjaZkRwJJ/XyVGC3jhlpJx/jM8W26t97ndGfRaz8E3Z/RNdAjVMMjU0bzyJSmLrkK6woOqyTBGnC08sGTMrori4LTigJ8dsIndxSjT2VK6h2MnzGU31bsYsNSX6ODK0qLK05HapiZlxffF3C+2+3FYnEEbG+ECExKjpnhe35W7fPvsuueraacq6BDDkhqKnhYXfg3Lsj+3m9rMHi86vHgvwLOTuiKQKi+d5KQGBDra6yRJAFK8MncYDgjlbKNOCPKESaNngd7TuaXCQ+z/rxHkETwZMlvFYtbvV5mWHcideoGVVEUYnW+FzshLYa45MDEBkBkbDiZXdtWNO7wOilzhn44d1S1ja3M6T5GbvFYnJ4cfH1FXuyu9fQ2FCBUZVRAEoEPzTc79vkZXICkzHLV10sI38vdKbY08NpW38N64qdM9s/vhOymkdlLUcAgGbmny5NtrrIYOmUA4dFBmNEEzH7y8jZdpyUqbSqMXc2g4ESWW+cmaLyeKzR3bQOqXIV8cuwhyhwn+OcHN/HE8rspeKATR27LJO+KVH67KJIhn7/HlkJ/w6jVatBoQnxnsr/BzOiSyvgrz6bgcDFHVu/3K3i22AJjo1KQZwXA7i3F3UyeXq8NXiMbZQpUGfmrINZg5oYOo31zRsPXUf9zp/IMIurZoVLTY4mNVw8TCgGTLzu9MvctcUaM7qmGEILrsp9upHps/s+laPks7/XGY1/67j7CIvwfWmOYnpeX3t/mz9MKbcilDkCSsW21mUWV96GmUaVBIVmjPhl1iw4sbn9v/eYAAxsWZQ9a0SQEaCUwapvi0QLQHWky6HV5UWx9ZSC73u/BsRUZnFjcg+f6fEiqqe0lPxqthvd3v0BUi8SFkAT3vnszqR1ab39Vg1HbljGElob54cR+xix+m55fvIhTJf4XDDJeVpV8hEeWmbXxe2r1sp+LWuN0MmPxAo5VN4VWJEkwdqw/yXpdmkThCD0FI/UUnGuiYnIGkYmRzH78Ml5b9xhavYbbhj1M5Fp/FYqdB9s3Pt8NaG0KLLY3JTozYl9RPUMSsUSbLw7Y/ldCX1cnxE+RUKkBp/D9/1MkB5dZWLLGV7kkhOChZy5DpwtcKXbp0Y4LZ6jXg58u/CUk2BMM6ZSpVB4A9ImegKzIrRq5/XU7cCp6QEaDjIJArhexqvVUYfdaMWnCSWwXy5cHX2T99zvI2ZJL576ZnHPhgJOqj9VKGkYm9Ofn0kBDB2BAR7/YtklH213qzSBCQIzGRpE3mgjhorO+hniNA40AW+lwZPP1hEXcgaj3esst1oBrOG0GlJjghhcgXOfE4dH5BCbn2nBYBJ74cL+YgrPKRFmlkbuvCSTBaQtik2NYUPAO+zcdZsuKnUQlRDLhmlEYjL9fdysl8lYKa18Pul8SZiRJPTxj87g4f8n7nLA20awV1kWQEVUV8rtqjhO2vby3fQt2j3r/l0eWeW/HFp4cNa5x2913nc+hQyXk5pZRm6GhLrtZuZmA2m7RKGO6Mu06n1rHzjX7qKuwoAXMa/OwnOMLsylILPp5CBeN2dgYk/Ai0AYZu1uWuGz9W/SMyuCdwTdjMvSiQ+L35FXchst7BJCIMJ5Lu9iXTqry40zgo8WbEMV6xLLAZ+edL39lykhf2Vi3XunMXXInC+etZ9tvuURFm7jo8qEMHdX1jGgjNsdfwuhOy3iINw/ditzC47N49Pxr39PoJT0Xpl7I+cnnI4RorK9t/uXZGhMqkioptmg2swshGDG5HyMmt15rGgw3d7yUE7ZiDlvy/AyvXuj4T5/b0bSxrEsIA4qiHuuLNvSkiyaK9mItEkozg1CHw/IyTttiYpN9SZHOSfFsPeHfqVR8NI74NDX+xobPhr7R7emQ3pnSZ3LYubuI9NnFaLpoqCuP5MS6VBSP7z50edX0+QOdVUIIup3ViW5nqScATxZaTRhJ5usosbyvur9TwhtBz73jl2/9DC7A3rJk0qOqWvUYG6AROr4/EjqEtKnIP8Sg02l4793r2Lb9GFd8p65esquwmGMVlWTFxbLtxFEKn2+PN90IAqSSSvRbPEjVWsoXO/jiwTQmPGOi39QMUkyDWLLxC9p1ONz4nDR4wmvKOwOCPTV5/GPnpzzbbxZhht50SV0ToMH3V0d5tX8cXwrzEDu6BGM7X+Lskd3Pc3un2SQa44hLiOTmEJUxZwp/CaMbZ0jlri4f8UPRexyz7qLOY6PMpccm+2Zdl+xiYf5C8o8U8vPlW6gs9r0w5phw7n7nRkZMPYteUYNZVvyF6vUNkhGj5tQKToZpTbzc715yao+ysWI3Vo+NXlGdGBrfG4OmbR6cw+WhuHIcCdFfqnpYKVEPE197j2rpC4Ai5+KwfowxfBb/nDiaqW9+4rdqdFiNOG06DGHuIPLlEv8dfAtCCKbXfUHXhXWNmm2RHWykDSpm/yOJeHbKSB6ZHxf8Rue+WW26t+Zwyw72Vi/DKVtID+tHiqnHKXnRs+IeIcI4gCPl96Dgi1nqpGS6JL5HuCFQLw3A6fXwY0Egl0ShJYY6l4FIvbNN3m7v6DFsVuV5bkKsZKDwWBkpmfF+99u5ewpyiJD0DzmHSM0w8Wz4VggzNnrDcooex2Qd5qUVhO2vwCspcKwbvWIv48P5x1i4chADhpoZfM5e9AYPFo+B1eVdyLU1JYjXlvoLof4vGVyAAd3TOVboy6cInUzKzGMIXZNDcqAul9u3P8rbA/9DlM4X0tqfV8riDXuwO9xMHdGTvh3alrz9s/CXMLoA4bpoLsm4lyOWIzyZ86TqMRtYR2Vpk1doqbLyxLQ5/PubexkyaQAdwrtzxBrYmHBpu+ClWX8EQgi6R7Wne1Rw+fNgqLbYmfLoB1gdGTw9O470hIpm14WosOmE6dtT5Q2d4LFZPsQYPgtrbg0x261U9fOfXA7/lk7Pc46CNtBw/63jTY1NCu0ebjK4DWNQJOj0QBkHpvkaUvS/I+u7tXwB68rfafx9I3OJ02czPu4ZFvy0jyV7D1Css4NWkBUVxcCwFHYdLMThdiN7ZUqrLeh1WsYN7MJDV4xGr/MfQ1z4+cSFt+7NuGUv60pyybNUg1BUstuC7w/1YlLHPUQaHK0YXi0fHd+JnKSQoTGQl5+AorRYXSkKxS/t5LqyrUREh/HIG1fRe4gvgWXUhW5ljzWbeHBLfQJZhdUm7CItF9xcjDFSRrCUb44vJayflozjA9n6aze22tOhh3pYSUFpU7jur4pbpo1g0c+78MoKEf0q/AxuA7yKl/eOfMHdXa7niU9/ZPGGPY1e/7cb99E9I5GPH5x5xiacM2Z07e58Dla+hNNTRLRpAB2i/45GMvJbRXB5dfQgRYDcrK5bURRevf1DhkwawI0dHmZp4edsqFiBW3ETrYvjknbX0y3y94cR/izc+eZiLHYXIPGP96+hb/tczu23C6dbx9ShL5IQ1RdZrkItyeYHxcfj8P38TUQdc2IuclGXacBtljCVuAkvcFO1KoFzX01mX+1+ZEUmIyyd67KvIiPcFyPMqduO0BKQWxECNFEKmigv3hoNE2aeHDNWvnWXn8FtQIXrKJ8VTsfbWaJTZCLFu3tgcxrZV17OPsox1IKhWYja4fKwZMNeftt3jKXP3oAQAqvrKKW2HwGFhLAxHKs1M3fvNk7UVTMkJYNZ3fsSb/JVTeytKmbW6o+xeVzIKGh0gAJeT/3MUg+b28DCnAF0iilhVFY5Ci1DM4Jql4njjhi8io8/JCHOTWyMhV17MpFlbeOaPmarFVOZL95bV23joave4e3l95GWnYBWkuicGMfB0gpaQitJGGNBzlVf3Zg0Tq7stBmBvz3WGzyMuXgT3340kooiE/RQ55PWCOl/1uACmMMMvP/4TG5/5itM7a1BJ8ddNTlsPpDH4vV7AvIu+06U8tyCVTww/fflKP4ozojRza+dT07FY42/17h2cKJmLkPTlhChDdGep4AcmC+i7EQ5siyjkbRMSZvFlLRZp37Qpxi7jjZvsxXsyO3AjlxfJ5CEi+vOA0mKQaPrjdcdnHnNYLoMjyxzPMrBifPD8RoE5uMu4rY7kOqfNqVCxz1dbg96DZunLmT6WwqHfgN7kN4p+WRukdWlrwXdJwRotTId04rJTinlo+Xn4nD5qkqciaA/5nNIAWStgjMajhgtzJy3gDvG5GL1zEdWPIDCocpXWXmiGwsPDaRHbAH5NYt4ckMYf+t/DxmRXZn284d+svWiviZWo1OQPTKK7G+EjtekMjv7CZLDInDLTuzeOsK10TyT8za59hZ6fQI0GujWvoK9exPQ1XhJ/rkGc4E/n6zXI/PW0ws4/4kiiqy/cNVrG5x6AAAgAElEQVQYmf0FiSzb3puKOh8JkyQEb8+YSokcqBjSgJHJBwMMbgM0WpleQw6xetEgcAgUY6AXeFG7M5u5PxXompXE8jdv4e4tT5HvVqtZ9tXtfvzj1qD1K4vW7/1/x+jKsoucin8HbFfwsK34ekYkL+Sbwm8CCqAVBZw7ZTV5NfQm/R8mqzmdUJRQXGRQWdfUTRMR8zrVpaMBNc8lDG3YNVz25RfsiK8CdKAoVMaZqOxrImOJBVO5l94Ds0OOp2fUYMhTTzwpHrjosolc+1DbGg3KbOs4Uv0adncBZtmKXYRhU4L33AsBGklmXP+dLNnYZBDc4aC3gFevUJeOb1KQoNy7jQr7UvRaT7NryIzL2M25abtpIKlTgMOVM9hcdgkuOXA532CMJK2C1+XflvD44Akkh/kmf51kQFcv+rajOqflZRphCrfS7bXgqxJzso1ud39EXrM8UOe0E3ROLSQn9xZSwrty9Vn9MOp0lNqbJywV0g0WXIpEiSuctLDqkGWAMfH1VJ3fxsJ5VRDpK2cTwOTUAdzb/YKgY/xfghCCadnn89LB91T3n5s4lJUhmlGc7uCsw382TrvRzbcsJFj9pMNbQJQ2ihnpM/gi7ws/w2v0GCm613+5p2gEtQMTMU7oyA0LFnHDWQMYnBm6e+2vAJ9yhh6rU13ye9LgpnIzjTadmOSdWKpux+1cQUO4QaMbRmTc28zdfoSdJc285mZSAicmmem90MY/np3md/3aSgvrlmxHAAPH9iAhLZbBMaPZVPVzwFguzJjFyIdaL5iXZRe/5E/C4W3yPEwSpOscVHjCKZfVm1Iahtwu0X+p3SD3ZU2ingrS9/vwrP3oNOovTPNwrwAk4SVOLGB4/ADWl4co4RM+DtzBSRk8O2QSGRHqemxKiFBPQ8tuMIz7zy4ktbdNeBjV8zdGpV3VuCnRFMGYlM700H7OtWn70Arfm5DnMPNdXZKPxljN8CpgrUoEQPJo6HOoN4/+fQIujYtUUyw6zV8mhXNKMCSuP10jOrK/7rDf9lhdFNPSJyMG7mbvcfWcSGzEqU2snwxO+1/BE9Ar3xIy45PH0ye6Dz+W/EiVq4qBMQMZFDuIt66dx+I3l/vicTpB/t39kMO0IDwUHjnKmiNHmdGvF4+fd+pleEKhtKyW9z9dx979RWS0i+G6K8+mQ1boluJ7Lz2Hf3/6Y8D2TqnxdM/0X8ZLUhiRce/hY4RzI4SeuhobN1/6Pmt72VDi/f+MEeE2MlPLiAi3kzwyhg321YyLnIDXJfPYFa+zfe1+v+NHXTKI+9+8kSRjGj+Vfo3Da8esjWJK6iz6xbQtjru+4CI/g9sAISBOa8XjlqhWgoeOhPA3WhoHeHQKshG/0EdYkCRXUO8PmJm5lVEJB9hU2YF8WwwH6lLqa7hBJ0l8P/l6ukQltnaLZIW146hNnVgprMiMLYhR1pvdRKaoV0VIyNicq1l1rCcKbgQ64k1jmNPZA669zct4yTBauN5gZ2FdPB6VskhJ6Lh34mM8MdXXOFJgL+f1g99wsC4fvaRjeFwPLsk8m0RjaJHP/xUIIXi8592sL9/CksKVuGQPoxOHMT75HAwaPTNG9ePdpZuw2J0tzoO7Lz3nDI36DFA72j0lrMsbpbpPI8IZndX0eTaXm9JaC4mRZsL0viViTXktP322jvesxzlE4PJBAN9fP4tOCaeH1HjPvgJu+8fnyM08HSHgH3dOZMJo9bKlBnyzbjcvfr0Wq8OFJAnG9O3Ek1dP8MvQ211uVh4+TIXNzvDMDDrX39eMc5+lusrK0YvNuGKaaoLbJZfTKdOnZNv8JZeQiJnbhV1f+OuUaU0exj19kIyhzfW2zAxIepKU8FEBY65zOim01JFiNhNp8IUNbO58fskPzcWgKHDCHYtdJdSgKGB1GPhoef1k6YWI41CTqYAeFKH4vF1F4ey+u7kie1Obmxiaf0bjz8DnxwayvqILAvhk5JUkhUWSaQ5dh3zCWsjdO58KCH1JSNwtbuCZ2R/icmpoGSA3RruYPv/XgDFr8BKj8SUpmu8zoDDIIKvykSgKLFmfQUXP+jLA+mP0kpnxac+SbPKRxi/O28DLB78OOF8vtLzQ70Z6x7S94qa4uo43V/7K5iMFpMZGcsd5w+mVcXLx/TOFGquDO15fxO76HIrJoOOeS0dy0YhTxx2shlDUjn+K0VUUBY/iRae6noKdJXdQagtkXOqV8ALJ5km4PV7+/tm3rDt0DBTfjHZBn648ftF4tBrfDN/lmTlB5csvP43e7uTLX6VOJXak0QhWfHkX2laUMBRFweZ0Y9Rr0TSLSyuKwr9//JlPtu/yO759RBTDjpr49UdfP35FLz0VA321nDqthxEDc1RfVgDZAUWXGsDrO0DSylz70ya0BnVPsW/8Y2RG+mK5hyrKuXP5UvaX+0hwFCA9MoqPLrwYtzKXozWhCWMAZAUOulJoaZQUBRatH0xBeQIoYCoEjUehtj0oOgXZVB+XNHjQmZz8d+B8QtEYtAWKAq8fGsm+2nZoEL6KgqhE3hh+Gcmm4KrQhbYS/nt4LkcsJxDAiJgoLolegywfRVFkrDV6vnm3L79827wJRGHmN+vQh/t7wrGa2gARDYAkSaazSilUA3btSOCRh0eRNaaITr3juHbanUTrm8Jqdo+TiWseDpo3SDBEsWD4I20qmdpfUMqM/36OpwWPxT2TzubaUao25S8Jl9uD1ekmKszoI8T5kxHK6J7S7JOsyLx3+DsmrnmA89bczwVrH+Lb/EDJjj5Jr9Ap5j60UjQCHWHabPonf0Cy2Ue3NuPtz/nl4DFffzk+Bq1FO3J45JsVzT4r+GRhdQUqkf4ZqKy0qhpcAK9XYfP2Y61eQwhBuFHvZ3ABXli7PsDgAuTWVvOVM68haEnMfjeS09eInxRfHaQIQQFkhEHGOKgpiz/qX4eDGlyAHeWPoygyiw/kMPGzeeSUl/lxjeTV1jDm4w85VKGeQQ64VyBc+H9figIH8lIbDa6mDnRO38GKRkEOkxtjukLvRUbL58cGqyqLnCxmZfmeTS8KTtnL3qoirlrzKaEckdSwJJ7pfT8Lh73GW91juSjyA2T5CD79MTBHu5hxxxZGTvWJqxqMWiZMG8yw9H+1+C5kVYPrG09wo6AoUFNjQHZryP2hHStfDKOuwH/18F3hbyETtbVuG8esJSGOaMLtc78NMLgAL33/C3bn6XnPTgZer8za77bz4PTXePDy11m/bCeyLKPXaYkxm06LwW0NpzSm+9TeT1hdtqPxd6vXwSuHvsTqsXN51hi/Y7OiZ5MVPTvgGsfKq8gpUhd3/G7nfh6eOIqv3/wZye5FNql7kZf0Cr2sP1Ww2oNrnQHY7eqJstZgc7n5YLM62TpCYE/VokgCoYAkQ/svreSPMxHevuUEoCA1p/sTkPioE9tmJxXPhtFhdGUry3SZ/LotPPjjBj+NrpZ4fbOXO4a27d5cDgmXVkIrydjdOlbldiP3SBoSkBYVwetXTKZdXDSRYQay33ih2X3Ljfy+68s7MTF1FzH61poYgkMIiND5Gw0ZKLHXsbUij4HxoQl1rLYlWK3q3r1OLzPr3sM88OI3iGbt4EZ9JFtKn8Yhl6NBBB17pRxae662l4Uht+xl98L22CpM7NpylPTsphyCxa1eo9sAASG1/Brg8coUVtWp7lOArzbv4coRf50aeK9X5p6pcziw83ijZ7Bz/UF6DGjPc1/d1mqF0zFLCe8dWcGWyiO4ZQ/JphiuyR7D2OQ+p7SR4pR5ujUui5/BbY6Pji1DDiqr6o+fcw4H3ScrCs889AVfvP4jiWsrfevVFugQF8uw7JMXvvs9SEmODknXN3TQ72PgP1FdjdvowRPrxhPnxhPpRpb8vz+voelzJS9k/GDHtEL4vagNBrchvtvwL2wQpM210RZ6iB0lR5BaYSQ4UJ5GlT14dUJzpIdX8/G2YczZMJ63N5/LwfJkPJFexnXuwA83XEOPjGSiwo0IIdBo62udJBlNhL+BfCFngvoHnCTCNIETZ4E1UFWhJaqq/xVyv6zUIcv+1TZp5rO5sP0ypnfczNT2awlmVmUEOW4R4M0rCuR59Fj1GjqNLeTCVzcQk2UjItJfUWVCauhlvxASWeGts7u1ZmesjtBOx+nGD59t4MCOJoMLgAJ7t+Ty45fqmoQN2FN9nNm/vcqq0j3Ueew4ZDfHrKU8uXcBz+eo82T8Xpwyo/tbxf6g+zyKTLmj9QcZIDkqeDwNYNuyPaCA+biD5JXlaKwe39PoVZjYqSNLrrvypMb9R6DVSFwbpEvrgvP6YDL9Phatf2z8AU+Y1/fXkQA9yLHeJsOrgOQOnHDsW83NiHaUoEX0QvgaHlyu1mdvoXTC6Q1d06ggeHrtJCJ1ob0e31hkRrVrVu9anwyaNbAvxmYJxAqbzZecVBQ04e5m5/tQ5TZTaI/8Q2EGATzeaxGppqZmBFlR6NyGSgZFaY1/18vda9bSc+4rdPtoDtcu/4q9FU1Leo0URpo5eBNPhSxRZLgGi+iEQxZUeyU2OMxsdPp4YoUAjV7mrBv3MvicLn7nppriMEpGVaOtKFDjdDPl07lYnaGNpkaSiAjBBDexX9uY9E4Xvn53VdB9X771U8hzn8/5BqccGC7xKF6WFW4hzxooyfR7ccqMbqQudN1b86Saoii4vV7V48Z064A2yDIg0xwBriaPz3zcQfanRXR4N5+O7+cziSR0mj8u2ngymDVtKA/ccR5R9d6GOVzPrbNHcfet40KfGASbivPYXlHUaIyg6X850guKQni+G0nl60uKi+WWjn9HE0RxojmEgCOu0CuCJOO5nJ3Rs7VGZAAsLhP9k+eSHB66+F6vUTgrKRfR3B0RcMRawdaCQq5e8CV9XnmVIW+8BQ6B0Ml+nBDN8XzOBMoc4QHcsm2FEGDUeLgm2xfblYDesSl0i24Lx29oYcNqp55FuYepc7uwedz8nJfL5EUfs6GwicK0c9w/SAu/WvX8dhFXkx7zMOscUXxri2WFPYYCr4Hm3rEQkNC5GoPRP0roVWRq3G68ivDjlvYq4JI1yIrEcVcBEz6a2+pdPnaZ+nM8tFMG6XF/rdIzuyX4JGKrC94oYfe6OBoixq0AmyoP/pGh+eGUxXQHxXZBIyS8KmGESG0YMYYIZFnhP1+v4ptfd+P2ykSFGbnngnO4cEhTDNag0/LGrAu5Zd4ivzhinNnEs+NH88DbgYQ2DY+hdIYILCaO7cXEsaemBGVuTpCWXwFowVTsJmlTU8xOgUYl2pvuP4+DOx3Ubh6GM+YYCb2PhWzvrSCVduG9yLcGUl7F6gcyOOV5JKFVlUISyIxpd4CJWXtIDqvjWF0yknIhfRKfId1+KVuKb0NWalSNZYzBygXZ21h8dEDjtp2FxTzx/Sq/hkPJK4UMSzllPY/umUqGqYJpmZvJMlcETU4FgyQgyVhLlM7G2ckDeXJg25QTIsyzqbO8orpPUWBlYaAXKCsKs1d8Rc7VdyKEQAiJLgkP0THuXsptK7G5j2PQJpEYNhGtpiFk0MpsIpo6RxxeF7mWwnpxU4FX0aBKUFc/oRfVWdhWUEj/tOCKKRN6dybsOh1PfP0zRVW1GHVarhjRj9vOa5sUld/9ywqrDhzhg/VbKauz0jMtib+fO4T2CaFltMAXr13/3XYWvfUTNoudAef2YOa9kwhvFlrpPbQTa75Vf3/6DAtOKaoREiLEiyIJ0SgYeypwSkvGVhZv4dmcz/2C9JKQeLnvrfSMbs+VL3/O7hYdIkLAw5eO4bLhvf2219gdfLt9H3mVNQztmMmoLr5W1un9/qkqOidpBJ9t+jdRZ1jT/o/izjVLWJSr3m4qCcGrEaOYP3cdxYU1eA0a0PuIsBUJarIMyPqmhyeldzHpAwuDGqHhcaOYlXUdFtdx9lS8RJ37COG6DHrE3EmUsXPjcVM+/5i9ZQ3SPgr39l3BBe33+IUvGrqkwsKuITrmaY5UvcOhqjlBP9vp0XDXLzPx1LeemWp1uFtaCI0MKY6gnm5zKAp0iCjjonbbaG8u9xubhIwJD040eFRWAQp6BiQvJs6UFfpD/D5Poah4FF7Z3wNSAJtbx6x111DqUPeGF06eweDkdCosVj78dRsHS8qJCTORFh3JwMw0hmRnNGbZ91fPZ1v5S0HHEWfoxYT093jr8GK+zl/T7N0TWF1aFJXFrCJDxYE4ZLeGW4YO4OI+XUgxxWI6hYalJbyyzM2fLGL94eN+04hGCJ6aOp4L+3UPeq7H7eWBqS+xb5O/VpxWr+GV5f+gfc92AFSU1HD1kMfwevwnao1WYt5v/yY2MXjo8oEdc1lflqOaYNQLLd+OfKRRDqgtOK11ukcshcw9upwT1hK6R2YyK3s8KaY49p4oYeZLn6meE6bXseHZv7UpQ3h0fyG3TXnJ74uVJMHNj17ElKtGnNRYTxfyi6tYtGInHo/MheN6k50evHFj85GjXLZmoaqHOiixHQsn+aTF77zzE3bvapJxscVpcMRp8S/SVWg3MJ/U3oECk3qh54W+b6CXWlcKOFRRznmfzkUBXj3nc/rGBzfkAJFRT7O/bh0VjsBywcaRKbDgeH9+zu3FhPTOrNuXh6OlCkOyHULUqza/Fgo0CEWkmaqYnbWWgZH5mFqc60FwwhPR2JEGYNK2Y3i7FSedoVYUmTrL+9TUvgjUAAI7A5m9tgd51uCNFo8MHkVdiZN31gUmd7SSRLuYSD6ZPY14cziy4uHbY5di8xYFHCvQcEHmIn4ozuGtI4tUxgdWt8/rbfgdwFIcjqPCRGz7KowRrnrHV3BeykDu73bpn8JCtmzPAe5buEy1Ckan0bD+gZuIMKo/iys+28Ccuz5WbbVOaBfD3G1PN/7tjuzN56mbPqA4z8e5m5wexyPvzKZ999AcuhXOOm7c9Dqljmo/mXed0PJYrxmMSjq5lWwoo3vK24A7mFN5vNe1Adu/3rg76Dk2l5s6u5PIsODEKA3I7prKwp1PseDNn9ix/hCpWfFccv2oVr9UAFnxIJAQp5Ha7vl3VrJoxc7G3xcu3cbIszrx9H0XBhx74kABj571OOaZyVgG1sfL6imxDFoNb4z2xUtzc0v9DC6AK1JDYFeEIH9LOqX7E8kecZzoNCtaNPSK7seszOvbZHABOsXFs2j6lTz7y5utGlyA2prHMehnhjxGCJicvhu3KZqL4y5k3T7/+yHc3SaD23AtBXw1XxJoFC8jovJVE4laRSFTW8dRTxSgQRJ6esQ/fdIG1/e5EpERNxAZcUPjNpvbRZ5VPezQAKvdpWpwwSfzc6Kyhvu+WsaHV1+KJLRMyVrIzvK3OFizABlfGWK8oQ8jUp4iTJvAvGMvEK2zodd48SoCq1uPQ9YhhCDBYKTM4URRwOOSqMuLxGPXk9ClDJ3JN8n55iyFpUWbkYTg/m6BGnx/FN9s3xe07FBRFNYeOsakXl1U9y+b90tQbouKohoKcktpV6+116FHOz5Y9y9qq3ydfpExQURRWyDOEMH84fexpnQP68tysHoc9IrJYmLKAGINoeP3J4vTxr0QZgi+dBH4YrmtweOV+eiXLSzZsR+DQcOVj4xjYp+ujV1qwbCncj7bKz7ErVgAQbZ5NMOS7sGgCV0p8UexaccxP4PbgDW/HeK7n3czebT/7PmvS1/G6/aSNreAuq21VJyfgNekwby3jhVzHiLB5AudrF4VnO1KDS6LgQM/dObRK8cxdVhPZMWLS3acFJl1r6Qk3hhnxtIadQYADtJMwyi0BLahNkeY5CZSW8FK62cImnn/UQ6IlE8qNttgeI3Czeu9vwxZuaFBwUw6NZbuROlmEGkYEHjg70SYTs9V3foyL0e9fNKk0bJlrzp/QwNkRWHr8UKqrHZiwk1ohI7+CbfRJ/5WAD8pqH01OUTpfWRBQoBGUdAbHHgVB+VOM4Pi2/Ng1yu5YN4nHKzwHafRe9CZPKrfz/eFm+ld2JWzh3QmPPzU6aXJIQmBFLwhlJu9QZLu4LtntyOw6qCtxrY5tJKGMcl9GJPc56TPPanP+VOv3gwXDOrOx6u3+XX7KIDbCEQIej33KjpJYnLPrjw9aVxAh5bV6WLcs+9TY2/KQv5j4XLm/7aLeTdNCzgewC3bWXx8NrXu5l6UwlHLT1Q6D3Fx1id/qtf7zhe/QLMYW3PM/Wqjn9FVFIWi3CZJ9Ig9dUTsaSpM/6nXWi6/z+fparSBYzbUeLHHCxVvtwm9shP49Nh/2F+3uXFcmaZuXJ75IOG60BOQonixWheGPKY5qiytZ8YBOhhL2WyJ4ImJF3L/ktXIcU4wBTO4CpMTjtA/sozDtiiWlrWn2tO0OlIUmGTeh0HyhjbYCqxYGcf3GzKADcAGbrlkGNdOGdLm+wuFJ4aPQ1EUPt7vP+GGabR8PmkGD88PbIFvCY0kqHM6iQk3ccKay3tHX6TG7VsyR2pjuK79XWSFd2LOoVeBpgmmGckcsXor2eEp6DQall17NUcqKlhz7BhLqzZSgnoDkqIo/OeNJTz7nIYLJvXjrttDc2q0FZP7dGVj7glVb1cIwYiOmUHPHXf5MA7vPKFaoWIwGcjoGjwR+FfEaTO6nVLjmXpWdxb/tq+xhddtBK8ZGgyAW5b5Ztc+dhYUsfSmq/2qEe757Hs/g9uAnSeK+GnvYcb36hywb2PpnBYGtwm17nwKbJtoF35qXrTmWFO6i9cOLqJ8XA26cfjI148a8K6NArvPS6kNwfWpAJ4eBuwXmvF01CO8sNR9gAmOOmKNEUydOoB5c9f5nWOs8uKM0NSzcgVanPYpUXxacSfuFiKYx+05vHjgRu7t+i5hIQjka2oeQ1HUv0s1WN2BOmQtIQTEauuXgVFWOvUwc7BWvZuqc1gF8/stxdisVu7BDpu5de8Y1lW1Q1FA4/FyTZeNbfKQc475U4C++dUGEmMimDTi1HQzPjliPP8aOoafThzhYFU5vROSGZGWiVZIXDNkI5mxy9BpvdhdOr7dNZhFuwbTfGI26LSkRUdS5azgpYOP+CV4aj1VzDn4KDdk/wOnrM5eJgToJIVzE5sm9g5xcZS6LDx7uILEUAUDdt8Fv/1+O9nZ8Uyd0v+Pfh1M6tWFL37byc4C/0S6JAR3jR1BbHjwktMJM4ez5L3V5B1qca5G4p7XrkLTykr3r4bTOtpHp4/j1RsuZHCndDomx0GkUE0Y5VZUsepQbuPviqKw/tBx1WsqwOcbA5fwsuLhSO3yoGNRkCl3hFZz/T34uWQ7T+35hHJXTWPWXUggtXeim1YOGt/L07W9fy2oEIKoOF/4wHZdFHX/jMfTzwiRGpQYDfsTqzh/1TMctZQSFRXGtOn+CgBCgajjLiJLA/PzXdMTmHj5sQCD2wC34mRdWWAipjls1nkn8S0IvFLrta6KArZ6gvESm4P8eoVek+Smb0QpV6Xu5e6sLbzVYwWLBizBWO/BNvzTSwqvdl+FUbjRCQ9zB32KUdtKI4cCVrue3MJAlqznPg7kE24Op9fGwdqNrCh6k+8K5rCzajlOb2AlTQP0Gg3nZ3fmjv7DODe9PTpJw/7CUXRO+g6DzoskIEzvZsbA9bw783XCdL4Jx6DV8NB5o9BIEl/nz1XNqCsorCj5KuR4tUKDQ/Zd02Z3UVFh4fnta3A6daq1zYoClElIzZQ05n68PuRntBU6jYaPr5/Gvy8YS3ZcNNEmIwMyU5l37WVcOzx0eEdv1PHaqoe55pGpxKVEExZhoufQjsxZ/gDDJv512pDbitPKpyuEYET3bEZ0zya/uoaxb3wY9Nivd+1lTGdfG61XVkIS3DTsc7g9rD6US5XNTt/02JCk0wDh2tCctycLRVF46/B3uAmMQQkBSpiM1MGGciicu68PZEG7772befC+V3CODgdt4Gwko3Dv1o/5auQ93HTTaPr2zeD1136ksKAKxe1Ba3WhL/ZiygGPWUtSx3ief+Vq4uOMPLF3Rsix76hazfgU9Q6pBh7ftkIIM2lRD1Jaekurxx60pSAh4XZF085QxROdV9MzotzPGwjlueqEl5FRBYxI2EOyqa5VL1cIOJSvvhwNRuBS56rgq/wnKWoxSe+uWcHSolfoHzOJcck3IwkNLq+Xr/fsZf7OXTg8Hka1b8/o2HSee2U5KWlbmX1drt8YG36OMLp44eK5vLb2Ie4aO5KzO2YBcNQWfMVQ5ixEQkIO8pwLBHK5kZm3vU1hkW9Cy5+iQISgpDySpIRavzJgWQbtfP92bos1+IrsZKHTaJg2sBfTBp58TbveoGP6Hecx/Y7zTtl4zhTOGJV8tMkUks0pXN+UeNNqJJKjIiiuUSffmH5WHzYePcH1n32D2+t7ACUBt00Nx2hQPwcksiLO/d3jV4NTdlPuDN7uLAQYOnl4/srpZKYFlhQNGt+bYdrR/ODYE/QaJ2xNCguDB3fgw9rv0RUH3qPO4qFyRzGL31nLjPsalq7Bv2+tFFyh1pfZ14DKZKKGmJg3MJpGkhR2HiW2H1SPURSo8Rgo9URzc4eriXIf4LL+X6ETbatYaIAkYEjUcYbEH2/TeYoCBWXqa2u182vd5bx9+AY8BO922l61DI3QEeOZzJXzF+BulhQ6VFHJu/JmEiwy156/Meg1hIA4s40ProonOiwLh9NNWYWFMCmCugCBTB8itFEMbzeaBflfBuyTkBgbPY6bbv3EL4llKAGPGdxuHQVFsYSZHOi0XpxOHfJxPUleDbKk0NBxHvs7ElL/H6FxxoIhZoOeAenqHockBNP6+c+Gj188tjHGG2600Tszl+zEIjokRnNut2xmf/J1o8EFHxfO95t6oCjqLbFjUp5EJ7W92Lkt0Eta9CL0PDZpWB/6dfePJzodbmxW30sdnRxFyDayZjiSU9hYjxgMP3y5GbM2Bk2IcSkK/Hhcz9QfPqDAov6Cm8KmqW5vCb1hJEaTj1Gub9JLDEz+gBjDYOcCZTkAACAASURBVLSiyYNSFCh2J3FMvoKne/2Ts+MH0EPzEnrp5Axuw7XOTt3X5vMURbD8N/Ulac/2gSGHlUVvhjS44AtVba9ayhXzP/UzuI0QUNFLwmh0tzrOwvKFXHrz24yd+QqX3/Y+6+YagrY4T0qdxqTU87m949+I1DYlQiO1kVyXfS07PlYCqgZit9NQI4aiCKw2E9U1ZuwOAzE7JRSNQDZIeAwSCnDrTWdGvPH/ZpxR0aR3pl/EuDc/pMLaFBeThOCS3t0Z0M7fIA/vnMX714/laNWtpMQWNjvexJc7XapZ0cMFaSxap+ea0cepdechkIg3duPc5McJ15/a0IJvLBJT2g3ly7y1QY+ZmdFEcVleUsPjf/+YQ3sLQIHoODPZ4xOgtwwGlflQUYjTNyW7yopqEK3wg7pdHnSSnmFxU/ilPLCES1GgzqVjf2USUMzIJW+y5oJbSQv3X2ZGRz+Nx30ItztYs4xEePh1REY92rhFlhX25ibyza+XUGWx0TU9kStH9qddvP+1FdcmX5vUSUJRYFdNDPFma5umKUWB95eMpqQyRnX/3VeMCth2yPJbm8bi9soYDS4sdpWJXAhkvcJvm7px/nlbQhreX7flU1zWtHKpPRJF6aYEEgf7N7iMiBtH7+hBAAyI7c+A2MBk11N7AjvZNG5B6vcKpeeAp/7PoLFD7CYJnU1q1lOv0H9wNiPP/muR2vzfgDNqdM0GPRvuuJFFu3NYsnc/0SYjVwzow4D0wEYHRVHQGO4lNa7Qfzt20mNfIilyOiW1gS/ToYIELsl++k+7h5a4ocMkci1FbKsKjMVdlTmOJJNvjC6nmxsnvYzd2uRFVVdY2PZZHQarE+fZRvzkEeoZ3S+u6Nm4qX3XlICWx5aIjPYtD8cmX4EkNKwr/wav4kFRfKuB/eWJ7CjNoOFtk1G4e8O3zB/nH98VwkBC4re43ftxOtaC0GEwjMXj2YPbW02pbQ3l1lXY6nayoHAAW6p01NmceG0KOHRQbGD70UI+X7uDc3pk8/J1FzSrrxbNOhzajmq3nm01ZiYEKbpoaE1uuNd5hy5ixeYOqgbaZNDRPTvFb5vVU9VqXqARwoPHG3rhuHxbP8aP24JGE4S83Kvh103dArbnr8ygaH0KSYPKGDO8C5f3mkGkvnUqTYNei90eGKfW1wnafQ+vvngFb366lt05hYEnC8HB4+plZf8ffwxnvNZCCMFFvbvzweUX89LUiaoGF6DOuR27+5jqPknyMqZroMoCQHiIpow/A3pJy4v9bua1/rcxNK4bKcZYBsV24a2Bd3Jth6YkwNL5m/wMbgMEEP6lBvOzlWiOunysam4Z6ZCbqAfKSLM0WZiktBiGjO6mWrfbgOvu832mJCRGJFzyf9g77/go6vSPv7+zNcmm95BAQu8C0nuRKorSLCAqKvaKJ556h57lTj3PE9uJXQRFFBRFqdJ7701IICG9J7vZOvP7Y0k2y84uQULz58cXwk75zszuzDPP9ymfDw2Dx1Jsb8HR8misTi0togsY0WQf8cHlNfvsKPJfvK/TtcQUOgWT6U50uhTshHG4+FlKrIuxuU4iybu4OeFjuoRvxaVzQbgMDW3QsxylXQWKUFizP51nv6wV69V14FxvRatLYvrePmRaolRJXaonPk4ZjlRGc8eWSXxxqhHXjfJN4khC8Noj1/t0pVU4Aoduzjxey4aBFTTyo0OYPv1OzGatF/uXooDTKdi1uw2HDqszv7ksOrJXJyIdbVkngwswepT/qgCdTqJ5swTMZv9E+7aLpMDy/w1XjCZzlSPdL9u9EJAYrp7Aur9P1wt5Wn7RJiKVVyLu9rt+8yo/XWVCIHRaDLvtGHZ7y5ILSdBxgHcd6VOvjufzt5byw6wNOBwuhEam7c3ptLw+E51BoVS7m4yKJ6hwJfPCgRkoKERozbQML60+HHqNhSGND7I1pyGHihLPSlpejYKqco7m3Y2uFsF6td0an7SdtcXNsMp6z5Q1zgndylC2hrF4x2Gev3kwQQYdQuiQIv6DXPIQYKdaZl5NalxRQBZGnMGPUqKEsSP7MBObbEGjIsteoUg40RATXE6LsGwyLVHsNOQy95Xb+ezHzRw+WUCr1DgmX9+dlHjfWZKujm3SAJIETZJy2fWbH8FHIXCYFMrLQvnr0w8QE1PEyGs3ExNVRlZmLNs3tOC3/BRVFzi0WSkp151EY5TZyW4e2PEhMfo4xiRP4KoI/4b15rHd+HXVQTJOnnEfCXjkgSEY9Fr6dm/K8ZPqXLFtr7CmgysFl4XRVRTlrL3vQbo0BBKKnwx6+6QBBOm0WB3OGqmT0R3acFePy1M8Lyo2FFu4lsrGwTiNGgyFdsLSLUhOxW+dwci7BpDU2JtgW6vTcNeTw5k8dRhOh4ufTtxOpXKixtBZXUWszX2WQ5VJKCQACg2CSlXLljonnuRwURwDknwbTc7EvIxtzDw0h5daqTczCKBX1G+sKGztvTBEgf5lKKd0rD74G8M6uKfTkqE/ImYRsmUOijMdNE0QtlU4XeloRXU4RKJMupXY+GcwCD1fjYQym5VKS3dslXd4cfSaFVHDKCYE3NV0Hctz27KzMIet5ixeuHfEWa8xUp+IQOP3njsTshLYW9fYqHHBCwuimPX+ULyU53UuCPF+JCOvKiT52iwfW1xoz+fD4zMYmTSaYQm+PB4Aer2WTz+4i0WL9zDnm42UlVWR2iiWyZN6c3XHVABuG9uDbxZux3JGK60kCf5yX/10o11I2Bzp5BdPo8q+FYSecNNEYsP/giTOzuNyqXDJjK6iKHz5xXq+/mojVquDsPAg7rt/EEOHtVfdPtTQkSBdKhaVTicJI50a3Me2aTFszciiwmajU0oSMabLt9xF1yWBIuvp6agQWEK1WBoFE7u+iPbNEsmoLHMTLysQHGrk/lcncM2EXlQ4Knn76FfsLDmEQNAzpgP3Nx2PQaOn2LWPSo6pFj+0CMnmN2ssGuFf9lAAqWEVvN49MKfssYp8/rnvZyK0gessDEKlUaF6hwYO3sldyjA8MUyhbYQm7K81nxXlCSTrUhTbSoQUiTZoLHE6b1KUcIORcMMQdlRo0J42uk4vBng39BoXesmBXdbx8vZVjG+m3l9/OK+AjemZNIoKp1fjRnSKGMn20h8CXKUbWmHgmgY38osmH6sKV4BRqyX4uBVhVxCKgnD5fndaJ3Rom8L2fdWdfwpJQ0/5TbzJuPg5ewF9Y64hWKt+rwshGDn8KkYOV79evU7Dtx9OYfq/f2LHvkwURaFpaix/f/xakhIuL5LyM2F3pJOR25eaUkbFRmnF+1iqfiU14dd61TWrT1wyo/vPl39gxXIPIXlZWRUvzfiZf3+/FicKaUnRTLt9IC0auTubhBC0S/iSg/kPUG7bQbUvaNCk0Cb+A/RatwfYo3Hd9dEOHs/jlU+WciyzEL1Ow8i+bXlsQv+zEuicL0rNVXy2bgc+7qakUNYtipf/fQ96o47inBJCwoMJDnVnxCsdFu7Y8nfstWRFVuRvZkvxXr7o9jJHSwNrOcXqKihyBn4RvdHjOkL1gb2EBSd24pBdFNhNWGUdQRrf2J8CbC9L9T+IgJOOIr7N2MbYVPXZiBA6RNC1EHRtwPOx2I8A4rSx9Q+95MQu6yizW5EVxavNPKesgtEzZ1Nc5fHcBfDOuGvRi2XYFX+dZwpGoSfW2JRBDcay6zENMzdv4d2Nm3HIMnqNBkVRuK9bF9Zs2k6h0z9jUGJiBP99fjzb9p7ki283kVOWi6QNnF2scrr456bveLH3JNX1dpedrzPns6Foi3v2EdOdccmjMGg8oZNQUxD/eb7+mcUuNLKLHkKtdtzuPIzZugJTkG8DUm1klZaRVVpOWnQk8aEXj4f7khjd8vIqL4MLYI3U4gzVYDutsLvvWA63/X02/3poJAO7uKe7ek00VyXOxe4qwu7Mw6BtgE5Tt6TCmThwPJfJz8+pSbhU2ZzMW7aLrftPMvfVO373tdUFi3ccVo9OC4HdIFC0EhqNRGyydxH/B8e+9TK41ahwWpiXuYxkQ+AHVEZgl7VuBkSVeKkkBK0jzh4DL7abq0+YmRl9eKTxr9WnD7hn0DvKUsizn53F7a2DS72Mrs1l5auTH7GjZBMKCho0DIgbwXUNxvswolU5jnMg51pkqhAKfpnFqmE/3XJskDReBldWFG74YBalZwgtKsCD8xbxwwOv83PhU9hkc82aToZ8uhgL3AEMAYKdVOWvxxjxMg/0GMD93btxIL+AMquVdgnxhBoMdNcn8NS0uarnJgS8+I8xCCHo0r4RXdo3osppYeqewIKKAD8fyaCBfgv3dfX+7Ryyg4d3TqPS5TH0v+QuY33hJt7q+Cr6AA0xVwJsDvXkOUCZ+Ru/RrfCamPi5/M4nF9dhifo1bgh742/Hr32wpvES1K9sG6tdzulrAFnqHodzd8/+MWnc02vicZkaP27DS7AizOXqBadZ2QXs2aHf0Xi+oDZGliavdKPyurWYv+daqvyt9Is7Ea/6xWgwB4KCDIt7m646uuv/vua+NvrxLHbO65ZjU+5uyKF5w9fy3FzNFaXhmJ7ELNPdeX9E/3POg5AmdPTZuqUnbx8YBrbSzbWJE1duFie/yP/OfSC136y7GJfzmBk3J6p4htRqIGiwN6SBthl9wN1Txtv47Q5I9PH4NbGV5uO82iLOYxMmkrjkM7cGGqle1AB2moeiOrjyFlUFd+OuehOQKZNfBw9GzUk1OD+Tjt3TuPBB30NQVJSBPO/e4S0NO/a8SBtMCGawFyusqyhpCyEd7b41hN/k7nAy+BWo9xZwY/ZvwQc90qHCOBPjvl4DofyC6p7RJAVhbXHTjB1wcX5Ti6Jpxsc7P1gO9UaAU7D7nBx+EQ+LVPrIhZYd2Rk+y8H+mHlPvp2alqvx6uN67q0YsYidSIRvVZDdKg645JW0vjtxNVLWhKCOxJlaEGxzZfIp23EAywt3olVtlHhCuJIZRxJxjKMGgcmbRQ3pzzOpu12Zmz6EqfsYkTnVtzaryNBBl9vaHBSK944EEq+1V3En2mN5uXfrj3nbjJw61NVY3fpVkoc6pn0E1W/kV55lDSTW+sqv/IzvL8Md/qxOolaDUUBs1PPjENuY9c4NJInOvTxHrtYvQuvGltOZqEROtpFDKJNaAcq8z4LuL3Ltgq7eRYG0x01yxYdOsSn23bikl3c8a8hJJUbsFjsdOjQkIYp/im/7m3yGG8eecmrcqdGaFIWbN/bBBBYHL4zoPVF/hs71hRsYExyYBHRyx1GfQes9u2q66JCp6gu/62gyO/vvezQb1gdTi9V6guBS2J0e/ZshiSJmhZFcZb68zMzq/UBSRLIqqp9oAtQ91ofiIsI5Zr2TVm+x9ujFsCj1/VW5QYGGBzfg2+zlqmuuyHZ3a45LOUj9hR9yMHSb3ApdkK0cXSPe4akkK58FD2exblrWZqzBptsJ1zfjrvSxmMv13Pr87OpsnsSX2//uJ6Pl2zhp+cnE3XGS0Anafl+wIPctvZjjlXWvYA+QlPJ7cmbaBOag1Y4kVCQBOzN/JwQQy8OVaonUauxOHcB9zd9CoDSqg1nlJSJ08IRp82T4i4821zQhLcODcJxOrTgVGSyMgrIP1VGszZJhEeG0Co+cHdiSoQnTOKoChw3d8OFw/xZjdEd++UcduV4aAn35uWTEh7G8rsn+/2tq9HU1ILprV/nu6zZHCg7SKXNRWlZMAUlEeQWhOOSNW4RABUVbIfdqer9V8sbXelIiJpBRu4AwHvmGGIYiNHQQXWfNccy/I6nAGVWK0bdhY3vXhKjqzdoeeTRobz138UoCmit/q2uViPROq1+vVyAHu1TWbvzuOq6W4bXn5LAmcjLLOL95+ZxbP0RolpEUNE8DKeiEGkK5skb+nJtZ9+OpGrc1mgkawt2kGfzrrtsYUplULybF1gjdHSMeYCOp1UGaiNIY+TGBoO5sYFHVltRFIa8MtPL4FbDYnfw+IcL+fwJX4ayUF0Q3w98iCPlefx0chefpfujAFQYFH2IW5K2YpTcx/D1iB2Ybatop11Flq4tWQ51z8/u8oQAMi0RhPvYK4Fcy+N968Ag1hd4VzucKihlyg1v13xOTInkP5/djU6S1HkTgCcG1tLeE3Xj61CUCk4Ul/LUosXsys31WZ9ZVs5/121gat+z6/rFGRO4v+lUHC4Xrd+egUNxgQHQAygoikK/hk189rNsNKL0sKgrMp9oAOdPk3tJYdCl0ThxHQWlL2O2rkEjhRIV9jDhIf6lohpHqbeAVyM6AK9vfeGSVS9cN6oTzVok8Nkna8g+VYIxwcSuPG+vSQjBQ+N7Y1SZ4p4vXrh/ONc/9hGVFu9YXq8OabRvdmGKwnNPFjKlz4tuDwQw7sjDuCOPBo1j+Wj942fdX5IkPuwynSW561mSuxGtkLi+wQD6xP7+p2d3eg4lAQjV92T4CiLWRvOweJ5oO5QdJSfYU+rbyXZr4laGxe1HI87uWgkBQ8P381lhL1wqqr3do/vX/Htl7gBuSPpOtYGi2r3bWHBGiEhRMOR6x2dyMkuYOPTfzJxzJ5Pnfu/jAA5u2ZRsuZQpC78ix1qOAoRqJvFIo82MTjzsJ6QicbC4BRO//xRZU53h893qm7376mR0q6HTaPh89GhuXTSv9mWCgMVZR9l+KgvNSSsul0ybq1Mpmmckoq2EFCp7JTnlIomKX4LAfwrgsoTVWYxdLiVEl4xGuDtNddpkkmLer/MYfZqmYtRqfUVQgU7JiWjPMvOoD1zS5oiWLZP412seL2rvb9m8MXslWXllJMaE8cC43vRol3pBjh0SZGDR21P4ctFWlm48TLBRz6SRXRjQpdkFq+979+m5NQa3Gs4QLXujYcAz75GcGMXEPp0Y3L5ZjQT3mRBCMCyxN8MS60f5uNRcFXCqGYB90wtf9L6b1/b9wvzM7dhcTsK1QXSNDmdY3IE6GdxqSCgk60s4YfdWTA7RmOgc3bPmcwNTNLOP9WBCk40+hldRYFFme+QzDLdwQuxG3ySmy6mw8qMtLHvmTl5evIot6ZmEanU80K87J4xFTFnnXXFQ4TLy8vG+LMhtyayOP/gYXpcSzFPLazWY+LmdHAG0v/xhyYnf/I5360ezaDzbHa+UNAIpLIiSv8URPKwCQ/cqkMG6PoSqJSa631D/hE/nCpfs4GDxq5yq/BFFcRBubE+b6L8SpveendhcpWzIeYgSm7viSaChZeQ9tIpSj9sGgkaS+GLSWCZ+Pg97re+/QXgoH906+vwuqI6odwn2iw2by8G7R1bw/cnt2GUXbSIa8Pf2o2gU4l/m/FJhVNpj2GvFp60JQRT2T/KpdWqdHMesh2+6KOUrheVmhv3tQ5x+hAOTUqqYensL4g0ptAzrGJB3tzYstp0cz78ZBfPZNz4DlZopzMs7jkNxIBC0MLXjrsaPYNS6p/YVVTY2ZpzgkR0/EmvM469tfyLG6K6jrXJpMSt3073BIzy54WfWZKcjEIgyJ3FLrZgy1A2dwajjrsk9eX/qLOTTFKGuCC0Z/20SoANE4YGU/dydlgGKDRBojQOYtrwZS466qypkSXHXCKmMcWObVrw+Yvg5fTdD537K4ZIi1XXCodDybU8iUjboQK9Obfr10r8QGX3hYpc7C06x6lQ6LSNjGN7Il6lMUVz8mjkEmyvPa7lAR68GX9cYXkVR+OXECKwu3xBN26jHaR6pXp98NlQ5HCw5eJTssgquTkmia6PkenW2AkmwX/ZGV5ZlPpu5msU/7cLhcNG5a2Mm3z+AxKRInLKL0atncNLiXYkgIZjd615aRZxdlv1iYnTTJ2pIbhQB2WMbo+jUpzOPX9ubyQO7/K7jVDis5FrKaGSKRq85u+H+17xf+XqNt+SR0VRF9xv2oDN4RB4Vl2DPl62oOBWGUa/l3lt6M354J5+bNaf0dQorAsuQB0LLpJ3oNL6emMslc+fMeWzPcLNiuUJkHA1dGPU6qhVlX+g2mJtqdZtZHHaqXE4eu/E9CrL9E8xrNALnwXSvZWUDwymcFH+W4l+IWBPCv6aMYNDVzRFCMOi9j8ksdRMIySie+WRNbZlbeHLrg/cTFnRu7ap3LPqOVZnpvisUBU2lTK9fMjBFWMlJD6fKbMAYa8Jqc9bMWIQQPDF9FEOuuzAyN2aHnSELP+KU2UOgZNRoWTBiEq0iPS3sx0o+5nDpm6pjRBk70z3xMwDyLdtZl6POYaIRRkY13lh/J1+PCGR0LwvuBX9wOl1MGvsOBfkeftFVKw6wcd0R/vfFFPZImT4GF9w3+tO75vFD/8cu5umeFT1HXMWKeVsAsEcaUFQkearx9Ybd52x0Kx1WJqz9mN8qThd9A4OTWvOfzuMCvsX/MrofPyzYSVWEAAmEUOhx4260em9VXqFRaDfxAJve6IrV7uStz1eRW1DGo7d7iK6dzuLzMrhhxmtVDS7A3R/NrzG4ABqzhHRQEJUQxEvjh9AlLplgnTerXLBOT7BOz6139+etf/hv5xUqdbouYx3iezp3dGbazJ8Z3Pk3Xrl7BO0TE2qMroRAdip4RTpcCo23llFxQylhTX2J0wPhqW59VI1ujLGCVzt/S9K4Uqqr8BQFDmxqRrNm77B1XS4RUSEMG9WJkNALx0tw69KvvAwugNXlZPTPX3BwwpM1y05W+qpdVKPU6ml6yLOs9budS7HWibflcsMlp3YMhC8+Wu1lcKthszmZ+fYyvs/c4XffE+YiHHJgkcKLjYdevRlTuHuK7EfQogb+9LoCYdSv79UYXHDn8Ffl7WHs2v/wa+5uquxHOZEznKNZSaf/JJNdcCfrt2xAX+oiPMNJ+HEnKboiH4NbDSFBfEfPVO+bn3dSUUtHq6By5jmftwcSDWP+p7qmyu5gy3FfNWKBoDi3ioa6CC+DW+VwUGnzxG+Hjb6a5FT1qgiDUYsjK99necjuOoRGauUgf91xlKXbDjNtUF+vTSQEkksgOQWSA7QVTpyLj/Ng92cpyik5+zFqoXVMHNO6edcZo8i81fUrGsSXImk8wp2SBG16HMWU+CCTHx7EmIk9L6jBdckyu4vUE69VLicrMmvzpvifYUvC8zsG6/54TGeXtdFd/NMuv+t2bMsgSBNA16vW/y8XGIMMzNrxCjfeN4gEKfDN36dV2jmNva/0FLlWj4ehlRyEB1kJ0jvIsebz8dEZnMzrh91VO4wgY7YtISZhIiaTh1sgKDqwGGFoomdbBTic7jFYih/F4bogxnSvX68lq9h/aABg50m3B3y8qJje78yk/Rvv0PHNd+n4n3dZfSwdIQQffv8okx8dgiksCEkSBAXrGTyqIx/98BioZLMNp+xQJfvPJioQdMRjIFyywvw1e0kIC2Xm+FHuh+sM4lxNsRmDsZisD9I48t8GPPjD51Q46vadVVXZ+fzztZT8nMcbif144uqePNypO+81iSTKZK55SQrAKCQMCIQAhzMds3VVnY5xPqhwBJY1OlrmiUU3MKkzowEkma6r+Xda2I0IP2YqytDhivNy4bIPLwQQUtRK3NmkD6vz1WXU24Qno5PO4k5eAhiD9UyZPpop00fz5BeLWLL7iM82GknwxMg+Knv7x+YCz5RToBBicHp5qo8lr/T7CtJqHYy8dhNfz3WHCSx5gWtRK7K9SXPCTZ4XSFTIRIoqPzqnc69GXNg0v+uSIgPzOLRPTqDYUsWwDz/38qEq7Xbunvc9X08cx9XJyYyf3Ifxk32/29TWyWTsd5e8OU06Cq9vjCM+GG22C2cTFWOigDZHQl/g/eKvnqH0b9qYNXfczg3X/R1bigldsRV9XgUFL6Vg1wSBcBNR7g4u5brl77F4yMMYAzgRGzce5W/PflfTEr9i+X6CgnTM/OgujC0Pkn+6ySpK0hIqPPe9AuS57FjtuzAFXVi9szCdAY0QqtJZAP0beLiG08JvI7NiHjaX9wzDoImndZTnPthVeoDN5c3oHHrYK9+sk8LpnvDver+Gi4HL2tPtM6CF33WDh7WnQ1QjRiT5UtYZJR2vd7rJ776WSitvPT2XMa2nMbLJE9zd/2U2LfXPa3Ch8M8JwxjYprGXMYwxBbPgyUnEhJ0bLWXbWklDg9Y7NKEXDsK1Vr/5ICGgR3fPy6v8WBhOi+Tj4CmKW8osZ7snDqnTSjRt5InBGvVNCdIHikWrGxaTfjCa00k/RVE4klXAvvScmrKqEIOeDo0SVfeNCw2haUIMz/681O+k9fEffPvqS4sqeWri/xjRYhony2VEfAyuIB05d7fBkRAMkkCyaBG5Wnd7W3Unlws0uYLgA76zlWs6e0rFImPCCD9USsziTMK3FFB6T6w7tnsGu1yetYIFJ3b7jFUNu93J35/zGNxqVFU5eObpb9DrWiIEhAsNoUKDEKLmjyQECRo9Wuns5EPnC0mSmNRCvWa8SVgULWsl0nSSib7JC2ka8SBGbQOCtQ1pHvEYA1OWIknu+yCnKo/XD7/LCVsYi4qv4qAlkZPWSLaWp5EU/gZGrf/26csZl7WnO+mufqxecQjzGcX70TEmBt3akbvmz2d7Th4uXQRh8Q6iQ40Ma9CWO5r0IVirTtxSWlTJlAGvUFHqmSKfOl7AS/d9woMvjmX4hJ6q+10I6DQa3po8ivyySo7lFZEQEUpanK80e13QLTaNMJ2RcocVjca7s0pztj5rwGBwotFIuFwyKII9/21Hu0f2oTe5iV8VBWSXYM8XbU6zy7jxwiPX+kzxmsTNJ7P4ccostZMlehpGf0yIoRvH8kZgd3laoCOCx5IS/V8ANh88ycPvzPcqYbtt8NU8Provn04Zy4R353Ig+7R3pCgIB9h3lDH0lrfI6eTfh8it8CZ+MVdUMbH3y+7rPQ2h11M6rIlb173WNekKDChlOkSkCywyugItmnIJccbcIcSoZ2w/TyuzTq/lmgl9WfrFahTAkWb0WwnxnjViAAAAIABJREFUw8nd3NJYneLy50W7fVR9q5GVVYzD2glBEOGS7He6HczvD/ucC6Z3HYxGkvj04HZcirs/sEd8Qz6/xtcJ0kkmmkfeT/PI+1XH+jT9q5p/W2U9+y3JNZ+/yVrE1VEXrnP0QuKyNrrRMaF8+vV9zP50HWtWHkSn0zB05FU4rgrhuq9mI1e/+W0CS6WeUq1E1xat/BpcgM9fX+RlcKvhcsp89PJCBo/vhlZ3ccMSceEm4sLPv2byu/73MW7VB1hddrSSpx5VWwejq9NF8tgD17BoyR6sVge9ezTjjm5PU+DIIt18kChdPLvWudhfuQWtxk64yUiXlimkxPkSXQshaBj9X5So16my70MIPUZd6xqD0CzhV+Zt3svO9FO0SIpldFe3bllWQSn3z/jOZ7xZy7YTHRrMpMGdmffoBB554Ws2H8lEsrmJvwEsVXasVRrQqxsd6Qxj9O4/vvcyuNWwJhhUDaOwS5An8ffhA5j57QbKzpBlT4wO44tnbiHE6F098dh7d3Fk+zHS92cFTDGYtP61/HJz/RPyKArYbU5SYmchiieobiOEAOeFZc6rjec6D+K5zoNwuFzoVDgh6ooTlpN+1+VXFVJaWUWEqW5t2ZcTLmujCxAZZeKhqcN4aKpbYPFkaSlDP/vMY3Brwep08tKqVXw/Qf3mA1gbIDnncDjJOVFIStP653q4GEgKjmD9iGlszv+NJ3Z/WLPcKp+9oSE0eATDh7Ygpes+0iu2Ikk7OWa20Cx8EEnBqQC0Ggl6WWLGBysoL3Ow4tQBVqw4QItmCbz3xkQf8nchdAQbvOtBD2Tlces7X+M8zXPw067DvPHzOt68bSQzv9/g9/zeW7iBSYM7U1VlZ+feExhbWBAN7Ch5OuTdJkAQkiVT2Vj9Ie+Xlur1ecPS/arbSQ7ZLbvjx2O0C5lf33wAh9PFpgMnKCwz0zYtgWbJ6mVuGq2G/237FyvmrOOZrGVUJKuPfW+Lvip7u9GxUyrzvtmiuk6v1xAXHwZ0xS404Ee9WKPvgKzImJ1WQrRGH27iC4HzMbgA4bowylVI3xUFSgsVhjz+Pn2uasLLU0acrtW+MnDZG90z8cvRo34D9QD7870D8zmF5fywdi8l5VX0v7opwk97LYDskjGFX3jCiwuNbnFNebHdBKbvnYOMgkPR4lQEugDtuIpuDJ8cHU7tUp68qr1sLnifmxvPxqAxcfBwDjM+WOG9oyRz+Ggu/31vGU8+PFR17FzLYY5XbiJIiuL+jw7jlH0fxsdn/YS2wn+Swe504ZJlvt21HsMDuR6vsZUV+lVgnx9FSJYReyLYjTWdAO4QhBPy52Ugj5WRTvfW+5uGRxwwU9hdnadZr9WQdJpxTKfV0Ke9HxHKM2B1OlgSb8ZSFQmU+6wfltSaLrGNUBSFtVkZ2F0ueic3wqh1G5IuXRoTFxdGfr7vvvfeN+j0tWgRxpEo1oVqZ84zR2xsLHYnqASCAXHtmd52wmWd/R+dPJIZR2eqxukLNsYjK7B61zH++sFPvPnwlUMkccUZXbvTqerlViNY53njfbtyF69/ubJm+/mr9xDWLR79snRVh6BFh0ZExgYmjb5SMDChPZ2jm7Iyby+VTisaoxbsn6puGxr8FD+celJ1nVUu5cfMRxmb+jEvvOp+oDVaJ9c/sJqkJu6WU9kl2LVqD+BtdM3OYr489gAW+XQDiwLjx8OWrc3Yt9+Xr1jWgxSA393msvOFbaFbqaE21wKgH12M7b0EorbZsZsUyptoQQhMp1yEnnRRDnw1ZyMTJvYCoM/wdiz7zpeLNSjTiqatCZfpDAE4WcFZbuPzO78g6vVb6dTXt7UVwGp1kHGikKTEcMLCgqlyOOj6yQdUOk5fWG4wGpMdTZCLhqYInus4nH4Jzfju8D6eWrWkxqEQwGOde/Bo515IkuCDDyfz0ovfs2N7BooCwSF67n/gGkaM8CSSdRGv4ijKQXFso+blKUJ5JecGNhZ7wgsKCr/m78a228E/O9zp/wu/xOgWdTXXxPVjef4aFJQaSsrCrTFY8zzO0fo96ZRUWIj0w0N9ueGKM7r9Gzfmf1u2qIr/aYTgpnbu+GBhmZnXvvzVJwNf7nIR1ikBw/Zcr3WRsSae+2DyhTz1i44wXTA9Q9vwy+ZDLLMMZ2jno8A6r20iTI+yqsS/7AlAke03zM5CcvPdtbK3PPML4dEeykCNVqHToIPsLXqHdtEPAe4KhNnHH/QYXDgtawNduxylqCiCnNxa/BgKBBLTNQXpmZWx2D3MGc6ZEG4To+lYiVgZjL4KTAWnKzgS7ejGVYJe4fvDa7hV6Yksy0y4rz/L52/3ugecBomSzjEYKsGGgqu6gEQBXbmD2BW5FFY4efaWd5nwxHAmTvUoCsuyzLPT57Np87GaZUmJEcTekuwxuACyhKvciKscMvJdNOgWxZHiQqauXOx1TQrw5raNtImJ55rUpoSFBfHa67fgdLpwOmWMRt/ptBAG9DFzkZ3pyPYDCG0ypXIqK/e/pPqdri86QKG5ghOF5USbgkmNDkx7eLEhhGBy4wlclzSU99cvYuWWY5hPhiI7vGdKsgI5ReV/Gt0LhXbx8VzbogWLjhzxomcTQJfkZJ7o2ZNyu5U5y7f7rWm3RBuZ+so4Vi7YjiRJDBzTmSHju13WU63fg7fnr+OzJVtrpmef/NKDXm378tyEcnTaMMJDJqEIA8V5gaW2FQXe3Psl5lQXjeQSL4NbDSHgcOmsGqObZz2K2aVOzALQretBvl/oXS+rJh5cjfZpiewo9VWCrn38xLZB5NeKfmivL0bT0ZPwsjTNYdTiR7HekIviUECnRdvII2Ra3jYSKchFbJ9cQlIsIIHTIlGwKZ6g2UVoXJ6wxez/LuHo/izu/fuNJKXG8sJLP3gZXIDsnFI2Hqjwm9xzKjKf79lJeqX/rrRXNq7mmlTPrECr1aDVBo6VSto0JK27uWZXrv8cBgoMePc9XBV6ENAoKoLPJo8jPuziiTTWBbHGGK7Sd2LhMf/3U4OY3y/ddbFxxRldgFeHDmVA48Z8uWsXeZWVtIiN5Y6OHakUVvosfI8iq8VNjt5JELpHi8bp7UI5XTIjJvRixIRel+gKLjyW7zjCp0t8RQ3X73MwY0EHXr7L7aVVOvKplroJhKWnsijuHUmnBF/O3GooOFEUF0JoyLP6Nn1UQwgID6+svSMAmgBGt2OzBhw3FHLcnO13m2FtOjFHuxunU0Y0tKHpaPN5OTiDFTSvReB8vAQcTpzpJwhvkoypQSyFoRINxx5DY1Q8RfghMokDcyjNDYYFDndv7emL2LL8AFuWH+COp69l7Xr161XO8h4/UVZKeoV/o5tn8a8eXBekhvhPCiuAq0pTo7SRUVTKuPdns/qpKRfNAamosLJj9wmCQwx0bJfi94Uy8OpmvPLlCtWKk1aN4gm/gqoYrkijK4RgePPmDG/uKUTfmp/JXcvmeUyHAGeEQkkPB9Fr9V41lWeW9fwR8epXK/2uW7L1MH+fNASDTkuQJgIJDTL+LZ6swClzBAjItwUushenu6HiDL5KBrWhKJLHzsugMZ8Ooao861pJMLxLS0q1cWwpPuh3zFsaX0P83xowfdaPhA1WbxsWQkAHA0jQ4cYChj+TRVDEZiRhYO2RNLYYG/oYHCEgfJyFyoV61RP85LVfUJqrG7eQbAVzQ1QrFrRComtSMiIfTlX6JskAGoX5luSdC5qEJhKhM1Hq8DbeinLa4Dq9jVxBpYXtGVl0Tks5r+OeDYqi8LeXF7B2oyfWrNdree2FMXRs38hn+9BgI6/cM4JnZv6Mq5bCR1yEiQ+eHHtBz7W+cVl3pJ0Lntuy2NdXE4AOqhp4x3/vveHiNUBcKhSV+9YiV0MByi3uYnmNpKd95Hj8FZEqCqzKbl6zfk9pMk7Ft1sNIMboKVZPDPYvOwSQFtoBbQXoKkBfDhoXno6vWmNrJYlpNw8kKTqc1uFp9Ixuqzreo03H8diGBTyYvZDCQQoi2uWv6guAHg/kM+b1EwRHutzOq7DRu/khxkSrCx0KHQh/FM1+ZH4Aog7K4MSXv0EBRVZoERrN33oM8FvCO73XQGTl7HXWgfBhl0cI1dbyBBWQbRIl29XL3HZn5akur0+8PmOxl8EFd+fd1OfmUVZepbrPoM7NWfT6PTw2ri83D+rIjEdvZNHr9xAcdHYF68sJfxijm17uX93XHu9CEgKtRuKu67px8+ALwyV6IbEy5wjXLH6HVvNfpsvC13nv4FqfttDakCVwGWr90XkKNrSShkiTJ+nQNfYeOkTdiiQ8yRlZAYcsWHGqOdsLU2uNLHj/UF+E8J4khWhT6Jv4ttcyk9a/OsGwBlPRyiC5PGo2QsYt8CtDYmQoD43qxY8vTWZMH0+X1wvt7uKJ5jcRow9HL2lpZkrmzQ4Ps+ZUIUtOeab4dqc2oOrFkIcya9i4aq5MQCNDESZJ/aHHz2IJ0DvVydE1dkhZ6iBsRym45JqXirCDphgemfMT+07k8t6Q673EJQ0GFympdqbsmEnHRdO5Ze37nDT7j2kGQnxQJIv6/YNn4scT/KOE/t9QvDEeXOqPf+vEONXl9QWHw8Uvy9Xb7l0umSUr/Lfkx4SHMHFoZ568ZQA926VdkXmYKzK8oAa9RovDqVJvJKB7izSmDe5NqJBIbBR72f9QZfYsDpT+iNVZQtPwQWwtNPLXbQtrHMAyh5X/HljF+rzjzO5/u8/+v+7+DeXMCIoEih4UO0wZ2c2rkUEIiW6xU+gcfQcWVzGy4mLEL+9QYAvGpVJSkG2JolPMIvTSDsod6SQG9yXK6OvZNg95kG2lz3sRlSgKHCxKRt88kp5tUlm/L8NrHwGgwMdPjichSj2UMSyxG91NmymtmIuiVCBZ5pBf0Q5oVrNNQUUojQy+sVJFUdCdqELbyL9Fbh6Uyw5zWq19gCpQKoXfrrIuyVEcsjspKDHjMghkSSA5FLRWF7qcUuK3lWBPjqqJ8dYOdz3/46/s+vvDDEtrxuHiQg6Wn+L5/d9RXivic6Asm3Fr3uWXgVOJMpwbLwe4jdkHNy9AWFzoAF1TB44InU/YI8Sgp0fThuqD1BPyC8sDOgwZmf5fLqdOFPL528vIyiik7dWp3HRXP6LjLjyvRH3iD+Ppjmvczu+68te2M/Wqadx71V8YHXc3s/+5IOCPfimxrfBzvk6/jT0lX3OkYgk/Zz3FsfJpSCqp/a1FJ9lW6Nsq+eI3y30HPu1Otmkcz+ThXVWPrZH0hOoSCNc3oH301aoGFyApOIwm4TGkhA6hTdS9qgYX4MkNW5h1oBu/lcZgdugotASz6HgbVmQ25aujO3lh8lDCQ7xJYyQhuGdkN78GV1EUsvNHUVL+LxSlFHAhy9k812IJY5M8KidWp54yi8HH2xV2sE7131YLYHb6Tlct7xpJaKzuAWp1Gm59ZCijbu6CI1yLbNCAXkIO1mCP1EG5GXMz9/WI0//Vhs3pJK+sAiEELaNjmXl8hdphsLocfJW+KeC5+8PmFfuxWjxOSfzKIiSb7KGdBIw6LXOn+Co/1zciwoMDOj4d2iSrLl/+407uvu5N1izZx/HDuSycs4lJQ1/nt4P+k6uXI/4wRveZqwfRyOSddBBA6NYyqta6fxRFAYfNyZcvzWf+DF/WqUuNYttxthd9xpmdG6G6Km5spM5CNTfdl8i9uNLPPFhARERQnTz9v3UaQqjGN+EYqjUw/5o7URQFl3U99qJJ2AtuwFn+JrLszUdQZLdQbg9icUZbPtnXi68OdyW93B1ymHtsJ5GhwSz852QeG9eHLi1TGNq1BTP/Mo77RvmPuVusy7E5fKsyJAF3p25CW+vllFseTlZJBJVWHanBSUxoOJz5A9/gfyv/g9Mcq1qwIdCw55dkFOdpkp98geXlIFx7dJjNdvrf0MlLhddgMtByQGtmfbWR9+esq1le87eAqi6NOJs2p+501t4pu8iu8v9SWFugTmV6NqQf9CYX11bJpMzPJXZNMeF7K3iqR3e2PPsATeIuPHNXSLCBHl3Uu/mCg3QM7N/aZ7nT6eI/z33n8xJ1OWWmP/TFhTjNC4Y/THhBJ2lYft29/HjiAHN/20WwVo/u25Okf5zpMyOUXTJfvjyfGx8eVtMWerGwKfckb+1Zx6GSfKKNwdzXtgejG7dFEoKDJYtQa5UTAlJCStBJThyy90+mE74lNhpJ4PLDShUVWrepaVJwGMtH3s8H+zby7eFdWC1OInKM3NmiM8GyFkfJQyi2xVRbLpdzDy7zTHQxi5F0vtnnM6E/Td9nCjJw25DO3DZEnWHrTJSb/T9gEtAjKp21RdVhBoHZZiDR0IB3O99bs11a24YkOReyJ3sgslL7BSU4tHgStk8rsH3K6Wvz3D0VpRamzZjEtBmTyMss4uN3lrN6xUH27MrEEmsAk+90vfqzpko6c7gahAcZiApxx9glIQIW8IVqf9/9mtYq0acyUCgQkmXFlG3j1oGdz5sr4Vzwt6eu46np37J3/6maWWdstIl3VTg8AFYGYForyq+gsrwKU9iVUTb2hzG64M5035jWlhvT3BnuMbdO8ZsVtlZasZptBIdenB/KJcvcu3oBy7M8Bf4lditPbljEwvT9fHHNzVTJgaVbgjQOL6MrIRid6ssnPLB9U5bt8m0kEMDkQXUzbgAluRZ+fnMvRlnCiB6Q+fTgFtIzFjJ90hKEj2mw4Si5E0PcrwBEG4IpsqlXUdzX+ndWkJzFYwzXeXv5Aninp6+0tkGbwNXJ+8irnEWFdSNGXSpJ4Q9xyLIOWFFr71pj1fqYX1jJ6hWe8jVF558kByGQI4KJ+jWH4kGJXoMJAf+bcEPNppKQ6BXbnHUFanW/CjrtOpblxDA48dzKpLoNbE1YRDDlJb6/x/CbumFQ6XC7kAgy6nn71VvJyS3lZFYxyQ0iaZDovyOuvMx/NQ7459K4HPGHCS+owRTh36vTGXUYQy5eqcmsIztZeUqdXm9NTgYbc0/QJNQ/s7+iCK9YowCGNGjJ1dG+9ZSv3DaMFJUOnfuH9yA1vm58vYqi8Og/v1P1Lvq32+VmM1eD6ySKyz2Vfbu3OglJg+BwVVnuuiA0ZJzfdUII7LSqqYZoGhbDoqH30CxcvYpCkiQSw26nedz/aBj5NFrJxKjJfbAkGinoEklhpwiqYvU1dr5dV8+UeNYHq7zG0pid/mV9gBFjuvLPaTfx5c03MrBlE5rGRnFjh9ase+peOjT01gF7scNoog0mPG8Yd8lD07ACTHoby/O/ocJxbtpqGq2GGT88RnJarEfWRxJcO6EHD7984epcq8xWCk4V4/JT3ZGYEEG3zo0DGlyAHgPcIQdZEriMOlxBOmSthAIYjLoLqv1W37jsPV1ZVigpMxMSbMBoOLe38ZhHhvP+1C9UDcfYx6+tU2jB7LSxKT+DKEMIV0U18OFlrSu+OLw9IDvaRwe28NGAMZi0CVQ6c89YK+gYdQv3tWjBsuzDhOuM3NOiJ/0Smqq+4fVaLT/9bTIbDmbw07aDRIQEMWnA1SRE1p3M59jJQkor1L0Lnc7h16mTXQqKbEZooHt8I+YMmsBTm34iy1yGRkjcmNqWf3RRZyOrC0KCRqDTNMXh8n2BaTUtea/vdOyy+wE31EF+vjYcLhcT3/+G4q6nX0yKgjU5CG25g6Z7qnh+pocc5kzPy1DhwBF5OgZ+xpej12m4bVwPkhPchuXqNmkEQpQhhHe7DePlfW+RX2XCILlICy/CqHFfl4LCofKddIk+N/md+OQoPlwxjfxTJZgrrKQ0ibtg3NFVZivP3/Yeeze4fyedXstt00Yy9sHALef+kJQSRXhyJMVlp2cyQqBoNSiywt2PD66v074ouKyN7sIlu3j341VYT+tOtW3VgFf/NgZTHT3Ua6dcw561B1n3w1YUl8fg9b+pBxOf9Z5y5pdXsiMjm8iQIDqnuY3ro5vnsTT7UM02wRo9n/SZSIco9exqIJxNtK/SYUcIiZvSZrE8ezonzZtQkNEIPR2iJnB19O30iBc82qZ/nY/Zs1UqPVulnvO5Alhs/um+ftnamh5tTqLGkmm1CDbNz2XATW6+gO7xjVgz6kFkRfndL6zaEEJLcsKv5Bc/gblqPtWBUoO+Lwkx7yGEOGdjW43pc5dxorBWEuv0+TrD9QRNSiWoVqVF7wEtST/qaSIQMoRkWTAnBsFpQyYAo0HHPx67tsbg1hUuxUZSSAVJIb5q2ACSSiy/rohrcP7ENoVVZg6XFBAfHErTCN/k2319XyQ/y1M7b7c5+PgfCzAY9Vx3V/9zPt7GdUcoLrf6FlZrBLmFgUMPlxsuW6O7bPUB3nhvmdeMbe+BU9x634f88MWDdYrhaDQSz815lIwDWWz+eScAQ27rS2S8Z+pdXmVl0v++4Wi+uzZQEgKTQU+ngTGsKDjkNZ7FZee21Z+zYeRUQnXnNp3pk5jG/OP+i76HN3LrwWklPcOS/4msuHDIFnRSsN8HzGK2sWLxXg4fPEXT5glcN6YLGpUkxO9B89Q4NJLA6fL1zjfsS+O39EiapZV41d/KMrz9dGN2Zv/Ibr1Mw9gIruvRBp1GUy8GtxpC6IiPfhtF+Q9OZyaSFIFG4wmb2G0Otv+6H0kSdOjbCkPQ2du+FUXh5x2H/K7flZ5NVlEZydHue2fcpN7Mm7URa5Xn5SQ5FcKyLLz52T2UO+3odVraNk+qqUw4F6SGtEAgoaiSkgtahqlrkV1oWBx2xv0ym/3Fbt5qCWgSEc3sITcTF+wmytm55qCXwa2NT1/5/ncZ3c8/XO133c8/7uCeBwed85iXCpet0X3no5WqIbLSsiqWrTrAkAFt6jxWautkUlv7eqeyrDBmxpdkl3q8CVlRKLNZWZF3SDXi7VBcfJexizuada/z8QEeb9+bpSePUKnSwBGhN3Jzsw5eyyShwaDxHw44fOAUT9z7GQ6He8q5lN3MnLGM19+7nTbtz79v3qjX8cht/fnPZ74cDooi8cT1bbn5jnRG3pFLUIiL7Awj77/SlLUpPVESdGQsd7fTvjL3V/59z0j6t/flzz1fCKFDp2vM5nVHWfjNzyhAKDJrZq9FOR1SEpJg0l9HccuTIwOOVWV3Bgz/KMDRnALS5Z0crjhEgjGB976/i7ee/YW9O0+gyJCSGsPTL42hSYsEv+PUFSZtOL1jR7C2YBFnZg8Hxt1IiPbS8D4PW/gpJys8swEZOFpaxC2Lv2LF6HsAWLvQt4yxGlWVgWd8/lAaIJHmcp5fm/TFxmVrdEsCfMkrNxw5J6PrD1uOZ5JTqjJ90yoB9awOl/nvTVcUhTKzlSCDDoPO8/WmhEbwy8jJTN+6jNXZx2se8F7xjXiz93UYz2FK7HLJPP3olzUGtxpOp8xTD37BguXT0BvO/6cdN7QT4aYgXv94OZWnPbrE2DBGNknim+93MPftZOa+7XmZFY9pgxLkXTblkhWmzvyJX166m7iI+qUMVBSFJ+7+lAN7MmsvhMQYxKkC90dZ4fNXvie1dQN6jPDf/m3UadFKwksQ02u9yco8679wnvAoLf/I99z18n28Fn1HvVzPmRiZOIkGxjSW5H5NhbOMcF0UI5Mm0SY8kNryhcOewhwvg1sb6RUlHCzOp1VUHFHx/mkWJc3vm/F06NSIFUvUZ4ptr7qwHXT1jcvW6Op0Gh+jUo3I8Pop8zqYna9egeQMfGNcHaPuSS7dephX5qygwmIDAW1TE3jroRtqxPNSQiP4ZKA7+251OtBIEjrp3KeeB/dmYTGrx1wdDhfrVx9kwBD/HXrngiG9WjGkVyusNgcul0xIsIGS/DK+e24ussvTiOAKNeCKCFItm5IVhfnr93LftT3q5Zyq8eO3W70NLriPbzSghIUgys3uZQp88sJ3AY2uJAnG97yKOevU+GcV2g35DecZTGwKCp+kf0DLsNaE6eq/FVUIQaeovnSK8q+fdjGxI99/55eiKOSYK2gVFcfIO/sx542fVbs+Ow9UJyw6G26/ux+rVxzAeYZXK0mCqX+99neNealw2ZaMXdPXP0vV+FF1rzUNhMSIUJ9Yo1brJCTYhihR/2qCNDpGNfStjV2/L52/fvwz5Rabu8BHgb3puVz/7CeqHKBGre53GVyAqip7wDbm/Dx1msDzgdGgIyTYncCMjAvn/jcmoatVTSIbA7+/T+afW4lTXTB/doCW2Ajv6XdhduDWX4CnRvWjTbJvq29kfBU6g7oDICOzrWTzWcf+IyA2OCTQBJAWke7SvIiYUO57ZbxP3iU6IZynP7jrdx07MSmStz+cTKO001RvAtKaxPLJV/cTHXNlcS9ctp7ukw8OYd/BU2Rmex5WIWDimG6kpvjj2Ds3DGjVhBCDjgqrnRCThc59DhEaXuVuR5c12JytWJYlUE7farEGE3P63alqLP/1lXoMutJq56uVO5l4zdW+K38nWrdLRgihanglSdC0+fnHFM+GkVOuoU3PFiz+dCVFOSW07N2Cfx085LcTrmuL+p8C2qwO9RVCcGZpRXIdFJ41ksTXj09gzYHjfLpqO2XmKga0acLVV0t8lumflN3qsp7TeV8OKHdU8EXGV+wvP4Re0jM0YSCD4wegk/yXZQ5KbopRq6PK6fu9t49OoIHJY/yun9yfzgPb8NOnayjKKabXiI70uq7TeSV6mzZP4KMv78PpdCFJElIAkdnLGZet0dVpNXz5/l1s3HacJSv3ExZq5MYRHWncyD9d4LnCoNMy696beGTe57Trs7OG6k8IQHIRpjvMe70mEiy1JiU4iqQQ/7Gq3GL/3uXy7Ufq1eiGmIzcMK4LC+b6ynInJEXS0U9fezUK8stZ+tMuSorNDBzSlta/M/GW1jaF+9+YVPM5a0EIny/f5rOdVpLApaAoSr12DnXv24KfF6jw3yoKmD2daULAva/cVOdx+7ZuTN/Wnu+wyFaIlCkhq7VoI2gddv5GGzKPAAAfYklEQVT5hYuJTMspntn7D6/rmX1yHmsKNvByu7+h8VMtY9RqmTvsFm5d8jVmh70mNNcmKo55Iyb4bJ+UGsuUF8bU+/mfTa7ococINE3t3Lmzsm2b70P0R8Obhx8nz5apui5MG8UzrWeedYzej7yDxabuefVpl8ZbD92guu58MHfWOr78eC02qwMhBF17NmXa8zdiCtCdM/uTNXzx4WovLzm5URTvfjaFoDqUVgWCoih8unQrH/6yGaujmjEGJMdp8qFgAwteuZOIehIQtJht3DT039htZzCwyTKkZyMUBUOQjifenUy/G88v+fTBsXfZXuL7kmsS0oxprZ47r7EvNh7a8RdKHCrhFgVcC8Oo+E6Dwain77VXccfU4UREeydAnbLMhpwTFFktdItPIcl0ZU3vLwaEENsVRVGNg/6/N7pWl5kX9t/ppx7SjVfafYMkAk+L3vx2DbOW+XpdAvjoL+Pp2LTB+Z6qX9htTjRa6axTt0P7T/HYPZ+oduh16prGv2bcVj/n43DS6/63kRXFJwbYLDmWr16on+MAlJWYeX36AnZty0ABOnRO5cnnb8BSYkaj1RDfMLpevGtZkfk+6zt+LViKXbajEzr6xg5gbMrNfj3DC4kD+flsOpVF48hIejds5J5N1AEFtiIe3/VXFD8kFrIZSqd42PoMQTo+XPIUsUnnJxv0/w2BjO5lG164eAj8QAZrQs9qcAEeG9OHLYdOcjizwGvk63u1uaAGF6hzedgP87b4ZWrauS2dygprQC+5rvhh3T53KEFl3dGsAmwOp1c53fkgPDKEl2ZM9FkeGVW/5WmSkBidMo7RKf65Hy4GCi0Wrp/zJblmj+ZZiE7HN+NuplXs2UNvZ4s/CyOIMBml3H3P26ocvDZ1Dq9/9cD5nfifqMH/e6Nr1ASTEtyUkxb1RMmg+Lo9ZEII5jw7gU0HTvD9hv0E67WM7XcVbVIvfFKrriguVG8pBXdssqS4sl6MbkFJYAVbq81Rb0b3/xMURWHknFnkm81ey80OBzd9O5ftU+4/Kz1jUlACekmHTfbT5l0tmVQL+7dl/P6T/hM+uGxLxi4mxiU/iFEK8WH07xI5iJ7Rw+s8jhCCHm1SefWea5l++9DLyuAC9BnoSw5dDUkjEZ9QP1PIwV1a+F2n12kIC7lyGKEuJ2zLzqbgDINbDYvdwdqTJ846hkZomNxIPbyjyODM0KCYvc3C5aqycqXiT3cDiDU24KmW77CtZCUZlYcI1UXSJ3YkMYbES31q9YrBI67isw9WUl52hrKEgPETe9ZLFxtAs5RYWjaK49CJfJ91D4/pc0Vxn15M2FwV7Cn6H4XWvZh0ybSJnESU0UOBeaykyC+dsIxCoaVuxC+947rjUBx8lOEhhFfsoNgE5g98k5zGYP8J1j1b0/lh9gZOnSyiZbsU7nhksE/i7U944/99Iu3/GyorqnjxmXns2nYCRVHQ6zVMmNyPm2/vVa/GUJYVXvx8Kb9sPIjTJRMWYuDRcX0Z1ad+OuX+aMgxb2VlzsPU5lkQSLSPvp82ke6yvE1ZmUyc/y2yyjMrgCW33U7TqLrL7ZQ7KlhdsI5TVTmElIbx1R2bUKp8idvvfnoko+/q57P/3I9W8eV7v3p3jgp46pWxDBx55Slu1yf+rF74Ez5QFAWHw4VOp/nT87zEcCl25h0biIxvyaFAw/WN5hOiS0BRFPp++jGnKnxrwjsmJPDdTbfW+ZhWq53NOzMINuq4qm0Kep2WgpxSnr7tA3JOFLq5piXBuLv7MemJYT73SFFBObcP/TdOlVZ9SRJ88vNUEuqBQvJKxZ/VC5cQsuJAICGEBkVRmHdgH69tXEtxVRVGrZZb2rTnmd790FxkrTYhBHr9nz//5YAc80ZkfNWeARRcZJnX0CLC3VY7/6ZbuPOH+RwsKKiRXBuYlsb7I0fV+Xivvr2YRcv31nzWaiWe/8t19O3enI+XTyM/u4SSwkoaNokjyA939fb1v9UwuZ0JWVZY/N1W7njk9xGW/9Hx51N3gVBpP8ihgmmYHW6OVoMmmUNFd/Hiurya6WGV08knu3ewMeskP90yqV45Z//ElQO77L+qBPCqIY8NCeGnW28j31xJqdVKakQk+nMQlPx49jovgwtudrq/v7qQOe/fTVJCBHFJkcQlBfZSNRopYIItO1OdT/dP/Fm9cEFQ5chkR/a4GoMLYHNlkfp/7Z13fBR1+sffM9uzm15IQhJK6CUBkd5EURGPoihFRTk4xVMOvVM57/ydgqeeWLjz7kQsqCgIKkcTQVGKtFCki9TQA6T3bLJtfn9ECMvObDYQKcn3/Q8v5jvlu0n2me888zyfT+gUmoad8dl/X24Oq48euZJTvGxOl+Xz6k9f8btNH/DOwZUUOK4v9f5riWhLB7TrxSUSrL19tsZYbbSIjKpRwFUUhbmLfLvqoHJ1+uVi7VSiy+Om2Gk/H2i79GmpmZaSdRJtO1bvCF1fEUH3V+Bk4UwUfOsgJQnG37BS5QiYv1/bVeJa44fM/QxeM40vTmxme94x3j+8mttXTSW92LdaQVA9wYYEGttuQy3wNgu5C5uhdppryuwOnA51tTSAIydyfLY5PS7e3LeEfiuf5/bVU7jp+//jrQNLsYWYGf9ndUnF4BAL/QfX7xdp/hDphV+BPPt6zbFISwlmvYNyl3cZTpD+ylpgXypuxcMz2+fiuah4yelxM2Hrxyy/edJVmtnV46w9na9P/4fs8mPIkp62oX25JXYsZp22G/XFdG/wAhGmNuzN/wiHpxizLoIOkb+nScjAWpun2WRAp5d9NGnP0aGdr/DRC7s/Z03WnvO/bYfiZt7x9ewvPMX0keNJaBzFW1MWkplRgCTBjb1a8Ie/DcFqE7XYWoigW8v8+YdvaGR10SZSVc8bkPAo3gOyJHFvm+ujlGpj1iFcivpqKbO8iBKnHZuhdkTmrweOFu/gsxPPc67Uy6242FnwLeklW3m0+XsY5cCCjyTJtAofQatwdTU0RVHYsPsoB45n0Twxmh4pTdDXUCZRp5MZOqAD85f62uno9TIjhniLAp21F7A2e69qbfDOgmPsLjjODd2bMeubZ3C7PZWKmlf4hfD1iAi6AeBR3PxcuJG03CU4PHbahPSkd/Q96C/SHn1y1VIWHd5Hs9BOtIlUV9lPz4/H4a46TkJiaIvWdG1Yc4fhq0FOhf+XPmVuR70JuoqisCjjDS72MAModuWxK+97Okf592YLhBNn8xk9ZTZlF+gHh1rNzHrhfhpGa8uNqvHY2H7kFZSyesPB8/nZsBAL/3ppJEEXNUEcLcn0+7JsxZmdpIY3Bqg1Q9T6gAi61aAoCp8df4nDJTs49+X6IXseaTmLeKLle1j1lX/0+eV2Fh/eB8DhwlgWH+nAkKaV1i+SVKlyaNCFcX/72Zj0WSw7fIgws4kxqZ3oEt/wuqmV7R2j3eKrkySiTFfHMPFqUOzKwe7Wvgn9XPTDZQddj0dh9OTZPrKhhaXlPP76fBa9VjMnBr1OZvIzg3nykTKOncwlJiqYeI327zhLuGYHHIAuACEogS8i6FZDeslO0i8IuOdwKOV8duzvPNzsDQC2ZWZ4jS9M78zqU635TZOdhJnsxFn7MbLNk+hkC6NTGjI65dd70VBa4WDdgaMU2Svo1aIx8eG1p3caZQ6mW2Qym3LTfcZGN+kVkCJbXUGu5utj9uPmHCgbdh/R1Gk+nVPEsTN5NI6LUB1XI9Oey+wTizlReobmwUkMC71dc9/GthjiLeFk2H3Lv/SSjv5xvrZVguoRQbcaduav1NQezSg/hMNTjlE2E2w0IUuSl413QYWN2ft7ATDz9rvRyb/+Y/eafUd4cvZSnO6qvGuHpDhmjR9e4xygFm91fpCX9ixiWcZOPIBBkhmb3IeHm99cK+e/XrAZwokwxpPrOKUyKnFj6GAWvr+aFZ+n4XZ56DukE3c/crNmw4Eah075VhQAeAwKxU1d3LFiFrIk0SO+EZO73eJlmXMxG7K388bBmef/f8J+mjXZW3i+zeOkhLVSPWZG50cZtXEaJa4qSUidJHNHXEfah15fLrzXCqINuBo+Pz6VfcUbVcckJP7U8kOCDRG4PR46ffo2BRW+eqUhRhM7HpxQbddZ1skcVny6Hr1RR99hXYlr4muS6I/s4lL6v/oBLhUjzB7Nk3h/XO1apzg9Loqd5YQag+rto2ZOxUlmpj+BS/EuEWwX1J/5w3N9DDGtIRbeXf0ckbGB5WLX7kjn6X8v9rrtu00KuZ3clQWfF2SljLKO5UPH0CzMV3/BrXgYvvEJVcshq87Cp11f10xxuTxulpzaysrM3YQZrQxJ6ELniGbXTUrsauCvDbh+flNqQOfIAZpjRtmCVV+ZD9PJMp8MvJcgvcGr2jLMaGb5sIeqDbivPfwuo1s/xaevLOSjyfP5bcok3nlmdo3munTHPtWAC5B26AR5Jb4NDC7nYQpyHyD7dDNyzrSnpGgaikZ1wsUYZD0RJlu9Crh2Vyl7CjaSXlwp1B5lSuSJlp/SN/oBGlpa0SK4O2ObvkXGzIaqDsSlRXb++5fPA75ej5QmhNm8n5CKm/oGXACHx83fNn6nep6NOdtUAy5AqdtOhj1Tcw56WcfdSd14u/MjvJx6P10im4uAexmI9EI1NLGmkGBpxSn7fq/tEjJ3xD7ilcNMiY5l24OP8+3RQ5wqKaRDdBw9Gzaq9g906cxVrJzrvZpWFIXFM76jy4AOdLqlXUBzPVvgp7JAgiNZeUTYqqT7XM7D5Gf3h1+EVhTFjr3kTRzl3xIR821A16xP/O/kO2zNr2pu0UsGHmr8LM2DU+kVM5JeMSPPj62cP0PzPFtW/hSwSadeJ/PJ5Pv5w5sLOH42D0UBRziaDWybM9W9/oqc/oXl61Mu/mojftLVIEkSv0ueSp+oEZjkICRkwgwxjEh8lg4R/Xz2t+gNDG3ehgkdu9MroXFAX6zZryxS3a4oMGfq4oDn2q2Zdo5NAmLDvF/sFBdOAhVlK7frJ8rtywK+7pVEURTyy9azL2si+7P+RGH5jwGLbJeXOdiRls7W9Qexl1XU6Lrrsr/yCrgALsXJh0dfotzl+wShZYtU+RlqdGliI0P48pUxLHn9d8x9cTRBJu1GGq2njnah2lUnBklPnLn2XLYF/hEr3QC5OfY+bo4NXDqvJhTnq7sBAORnFgZ8nj6tmhBpCyJXJY2QkhhPQoR3HtHl8C2SP4e9dCZmS+11Q9UGiuJh99n7KarYen5bdtlXRAYNoHX0v/3e4OZ/tJaZ075FueAJu99vUnj65XvQBWDpvSrzf+pzQmFF5jwGNxzrtb377e1ZvUD9fUiHntq6Bf6IjQyBSLg7uS2zD+xU3Wdw09aq2xtZ42lpacKBsqNeq2RFUTB8pqOigwNzUOAv+ASXjljpXgPEJGiX/HTsp22xczE6WWbpUw/RvEHVixRZkkhNjOXtMWrSf36CjaJepnQ1OV082yvgVqKQW7acPPsazePSVv3sE3ABVi/dzT+eCSy/Wu7RFvTJLPd9pB/33FCCVPzmjGYDE1/z7TpTnPvx5E/Ak9UXT+4YPI6fffbJKS/hpV3LySOPCJuBi8sYY4NsPN9Vu4Ik/D0bzkUOlAoFRVHw5HiomGInZ1Yun01donmcoHYRK91rgPGv3sfkEW/5PCbr9DJjp9TMfTbEYmbRHx9k056jzFiYRmFOKY3LrBTmlxEW5B0EDKZ+OCuWq57HHDRSdfvVJKNwpubYiYK3iQzyTfcAzHlnlU/APcf6FXspKbJjDTZysmQxeWWbCDI0omnob9HrqvLfQfpgSl3qTx0JlmSfbZGxoXycNplP3/iadUt3oigKXW9tx9jnhhB6kVOxp/wHKHiU846QnjOQNxRPyIvIv/weZh7YyGt7v686SA9hEXqSdPEYJT13N2vH8BbtMcjaN9J1X27FWeHCOcNXjGn5R2tq/LcmuDRE0L0G6DawI3946yFmTJqDs8KJokB0QgQvL3oaW2jgoinnmLNoC+98tu587jDrTBFrtxxi0vhbGdw/5fx+IWH/JDfrB1C8V3GyLglz0OWVl9krnLy3JI21O9LR6WTu6tOeu/q2x2y8dGEflx/dWZcnX3MsU6WK4ELWrU5DTn0W9wU/h8OF73Jjg+nEBFXWWd8R+wDzT73tc6yEzM0N7lE9b3C4lcdeHs5jLw/XvLaiKFDwR3wseAGKJuMx382xkiLvgPsL5R4XIeEwp6+vBb0abpd2VYrLqS6iLqh9RNC9RrhzbD8GPNiHM0ezsYUGERZzaV1k2XklzJi73udljQK88f733NS1OSHBlSVIsi6YyAZbKSl8EUf5N4AOc9AogoInIkmXnt87mZnPqMmfUu6o+iK/OW8N81bu4LMXRmO1aBsd+sNmakdhuXrNdJi5h+ZxTVvGsWuztl6x0vxVr4BbiYdtmRO4vdF2ZFnmxoh+FDiyWZk1/7youFkOYlyTv2HSXYailvskoJXT90D5Yqbv1169bss9QbnbiVlX/c2sbbfm7NlwUHWs20AhxXilEDndawidXkdC89hLDrgAG7ala74593gU1v3o3b4ry2GEhE8jKu5nouL2YAv9K7J8eW6uk975yivgnuN0diFzVmy75PMmRzyPWq2UhJ7GYU9pHjfmiVs1x/QGBSzq4kQKLs6UVqVf+scO56X2c5mQPJWnWvyHye0+IdHaPPAPoIZkRE0w5xxuZyEHs7RraBXA5dHInVzExLd+i97gG8DNVhPj/q69GhfULiLo1jH8lSMpgEMlGNYmBcV20jNyNa//1Ya9l3xuq7EZ7Rt8gkGOoDL4Sph0DekQtxCDXl20BaB1ahKPPTfIZ7uskxn8QBe/1yx3eTt96CQdCdZkos1xl/IRfJB0sYCv7TmAokg8/sQhzqzK0YzL0SYbNkNgTyVJreL5cNdrdLk9FbPVhMVmpu89XZn10xtExQeu3yC4PER6oY7Rs1NTpn0AaotdWZbo1rGJ5rHbMzLYnJFBx9hYuiVdWl+9w+XSNJ4BcPnJKwZCmKU7XRM343CfQUKPUa/eKq0oCtuPnmbvqUxSkmIZNKorHbo2Zea0b9m/+ySR0cHcO64PNw1M4ZtjU1BU6pUBYq3aq2R/nDx4hkUzvufs8Rw639aeQeP6aZemhb4ChU9eNH/YuTeW9GOh2E4rFHd2oxjxWei/emPghpQADZKi+Pv//lSjYwS1iwi6dYyYyGDuG9qF2Qu9vbAk4K7bUomL8e35z7fbGTBrFjllVXnNMLOZZaNH0yC4ZkpZ0WE2osJsZOapvPRSFApOFrFpSzrduvi+8a+O1UeO8L+9ezHqdNzTti3dk9RXm4fP5vDg9C8oslc1QERYLcyeMJIpb4/22b9F+B84kD/NZ3uoKRWrseZeX0veX8k7z84775a7bdVePn5xAe9t/jsxCb66CLJlIB45Aor+D9ynQDKzaVdXXnytUmNZdkg0mKMj7zY3jl+cexqYbUztfBfdY7RvooJrEyF4U0fZtP0o0+es5Ux2EdERNh4e0ZObuqn3zPebOZMThb7lUDFWK2njx9f42j/uP8ljb8zH4/FU2WcoCpIbgrJcWC1Gls5/ElkOrEGgxOFg8KezOV5QVYUgAd2SEvl42DD0F+haOFwuer0wgzKH78o1LMjMDy+MV9XBOFY0lwN5/8KtlCKhJ956JynRLyJJgRs/AuSeLeCBts+o2pPHJEbyye6pAZ3n8d9/zP59vrlmBQW9Xuaf/xpN23bXh/B9fcSf4I1Y6dZRut3QhG43VL8KyiopIaM4j44Jx0iNP47TrSftWAuO5MaSVVpKel4eyRE1y/fd2CqRHnFxpKWfxG2UQQFDqRtjiYIElJc7ST+aRfPkBgGd76lly70CLlSmODefOMkXe/ZwX2qVruuan49gd6qnCorsFaQdPEGvVo19xhqHjKJxyKhAP6Im65dsUw24AFmncjl7PJvYRtW33KakJKoGXQkJxQNxcdo5bMG1jXiRVs85WnCMv966kFE3bKRtXAapDY/zaM/vmNB7GTrZTXqu+kux6oiyWbHkebCddWHLdGH6JeBC5eI30DbYMqeT1UfUy708wCfbd3htO51frPky0aMonC30bzdU6nQEXA1wMYqiUFxs16xFkIDcM/5rhs9x9z1dMKhUGgD06NmCiMjLqzARXD3ESreOkV5ygtVZmylz2ekX05X2YdpCJwAO6T2CTfbzWYBz/yaF53Jry13c2PDxS5rHoIEdWLlmn6oYTZDFRNPGgQmsFFdUVAZojUhadtGqtlV8NLIk4VHZX5YkWsarX/f93Vt5fcs6HJ7KF31NQ8OZN2gkMUHVBzdFUfj8+x28uyiNUnsFyqCWWA7kYjmY4/XeS5JlkloGVvUQHR3M9Bm/5bm/fEFWVlHl8ZJEn74tefYvgwM6h+DaRATdOoKiKLyT/hkrM9PO26Ovzt5MgiWWaR3+gkH2/VV7FCc55RtVXYslCXo3PUREkHo5U3Wktk/k9lva8s33P104SQDceeUcPnCGFq3jqz1PtNVKkMFAUYW6KthvWnnfVLokJxIfEcKp3KocdXxULu2bHSXCZqFhdDYQ63XMvH27eXnTGq9tRwrz6TfvA/aMmVitw+2HX2/mvYVpVYFer8PeJhp3sJHgbb+kCCQYMLoXweGBr1CbJscw94sJlJZUkJtbTFR0MEFClOa6R6QX6gi7CvZ7BdxznLKfZdqBj1SPURTP+e4qNUz6yyvvevbpO2kRGw5OD7g9SBVuDIUOXCUVPDHuQwoLtEVkziFLEs/d1Bedyp0hxGRifBfvOltZlvh84n10bpoAKAzps4FHhy2ld4e9tGv+I+tOP8TWzEleK/B/bF6jeu1Sl5N5B3b7nZ/T5eaDxZt8V9aShCMxFI9RxmQxcv+kQUx4M7B23Yux2kwkNYoSAbeOIIJuHWHZmR98Au45tuTtosLtK3Kik02EGLU7qsJN7S9rTvm5JZw8kIWx2Imx0ImhzI30S4x3uzwsnLc5oPPc064d/7rzTmKsViTAIMvclpzMmnFjCTX7tuCGBpn56Pf3suKFrnRpm45O5oLVvIfTpSvIKPnm/P5FDm1t3Y0ZJ/zOLSu/BLeGWweSxOiPH2bx6emMfnZItStmQf1ApBfqCCUubU1eBShxlWHS+WoedIp+njUZY1Dw7lST0JMS9XS11z16JIvF83/kdEY+KR2S+M3QGwgLrxTpKSgo0zT1BNi7S93lQI2BLVuQmuhiW94CKjxlJNuaYDP7//M9UjgTrVau/fnvkhB8BwAGnQ6HW31V3zg03O81wmwWvzblUoBlcYL6g7j11hH6RHfWHDPIekIN6k0O4ea29E+cT4SpIzIGJPREmFLolzCbCHNbv9f87pvdPDrmA75esoMd247xyYfrGDNqOhmnKi27GyZEIPupUmjSLLCSMUXxsOjkFOYe/yMHi9dxvHQbqzKn898Dwyl0nNU8zu7O0hyrcFdVZYxsmaK6jwQ81rGr37lZLUbiItV/trIk0Tu15k0ggrqNCLp1hFsadCdU7/vll5AYnnAHej86q8HGRtyU8CFDk7dwV/JWbkqYRZjJf9VDud3J668s9RLXURSF0hIH0179GgCjSc/d93VTPV6nlxk6QvtGcSEHitZyuCTNZ7tTKWPRySmax4WbtL3lbIaqTrPJPW6ma5x3o4FekvlwwDCC9NUror05cSi6i1a0siRxa5eWNIr1v1IW1D9EeqGOYJANTO80mdf3f8Cuwv0oKBglAyOT7mRow/61fr0FX27RbALYvfMEHo+CLEuMe/wW3C43C+Zu+cWMEYwmA89PvZd4P44ZF7ItbxFaaYIcxzFKnLnYDL7tta0jJpJRugIuelnoUeC/CxrxdfgyJg3pS1SIlc8HjeJkcSFfp+8n1hrM4ORWAedgmydG8+XLY/j3F+vYfvAUoVYzDwzoxNA+6itoQf1GtAHXQdyKh3J3ORad+VdzeX1l8kJWf+9rKXOOb9f+1avN1+3ycHB/ZflUi1bx6PSBz+uj9EfIdRxXHZPQMTb5fcKNDVXHz5auZWvWM3gUJx6PgkeRWL7uRjb/VOklZjboWfTnB2kY4atJIRBcKv7agEV6oQ6ik2Ss+qBf1Va7WYtYzbGQUIuProJOL9O6XQKt2yXUKOACtAq5SXPMKJsJM2g3HMRa+zCw8Tq277yPOcv68coHI88HXIByp4vXF6+t0XwEgstBBF3BJXHrgPaabaq/+722OeKlcEPEEIJ0aloDErfF/hGpmpuLTjKyNM3IgWOJOJy+Dgvr9x2tpZkKBNUjgq7gkgiPsPH8S8MwmfTo9TKyLKHTyQwb0YUBd6ZWf4IaYNJZGZs8k3ahAzBIZmT0RBobMTLpdVqG9g7oHLJOu4pCrxNfA8GVQ+R0BZeFvczBlrTDlFc46dS5CVHRl2419GvywhffsWDTT6pjo3qm8tdhtbs6F9RvhLSj4FfDEmSk7y1trvY0quWpQb3ZsO8YmYUlXtvjwoOZeGfPqzQrQX1EBF1BvSDEYmb5c2NZsOUnFm35GRm4q2s7hnZpK9ILgiuKCLqCeoNBr2NEj1RG9KjdnLNAUBPELV4gEAiuICLoCgQCwRVEBF2BQCC4goigKxAIBFcQEXQFAoHgCuK3OUKSpGxAXWlEIBAIBFo0UhRF1QXVb9AVCAQCQe0i0gsCgUBwBRFBVyAQCK4gIugKBALBFUQEXYFAILiCiKArEAgEV5D/B/uivgUWGfSBAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# loop Louvain community clustering\n",
    "# Nutures\n",
    "# 20220713\n",
    "\n",
    "import community.community_louvain as community_louvain\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.pyplot as plt\n",
    "import networkx as nx\n",
    "\n",
    "# load the karate club graph\n",
    "G = nx.read_edgelist(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/peakachu_hiACT/louvain_clusting/A01_loop.nx\", nodetype=str, data=(('weight', float),), create_using=nx.DiGraph())\n",
    "# nx.draw(G)\n",
    "# plt.show()\n",
    "H = nx.Graph(G)\n",
    "# #first compute the best partition\n",
    "partition = community_louvain.best_partition(H)\n",
    "\n",
    "# # compute the best partition\n",
    "# partition = community_louvain.best_partition(G)\n",
    "\n",
    "# # draw the graph\n",
    "pos = nx.spring_layout(H)\n",
    "# # color the nodes according to their partition\n",
    "cmap = cm.get_cmap('viridis', max(partition.values()) + 1)\n",
    "nx.draw_networkx_nodes(H, pos, partition.keys(), node_size=40,\n",
    "                        cmap=cmap, node_color=list(partition.values()))\n",
    "nx.draw_networkx_edges(H, pos, alpha=0.5)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>r1_GGCCGCAGTCCG</th>\n",
       "      <th>r1_CTTGTGCGGGAA</th>\n",
       "      <th>r1_GCGCAACTGCTC</th>\n",
       "      <th>r1_GATTGGGAGGCA</th>\n",
       "      <th>r1_CCTCCTAGTTGG</th>\n",
       "      <th>r1_AGTCAAGCCCTC</th>\n",
       "      <th>r1_GTGCCGCCTCTC</th>\n",
       "      <th>r1_CCTGTGACACAC</th>\n",
       "      <th>r1_AATCTCGTTAAT</th>\n",
       "      <th>r1_GATTTCCTCTGA</th>\n",
       "      <th>...</th>\n",
       "      <th>p1_GAGGGGCTCTAA</th>\n",
       "      <th>p1_AGCCAAGGCTCG</th>\n",
       "      <th>p1_TGAGTCGTCTTA</th>\n",
       "      <th>p1_AACGGTCGCTTT</th>\n",
       "      <th>p1_CGAATACGTGTC</th>\n",
       "      <th>p1_TCAAAAGCCGGG</th>\n",
       "      <th>p1_ATTAAGTTCCAA</th>\n",
       "      <th>p1_CTGTCTGAGACC</th>\n",
       "      <th>p1_TAACGCGCTCCT</th>\n",
       "      <th>p1_ATTCTTGTTCTT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KITL</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMTC3</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CEP290</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4930430F08RIK</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700017N19RIK</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 49300 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               r1_GGCCGCAGTCCG  r1_CTTGTGCGGGAA  r1_GCGCAACTGCTC  \\\n",
       "gene                                                               \n",
       "KITL                         0                0                1   \n",
       "TMTC3                        3                0                0   \n",
       "CEP290                       1                3                0   \n",
       "4930430F08RIK                2                1                2   \n",
       "1700017N19RIK                0                0                0   \n",
       "\n",
       "               r1_GATTGGGAGGCA  r1_CCTCCTAGTTGG  r1_AGTCAAGCCCTC  \\\n",
       "gene                                                               \n",
       "KITL                         0                0                0   \n",
       "TMTC3                        0                2                0   \n",
       "CEP290                       2                1               18   \n",
       "4930430F08RIK                0                1                1   \n",
       "1700017N19RIK                0                0                0   \n",
       "\n",
       "               r1_GTGCCGCCTCTC  r1_CCTGTGACACAC  r1_AATCTCGTTAAT  \\\n",
       "gene                                                               \n",
       "KITL                         0                1                0   \n",
       "TMTC3                        1                1                0   \n",
       "CEP290                      10                3                4   \n",
       "4930430F08RIK                0                1                1   \n",
       "1700017N19RIK                0                0                0   \n",
       "\n",
       "               r1_GATTTCCTCTGA  ...  p1_GAGGGGCTCTAA  p1_AGCCAAGGCTCG  \\\n",
       "gene                            ...                                     \n",
       "KITL                         0  ...                0                0   \n",
       "TMTC3                        0  ...                0                0   \n",
       "CEP290                       3  ...                0                0   \n",
       "4930430F08RIK                1  ...                0                0   \n",
       "1700017N19RIK                0  ...                0                0   \n",
       "\n",
       "               p1_TGAGTCGTCTTA  p1_AACGGTCGCTTT  p1_CGAATACGTGTC  \\\n",
       "gene                                                               \n",
       "KITL                         0                0                0   \n",
       "TMTC3                        0                0                0   \n",
       "CEP290                       0                0                0   \n",
       "4930430F08RIK                0                0                0   \n",
       "1700017N19RIK                0                0                0   \n",
       "\n",
       "               p1_TCAAAAGCCGGG  p1_ATTAAGTTCCAA  p1_CTGTCTGAGACC  \\\n",
       "gene                                                               \n",
       "KITL                         0                0                0   \n",
       "TMTC3                        0                0                0   \n",
       "CEP290                       0                0                0   \n",
       "4930430F08RIK                0                1                0   \n",
       "1700017N19RIK                0                0                0   \n",
       "\n",
       "               p1_TAACGCGCTCCT  p1_ATTCTTGTTCTT  \n",
       "gene                                             \n",
       "KITL                         0                0  \n",
       "TMTC3                        0                0  \n",
       "CEP290                       0                0  \n",
       "4930430F08RIK                0                0  \n",
       "1700017N19RIK                0                0  \n",
       "\n",
       "[5 rows x 49300 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pickle\n",
    "import gzip\n",
    "\n",
    "from examples import utils\n",
    "\n",
    "import loompy\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.sparse as sp\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "data = pd.read_table(\"G:/Billfish/VS_code_program/openTSNE-master/data/GSE63472_P14Retina_merged_digital_expression.txt.gz\", index_col=0)\n",
    "data.head()\n",
    "cluster_ids = pd.read_table(\"G:/Billfish/VS_code_program/openTSNE-master/data/retina_clusteridentities.txt\",  header=None, index_col=0, squeeze=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'data/macosko_2015.pkl.gz'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-12-89d31db229e3>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     10\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mpickle\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     11\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 12\u001b[1;33m \u001b[1;32mwith\u001b[0m \u001b[0mgzip\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"data/macosko_2015.pkl.gz\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"rb\"\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0mf\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     13\u001b[0m     \u001b[0mdata\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpickle\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mload\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mf\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32md:\\software\\Anaconda3\\lib\\gzip.py\u001b[0m in \u001b[0;36mopen\u001b[1;34m(filename, mode, compresslevel, encoding, errors, newline)\u001b[0m\n\u001b[0;32m     56\u001b[0m     \u001b[0mgz_mode\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mmode\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"t\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     57\u001b[0m     \u001b[1;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfilename\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mbytes\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mos\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mPathLike\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 58\u001b[1;33m         \u001b[0mbinary_file\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mGzipFile\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfilename\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mgz_mode\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcompresslevel\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     59\u001b[0m     \u001b[1;32melif\u001b[0m \u001b[0mhasattr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfilename\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"read\"\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mor\u001b[0m \u001b[0mhasattr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfilename\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"write\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     60\u001b[0m         \u001b[0mbinary_file\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mGzipFile\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;32mNone\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mgz_mode\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcompresslevel\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mfilename\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32md:\\software\\Anaconda3\\lib\\gzip.py\u001b[0m in \u001b[0;36m__init__\u001b[1;34m(self, filename, mode, compresslevel, fileobj, mtime)\u001b[0m\n\u001b[0;32m    171\u001b[0m             \u001b[0mmode\u001b[0m \u001b[1;33m+=\u001b[0m \u001b[1;34m'b'\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    172\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mfileobj\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 173\u001b[1;33m             \u001b[0mfileobj\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmyfileobj\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mbuiltins\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfilename\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mmode\u001b[0m \u001b[1;32mor\u001b[0m \u001b[1;34m'rb'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    174\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mfilename\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    175\u001b[0m             \u001b[0mfilename\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfileobj\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'name'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m''\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'data/macosko_2015.pkl.gz'"
     ]
    }
   ],
   "source": [
    "from openTSNE import TSNE\n",
    "from examples import utils\n",
    "import numpy as np\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "import gzip\n",
    "import pickle\n",
    "\n",
    "with gzip.open(\"data/macosko_2015.pkl.gz\", \"rb\") as f:\n",
    "    data = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 统计各loops 类型\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type.bed\", sep=\"\\t\", names=['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tags1', 'tags2', 'tissue'])\n",
    "df['tags'] = df['tags1'].map(str) + \"-\" + df['tags2'].map(str)\n",
    "#df['tags'].replace(\"N-G\", \"G-N\", inplace=True)\n",
    "df.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type.bed\", sep=\"\\t\", index=False)\n",
    "# df['number'] = [1]*df.shape[0]\n",
    "# df_result = df.groupby(by=['tissue', 'tags'], as_index=False)['number'].sum()\n",
    "# df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type_statis.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 统计loop类型分为At和Dt两个亚基因组\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type.bed\", sep=\"\\t\")\n",
    "df['subg'] = df['chrom1'].apply(lambda x: 'At' if 'Ghir_A' in x else 'Dt')\n",
    "df['tags'].replace(['N-G'], ['G-N'], inplace=True)\n",
    "df_sta = pd.DataFrame(df[['tissues', 'subg', 'tags']].value_counts(), columns=['number'])\n",
    "df_sta.reset_index(inplace=True)\n",
    "df_sta.rename(columns={'tags':'sigs'}, inplace=True)\n",
    "df_sta.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type_statis_subg.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type.bed\", sep=\"\\t\")\n",
    "df.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type.bed\", sep=\"\\t\", columns=['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tags1', 'tags2', 'tissue', 'tags'], index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 分析各组织特异高活性TAD区间所包含的loops\n",
    "tissues = ['anther', 'cotyledon', 'fiber_10DPA', 'fiber_20DPA', 'hypocotyl', 'leaf', 'ovule', 'petal', 'radicle', 'root', 'stem', 'stigma']\n",
    "df_loop = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type.bed\", sep=\"\\t\")\n",
    "df_result = pd.DataFrame()\n",
    "for i in tissues:\n",
    "    input_region = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TSG_TAD_region/{}_TAD_TSG_region.bed\".format(i)\n",
    "    df_region = pd.read_csv(input_region, sep=\"\\t\", names=['chrom', 'start', 'end'])\n",
    "    for j in tissues:\n",
    "        print(j)\n",
    "        df_loop_tissue = df_loop.query(\"tissue==@j\").copy()\n",
    "        for index, row in df_region.iterrows():\n",
    "            chrom, start, end = row['chrom'], row['start'], row['end']\n",
    "            tad_tag = \"{0}_{1}_{2}\".format(chrom, start, end)\n",
    "            df_tmp = df_loop_tissue.query(\"chrom1==@chrom & start1>=@start & end1<=@end & start2>=@start & end2<=@end\")[['tissue', 'tags']].copy()\n",
    "            df_tmp.columns = ['loop_tissue', 'gene_tags']\n",
    "            df_tmp['TAD_tissue'] = [i]*df_tmp.shape[0]\n",
    "            df_tmp['tad_tag'] = [tad_tag]*df_tmp.shape[0]\n",
    "            df_result = pd.concat([df_result, df_tmp], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 统计loops数量以及loops在组织特异高活性TAD范围内出现互作的概率\n",
    "df_result['number'] = [1]*df_result.shape[0]\n",
    "df_statistics = df_result.groupby(by=['TAD_tissue', 'loop_tissue', 'gene_tags'], as_index=False)['number'].sum()\n",
    "tissue = ['anther', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'root', 'radicle', 'ovule', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "loop_number = [118041, 163830, 139540, 180537, 154797, 165406, 175759, 157033, 160463, 156524, 142253, 115527]\n",
    "df_out = pd.DataFrame()\n",
    "for i in range(12):\n",
    "    T, N = tissue[i], loop_number[i]\n",
    "    df_tmp = df_statistics.query(\"loop_tissue == @T\").copy()\n",
    "    df_tmp['total_number'] = [N]*df_tmp.shape[0]\n",
    "    df_out = pd.concat([df_out, df_tmp], axis=0)\n",
    "df_out['ratio'] = df_out['number']/df_out['total_number']\n",
    "df_loop_type = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type_statis.csv\")\n",
    "df_loop_type.columns = ['loop_tissue', 'gene_tags', 'loop_type_number']\n",
    "df_out = pd.merge(df_out, df_loop_type, how='left', on=['loop_tissue', 'gene_tags'])\n",
    "df_out['type_ratio'] = df_out['number']/df_out['loop_type_number']\n",
    "df_out.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/TSG/J668_TSG_TAD_loop_number_ratio.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 分析各loop anchor 位点所出现的次数\n",
    "import pandas as pd\n",
    "def loop_anchor_number(df_loop):\n",
    "    df_loop['anchor1'] = df_loop['chrom1'].map(str) + \"-\" + df_loop['start1'].map(str) + \"-\" + df_loop['end1'].map(str)\n",
    "    df_loop['anchor2'] = df_loop['chrom2'].map(str) + \"-\" + df_loop['start2'].map(str) + \"-\" + df_loop['end2'].map(str)\n",
    "    df_a, df_b = df_loop[['anchor1']].copy(), df_loop[['anchor2']].copy()\n",
    "    df_a.rename(columns={'anchor1':'anchor'}, inplace=True)\n",
    "    df_b.rename(columns={'anchor2':'anchor'}, inplace=True)\n",
    "    df_c = pd.concat([df_a, df_b], axis=0)\n",
    "    df_c['number'] = [1]*df_c.shape[0]\n",
    "    df_return = df_c.groupby(by=['anchor'], as_index=False)['number'].sum()\n",
    "    df_return.sort_values(by=['number'], ascending=False, inplace=True)\n",
    "    return df_return\n",
    "tissues = ['anther', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'root', 'radicle', 'ovule', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df_result = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/final_loops/{}_final_loop.bed\".format(t)\n",
    "    df = pd.read_csv(input_f, sep=\"\\t\")\n",
    "    df_tmp = loop_anchor_number(df)\n",
    "    df_tmp['tissues'] = [t]*df_tmp.shape[0] \n",
    "    df_result = pd.concat([df_result, df_tmp], axis=0)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/J668_loop_anchor_number.csv\".format(t), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 各loop anchor 打上不同的标签\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/J668_loop_anchor_number.csv\")\n",
    "tissues =  ['anther', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'root', 'radicle', 'ovule', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df_result = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    df_tissue = df.query(\"tissues==@t\").copy()\n",
    "    df_tissue.sort_values(by='number', ascending=False)\n",
    "    #N = df_tissue.shape[0]*0.03\n",
    "    ## 选择大于5的视为高频互作区域，并将高频互作区域设置等级，再分析其区间所包含的基因，在分析其表达, 梯度设置： 1-5， 6-10， 10-15，15-20，大于20\n",
    "    df_tissue['tags'] = df_tissue['number'].apply(lambda x: 1 if x<5 else (2 if (5<=x<10) else (3 if (10<=x<15) else (4 if (15<=x<20) else 5))))\n",
    "    df_tissue[['chr', 'start', 'end']] = df_tissue['anchor'].str.split(\"-\", expand=True)\n",
    "    df_tissue = df_tissue[['chr', 'start', 'end', 'tags', 'tissues']].copy()\n",
    "    df_result = pd.concat([df_result, df_tissue], axis=0)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/J668_all_loop_anchor.bed\", sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 对loop anchor 富集到的基因打上loop anchor的标签，然后再分析其表达水平\n",
    "import pandas as pd\n",
    "df_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data/J668_all_Tissue_mean_TPM_long.csv\")\n",
    "df_loopanchor_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/J668_loop_anchor_gene.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'GeneId', 'tissues', 'anchor_order'])\n",
    "df_loopanchor_gene = df_loopanchor_gene[['GeneId', 'tissues', 'anchor_order']].copy()\n",
    "df = pd.merge(df_gene, df_loopanchor_gene, how='left', on=['GeneId', 'tissues'])\n",
    "df.fillna(0, inplace=True)\n",
    "df.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/J668_gene_loop_anchor_number.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        tissues  size  number\n",
      "0        anther     1   32856\n",
      "1        anther     2     304\n",
      "2        anther     3     105\n",
      "3        anther     4      33\n",
      "4        anther     5       8\n",
      "5     cotyledon     1   42568\n",
      "6     cotyledon     2     871\n",
      "7     cotyledon     3     128\n",
      "8     cotyledon     4      72\n",
      "9     cotyledon     5      27\n",
      "10  fiber_10DPA     1   36898\n",
      "11  fiber_10DPA     2     389\n",
      "12  fiber_10DPA     3     108\n",
      "13  fiber_10DPA     4      46\n",
      "14  fiber_10DPA     5       6\n",
      "15  fiber_20DPA     1   34458\n",
      "16  fiber_20DPA     2     380\n",
      "17  fiber_20DPA     3     114\n",
      "18  fiber_20DPA     4      52\n",
      "19  fiber_20DPA     5      17\n",
      "20    hypocotyl     1   38886\n",
      "21    hypocotyl     2     632\n",
      "22    hypocotyl     3     139\n",
      "23    hypocotyl     4      68\n",
      "24    hypocotyl     5      15\n",
      "25         leaf     1   42556\n",
      "26         leaf     2     957\n",
      "27         leaf     3     140\n",
      "28         leaf     4      88\n",
      "29         leaf     5      38\n",
      "30        ovule     1   46257\n",
      "31        ovule     2     881\n",
      "32        ovule     3     113\n",
      "33        ovule     4      20\n",
      "34        petal     1   39035\n",
      "35        petal     2     695\n",
      "36        petal     3     122\n",
      "37        petal     4      35\n",
      "38        petal     5       6\n",
      "39      radicle     1   46792\n",
      "40      radicle     2    1026\n",
      "41      radicle     3      98\n",
      "42      radicle     4      36\n",
      "43      radicle     5       5\n",
      "44         root     1   42556\n",
      "45         root     2     809\n",
      "46         root     3     167\n",
      "47         root     4      62\n",
      "48         root     5      19\n",
      "49         stem     1   40166\n",
      "50         stem     2     558\n",
      "51         stem     3     124\n",
      "52         stem     4      79\n",
      "53         stem     5      18\n",
      "54       stigma     1   33917\n",
      "55       stigma     2     278\n",
      "56       stigma     3     100\n",
      "57       stigma     4      41\n",
      "58       stigma     5       1\n"
     ]
    }
   ],
   "source": [
    "## 统计基因所受到的loop调控数目 (所有)\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/J668_loop_anchor_gene.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'GeneId', 'tissues', 'size'])\n",
    "df['number'] = [1]*df.shape[0]\n",
    "df_result = df.groupby(by=['tissues', 'size'], as_index=False)['number'].sum()\n",
    "print(df_result)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 统计基因受到loop的调控次数，所有组织都用7K 个锚点，并统计平均数\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/7K_anchor/J668_gene_7Kanchor_number.bed\", sep=\"\\t\", names=['anchor_id', 'anchor_number', 'tissues', 'sampleid'])\n",
    "df_result = df.groupby(by=['tissues', 'anchor_id'], as_index=False)['anchor_number'].mean()\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/7K_anchor/J668_gene_7Kanchor_ratio.bed\", sep=\"\\t\", index=False)\n",
    "## 求和\n",
    "df_result['tags'] = df_result['anchor_id'].apply(lambda x: 1 if x<5 else (2 if (5<=x<10) else (3 if (10<=x<15) else (4 if (15<=x<20) else (5 if (20<=x<=25) else 6 )))))\n",
    "df_result2 = df_result.groupby(by=['tissues', 'tags'], as_index=False)['anchor_number'].sum()\n",
    "df_result2.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/7K_anchor/J668_gene_7Kanchor_ratio_order.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 随着loop互作增加，基因表达水平的变化情况\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/step2/J668_gene_loop_anchor_number_final.bed\", sep=\"\\t\")\n",
    "df_tmp = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data/J668_all_Tissue_mean_TPM_long.csv\")\n",
    "df_result = pd.merge(df, df_tmp, how='inner', on=['GeneId', 'tissues'])\n",
    "df_result['tags'] = df_result['number'].apply(lambda x: x if x<=7 else \">=8\")\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/step2/J668_gene_loop_anchor_tpm.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      tissues  number   size\n",
      "0      anther       0  37043\n",
      "1      anther       1  18442\n",
      "2      anther       2   8386\n",
      "3      anther       3   3680\n",
      "4      anther       4   1446\n",
      "5      anther       5    547\n",
      "6      anther       6    242\n",
      "7      anther       7    131\n",
      "8      anther       8     64\n",
      "9      anther       9     49\n",
      "10     anther      10     40\n",
      "11     anther      11     25\n",
      "12     anther      12     25\n",
      "13     anther      13     17\n",
      "14     anther      14     13\n",
      "15     anther      15     16\n",
      "16     anther      16     11\n",
      "17     anther      17      4\n",
      "18     anther      18      6\n",
      "19     anther      19      3\n",
      "20     anther      20      2\n",
      "21     anther      21      2\n",
      "22     anther      22      1\n",
      "23     anther      24      3\n",
      "24     anther      30      1\n",
      "25  cotyledon       0  27129\n",
      "26  cotyledon       1  18320\n",
      "27  cotyledon       2  11174\n",
      "28  cotyledon       3   6170\n",
      "29  cotyledon       4   3333\n",
      "30  cotyledon       5   1790\n",
      "31  cotyledon       6    944\n",
      "32  cotyledon       7    502\n",
      "33  cotyledon       8    269\n",
      "34  cotyledon       9    143\n",
      "35  cotyledon      10     94\n",
      "36  cotyledon      11     67\n",
      "37  cotyledon      12     53\n",
      "38  cotyledon      13     39\n",
      "39  cotyledon      14     27\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/step2/J668_gene_loop_anchor_number_final.bed\", sep=\"\\t\")\n",
    "df['size'] = [1]*df.shape[0]\n",
    "df_result = df.groupby(by=['tissues', 'number'], as_index=False)['size'].sum()\n",
    "print(df_result.head(40))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "695\n",
      "539\n",
      "            GeneId      tissues        TPM\n",
      "0  Ghir_A01G001070        ovule  13.161481\n",
      "1  Ghir_A01G001070  fiber_20DPA   4.722409\n",
      "2  Ghir_A01G002570        ovule   0.626373\n",
      "3  Ghir_A01G002570  fiber_20DPA   0.174301\n",
      "4  Ghir_A01G003990        ovule   4.545330\n",
      "1078\n",
      "                 GeneId tissues  freq\n",
      "491499  Ghir_A01G001070   ovule  high\n",
      "491649  Ghir_A01G002570   ovule  high\n",
      "491791  Ghir_A01G003990   ovule  high\n",
      "491813  Ghir_A01G004210   ovule  high\n",
      "492163  Ghir_A01G007710   ovule  high\n",
      "1078\n"
     ]
    }
   ],
   "source": [
    "## 分析高频loop互作基因差异表达\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/step2/J668_gene_loop_anchor_tpm.csv\")\n",
    "df['freq'] = df['tags'].apply(lambda x: 'high' if x>=6 else 'low')\n",
    "df_20DPA_ovule = df.query(\"(tissues=='fiber_20DPA' | tissues=='ovule') & (freq=='high')\")[['GeneId', 'tissues', 'freq']].copy()\n",
    "#print(df_20DPA_ovule.drop_duplicates(keep=False, subset=['GeneId', 'freq']))\n",
    "print(df_20DPA_ovule.shape[0])\n",
    "df_20DPA_ovule.drop_duplicates(keep=False, subset=['GeneId', 'freq'], inplace=True)\n",
    "print(df_20DPA_ovule.shape[0])\n",
    "df_tpm = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data/J668_all_Tissue_mean_TPM_long.csv\")\n",
    "df_20DPA_ovule_tpm = df_tpm.query(\"tissues=='fiber_20DPA' | tissues=='ovule'\").copy()\n",
    "df_tmp = pd.merge(df_20DPA_ovule_tpm, df_20DPA_ovule[['GeneId']], how='inner', on=['GeneId'])\n",
    "print(df_tmp.head())\n",
    "print(df_tmp.shape[0])\n",
    "print(df_20DPA_ovule.head())\n",
    "df_result = pd.merge(df_tmp, df_20DPA_ovule, how='left', on=['GeneId', 'tissues'])\n",
    "print(df_result.shape[0])\n",
    "df_result.fillna('low', inplace=True)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/step2/J668_20DPA_ovule_diff_anchor.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 分析不表达的基因，与表达的基因其各loop anchor的比例\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/step2/J668_gene_loop_anchor_tpm.csv\")\n",
    "df['tags2'] = df['TPM'].apply(lambda x: \"exp\" if x>=0.5 else 'Nexp')\n",
    "df['number'] = [1]*df.shape[0]\n",
    "df_result = df.groupby(by=['tissues', 'tags2', 'tags'], as_index=False)['number'].sum()\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/step2/J668_exp_Nexp_loopanchor.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 举例说明\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/step2/J668_gene_loop_anchor_number_final.bed\", sep=\"\\t\")\n",
    "df_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data/J668_all_Tissue_mean_TPM.csv\")\n",
    "df_20DPA_ovule = df_gene[['GeneId', 'fiber_20DPA', 'ovule']].copy()\n",
    "df_20DPA = df.query(\"tissues=='fiber_20DPA'\")[['chrom', 'start', 'end', 'GeneId', 'number']].copy()\n",
    "df_ovule = df.query(\"tissues=='ovule'\")[['chrom', 'start', 'end', 'GeneId', 'number']].copy()\n",
    "df_20DPA.rename(columns={'number':'fiber_20DPA_n'}, inplace=True)\n",
    "df_ovule.rename(columns={'number':'ovule_n'}, inplace=True)\n",
    "df_20DPA_ovule_n = pd.merge(df_20DPA, df_ovule, how='inner', on=['chrom', 'start', 'end', 'GeneId'])\n",
    "df_tmp = pd.merge(df_20DPA_ovule_n, df_20DPA_ovule, how='inner', on=['GeneId'])\n",
    "# df_ovule_high = df_tmp.query(\"0<ovule_n<=2 & fiber_20DPA_n==0 & ovule>fiber_20DPA\")\n",
    "# df_20DPA_high = df_tmp.query(\"0<ovule_n<=2 & fiber_20DPA_n>=6 & ovule>fiber_20DPA\")\n",
    "df_ovule_high = df_tmp.query(\"ovule_n>fiber_20DPA_n & ovule>fiber_20DPA\")\n",
    "df_20DPA_high = df_tmp.query(\"ovule_n<fiber_20DPA_n & ovule>fiber_20DPA\")\n",
    "df_ovule_high.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/step2/ovule_HTPM_Hloop_20DPA.bed\", sep=\"\\t\", index=False)\n",
    "df_20DPA_high.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/step2/ovule_HTPM_Lloop_20DPA.bed\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析各组织loop长度的变化及其密度\n",
    "import pandas as pd\n",
    "tissues =  ['anther', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'root', 'radicle', 'ovule', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df_result = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/final_loops/{}_final_loop.bed\".format(t)\n",
    "    df = pd.read_csv(input_f, sep=\"\\t\")\n",
    "    df['length'] = df['start2'] - df['start1']\n",
    "    df['subg'] = df['chrom1'].apply(lambda x: 'At' if 'Ghir_A' in str(x) else 'Dt')\n",
    "    df['tissues'] = [t]*df.shape[0]\n",
    "    df_result = pd.concat([df_result, df], axis=0)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/length_density/J668_loop_length.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/length_density/J668_loop_length.csv\")\n",
    "df_result = pd.DataFrame(columns=['start', 'end', 'number','tissues' ])\n",
    "for t in tissues:\n",
    "    df_tissue = df.query(\"tissues==@t\")\n",
    "    for i in range(0, 4000000, 100000):\n",
    "        start, end = i, i+100000\n",
    "        df_tmp = df_tissue.query(\"@start<=length<@end\")\n",
    "        result_list = [start, end, df_tmp.shape[0], t]\n",
    "        #print(result_list)\n",
    "        df_result.loc[df_result.shape[0] ,: ] = result_list\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/length_density/J668_density.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将染色质loop分为小于100Kb的短程互作，和大于100Kb的长程互作，并分析短程互作和长程互作中GG GN NN 之间的占比\n",
    "import pandas as pd\n",
    "df_loop = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/length_density/J668_loop_length.csv\")\n",
    "df_loop['tags'] = df_loop['length'].apply(lambda x: \"long\" if x>=100000 else 'short')\n",
    "df_loop_type = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type.bed\", sep=\"\\t\")\n",
    "df_loop_type = df_loop_type[['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tissue', 'tags']]\n",
    "df_loop_type.rename(columns={'tissue':'tissues', 'tags':'loop_type'}, inplace=True)\n",
    "df_tmp = pd.merge(df_loop, df_loop_type, how='inner', on=['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tissues'])\n",
    "df_tmp.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type_all.csv\", index=False)\n",
    "df_tmp['number'] = [1]*df_tmp.shape[0]\n",
    "## 长程与短程loop 关于不同类型的loop占比\n",
    "df_result = df_tmp.groupby(by=['tissues', 'tags', 'loop_type'], as_index=False)['number'].sum()\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/length_density/J668_loop_length_type.csv\", index=False)\n",
    "\n",
    "## 不同组织的两个亚基因组关于长程和短程互作的占比情况\n",
    "df_result2 = df_tmp.groupby(by=['tissues', 'tags', 'subg'], as_index=False)['number'].sum()\n",
    "df_result2.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/length_density/J668_loop_length_type2.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           chrom1    start1      end1    chrom2    start2      end2  length  \\\n",
      "0        Ghir_A01     24000     27000  Ghir_A01     45000     48000   21000   \n",
      "1        Ghir_A01    546000    549000  Ghir_A01    576000    579000   30000   \n",
      "2        Ghir_A01   1059000   1062000  Ghir_A01   1092000   1095000   33000   \n",
      "3        Ghir_A01   1074000   1077000  Ghir_A01   1209000   1212000  135000   \n",
      "4        Ghir_A01   1248000   1251000  Ghir_A01   1563000   1566000  315000   \n",
      "...           ...       ...       ...       ...       ...       ...     ...   \n",
      "1829705  Ghir_D07  22206000  22209000  Ghir_D07  22380000  22383000  174000   \n",
      "1829706  Ghir_A13  33291000  33294000  Ghir_A13  33324000  33327000   33000   \n",
      "1829707  Ghir_D07  46671000  46674000  Ghir_D07  46782000  46785000  111000   \n",
      "1829708  Ghir_D07  38325000  38328000  Ghir_D07  38370000  38373000   45000   \n",
      "1829709  Ghir_A03  69513000  69516000  Ghir_A03  69534000  69537000   21000   \n",
      "\n",
      "        subg tissues tags  \n",
      "0         At  anther  N-N  \n",
      "1         At  anther  G-N  \n",
      "2         At  anther  G-N  \n",
      "3         At  anther  N-N  \n",
      "4         At  anther  G-N  \n",
      "...      ...     ...  ...  \n",
      "1829705   Dt  stigma  N-N  \n",
      "1829706   At  stigma  N-N  \n",
      "1829707   Dt  stigma  N-N  \n",
      "1829708   Dt  stigma  N-N  \n",
      "1829709   At  stigma  N-N  \n",
      "\n",
      "[1829710 rows x 10 columns]\n"
     ]
    }
   ],
   "source": [
    "df_loop = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/length_density/J668_loop_length.csv\")\n",
    "df_loop_type = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type.bed\", sep=\"\\t\")\n",
    "df_loop_type = df_loop_type[['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tissue', 'tags']]\n",
    "df_loop_type.columns = ['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tissues', 'tags']\n",
    "df_result = pd.merge(df_loop, df_loop_type, how='inner', on=['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tissues'])\n",
    "print(df_result)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/length_density/J668_loop_length_type3.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析不同长度下G-G, G-N 比例的变化\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/length_density/J668_loop_length_type3.csv\")\n",
    "tissues = ['anther', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'root', 'radicle', 'ovule', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df_result = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    for i in range(0, 980000, 30000):\n",
    "        j = 30000 + i\n",
    "        df_tmp = df.query(\"length>=@i & length<=@j & tissues==@t\").copy()\n",
    "        df_tmp['number'] = [1]*df_tmp.shape[0]\n",
    "        df_tmp2 = df_tmp.groupby(by=['tags'], as_index=False)['number'].sum()\n",
    "        total_number = df_tmp.shape[0]\n",
    "        df_tmp2['ratio'] = df_tmp2['number']/total_number\n",
    "        df_tmp2['x_lab'] = [i]*df_tmp2.shape[0]\n",
    "        df_tmp2['tissues'] = [t]*df_tmp2.shape[0]\n",
    "        df_result = pd.concat([df_result, df_tmp2], axis=0)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/length_density/J668_loop_gradient_ratio.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 调整A/B compartment 格式\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/length_density/compartment/J668_all_loop_cpt.bed\", sep=\"\\t\", names =['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tissues', 'tags1', 'tags2'])\n",
    "df['tags'] = df['tags1'].map(str) + '-' + df['tags2'].map(str)\n",
    "df['tags'].replace('B-A', 'A-B', inplace=True)\n",
    "df.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/length_density/compartment/J668_all_loop_cpt.bed\", sep=\"\\t\", columns=['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tissues', 'tags'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/length_density/compartment/J668_all_loop_cpt.bed\", sep=\"\\t\")\n",
    "df['length'] = df['start2'] - df['start1']\n",
    "df['subg'] = df['chrom1'].apply(lambda x: 'At' if 'Ghir_A' in str(x) else 'Dt')\n",
    "df.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/length_density/compartment/J668_all_loop_cpt3.bed\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 不同长度下A-A, A-B, B-B 比例的变化\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/length_density/compartment/J668_all_loop_cpt3.bed\", sep=\"\\t\")\n",
    "tissues = ['anther', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'root', 'radicle', 'ovule', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df_result = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    for i in range(0, 980000, 30000):\n",
    "        j = 30000 + i\n",
    "        df_tmp = df.query(\"length>=@i & length<=@j & tissues==@t\").copy()\n",
    "        df_tmp['number'] = [1]*df_tmp.shape[0]\n",
    "        df_tmp2 = df_tmp.groupby(by=['tags'], as_index=False)['number'].sum()\n",
    "        total_number = df_tmp.shape[0]\n",
    "        df_tmp2['ratio'] = df_tmp2['number']/total_number\n",
    "        df_tmp2['x_lab'] = [i]*df_tmp2.shape[0]\n",
    "        df_tmp2['tissues'] = [t]*df_tmp2.shape[0]\n",
    "        df_result = pd.concat([df_result, df_tmp2], axis=0)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/length_density/compartment/J668_all_loop_cpt_density.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "685775\n",
      "403730\n"
     ]
    }
   ],
   "source": [
    "## 分析保守的loops\n",
    "import pandas as pd\n",
    "tissues = ['anther', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'root', 'radicle', 'ovule', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df_result = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/final_loops/{}_gene_loop.bed\".format(t)\n",
    "    df_t = pd.read_csv(input_f, sep=\"\\t\")\n",
    "    df_t['tissues'] = [t]*df_t.shape[0]\n",
    "    df_result = pd.concat([df_result, df_t], axis=0)\n",
    "print(df_result.shape[0])\n",
    "# df_result['number'] = [1]*df_result.shape[0]\n",
    "# df_a = df_result.groupby(by=['chrom1','start1','end1','chrom2','start2','end2'], as_index=False)['number'].sum()\n",
    "# print(df_a)\n",
    "df_result.drop_duplicates(keep=False, subset=['chrom1','start1','end1','chrom2','start2','end2'], inplace=True)\n",
    "print(df_result.shape[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将loop anchor划分为以tab为分隔符的文件\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/J668_loop_anchor_number.csv\")\n",
    "df[['chrom', 'start', 'end']] = df['anchor'].str.split(\"-\", expand=True)\n",
    "df.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/J668_loop_anchor_number.bed\", sep=\"\\t\", index=False, header=False, columns=['chrom', 'start', 'end', 'tissues', 'number', 'anchor'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 统计各层级下ATAC-peaks的分布\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/ATAC/J668_loop_anchor_ATAC.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'tissues', 'anchor_number', 'ATAC_number'])\n",
    "df['anchor_tags'] = df['anchor_number'].apply(lambda x: x if x<=7 else \">=8\")\n",
    "df['ATAC_peaks'] = df['ATAC_number'].apply(lambda x: 'peak' if x>0 else 'nopeak')\n",
    "df['number'] = [1]*df.shape[0]\n",
    "#df['number'] = df['ATAC_number'].apply(lambda x: x if x>0 else 1)\n",
    "df_result = df.groupby(by=['tissues', 'anchor_tags', 'ATAC_peaks'], as_index=False)['number'].sum()\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/ATAC/J668_loop_anchor_ATAC_number.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 统计各层级下H3K27ac的分布\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/ChIP/J668_loop_anchor_H3K27ac.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'tissues', 'anchor_number', 'ChIP_number'])\n",
    "df['anchor_tags'] = df['anchor_number'].apply(lambda x: x if x<=7 else \">=8\")\n",
    "df['ChIP_peaks'] = df['ChIP_number'].apply(lambda x: 'peak' if x>0 else 'nopeak')\n",
    "df['number'] = [1]*df.shape[0]\n",
    "#df['number'] = df['ATAC_number'].apply(lambda x: x if x>0 else 1)\n",
    "df_result = df.groupby(by=['tissues', 'anchor_tags', 'ChIP_peaks'], as_index=False)['number'].sum()\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/ChIP/J668_loop_anchor_H3K27ac_number.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 统计各层级下H3K4me3的分布\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/ChIP/J668_loop_anchor_H3K4me3.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'tissues', 'anchor_number', 'ChIP_number'])\n",
    "df['anchor_tags'] = df['anchor_number'].apply(lambda x: x if x<=7 else \">=8\")\n",
    "df['ChIP_peaks'] = df['ChIP_number'].apply(lambda x: 'peak' if x>0 else 'nopeak')\n",
    "df['number'] = [1]*df.shape[0]\n",
    "#df['number'] = df['ATAC_number'].apply(lambda x: x if x>0 else 1)\n",
    "df_result = df.groupby(by=['tissues', 'anchor_tags', 'ChIP_peaks'], as_index=False)['number'].sum()\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/ChIP/J668_loop_anchor_H3K4me3_number.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "599057\n",
      "556338\n"
     ]
    }
   ],
   "source": [
    "## 统计各层级下subCpt的分布\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/subCpt/J668_loop_anchor_subcpt.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'tissues', 'anchor_number', 'subCpt_tags'])\n",
    "print(df.shape[0])\n",
    "df.drop_duplicates(keep='last', inplace=True, subset=['chrom', 'start', 'end', 'tissues'])\n",
    "print(df.shape[0])\n",
    "df['anchor_tags'] = df['anchor_number'].apply(lambda x: x if x<=7 else \">=8\")\n",
    "df['number'] = [1]*df.shape[0]\n",
    "#df['number'] = df['ATAC_number'].apply(lambda x: x if x>0 else 1)\n",
    "df_result = df.groupby(by=['tissues', 'anchor_tags', 'subCpt_tags'], as_index=False)['number'].sum()\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/subCpt/J668_loop_anchor_subCpt.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "40301\n",
      "17466\n",
      "2624\n",
      "      chrom     start       end           GeneId  hypocotyl_anchor  \\\n",
      "0  Ghir_A01   1665417   1669273  Ghir_A01G001860                 1   \n",
      "1  Ghir_A01   1665417   1669273  Ghir_A01G001860                 1   \n",
      "2  Ghir_A01   1665417   1669273  Ghir_A01G001860                 1   \n",
      "3  Ghir_A01   1665417   1669273  Ghir_A01G001860                 1   \n",
      "4  Ghir_A01  12650994  12656148  Ghir_A01G007520                 0   \n",
      "\n",
      "   stigma_anchor  number   anchor_N        DEG       ATAC    H3K27ac  \n",
      "0              0       1  hypocotyl  hypocotyl  hypocotyl  hypocotyl  \n",
      "1              0       1  hypocotyl  hypocotyl  hypocotyl  hypocotyl  \n",
      "2              0       1  hypocotyl  hypocotyl  hypocotyl  hypocotyl  \n",
      "3              0       1  hypocotyl  hypocotyl  hypocotyl  hypocotyl  \n",
      "4              2      -2     stigma  hypocotyl  hypocotyl  hypocotyl  \n",
      "135\n",
      "82\n",
      "65\n"
     ]
    }
   ],
   "source": [
    "## 以stigma 和 hypocotyl 为例子\n",
    "## 主要目的是找到一个基因在stigma和hypocotyl上具有loop anchor 差异，并且同时具有ATAC-Seq、H3K27ac、H3K4me3 peaks 修饰上的差异\n",
    "# 1、loop anchor 数目有差异， hypocotyl stigma 为例子\n",
    "df_anchor = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/step2/J668_gene_loop_anchor_tpm.csv\")\n",
    "df_hypocotyl = df_anchor.query(\"tissues=='hypocotyl'\")[['chrom', 'start', 'end', 'GeneId', 'number']].copy()\n",
    "df_stigma = df_anchor.query(\"tissues=='stigma'\")[['chrom', 'start', 'end', 'GeneId',  'number']].copy()\n",
    "\n",
    "df_hypocotyl.columns = ['chrom', 'start', 'end', 'GeneId', 'hypocotyl_anchor']\n",
    "df_stigma.columns = ['chrom', 'start', 'end', 'GeneId', 'stigma_anchor']\n",
    "df_hypo_stig = pd.merge(df_hypocotyl, df_stigma, how='inner', on=['chrom', 'start', 'end', 'GeneId'])\n",
    "df_hypo_stig['number'] = df_hypo_stig['hypocotyl_anchor'] - df_hypo_stig['stigma_anchor']\n",
    "df_hypo_stig['anchor_N'] = df_hypo_stig['number'].apply(lambda x: 'hypocotyl' if x>0 else ('stigma' if x<0 else None))\n",
    "df_hypo_stig.dropna(inplace=True)\n",
    "print(df_hypo_stig.shape[0])\n",
    "# 2、打上差异表达标签\n",
    "df_DEG = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/hypocotyl_stigma_DEG.csv\")\n",
    "old_tags, new_tags = ['Up', 'Down', 'Not'], ['stigma', 'hypocotyl', None]\n",
    "df_DEG['sig'].replace(old_tags, new_tags, inplace=True)\n",
    "df_DEG.dropna(inplace=True)\n",
    "df_DEG = df_DEG[['GeneId', 'sig']].copy()\n",
    "df_DEG.columns = ['GeneId', 'DEG']\n",
    "df_hypo_stig = pd.merge(df_hypo_stig, df_DEG, how='inner', on=['GeneId'])\n",
    "print(df_hypo_stig.shape[0])\n",
    "# 3、打上差异的ATAC-Seq 标签\n",
    "df_ATAC = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/diffpeaks/gene/hypocotyl_stigma_annotation_gene.bed\", sep=\" \")\n",
    "df_ATAC = df_ATAC.query(\"annotation!='Distal Intergenic'\")[['geneId', 'V4']].copy()\n",
    "df_ATAC.columns = ['GeneId', 'ATAC']\n",
    "old_tags = ['-', '+']\n",
    "new_tags = ['hypocotyl', 'stigma']\n",
    "df_ATAC['ATAC'].replace(old_tags, new_tags, inplace=True)\n",
    "df_hypo_stig = pd.merge(df_hypo_stig, df_ATAC, how='inner', on=['GeneId'])\n",
    "print(df_hypo_stig.shape[0])\n",
    "\n",
    "# 4、打上H3K27ac的标签\n",
    "df_H3K27ac = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/Chip_Seq/diffbind/H3K27ac/rawdata/hypocotyl_stigma_H3K27ac_DiffPeaks_annotation.bed\", sep=\" \")\n",
    "df_H3K27ac = df_H3K27ac.query(\"annotation!='Distal Intergenic'\")[['geneId', 'V4']].copy()\n",
    "df_H3K27ac.columns = ['GeneId', 'H3K27ac']\n",
    "old_tags = ['-', '+']\n",
    "new_tags = ['hypocotyl', 'stigma']\n",
    "df_H3K27ac['H3K27ac'].replace(old_tags, new_tags, inplace=True)\n",
    "df_hypo_stig = pd.merge(df_hypo_stig, df_H3K27ac, how='inner', on=['GeneId'])\n",
    "print(df_hypo_stig.head())\n",
    "df_hypo_stig_hypocotyl = df_hypo_stig.query(\"(number>0 & DEG=='hypocotyl' & ATAC=='hypocotyl' & H3K27ac=='hypocotyl') | (number>0 & DEG=='stigma' & ATAC=='stigma' & H3K27ac=='stigma')\")\n",
    "print(df_hypo_stig_hypocotyl.shape[0])\n",
    "df_hypo_stig_stigma = df_hypo_stig.query(\"(number<0 & DEG=='stigma' & ATAC=='stigma' & H3K27ac=='stigma') | (number<0 & DEG=='hypocotyl' & ATAC=='hypocotyl' & H3K27ac=='hypocotyl')\")\n",
    "print(df_hypo_stig_stigma.shape[0])\n",
    "print(df_hypo_stig_stigma.drop_duplicates(keep='first', subset=['GeneId']).shape[0])\n",
    "df_hypo_stig_stigma.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/hypocotyl_stigma_example/hypo_stig_stigma.bed\", sep=\"\\t\", columns=['chrom', 'start', 'end', 'GeneId'], index=False, header=False)\n",
    "df_hypo_stig_hypocotyl.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/hypocotyl_stigma_example/hypo_stig_hypocotyl.bed\", sep=\"\\t\", columns=['chrom', 'start', 'end', 'GeneId'], index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 整理所有bin所能包含的信号\n",
    "import sys\n",
    "import pandas as pd\n",
    "inp_f=sys.argv[1]\n",
    "out_f=sys.argv[2]\n",
    "df_result = pd.DataFrame()\n",
    "for line in open(inp_f):\n",
    "    f1, name = line.strip().split(\"\\t\")[0], line.strip().split(\"\\t\")[1]\n",
    "    df = pd.read_csv(f1, sep=\"\\t\", name=['chrom', 'start', 'end', name])\n",
    "    if df_result.shape[0] == 0:\n",
    "        df_result = df.copy()\n",
    "    else:\n",
    "        df_result = pd.merge(df_result, df, how='inner', on=['chrom', 'start', 'end'])\n",
    "df_result.to_csv(out_f, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析cis-element 原件\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_3kbin_all_signal_simple.bed\", sep=\"\\t\")\n",
    "df_cis = df.query(\"gene==0 & loop>0 & (H3K4me3>0 | H3K27ac>0 | ATAC>0)\").copy()\n",
    "df_cis.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_cis.bed\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 找到每个组织的cis调控元件\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "df = pd.read_csv(\"G:\\Billfish\\J668_multip_tissue_3D-genome\\HiC\\loops\\Cis-element\\J668_3kbin_all_signal.bed\", sep=\"\\t\")\n",
    "all_name = df.columns.to_list()\n",
    "tissues = ['anther', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'root', 'radicle', 'ovule', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "for t in tissues:\n",
    "    for j in 'ATAC', 'H3K27ac', 'H3K4me3', 'loop':\n",
    "        N = \"{0}_{1}\".format(t, j)\n",
    "        if N not in all_name:\n",
    "            df[N] = [0]*df.shape[0]\n",
    "df_NG = df.query(\"gene==0\").copy()\n",
    "for t in tissues:\n",
    "    output_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/{}_Cis.bed\".format(t)\n",
    "    A_tag, K27_tag, K4_tag, l_tag = \"{}_ATAC\".format(t), \"{}_H3K27ac\".format(t), \"{}_H3K4me3\".format(t), \"{}_loop\".format(t)\n",
    "    df_tmp = df_NG[(df_NG[l_tag]>0) & ((df_NG[A_tag]>0) | (df_NG[K27_tag] > 0) | (df_NG[K4_tag] > 0))]\n",
    "    df_tmp.to_csv(output_f, sep=\"\\t\", columns=['chrom', 'start', 'end'], index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 总结各个组织的活性cis调控元件和非活性cis调控元件\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_cis.bed\", sep=\"\\t\")\n",
    "df = df[['chrom', 'start', 'end']].copy()\n",
    "tissues = ['anther', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'root', 'radicle', 'ovule', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df_result = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/{}_Cis.bed\".format(t)\n",
    "    df_T = pd.read_csv(input_f, sep=\"\\t\")\n",
    "    df_T['sigs'] = ['active']*df_T.shape[0]\n",
    "    df_tmp = pd.merge(df, df_T, how='left', on=['chrom', 'start', 'end'])\n",
    "    df_tmp['tissues'] = [t]*df_tmp.shape[0]\n",
    "    df_tmp.fillna(\"inactive\", inplace=True)\n",
    "    df_result = pd.concat([df_result, df_tmp], axis=0)\n",
    "df_result['subg'] = df_result['chrom'].apply(lambda x: 'At' if 'Ghir_A' in x else 'Dt')\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_active_cis.csv\", index=False)\n",
    "\n",
    "## 统计各组织active 和inactive cis的占比\n",
    "df_sta = pd.DataFrame(df_result[['tissues', 'subg', 'sigs']].value_counts(), columns=['number'])\n",
    "df_sta.reset_index(inplace=True)\n",
    "df_sta.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_active_cis_sta.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "anther\n",
      "leaf\n",
      "hypocotyl\n",
      "cotyledon\n",
      "stem\n",
      "root\n",
      "radicle\n",
      "ovule\n",
      "fiber_10DPA\n",
      "fiber_20DPA\n",
      "petal\n",
      "stigma\n"
     ]
    }
   ],
   "source": [
    "## 分析基因受到的loop 调控方式，1：G-G、2：G-Cis、3: G-N、4：G-N-G、5：G-N-Cis、6：G-N-N、7：G-noloop\n",
    "\n",
    "import pandas as pd\n",
    "import sys\n",
    "\n",
    "## 对非基因区域打上标签，1、间接与基因相关；2、无相关；3、间接与Cis-element相关；4、间接与非基因区相关及其它\n",
    "\n",
    "# 1、间接与基因相关\n",
    "def NG(df, df_cis):\n",
    "    df, df_cis =df.copy(), df_cis.copy()\n",
    "    df_GN = df.query(\"tags1=='G' & tags2=='N'\")[['chrom2', 'start2', 'end2']].copy()\n",
    "    df_NG = df.query(\"tags1=='N' & tags2=='G'\")[['chrom1', 'start1', 'end1']].copy()\n",
    "    df_GN.columns = ['chrom', 'start', 'end']\n",
    "    df_NG.columns = ['chrom', 'start', 'end']\n",
    "    df_GN1 = pd.concat([df_GN, df_NG], axis=0)\n",
    "    df_GN1['number'] = [1]*df_GN1.shape[0]\n",
    "    df_GN = df_GN1.groupby(by=['chrom', 'start', 'end'], as_index=False)['number'].sum()\n",
    "    df_GNG = df_GN.query(\"number>1\")[['chrom', 'start', 'end']].copy()\n",
    "    df_GNG.drop_duplicates(keep='first', inplace=True)\n",
    "    ## 去除其中N端是Cis anchor\n",
    "    df_return = pd.concat([df_GNG, df_cis, df_cis], axis=0)\n",
    "    df_return.drop_duplicates(keep=False, inplace=True)\n",
    "    #print(df_return)\n",
    "    return df_return\n",
    "\n",
    "# 2、确定仅与基因相关的loop anchoer\n",
    "def Gnone(df, df_cis):\n",
    "    df = df.copy()\n",
    "    df_cis = df_cis.copy()\n",
    "    df_GN = df.query(\"tags1=='G' & tags2=='N'\")[['chrom2', 'start2', 'end2']].copy()\n",
    "    df_NG = df.query(\"tags1=='N' & tags2=='G'\")[['chrom1', 'start1', 'end1']].copy()\n",
    "    df_GN.columns = ['chrom', 'start', 'end']\n",
    "    df_NG.columns = ['chrom', 'start', 'end']\n",
    "    df_GN1 = pd.concat([df_GN, df_NG], axis=0) ## 1、找到与基因相关的loop anchor\n",
    "    # df_test = df_GN1.copy()\n",
    "    # df_test.drop_duplicates(inplace=True)\n",
    "    # print(df_test.shape[0])\n",
    "\n",
    "# 3、去掉与NN loop相关重合的anchor\n",
    "    df_NN = df.query(\"tags1=='N' & tags2=='N'\").copy()\n",
    "    df_NN1 = df_NN[['chrom1', 'start1', 'end1']].copy()\n",
    "    df_NN2 = df_NN[['chrom2', 'start2', 'end2']].copy()\n",
    "    df_NN1.columns = ['chrom', 'start', 'end']\n",
    "    df_NN2.columns = ['chrom', 'start', 'end']\n",
    "    #print(df_NN1)\n",
    "    df_NN = pd.concat([df_NN1, df_NN2], axis=0)\n",
    "    df_Gnone = pd.concat([df_NN, df_NN, df_GN1], axis=0)\n",
    "    df_Gnone.drop_duplicates(keep=False, inplace=True)\n",
    "    #去掉与Gcis相关的loop anchor\n",
    "    df_return = pd.concat([df_Gnone, df_cis, df_cis], axis=0).drop_duplicates(keep=False) \n",
    "    df_return.drop_duplicates(keep='first', inplace=True)\n",
    "    return df_return\n",
    "\n",
    "# 4、间接与Cis-element相关\n",
    "# 想法： 将所有与cis-element 相关的N-N loop选出来，然后判断非cis-element出现的次数，大于1次则说明是存在间接cis-element调控关系\n",
    "def Ncis(df, df_cis):\n",
    "    df = df.copy()\n",
    "    df_cis = df_cis.copy()\n",
    "    ## 1、找到与基因相关的loops GN 处的N锚点\n",
    "    df_GN = df.query(\"tags1=='G' & tags2=='N'\")[['chrom2', 'start2', 'end2']].copy()\n",
    "    df_NG = df.query(\"tags1=='N' & tags2=='G'\")[['chrom1', 'start1', 'end1']].copy()\n",
    "    df_GN.columns = ['chrom', 'start', 'end']\n",
    "    df_NG.columns = ['chrom', 'start', 'end']\n",
    "    df_GN1 = pd.concat([df_GN, df_NG], axis=0) ## 1、找到与基因相关的loop anchor\n",
    "    df_GN1.columns=['chrom1', 'start1', 'end1']\n",
    "    ## 1、所有N-N 存在cis 多互作\n",
    "    df_NN = df.query(\"tags1=='N' and tags2=='N'\").copy()\n",
    "    #print(df_NN.shape[0])\n",
    "    df_cis.columns = ['chrom1', 'start1', 'end1']\n",
    "    df_CisN = pd.merge(df_NN, df_cis, how='inner', on=['chrom1', 'start1', 'end1'])[['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2']].copy()\n",
    "    #print(df_CisN.shape[0])\n",
    "    df_cis.columns = ['chrom2', 'start2', 'end2']\n",
    "    df_NCis = pd.merge(df_NN, df_cis, how='inner', on=['chrom2', 'start2', 'end2'])[['chrom2', 'start2', 'end2', 'chrom1', 'start1', 'end1']].copy()\n",
    "    #print(df_NCis.shape[0])\n",
    "    df_CisN.columns = ['chrom', 'start', 'end', 'chrom1', 'start1', 'end1']\n",
    "    df_NCis.columns = ['chrom', 'start', 'end', 'chrom1', 'start1', 'end1']\n",
    "    df_tmp = pd.concat([df_CisN, df_NCis], axis=0)[['chrom1', 'start1', 'end1']]\n",
    "    df_tmp = pd.merge(df_tmp, df_GN1, how='inner', on=['chrom1', 'start1', 'end1'])[['chrom1', 'start1', 'end1']].copy()\n",
    "    df_tmp.columns = ['chrom', 'start', 'end']\n",
    "    #print(df_tmp, 'h')\n",
    "    df_return = df_tmp\n",
    "    df_return.drop_duplicates(keep=False, inplace=True)\n",
    "    #print(df_return.shape[0])\n",
    "    return df_return\n",
    "\n",
    "## 5、找到所有与基因相关的非基因loop anchor位点\n",
    "def NGene_anchor(df):\n",
    "    df = df.copy()\n",
    "    ## 1、找到与基因相关的loop anchor\n",
    "    df_GN = df.query(\"tags1=='G' & tags2=='N'\")[['chrom2', 'start2', 'end2']].copy()\n",
    "    df_NG = df.query(\"tags1=='N' & tags2=='G'\")[['chrom1', 'start1', 'end1']].copy()\n",
    "    df_GN.columns = ['chrom', 'start', 'end']\n",
    "    df_NG.columns = ['chrom', 'start', 'end']\n",
    "    df_return = pd.concat([df_GN, df_NG], axis=0)\n",
    "    #print(df_return)\n",
    "    df_return.drop_duplicates(keep='first', inplace=True)\n",
    "    #print(df_return.shape[0], 'NGene_anchor')\n",
    "    return df_return\n",
    "\n",
    "def Gcis(df, df_cis):\n",
    "    df, df_cis = df.copy(), df_cis.copy()\n",
    "    df_GN = df.query(\"tags1=='G' & tags2=='N'\")[['chrom2', 'start2', 'end2']].copy()\n",
    "    df_NG = df.query(\"tags1=='N' & tags2=='G'\")[['chrom1', 'start1', 'end1']].copy()\n",
    "    df_GN.columns = ['chrom', 'start', 'end']\n",
    "    df_NG.columns = ['chrom', 'start', 'end']\n",
    "    df_GN1 = pd.concat([df_GN, df_NG], axis=0) ## 1、找到与基因相关的loop anchor\n",
    "    df_GN1.drop_duplicates(inplace=True)\n",
    "    #print(df_GN1.shape[0], 'GCis')\n",
    "    df_return = pd.merge(df_GN1, df_cis, how='inner')\n",
    "    df_return.drop_duplicates(keep='first', inplace=True)\n",
    "    return df_return\n",
    "# 6、间接与N相关以及其它, 整理为一个文件\n",
    "def NN(df, df_cis, tissue):\n",
    "    df, df_cis = df.copy(), df_cis.copy()\n",
    "    ## 1、找到所有与基因相关的非基因anchor位点\n",
    "    df_NGene = NGene_anchor(df)\n",
    "    ## 2、间接与基因相关的非基因anchor位点\n",
    "    df_GNG = NG(df, df_cis)\n",
    "    ## 3、仅与基因相关的非基因anchor位点\n",
    "    df_Gnone = Gnone(df, df_cis)\n",
    "    ## 4、间接与cis元件相关的非基因anchor位点\n",
    "    df_GNCis = Ncis(df, df_cis)\n",
    "    ## 5、间接与无功能区相关的非基因anchor位点\n",
    "    #print(df_NGene.shape[0])\n",
    "    df_GNN = pd.concat([df_NGene, df_GNG, df_GNCis, df_Gnone, df_cis, df_cis], axis=0)\n",
    "    #print(df_NGene.shape[0], df_GNG.shape[0], df_Gnone.shape[0], df_GNCis.shape[0], df_cis.shape[0])\n",
    "    df_GNN.drop_duplicates(inplace=True, keep=False)\n",
    "    ## 6、直接与基因相关的Cis\n",
    "    df_GCis = Gcis(df, df_cis)\n",
    "    df_GCis['tags'] = ['GCis']*df_GCis.shape[0]\n",
    "    df_GNG['tags'] = ['GNG']*df_GNG.shape[0]\n",
    "    df_Gnone['tags'] = ['Gnone']*df_Gnone.shape[0]\n",
    "    df_GNCis['tags'] = ['GNCis']*df_GNCis.shape[0]\n",
    "    df_GNN['tags'] = ['GNN']*df_GNN.shape[0]\n",
    "    \n",
    "    df_result = pd.concat([df_GCis, df_GNG, df_GNCis, df_Gnone, df_GNN], ignore_index=True)\n",
    "    \n",
    "    df_result.drop_duplicates(keep='first', inplace=True, subset=['chrom', 'start', 'end'])\n",
    "    #print(df_GCis.shape[0], df_GNG.shape[0], df_GNCis.shape[0], df_Gnone.shape[0], df_GNN.shape[0])\n",
    "    df_result['number'] = [1]*df_result.shape[0]\n",
    "    #print(df_result.shape[0], 'NN')\n",
    "    ##  将与基因相关的非基因区域anchor 标签写出来\n",
    "    output_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_level/{}_level_loop_GN_Nanchor.csv\".format(tissue)\n",
    "    df_result.to_csv(output_f, index=False)\n",
    "    \n",
    "    return df_result\n",
    "\n",
    "## 7、找到所有基因锚点\n",
    "def Gene_anchor(df):\n",
    "    df = df.copy()\n",
    "    df_GG = df.query(\"tags=='G-G'\")\n",
    "    #print(df_GG.shape[0])\n",
    "    df1 = df_GG[['chrom1', 'start1', 'end1']].copy()\n",
    "    df2 = df_GG[['chrom2', 'start2', 'end2']].copy()\n",
    "    df1.columns = ['chrom', 'start', 'end']\n",
    "    df2.columns = ['chrom', 'start', 'end']\n",
    "    df_return = pd.concat([df1, df2], axis=0)\n",
    "    #print(df_return.shape[0])\n",
    "    df_return.drop_duplicates(keep='first', inplace=True)\n",
    "    df_return['tags'] = ['GG']*df_return.shape[0]\n",
    "    return df_return\n",
    "## 找到所有G-N loop并调整\n",
    "def GN_loop(df):\n",
    "    df = df.copy()\n",
    "    df_GN = df.query(\"tags1=='G' & tags2=='N'\")[['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2']].copy()\n",
    "    df_NG = df.query(\"tags1=='N' & tags2=='G'\")[['chrom2', 'start2', 'end2', 'chrom1', 'start1', 'end1']].copy()\n",
    "    df_GN.columns = ['chrom1', 'start1', 'end1', 'chrom', 'start', 'end']\n",
    "    df_NG.columns = ['chrom1', 'start1', 'end1', 'chrom', 'start', 'end']\n",
    "    df_return = pd.concat([df_GN, df_NG], axis=0)\n",
    "    # print(df_return)\n",
    "    # print(df_return.shape[0])\n",
    "    return df_return\n",
    "## 对G-N loop 打上标签\n",
    "def GN_anchor(df, df_cis, tissue):\n",
    "    df, df_cis = df.copy(), df_cis.copy()\n",
    "    df_anchor_tags = NN(df, df_cis, tissue)\n",
    "    df_GNanchor = GN_loop(df)\n",
    "    df_anchor_tags['number'] = [1]*df_anchor_tags.shape[0]\n",
    "\n",
    "    df_tmp = pd.merge(df_GNanchor, df_anchor_tags, how='left', on=['chrom', 'start', 'end'])    #print(df_tmp)\n",
    "    df_tmp['number'] = [1]*df_tmp.shape[0]\n",
    "    df_tmp.drop_duplicates(keep='first', inplace=True, subset=['chrom', 'start', 'end'])\n",
    "\n",
    "    \n",
    "    df_Ganchor = df_tmp[['chrom1', 'start1', 'end1', 'tags']].copy()\n",
    "    df_Ganchor.columns = ['chrom', 'start', 'end', 'tags']\n",
    "    ## 排序\n",
    "    df_mapping = pd.DataFrame({\"tags_sort\":['GCis', 'GNG', 'GNCis', 'Gnone', 'GNN']})\n",
    "    sort_mapping = df_mapping.reset_index().set_index('tags_sort')\n",
    "    df_Ganchor['tags_sort'] = df_Ganchor['tags'].map(sort_mapping['index'])\n",
    "    df_Ganchor.sort_values(by=['tags_sort'], inplace=True)\n",
    "    df_Ganchor = df_Ganchor[['chrom', 'start', 'end', 'tags']].copy()\n",
    "    return df_Ganchor\n",
    "\n",
    "##  对所有基因区域打上标签\n",
    "def Gene_anchor_tags(df, df_cis, tissue):\n",
    "    df, df_cis = df.copy(), df_cis.copy()\n",
    "    df_GG = Gene_anchor(df)\n",
    "    #print(df_GG.shape[0])\n",
    "    df_GN = GN_anchor(df, df_cis, tissue)\n",
    "    df_GN['number'] = [1]*df_GN.shape[0]\n",
    "    df_result = pd.concat([df_GG, df_GN], axis=0)\n",
    "    #print(df_result)\n",
    "    df_result['number'] = [1]*df_result.shape[0]\n",
    "    df_groupby = df_result.groupby(by=['tags'], as_index=False)['number'].sum()\n",
    "    #print(df_groupby)\n",
    "    return df_result\n",
    "tissues = ['anther', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'root', 'radicle', 'ovule', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df_loop_type = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type.bed\", sep=\"\\t\")\n",
    "for t in tissues:\n",
    "    print(t)\n",
    "    cis_input = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/{0}_cis.bed\".format(t)\n",
    "    level_loop_output = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_level/{0}_level_loop_gene_anchor.bed\".format(t)\n",
    "    df_loop_tissue = df_loop_type.query(\"tissue==@t\").copy()\n",
    "    df_cis = pd.read_csv(cis_input, sep=\"\\t\")\n",
    "    #print(df_cis)\n",
    "    df_result = Gene_anchor_tags(df_loop_tissue, df_cis, t)\n",
    "    df_result.to_csv(level_loop_output, sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 给基因锚点打上TPM \n",
    "import pandas as pd\n",
    "df_anchor_gene = pd.read_csv(\"G:\\Billfish\\J668_multip_tissue_3D-genome\\HiC\\loops\\loop_level/J668_level_loop.bed\", sep=\"\\t\", names=['GeneId', 'tags', 'tissues'])\n",
    "df_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data/J668_all_Tissue_mean_TPM_long.csv\")\n",
    "df_result = pd.merge(df_anchor_gene, df_gene, how='right', on=['GeneId', 'tissues'])\n",
    "df_result.fillna(\"None\", inplace=True)\n",
    "df_result.to_csv(\"G:\\Billfish\\J668_multip_tissue_3D-genome\\HiC\\loops\\loop_level/J668_level_loop_TPM.csv\",  index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21812 21812\n",
      "33638 33638\n",
      "27231 27231\n",
      "34682 34682\n",
      "30433 30433\n",
      "31369 31369\n",
      "36326 36326\n",
      "33959 33959\n",
      "26533 26533\n",
      "25040 25040\n",
      "27793 27793\n",
      "21708 21708\n"
     ]
    }
   ],
   "source": [
    "## 分析GN 的N端所包含的anchor 锚点数目\n",
    "tissues = ['anther', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'root', 'radicle', 'ovule', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df_result = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    GN_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_level/{}_level_loop_GN_Nanchor.csv\".format(t)\n",
    "    anchor_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/J668_loop_anchor_number.bed\"\n",
    "    df_GN = pd.read_csv(GN_f)\n",
    "    df_anchor = pd.read_csv(anchor_f, sep=\"\\t\", names=['chrom', 'start', 'end', 'tissue', 'loop_number', 'tags'])\n",
    "    df_tissue = df_anchor.query(\"tissue==@t\")[['chrom', 'start', 'end', 'tissue', 'loop_number']].copy()\n",
    "    df_tmp = pd.merge(df_GN, df_tissue, how='inner', on=['chrom', 'start', 'end'])\n",
    "    df_result = pd.concat([df_result, df_tmp], axis=0)\n",
    "    print(df_GN.shape[0], df_tmp.shape[0])\n",
    "output_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_level/Nanchor_number/J668_GN_Nanchor_number.csv\"\n",
    "df_result.to_csv(output_f, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 各种类型的loops 随机选择100条，来作图分析\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_level/hypocotyl_level_loop_gene_anchor.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'tags', 'number'])\n",
    "df_result = pd.DataFrame()\n",
    "for t in ['GG', 'Gnone', 'GCis', 'GNG', 'GNCis', 'GNN']:\n",
    "    df_tags = df.query(\"tags==@t\").copy()\n",
    "    df_tmp = df_tags.sample(100)\n",
    "    df_result = pd.concat([df_result, df_tmp], axis=0)\n",
    "df_GNG = df.query(\"tags=='GNCis'\").copy()\n",
    "df_GNG.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_level/hypocotyl_example_GNCis.bed\", sep=\"\\t\", index=False, header=False)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_level/hypocotyl_example_sample.bed\", sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "39994\n"
     ]
    }
   ],
   "source": [
    "## 分析各cis-element loop 的调控关系\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_3kbin_all_signal.bed\", sep=\"\\t\")\n",
    "df_cis = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_cis.bed\", sep=\"\\t\")\n",
    "df_cis_position = df_cis[['chrom', 'start', 'end']].copy()\n",
    "tissues = ['anther', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'root', 'radicle', 'ovule', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "loop_list = [\"{}_loop\".format(i) for i in tissues]\n",
    "chr_list = ['chrom', 'start', 'end']\n",
    "tags_list = chr_list + loop_list\n",
    "df_loop_anchor = df[tags_list]\n",
    "df_cis_anthor = pd.merge(df_loop_anchor, df_cis_position, how='inner', on=['chrom', 'start', 'end'])\n",
    "print(df_cis_anthor.shape[0])\n",
    "df_cis_anthor.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_loop_anthor.bed\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_loop_anther = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/J668_loop_anchor_number.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'tissues', 'number', 'tags'])\n",
    "df_loop_anther_wide = df_loop_anther.pivot(index='tags', columns=\"tissues\", values='number')\n",
    "df_loop_anther_wide.reset_index(inplace=True)\n",
    "df_loop_anther_wide[['chrom', 'start', 'end']] = df_loop_anther_wide['tags'].str.split(\"-\", expand=True)\n",
    "df_loop_anther_wide.fillna(0, inplace=True)\n",
    "df_loop_anther_wide['start'] = df_loop_anther_wide['start'].map(int)\n",
    "df_loop_anther_wide['end'] = df_loop_anther_wide['end'].map(int)\n",
    "df_cis = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_cis.bed\", sep=\"\\t\")\n",
    "df_cis_position = df_cis[['chrom', 'start', 'end']].copy()\n",
    "df_cis_loop_anther = pd.merge(df_cis_position, df_loop_anther_wide, how='inner', on=['chrom', 'start', 'end'])\n",
    "df_cis_loop_anther.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_cis_element_loop_anther.bed\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析高热点cis调控元件的有无\n",
    "import pandas as pd\n",
    "# 找两个组织间差异的高hot 热点\n",
    "def diff_hot_cis(df, tissue1, tissue2):\n",
    "    df_two_tissue = df[['chrom', 'start', 'end', tissue1, tissue2]].copy()\n",
    "    df_tissue_diff = df_two_tissue[df_two_tissue[tissue1] != df_two_tissue[tissue2]]\n",
    "    df_return= df_tissue_diff[((df_tissue_diff[tissue1] == 0) & (df_tissue_diff[tissue2] >= 4)) | ((df_tissue_diff[tissue1] >= 4) & (df_tissue_diff[tissue2] == 0))]\n",
    "    return df_return\n",
    "# # 找差异的高hot cis 热点互作的基因区域\n",
    "# def gene_anchor(df_loop, tissue1, tissue2):\n",
    "\n",
    "# 执行命令\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_cis_element_loop_anther.bed\", sep=\"\\t\")\n",
    "df_tissue = diff_hot_cis(df, 'hypocotyl', 'stigma')\n",
    "df_tissue.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/HotCis/tissue_diff/J668_diff_hot_cis.bed\", sep=\"\\t\", columns=['chrom', 'start', 'end'], index=False, header=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissues = ['anther', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'root', 'radicle', 'ovule', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/final_loops/{}_final_loop.bed\".format(t)\n",
    "    df_tmp = pd.read_csv(input_f, sep=\"\\t\")\n",
    "    df_tmp['tissues'] = [t]*df_tmp.shape[0]\n",
    "    df = pd.concat([df, df_tmp], axis=0)\n",
    "df.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/final_loops/J668_all_loop.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1242957 28 235082\n",
      "        tissues  number\n",
      "0        anther   79349\n",
      "1     cotyledon  126678\n",
      "2   fiber_10DPA  109858\n",
      "3   fiber_20DPA  112708\n",
      "4     hypocotyl   99745\n",
      "5          leaf  116636\n",
      "6         ovule  100371\n",
      "7         petal   90424\n",
      "8       radicle  109051\n",
      "9          root  107318\n",
      "10         stem  112177\n",
      "11       stigma   78642\n",
      "   number  number2\n",
      "0       2   166912\n",
      "1       3    41945\n",
      "2       4    14661\n",
      "3       5     6151\n",
      "4       6     2749\n",
      "5       7     1352\n",
      "6       8      709\n",
      "7       9      370\n",
      "8      10      164\n",
      "9      11       69\n"
     ]
    }
   ],
   "source": [
    "## 分析完全保守，动态保守和unique loops\n",
    "import pandas as pd\n",
    "#df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/final_loops/J668_all_loop.csv\")\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type.bed\", sep=\"\\t\")\n",
    "#df = df.query(\"tags=='G-G' | tags=='G-N'\")[['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tissues']].copy()\n",
    "#df = df.query(\"tags=='G-G' \")[['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tissues']].copy()\n",
    "#df = df.query(\"tags=='G-N' \")[['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tissues']].copy()\n",
    "#df = df.query(\"tags=='N-N'\")[['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tissues']].copy()\n",
    "\n",
    "def tissue_loop_type(df):\n",
    "    df = df.copy()\n",
    "    df['number'] = [1]*df.shape[0]\n",
    "    df_tiss_loop_number = df.groupby(by=['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2'], as_index=False)['number'].sum()\n",
    "    df_uniq = df_tiss_loop_number.query(\"number==1\").copy()\n",
    "    df_conserved = df_tiss_loop_number.query(\"number==12\").copy()\n",
    "    df_dynamic = df_tiss_loop_number.query(\"number>1 & number<12\").copy()\n",
    "    print(df_uniq.shape[0], df_conserved.shape[0], df_dynamic.shape[0])\n",
    "    return df_uniq, df_conserved, df_dynamic\n",
    "df_uniq, df_conserved, df_dynamic = tissue_loop_type(df)\n",
    "\n",
    "## 1、分析uniq 的数目\n",
    "df_uniq = pd.merge(df, df_uniq, how='inner', on=['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2'])\n",
    "df_uniq_group = df_uniq.groupby(by=['tissues'], as_index=False)['number'].sum()\n",
    "print(df_uniq_group)\n",
    "\n",
    "## 2、分析保守的loop数目\n",
    "df_dynamic['number2'] = [1]*df_dynamic.shape[0]\n",
    "df_dynamic = df_dynamic.groupby(by=['number'], as_index=False)['number2'].sum()\n",
    "print(df_dynamic)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "与基因重复的cis数目:  950\n"
     ]
    }
   ],
   "source": [
    "# 根据cis元件和基因的同源性来去除一部分cis元件\n",
    "import pandas as pd\n",
    "import re\n",
    "pattern = r'[A-Z][0-9]{2}'\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/cis_gene_homo/Ghir_AD_cis_blast.txt\", sep=\"\\t\", names=['q_id', 's_id', 'identity', 'alig_length', 'mismatch', 'gap', 'q_start', 'q_end', 's_start', 's_end', 'e-value', 'bin_score'])\n",
    "df['q_tags'] = df['q_id'].apply(lambda x: re.findall(pattern,x)[0] if re.findall(pattern,x) else \"\")\n",
    "df['s_tags'] = df['s_id'].apply(lambda x: re.findall(pattern,x)[0] if re.findall(pattern,x) else \"\")\n",
    "df['q_length'] = df['q_end'] - df['q_start']\n",
    "df_need = df.query(\"q_length>=1000 & q_tags==s_tags\")[['q_id']].drop_duplicates(keep='first', subset=['q_id']).copy() ## 确定可能同源的cis元件\n",
    "# 去除cis元件\n",
    "df_cis = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_cis.bed\", sep=\"\\t\")\n",
    "df_no_H3K4me3 = df_cis.query(\"H3K4me3==0\")[['chrom', 'start', 'end']].copy()\n",
    "df_cis = df_cis.query(\"H3K4me3>0\").copy()\n",
    "df_cis['q_id'] = df_cis['chrom'].map(str) + \"_\" + (df_cis['start']+1).map(str) + \"-\" + df_cis['end'].map(str) + \":.\"\n",
    "df_tags = df_cis[['q_id']].copy()\n",
    "df_total = pd.concat([df_need, df_need, df_tags], axis=0)\n",
    "df_total.drop_duplicates(keep=False, inplace=True)\n",
    "\n",
    "df_total[['chr-start', 'end']] = df_total['q_id'].str.split(\"-\", expand=True)\n",
    "df_total[['chrs','id', 'start']] = df_total['chr-start'].str.split(\"_\", expand=True)\n",
    "df_total['chr'] = df_total['chrs'].map(str) + \"_\" + df_total['id'].map(str)\n",
    "\n",
    "print(\"与基因重复的cis数目: \", df_cis.shape[0] - df_total.shape[0])\n",
    "df_total = df_total[['chr', 'start', 'end']].copy()\n",
    "df_total.columns = ['chrom', 'start', 'end']\n",
    "df_total['start'] = df_total['start'].map(int)-1\n",
    "df_total = pd.concat([df_total, df_no_H3K4me3], axis=0)\n",
    "df_total.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/cis_gene_homo/J668_cis_hg_filter.bed\", sep=\"\\t\", index=False, columns=['chrom', 'start', 'end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_total.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Tissue_diffloop_gene/result.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "39994\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_cis.bed\", sep=\"\\t\")\n",
    "df.drop_duplicates(keep=False, inplace=True, subset=['chrom', 'start', 'end'])\n",
    "print(df.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "685775\n",
      "403730 107326\n"
     ]
    }
   ],
   "source": [
    "## 划分保守loops和动态loops anchor\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type.bed\", sep=\"\\t\")\n",
    "df_Gloop = df.query(\"tags=='G-G' | tags=='G-N'\")[['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2']].copy()\n",
    "print(df_Gloop.shape[0])\n",
    "def tissue_loop_type(df):\n",
    "    df = df.copy()\n",
    "    df['number'] = [1]*df.shape[0]\n",
    "    df_tiss_loop_number = df.groupby(by=['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2'], as_index=False)['number'].sum()\n",
    "    df_uniq = df_tiss_loop_number.query(\"number==1\").copy()\n",
    "    df_conserved = df_tiss_loop_number.query(\"number>1\").copy()\n",
    "    print(df_uniq.shape[0], df_conserved.shape[0])\n",
    "    return df_uniq, df_conserved\n",
    "def loop_anchor(df):\n",
    "    df1 = df[['chrom1', 'start1', 'end1']].copy()\n",
    "    df2 = df[['chrom2', 'start2', 'end2']].copy()\n",
    "    df2.columns = ['chrom1', 'start1', 'end1']\n",
    "    df_return = pd.concat([df1, df2], axis=0)\n",
    "    df_return.drop_duplicates(keep='first', inplace=True)\n",
    "    return df_return\n",
    "df_uniq_loop, df_conserved_loop = tissue_loop_type(df_Gloop)\n",
    "df_uniq_anchor = loop_anchor(df_uniq_loop)\n",
    "df_conserved_anchor = loop_anchor(df_conserved_loop)\n",
    "df_uniq_anchor.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_conserved_dynamic/J668_uniq_loop_anchor.bed\", sep=\"\\t\", index=False, header=False)\n",
    "df_conserved_anchor.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_conserved_dynamic/J668_conserved_loop_anchor.bed\", sep=\"\\t\", index=False, header=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "146816 127511.36\n"
     ]
    }
   ],
   "source": [
    "# 分析保守的loops和特有的loops 其稳定的loop anchor数目\n",
    "# 1、对于特异的loop anchor 数目较多，选择与保守的loop 数目相同，并且取数目平均值\n",
    "import numpy as np\n",
    "def loop_anchor(df):\n",
    "    df1 = df[['chrom1', 'start1', 'end1']].copy()\n",
    "    df2 = df[['chrom2', 'start2', 'end2']].copy()\n",
    "    df2.columns = ['chrom1', 'start1', 'end1']\n",
    "    df_return = pd.concat([df1, df2], axis=0)\n",
    "    df_return.drop_duplicates(keep=False, inplace=True)\n",
    "    return df_return\n",
    "def loop_sample(df):\n",
    "    all_number = []\n",
    "    for i in range(100):\n",
    "        df_tmp = df.sample(107326)\n",
    "        df_tmp = loop_anchor(df_tmp)\n",
    "        number = 107326*2 - df_tmp.shape[0]\n",
    "        all_number.append(number)\n",
    "    return np.mean(all_number)\n",
    "df_conserved_anchor = loop_anchor(df_conserved_loop)\n",
    "conserved_keep_anchor = 107326*2 - df_conserved_anchor.shape[0]\n",
    "uniq_keep_anchor = loop_sample(df_uniq_loop)\n",
    "print(conserved_keep_anchor, uniq_keep_anchor)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "146816 67836 127511.36 87140.64\n"
     ]
    }
   ],
   "source": [
    "a, b = conserved_keep_anchor, 107326*2-conserved_keep_anchor\n",
    "c, d = uniq_keep_anchor, 107326*2-uniq_keep_anchor\n",
    "print(a, b, c, d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "## bnmapper 鉴定出的同源cis元件\n",
    "import pandas as pd\n",
    "bn_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/bnmapper/At_Cis_Dt_25.bed\"\n",
    "cis_f1 = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_Cis_At.bed\"\n",
    "cis_f2 = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_Cis_Dt.bed\"\n",
    "df_bn = pd.read_csv(bn_f, sep=\"\\t\", names =['chrom', 'start', 'end', 'bn_tags'])\n",
    "df_cis1 = pd.read_csv(cis_f1, sep=\"\\t\", names=['chrom', 'start', 'end', 'cis1'])\n",
    "df_cis2 = pd.read_csv(cis_f2, sep=\"\\t\", names=['chrom', 'start', 'end', 'cis2'])\n",
    "\n",
    "df_result = pd.DataFrame()\n",
    "for index, row in df_bn.iterrows():\n",
    "    C, S, E, T = row['chrom'], row['start'], row['end'], row['bn_tags']\n",
    "    df_tmp = df_cis2.query(\"chrom==@C & end>=@S & start<=@E\").copy()\n",
    "    df_tmp['cis1'] = [T]*df_tmp.shape[0]\n",
    "    df_result = pd.concat([df_result, df_tmp], axis=0)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/bnmapper/At_vs_Dt_homo.txt\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "## bnmapper 鉴定出的同源cis元件\n",
    "import pandas as pd\n",
    "bn_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/bnmapper/Dt_Cis_At_25.bed\"\n",
    "cis_f2 = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_Cis_At.bed\"\n",
    "cis_f1 = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_Cis_Dt.bed\"\n",
    "df_bn = pd.read_csv(bn_f, sep=\"\\t\", names =['chrom', 'start', 'end', 'bn_tags'])\n",
    "df_cis1 = pd.read_csv(cis_f1, sep=\"\\t\", names=['chrom', 'start', 'end', 'cis1'])\n",
    "df_cis2 = pd.read_csv(cis_f2, sep=\"\\t\", names=['chrom', 'start', 'end', 'cis2'])\n",
    "\n",
    "df_result = pd.DataFrame()\n",
    "for index, row in df_bn.iterrows():\n",
    "    C, S, E, T = row['chrom'], row['start'], row['end'], row['bn_tags']\n",
    "    df_tmp = df_cis2.query(\"chrom==@C & end>=@S & start<=@E\").copy()\n",
    "    df_tmp['cis1'] = [T]*df_tmp.shape[0]\n",
    "    df_result = pd.concat([df_result, df_tmp], axis=0)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/bnmapper/Dt_vs_At_homo.txt\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "# bnmapper 结果合并\n",
    "df1 = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/bnmapper/At_vs_Dt_homo.txt\", sep=\"\\t\")\n",
    "df2 = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/bnmapper/Dt_vs_At_homo.txt\", sep=\"\\t\")\n",
    "df1.columns = ['chrom', 'start', 'end', 'cis1', 'cis2']\n",
    "df = pd.concat([df1, df2], axis=0)[['cis1', 'cis2']].copy()\n",
    "df.drop_duplicates(keep='first', subset=['cis1', 'cis2'], inplace=True)\n",
    "\n",
    "df_cis1 = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_Cis_Dt.bed\", sep=\"\\t\", names=['chrom1', 'start1', 'end1','cis1'])\n",
    "df_cis2 = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_Cis_At.bed\", sep=\"\\t\", names=['chrom2', 'start2', 'end2','cis2'])\n",
    "\n",
    "df_a = pd.merge( df, df_cis1, how='left', on=['cis1'])\n",
    "df_ab = pd.merge(df_a, df_cis2, how='left', on=['cis2'])\n",
    "df_ab.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/bnmapper/J668_cis_hg_bnmapper.bed\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 同源cis 元件\n",
    "import pandas as pd\n",
    "blast_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/blastx/Ghir_HCis_At_blast_25_signel.txt\"\n",
    "cis_f1 = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_Cis_At.bed\"\n",
    "cis_f2 = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_Cis_Dt.bed\"\n",
    "\n",
    "df = pd.read_csv(blast_f, sep=\"\\t\", names=['chrom', 'start', 'end', 'tags'])\n",
    "df_cis1 = pd.read_csv(cis_f1, sep=\"\\t\", names=['chrom2', 'start2', 'end2', 'cis2_tags'])\n",
    "df_cis2 = pd.read_csv(cis_f2, sep=\"\\t\", names=['chrom', 'start', 'end', 'cis_tags'])\n",
    "\n",
    "df_result = pd.DataFrame()\n",
    "for index, row in df.iterrows():\n",
    "    C, S, E, T = row['chrom'], row['start'], row['end'], row['tags']\n",
    "    df_tmp = df_cis2.query(\"chrom==@C & end>=@S & start<=@E\").copy()\n",
    "    df_tmp['tags'] = [T]*df_tmp.shape[0]\n",
    "    df_result = pd.concat([df_result, df_tmp], axis=0)\n",
    "df_result.columns = ['chrom1', 'start1', 'end1', 'cis1_tags', 'tags']\n",
    "df_result[['chrom2_start2', 'end2']] = df_result['tags'].str.split(\"-\", expand = True)\n",
    "df_result[['chroma', 'chromb', 'starta']] = df_result['chrom2_start2'].str.split(\"_\", expand = True)\n",
    "df_result['start2'] = df_result['starta'].map(int) - 1\n",
    "\n",
    "df_result['end2'] = df_result['end2'].map(int)\n",
    "df_result['chrom2'] = df_result['chroma'].map(str) + \"_\" + df_result['chromb'].map(str)\n",
    "df_result = pd.merge(df_result, df_cis1, how='inner', on=['chrom2', 'start2', 'end2'])\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/blastx/Ghir_HCis_At_vs_Dt.bed\", sep=\"\\t\", index=False, columns=['chrom1', 'start1', 'end1', 'cis1_tags', 'chrom2', 'start2', 'end2', 'cis2_tags'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 同源cis元件\n",
    "import pandas as pd\n",
    "blast_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/blastx/Ghir_HCis_Dt_blast_25_signel.txt\"\n",
    "cis_f2 = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_Cis_At.bed\"\n",
    "cis_f1 = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_Cis_Dt.bed\"\n",
    "\n",
    "df = pd.read_csv(blast_f, sep=\"\\t\", names=['chrom', 'start', 'end', 'tags'])\n",
    "df_cis1 = pd.read_csv(cis_f1, sep=\"\\t\", names=['chrom2', 'start2', 'end2', 'cis2_tags'])\n",
    "df_cis2 = pd.read_csv(cis_f2, sep=\"\\t\", names=['chrom', 'start', 'end', 'cis_tags'])\n",
    "\n",
    "df_result = pd.DataFrame()\n",
    "for index, row in df.iterrows():\n",
    "    C, S, E, T = row['chrom'], row['start'], row['end'], row['tags']\n",
    "    df_tmp = df_cis2.query(\"chrom==@C & end>=@S & start<=@E\").copy()\n",
    "    df_tmp['tags'] = [T]*df_tmp.shape[0]\n",
    "    df_result = pd.concat([df_result, df_tmp], axis=0)\n",
    "df_result.columns = ['chrom1', 'start1', 'end1', 'cis1_tags', 'tags']\n",
    "df_result[['chrom2_start2', 'end2']] = df_result['tags'].str.split(\"-\", expand = True)\n",
    "df_result[['chroma', 'chromb', 'starta']] = df_result['chrom2_start2'].str.split(\"_\", expand = True)\n",
    "df_result['start2'] = df_result['starta'].map(int) - 1\n",
    "\n",
    "df_result['end2'] = df_result['end2'].map(int)\n",
    "df_result['chrom2'] = df_result['chroma'].map(str) + \"_\" + df_result['chromb'].map(str)\n",
    "df_result = pd.merge(df_result, df_cis1, how='inner', on=['chrom2', 'start2', 'end2'])\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/blastx/Ghir_HCis_Dt_vs_At.bed\", sep=\"\\t\", index=False, columns=['chrom1', 'start1', 'end1', 'cis1_tags', 'chrom2', 'start2', 'end2', 'cis2_tags'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 同源非基因区anchor\n",
    "import pandas as pd\n",
    "blast_f = \"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/Ghir_HNgene_At_blast.txt\"\n",
    "#blast_f = \"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/test.txt\"\n",
    "cis_f1 = \"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/At_nGene_anchor.bed\"\n",
    "cis_f2 = \"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/Dt_nGene_anchor.bed\"\n",
    "\n",
    "df = pd.read_csv(blast_f, sep=\"\\t\", names=['chrom', 'start', 'end', 'tags'])\n",
    "df_cis1 = pd.read_csv(cis_f1, sep=\"\\t\", names=['chrom', 'start', 'end'])\n",
    "df_cis2 = pd.read_csv(cis_f2, sep=\"\\t\", names=['chrom', 'start', 'end'])\n",
    "\n",
    "## 整理非基因anchor格式，也就是给非基因anchor打标签\n",
    "def NGeneAnchor_format(df, subg_tags):\n",
    "    df = df.copy()\n",
    "    df['subg'] = [subg_tags]*df.shape[0]\n",
    "    df['number'] = range(df.shape[0])\n",
    "    df['sigs'] = df['subg'] + df['number'].map(str)\n",
    "    df_return = df[['chrom', 'start', 'end', 'sigs']].copy()\n",
    "    return df_return\n",
    "df_cis1 = NGeneAnchor_format(df_cis1, 'At')\n",
    "df_cis2 = NGeneAnchor_format(df_cis2, 'Dt')\n",
    "## 切分为单条染色体，并查找对应的区段\n",
    "chrs_At = ['Ghir_A01', 'Ghir_A02', 'Ghir_A03', 'Ghir_A04', 'Ghir_A05', 'Ghir_A06', 'Ghir_A07', 'Ghir_A08', 'Ghir_A09', 'Ghir_A10', 'Ghir_A11', 'Ghir_A12', 'Ghir_A13']\n",
    "chrs_Dt = ['Ghir_D01', 'Ghir_D02', 'Ghir_D03', 'Ghir_D04', 'Ghir_D05', 'Ghir_D06', 'Ghir_D07', 'Ghir_D08', 'Ghir_D09', 'Ghir_D10', 'Ghir_D11', 'Ghir_D12', 'Ghir_D13']\n",
    "# 对格式进行调整\n",
    "def tags_format(df):\n",
    "    df[['a', 'b']] = df['tags'].str.split(\"-\", expand=True )\n",
    "    df[['a1', 'a2', 'a3']] = df['a'].str.split(\"_\", expand=True)\n",
    "    df[['b1', 'b2']] = df['b'].str.split(\":\", expand=True)\n",
    "    df['chrom2'] = df['a1'] + \"_\" + df['a2']\n",
    "    df['start2'] = df['a3'].map(int) - 1\n",
    "    df['end2'] = df['b1'].map(int)\n",
    "    df_return = df[['chrom', 'start', 'end', 'sigs', 'chrom2', 'start2', 'end2']].copy()\n",
    "    return df_return\n",
    "\n",
    "def region_pair(df, df_tags, chr_tags):\n",
    "    df_return = pd.DataFrame()\n",
    "    for c in chr_tags:\n",
    "        df_chr = df.query(\"chrom == @c\").copy()\n",
    "        df_tags_chr = df_tags.query(\"chrom == @c\").copy()\n",
    "        for index, row in df_chr.iterrows():\n",
    "            C, S, E, T = row['chrom'], row['start'], row['end'], row['tags']\n",
    "            df_tmp = df_tags_chr.query(\"chrom==@C & end>=@S & start<=@E\").copy()\n",
    "            df_tmp['tags'] = [T]*df_tmp.shape[0]\n",
    "            df_return = pd.concat([df_return, df_tmp], axis=0)\n",
    "    df_return = tags_format(df_return)\n",
    "    return df_return\n",
    "\n",
    "def final_format(df, df_cis):\n",
    "    df, df_cis = df.copy(), df_cis.copy()\n",
    "    df_cis.columns = ['chrom2', 'start2', 'end2', 'sigs2']\n",
    "    df_return = pd.merge(df, df_cis, how='inner', on=['chrom2', 'start2', 'end2'])\n",
    "    return df_return\n",
    "df_result = region_pair(df, df_cis2, chrs_Dt)\n",
    "df_result = final_format(df_result, df_cis1)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/At_NgeneAnchor_blast_Dt.bed\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 同源非基因区anchor, 从At基因组的非基因anchor 找对应于Dt基因组上的区域\n",
    "import pandas as pd\n",
    "blast_f = \"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/Ghir_HNgene_Dt_blast.txt\"\n",
    "#blast_f = \"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/test.txt\"\n",
    "\n",
    "cis_f1 = \"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/Dt_nGene_anchor.bed\"\n",
    "cis_f2 = \"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/At_nGene_anchor.bed\"\n",
    "\n",
    "df = pd.read_csv(blast_f, sep=\"\\t\", names=['chrom', 'start', 'end', 'tags'])\n",
    "df_cis1 = pd.read_csv(cis_f1, sep=\"\\t\", names=['chrom', 'start', 'end'])\n",
    "df_cis2 = pd.read_csv(cis_f2, sep=\"\\t\", names=['chrom', 'start', 'end'])\n",
    "\n",
    "## 整理非基因anchor格式，也就是给非基因anchor打标签\n",
    "def NGeneAnchor_format(df, subg_tags):\n",
    "    df = df.copy()\n",
    "    df['subg'] = [subg_tags]*df.shape[0]\n",
    "    df['number'] = range(df.shape[0])\n",
    "    df['sigs'] = df['subg'] + df['number'].map(str)\n",
    "    df_return = df[['chrom', 'start', 'end', 'sigs']].copy()\n",
    "    return df_return\n",
    "df_cis1 = NGeneAnchor_format(df_cis1, 'Dt')\n",
    "df_cis2 = NGeneAnchor_format(df_cis2, 'At')\n",
    "## 切分为单条染色体，并查找对应的区段\n",
    "chrs_At = ['Ghir_A01', 'Ghir_A02', 'Ghir_A03', 'Ghir_A04', 'Ghir_A05', 'Ghir_A06', 'Ghir_A07', 'Ghir_A08', 'Ghir_A09', 'Ghir_A10', 'Ghir_A11', 'Ghir_A12', 'Ghir_A13']\n",
    "chrs_Dt = ['Ghir_D01', 'Ghir_D02', 'Ghir_D03', 'Ghir_D04', 'Ghir_D05', 'Ghir_D06', 'Ghir_D07', 'Ghir_D08', 'Ghir_D09', 'Ghir_D10', 'Ghir_D11', 'Ghir_D12', 'Ghir_D13']\n",
    "# 对格式进行调整\n",
    "def tags_format(df):\n",
    "    df[['a', 'b']] = df['tags'].str.split(\"-\", expand=True )\n",
    "    df[['a1', 'a2', 'a3']] = df['a'].str.split(\"_\", expand=True)\n",
    "    df[['b1', 'b2']] = df['b'].str.split(\":\", expand=True)\n",
    "    df['chrom2'] = df['a1'] + \"_\" + df['a2']\n",
    "    df['start2'] = df['a3'].map(int) - 1\n",
    "    df['end2'] = df['b1'].map(int)\n",
    "    df_return = df[['chrom', 'start', 'end', 'sigs', 'chrom2', 'start2', 'end2']].copy()\n",
    "    return df_return\n",
    "\n",
    "def region_pair(df, df_tags, chr_tags):\n",
    "    df_return = pd.DataFrame()\n",
    "    for c in chr_tags:\n",
    "        df_chr = df.query(\"chrom == @c\").copy()\n",
    "        df_tags_chr = df_tags.query(\"chrom == @c\").copy()\n",
    "        for index, row in df_chr.iterrows():\n",
    "            C, S, E, T = row['chrom'], row['start'], row['end'], row['tags']\n",
    "            df_tmp = df_tags_chr.query(\"chrom==@C & end>=@S & start<=@E\").copy()\n",
    "            df_tmp['tags'] = [T]*df_tmp.shape[0]\n",
    "            df_return = pd.concat([df_return, df_tmp], axis=0)\n",
    "    df_return = tags_format(df_return)\n",
    "    return df_return\n",
    "\n",
    "def final_format(df, df_cis):\n",
    "    df, df_cis = df.copy(), df_cis.copy()\n",
    "    df_cis.columns = ['chrom2', 'start2', 'end2', 'sigs2']\n",
    "    df_return = pd.merge(df, df_cis, how='inner', on=['chrom2', 'start2', 'end2'])\n",
    "    return df_return\n",
    "df_result = region_pair(df, df_cis2, chrs_At)\n",
    "df_result = final_format(df_result, df_cis1)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/Dt_NgeneAnchor_blast_At.bed\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 同源非基因区anchor， 从Dt基因组的非基因anchor 找对应于At基因组上的区域\n",
    "import pandas as pd\n",
    "blast_f = \"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/Ghir_HNgene_At_blast.txt\"\n",
    "#blast_f = \"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/test.txt\"\n",
    "cis_f1 = \"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/At_nGene_anchor.bed\"\n",
    "cis_f2 = \"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/Dt_nGene_anchor.bed\"\n",
    "\n",
    "df = pd.read_csv(blast_f, sep=\"\\t\", names=['chrom', 'start', 'end', 'tags'])\n",
    "df_cis1 = pd.read_csv(cis_f1, sep=\"\\t\", names=['chrom', 'start', 'end'])\n",
    "df_cis2 = pd.read_csv(cis_f2, sep=\"\\t\", names=['chrom', 'start', 'end'])\n",
    "\n",
    "## 整理非基因anchor格式，也就是给非基因anchor打标签\n",
    "def NGeneAnchor_format(df, subg_tags):\n",
    "    df = df.copy()\n",
    "    df['subg'] = [subg_tags]*df.shape[0]\n",
    "    df['number'] = range(df.shape[0])\n",
    "    df['sigs'] = df['subg'] + df['number'].map(str)\n",
    "    df_return = df[['chrom', 'start', 'end', 'sigs']].copy()\n",
    "    return df_return\n",
    "df_cis1 = NGeneAnchor_format(df_cis1, 'At')\n",
    "df_cis2 = NGeneAnchor_format(df_cis2, 'Dt')\n",
    "## 切分为单条染色体，并查找对应的区段\n",
    "chrs_At = ['Ghir_A01', 'Ghir_A02', 'Ghir_A03', 'Ghir_A04', 'Ghir_A05', 'Ghir_A06', 'Ghir_A07', 'Ghir_A08', 'Ghir_A09', 'Ghir_A10', 'Ghir_A11', 'Ghir_A12', 'Ghir_A13']\n",
    "chrs_Dt = ['Ghir_D01', 'Ghir_D02', 'Ghir_D03', 'Ghir_D04', 'Ghir_D05', 'Ghir_D06', 'Ghir_D07', 'Ghir_D08', 'Ghir_D09', 'Ghir_D10', 'Ghir_D11', 'Ghir_D12', 'Ghir_D13']\n",
    "# 对格式进行调整\n",
    "def tags_format(df):\n",
    "    df[['a', 'b']] = df['tags'].str.split(\"-\", expand=True )\n",
    "    df[['a1', 'a2', 'a3']] = df['a'].str.split(\"_\", expand=True)\n",
    "    df[['b1', 'b2']] = df['b'].str.split(\":\", expand=True)\n",
    "    df['chrom2'] = df['a1'] + \"_\" + df['a2']\n",
    "    df['start2'] = df['a3'].map(int) - 1\n",
    "    df['end2'] = df['b1'].map(int)\n",
    "    df_return = df[['chrom', 'start', 'end', 'sigs', 'chrom2', 'start2', 'end2']].copy()\n",
    "    return df_return\n",
    "\n",
    "def region_pair(df, df_tags, chr_tags):\n",
    "    df_return = pd.DataFrame()\n",
    "    for c in chr_tags:\n",
    "        df_chr = df.query(\"chrom == @c\").copy()\n",
    "        df_tags_chr = df_tags.query(\"chrom == @c\").copy()\n",
    "        for index, row in df_chr.iterrows():\n",
    "            C, S, E, T = row['chrom'], row['start'], row['end'], row['tags']\n",
    "            df_tmp = df_tags_chr.query(\"chrom==@C & end>=@S & start<=@E\").copy()\n",
    "            df_tmp['tags'] = [T]*df_tmp.shape[0]\n",
    "            df_return = pd.concat([df_return, df_tmp], axis=0)\n",
    "    df_return = tags_format(df_return)\n",
    "    return df_return\n",
    "\n",
    "def final_format(df, df_cis):\n",
    "    df, df_cis = df.copy(), df_cis.copy()\n",
    "    df_cis.columns = ['chrom2', 'start2', 'end2', 'sigs2']\n",
    "    df_return = pd.merge(df, df_cis, how='inner', on=['chrom2', 'start2', 'end2'])\n",
    "    return df_return\n",
    "df_result = region_pair(df, df_cis2, chrs_Dt)\n",
    "df_result = final_format(df_result, df_cis1)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/At_NgeneAnchor_blast_Dt.bed\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "同源染色体间的同源非基因区域中At 亚基因组的anchor 数目： 78240\n",
      "同源染色体间的同源非基因区域中Dt 亚基因组的anchor 数目： 76257\n"
     ]
    }
   ],
   "source": [
    "## 将以At 找Dt和以Dt 找At 的非基因区域对应起来\n",
    "import pandas as pd\n",
    "df_At = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/At_NgeneAnchor_blast_Dt.bed\", sep=\"\\t\")\n",
    "df_Dt = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/Dt_NgeneAnchor_blast_At.bed\", sep=\"\\t\")\n",
    "\n",
    "df_At.columns = ['chrom_Dt', 'start_Dt', 'end_Dt', 'sigs_Dt', 'chrom_At', 'start_At', 'end_At', 'sigs_At']\n",
    "df_Dt.columns = ['chrom_At', 'start_At', 'end_At', 'sigs_At', 'chrom_Dt', 'start_Dt', 'end_Dt', 'sigs_Dt']\n",
    "df_At_Dt = pd.concat([df_At, df_Dt], axis=0)\n",
    "df_At_Dt['tags'] = range(df_At_Dt.shape[0])\n",
    "\n",
    "df_At_Dt_single = df_At_Dt.drop_duplicates(keep='first', subset=['chrom_Dt', 'start_Dt', 'end_Dt', 'sigs_Dt', 'chrom_At', 'start_At', 'end_At', 'sigs_At'])\n",
    "df_duplicates = pd.concat([df_At_Dt, df_At_Dt_single], axis=0)\n",
    "df_duplicates.drop_duplicates(keep=False, inplace=True)\n",
    "\n",
    "df_Dt_anchor = df_duplicates[['chrom_Dt', 'start_Dt', 'end_Dt']].copy()\n",
    "df_At_anchor = df_duplicates[['chrom_At', 'start_At', 'end_At']].copy()\n",
    "\n",
    "df_At_anchor.drop_duplicates(keep='first', inplace=True)\n",
    "df_Dt_anchor.drop_duplicates(keep='first', inplace=True)\n",
    "print(\"同源染色体间的同源非基因区域中At 亚基因组的anchor 数目：\", df_At_anchor.shape[0])\n",
    "print(\"同源染色体间的同源非基因区域中Dt 亚基因组的anchor 数目：\", df_Dt_anchor.shape[0])\n",
    "\n",
    "df_duplicates['At_tags'] = ['At']*df_duplicates.shape[0]\n",
    "df_duplicates['Dt_tags'] = ['Dt']*df_duplicates.shape[0]\n",
    "df_duplicates['At_number'] = range(df_duplicates.shape[0])\n",
    "df_duplicates['Dt_number'] = range(df_duplicates.shape[0])\n",
    "df_duplicates['At_homo_tags'] = df_duplicates['At_tags'] + df_duplicates['At_number'].map(str)\n",
    "df_duplicates['Dt_homo_tags'] = df_duplicates['Dt_tags'] + df_duplicates['Dt_number'].map(str)\n",
    "df_duplicates.to_csv('G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/strict/At_homo_Dt_homo_NgeneAnchor_strict.bed', sep=\"\\t\", index=False, columns = ['chrom_At', 'start_At', 'end_At', 'At_homo_tags', 'sigs_At', 'chrom_Dt', 'start_Dt', 'end_Dt', 'Dt_homo_tags', 'sigs_Dt'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "同源染色体间的同源非基因区域中At 亚基因组的anchor 数目： 359887\n",
      "同源染色体间的同源非基因区域中Dt 亚基因组的anchor 数目： 186362\n"
     ]
    }
   ],
   "source": [
    "## 取并集\n",
    "import pandas as pd\n",
    "df_At = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/At_NgeneAnchor_blast_Dt.bed\", sep=\"\\t\")\n",
    "df_Dt = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/Dt_NgeneAnchor_blast_At.bed\", sep=\"\\t\")\n",
    "df_At = df_At[['chrom', 'start', 'end', 'chrom2', 'start2', 'end2']].copy()\n",
    "df_Dt = df_Dt[['chrom', 'start', 'end', 'chrom2', 'start2', 'end2']].copy()\n",
    "\n",
    "## 将At参考和Dt参考的结果合并\n",
    "df_At.columns = ['chrom_Dt', 'start_Dt', 'end_Dt', 'chrom_At', 'start_At', 'end_At']\n",
    "df_Dt.columns = ['chrom_At', 'start_At', 'end_At', 'chrom_Dt', 'start_Dt', 'end_Dt']\n",
    "df_At_Dt = pd.concat([df_At, df_Dt], axis=0)\n",
    "df_At_Dt.drop_duplicates(keep='first', inplace=True) ## 获得合集 同源bin\n",
    "\n",
    "## 打上同源tags标签\n",
    "df_At_Dt['number'] = range(df_At_Dt.shape[0])\n",
    "df_At_Dt['Dt'] = ['Dt']*df_At_Dt.shape[0]\n",
    "df_At_Dt['At'] = ['At']*df_At_Dt.shape[0]\n",
    "df_At_Dt['At_tags'] = df_At_Dt['At'].map(str) + df_At_Dt['number'].map(str)\n",
    "df_At_Dt['Dt_tags'] = df_At_Dt['Dt'].map(str) + df_At_Dt['number'].map(str)\n",
    "df_At_Dt = df_At_Dt[['chrom_Dt', 'start_Dt', 'end_Dt', 'chrom_At', 'start_At', 'end_At', 'At_tags', 'Dt_tags']].copy()\n",
    "\n",
    "## 统计所存在的同源anchor数目\n",
    "df_Dt_anchor = df_At_Dt[['chrom_Dt', 'start_Dt', 'end_Dt']].copy()\n",
    "df_At_anchor = df_At_Dt[['chrom_At', 'start_At', 'end_At']].copy()\n",
    "\n",
    "df_At_anchor.drop_duplicates(keep='first', inplace=True)\n",
    "df_Dt_anchor.drop_duplicates(keep='first', inplace=True)\n",
    "print(\"同源染色体间的同源非基因区域中At 亚基因组的anchor 数目：\", df_At_anchor.shape[0])\n",
    "print(\"同源染色体间的同源非基因区域中Dt 亚基因组的anchor 数目：\", df_Dt_anchor.shape[0])\n",
    "\n",
    "df_At_Dt.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/loose/At_homo_Dt_homo_NgeneAnchor_loose.bed\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "其中同源染色体间的非基因区域同源anchor数目是: 73446\n",
      "其中非同源染色体间的非基因区域同源anchor数目是: 27420\n"
     ]
    }
   ],
   "source": [
    "## 分析有多少同源anchor非基因anchor区域是满足 两个同源染色体间的对应关系\n",
    "import pandas as pd\n",
    "import re\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/At_homo_Dt_homo_NgeneAnchor.bed\", sep=\"\\t\")\n",
    "pattern = r'[0-9]+'\n",
    "df['At_chr_number'] = df['chrom_At'].str.findall(pattern)\n",
    "df['Dt_chr_number'] = df['chrom_Dt'].str.findall(pattern)\n",
    "df_homo_chr = df.query(\"At_chr_number == Dt_chr_number\")\n",
    "df_Nhomo_chr = df.query(\"At_chr_number != Dt_chr_number\")\n",
    "print(\"其中同源染色体间的非基因区域同源anchor数目是:\",df_homo_chr.shape[0])\n",
    "print(\"其中非同源染色体间的非基因区域同源anchor数目是:\",df_Nhomo_chr.shape[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "确定同源cis元件数目： 3311\n"
     ]
    }
   ],
   "source": [
    "## 分析非基因同源anchor区域的同源cis元件\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/At_homo_Dt_homo_NgeneAnchor.bed\", sep=\"\\t\")\n",
    "## 给非基因anchor 区域打标签\n",
    "df_cis_At = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_Cis_At.bed\", sep=\"\\t\", names=['chrom_At', 'start_At', 'end_At', 'tags_At'])\n",
    "df_cis_Dt = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_Cis_Dt.bed\", sep=\"\\t\", names=['chrom_Dt', 'start_Dt', 'end_Dt', 'tags_Dt'])\n",
    "df_cis_At = df_cis_At[['chrom_At', 'start_At', 'end_At']].copy()\n",
    "df_cis_Dt = df_cis_Dt[['chrom_Dt', 'start_Dt', 'end_Dt']].copy()\n",
    "df_cis_At['cis_At'] = ['Cis']*df_cis_At.shape[0]\n",
    "df_cis_Dt['cis_Dt'] = ['Cis']*df_cis_Dt.shape[0]\n",
    "## 同源cis元件\n",
    "df_a = pd.merge(df, df_cis_At, how='left', on=['chrom_At', 'start_At', 'end_At'])\n",
    "df_ab = pd.merge(df_a, df_cis_Dt, how='left', on=['chrom_Dt', 'start_Dt', 'end_Dt'])\n",
    "df_homo_cis = df_ab.query(\"cis_At=='Cis' & cis_Dt=='Cis'\").copy()\n",
    "df_homo_cis.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/J668_homo_cis2.bed\", sep=\"\\t\", index=False)\n",
    "\n",
    "print(\"确定同源cis元件数目：\", df_homo_cis.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25293\n",
      "57905\n",
      "158175\n",
      "154054\n"
     ]
    }
   ],
   "source": [
    "## 1、整理出同源基因anchor, 这些bin包含包含了超过25%的基因占比\n",
    "import pandas as pd\n",
    "df_hg_pairs = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/Ghir_homo_gene_pairs.txt\", sep=\"\\t\")\n",
    "df_gene_bin = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/3kb_bin_gene.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'GeneId'])\n",
    "print(df_hg_pairs.shape[0])\n",
    "## At_homo_gene\n",
    "df_gene_bin.columns = ['chrom_At', 'start_At', 'end_At', 'GeneId_At']\n",
    "df_a = pd.merge(df_hg_pairs, df_gene_bin, how='left', on=['GeneId_At'])\n",
    "print(df_a.shape[0])\n",
    "df_gene_bin.columns = ['chrom_Dt', 'start_Dt', 'end_Dt', 'GeneId_Dt']\n",
    "df_ab = pd.merge(df_a, df_gene_bin, how='left', on=['GeneId_Dt'])\n",
    "print(df_ab.shape[0])\n",
    "df_ab.dropna(inplace=True)\n",
    "df_ab.drop_duplicates(inplace=True, subset=['chrom_At', 'start_At', 'end_At', 'chrom_Dt', 'start_Dt', 'end_Dt'])\n",
    "print(df_ab.shape[0])\n",
    "df_ab['number'] = range(df_ab.shape[0])\n",
    "df_ab['At_tags'] = ['AtG']*df_ab.shape[0]\n",
    "df_ab['Dt_tags'] = ['DtG']*df_ab.shape[0]\n",
    "df_ab['At_homo_tags'] = df_ab['At_tags'] + df_ab['number'].map(str)\n",
    "df_ab['Dt_homo_tags'] = df_ab['Dt_tags'] + df_ab['number'].map(str)\n",
    "df_ab.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/homo_gene_anchor/J668_homo_gene_bin1.bed\", sep=\"\\t\", index=False, columns=['chrom_At', 'start_At', 'end_At', 'At_homo_tags', 'chrom_Dt', 'start_Dt', 'end_Dt', 'Dt_homo_tags'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25293\n",
      "49882\n",
      "122801\n",
      "117007\n"
     ]
    }
   ],
   "source": [
    "## 1、整理出同源基因anchor, 这些bin包含包含了超过25%的基因占比\n",
    "import pandas as pd\n",
    "df_hg_pairs = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/Ghir_homo_gene_pairs.txt\", sep=\"\\t\")\n",
    "df_gene_bin = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/3kb_bin_gene_25ratio.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'GeneId'])\n",
    "print(df_hg_pairs.shape[0])\n",
    "## At_homo_gene\n",
    "df_gene_bin.columns = ['chrom_At', 'start_At', 'end_At', 'GeneId_At']\n",
    "df_a = pd.merge(df_hg_pairs, df_gene_bin, how='left', on=['GeneId_At'])\n",
    "print(df_a.shape[0])\n",
    "df_gene_bin.columns = ['chrom_Dt', 'start_Dt', 'end_Dt', 'GeneId_Dt']\n",
    "df_ab = pd.merge(df_a, df_gene_bin, how='left', on=['GeneId_Dt'])\n",
    "print(df_ab.shape[0])\n",
    "df_ab.dropna(inplace=True)\n",
    "df_ab.drop_duplicates(inplace=True, subset=['chrom_At', 'start_At', 'end_At', 'chrom_Dt', 'start_Dt', 'end_Dt'])\n",
    "print(df_ab.shape[0])\n",
    "df_ab['number'] = range(df_ab.shape[0])\n",
    "df_ab['At_tags'] = ['AtG']*df_ab.shape[0]\n",
    "df_ab['Dt_tags'] = ['DtG']*df_ab.shape[0]\n",
    "df_ab['At_homo_tags'] = df_ab['At_tags'] + df_ab['number'].map(str)\n",
    "df_ab['Dt_homo_tags'] = df_ab['Dt_tags'] + df_ab['number'].map(str)\n",
    "df_ab.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/homo_gene_anchor/J668_homo_gene_bin.bed\", sep=\"\\t\", index=False, columns=['chrom_At', 'start_At', 'end_At', 'At_homo_tags', 'chrom_Dt', 'start_Dt', 'end_Dt', 'Dt_homo_tags'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "非冗余bin的数目:  85799\n",
      "非冗余bin的数目:  154497\n"
     ]
    }
   ],
   "source": [
    "## 分析同源基因bin，和同源非基因bin 分别包含多少anchor数目\n",
    "import pandas as pd\n",
    "def bin_number(df):\n",
    "    df = df.copy()\n",
    "    df_a = df[['chrom_At', 'start_At', 'end_At']].copy()\n",
    "    df_b = df[['chrom_Dt', 'start_Dt', 'end_Dt']].copy()\n",
    "    df_ab = pd.concat([df_a, df_b], axis=0)\n",
    "    df_ab.drop_duplicates(keep='first', inplace=True)\n",
    "    print(\"非冗余bin的数目: \", df_ab.shape[0])\n",
    "df_HGbin = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/homo_gene_anchor/J668_homo_gene_bin.bed\", sep=\"\\t\")\n",
    "df_HNGbin = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/At_homo_Dt_homo_NgeneAnchor.bed\", sep=\"\\t\")\n",
    "\n",
    "bin_number(df_HGbin)\n",
    "bin_number(df_HNGbin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将同源非基因anchor 和同源基因anchor 整理为一个文件\n",
    "import pandas as pd\n",
    "df_HGbin = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/homo_gene_anchor/J668_homo_gene_bin.bed\", sep=\"\\t\")\n",
    "df_HNGbin = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/strict/At_homo_Dt_homo_NgeneAnchor_strict.bed\", sep=\"\\t\")\n",
    "df_HNGbin = df_HNGbin[['chrom_At', 'start_At', 'end_At', 'At_homo_tags', 'chrom_Dt', 'start_Dt', 'end_Dt', 'Dt_homo_tags']].copy()\n",
    "df_Hbin = pd.concat([df_HGbin, df_HNGbin], axis=0)\n",
    "df_Hbin.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/strict/J668_Hbin_pairs_strict.bed\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将同源非基因anchor 和同源基因anchor 整理为一个文件\n",
    "#! strict\n",
    "# import pandas as pd\n",
    "# df_HGbin = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/homo_gene_anchor/J668_homo_gene_bin.bed\", sep=\"\\t\")\n",
    "# df_HNGbin = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/strict/At_homo_Dt_homo_NgeneAnchor_strict.bed\", sep=\"\\t\")\n",
    "# df_HNGbin = df_HNGbin[['chrom_At', 'start_At', 'end_At', 'At_homo_tags', 'chrom_Dt', 'start_Dt', 'end_Dt', 'Dt_homo_tags']].copy()\n",
    "# df_Hbin = pd.concat([df_HGbin, df_HNGbin], axis=0)\n",
    "# df_Hbin.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/strict/J668_Hbin_pairs_strict.bed\", sep=\"\\t\", index=False)\n",
    "\n",
    "#! loose\n",
    "import pandas as pd\n",
    "df_HGbin = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/homo_gene_anchor/J668_homo_gene_bin.bed\", sep=\"\\t\")\n",
    "df_HNGbin = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/loose/At_homo_Dt_homo_NgeneAnchor_loose.bed\", sep=\"\\t\")\n",
    "df_HNGbin.columns = ['chrom_Dt', 'start_Dt', 'end_Dt', 'chrom_At', 'start_At', 'end_At', 'At_homo_tags', 'Dt_homo_tags']\n",
    "df_Hbin_pairs = pd.concat([df_HGbin, df_HNGbin], axis=0)\n",
    "df_Hbin_pairs.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/loose/J668_Hbin_pairs_loose.bed\", sep=\"\\t\", index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 确定同源bin区间 \n",
    "#! strict\n",
    "# import pandas as pd\n",
    "# df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/strict/J668_Hbin_pairs_strict.bed\", sep=\"\\t\")\n",
    "# df_At = df[['chrom_At', 'start_At', 'end_At', 'At_homo_tags']].copy()\n",
    "# df_Dt = df[['chrom_Dt', 'start_Dt', 'end_Dt', 'Dt_homo_tags']].copy()\n",
    "# df_At.columns = ['chrom', 'start', 'end', 'tags']\n",
    "# df_Dt.columns = ['chrom', 'start', 'end', 'tags']\n",
    "# df_AtDt = pd.concat([df_At, df_Dt], axis=0)\n",
    "# df_AtDt.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/strict/J668_Hbin_strict.bed\", sep=\"\\t\", index=False)\n",
    "#! loose\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/loose/J668_Hbin_pairs_loose.bed\", sep=\"\\t\")\n",
    "df_At = df[['chrom_At', 'start_At', 'end_At', 'At_homo_tags']].copy()\n",
    "df_Dt = df[['chrom_Dt', 'start_Dt', 'end_Dt', 'Dt_homo_tags']].copy()\n",
    "df_At.columns = ['chrom', 'start', 'end', 'tags']\n",
    "df_Dt.columns = ['chrom', 'start', 'end', 'tags']\n",
    "df_AtDt = pd.concat([df_At, df_Dt], axis=0)\n",
    "df_AtDt.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/loose/J668_Hbin_loose.bed\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将同源染色体loop 分为不同的类型，1、同源和非同源互作。2、同源非loop anchor。 3、保守。 4、同源，多loop调控\n",
    "import pandas as pd\n",
    "## 给loop anchor 打上 同源标签\n",
    "def loop_anchor_tags(df_loop, df_bin_tags):\n",
    "    df_loop, df_bin_tags = df_loop.copy(), df_bin_tags.copy()\n",
    "    ## 由于存在一个bin打上了多个标签的情况，因此将所有的bin去重\n",
    "    df_bin_tags['tags'] = ['Hbin']*df_bin_tags.shape[0]\n",
    "    df_bin_tags.drop_duplicates(inplace=True)\n",
    "\n",
    "    df_bin_tags.columns = ['chrom1', 'start1', 'end1', 'tags1']\n",
    "    df_a = pd.merge(df_loop, df_bin_tags, how='left', on=['chrom1', 'start1', 'end1'])\n",
    "    df_bin_tags.columns = ['chrom2', 'start2', 'end2', 'tags2']\n",
    "    df_ab = pd.merge(df_a, df_bin_tags, how='left', on=['chrom2', 'start2', 'end2'])\n",
    "    df_ab.fillna(\"NHbin\", inplace=True)\n",
    "    return df_ab\n",
    "\n",
    "#! 为了找同源anchor 但是对应的anchor没有loops这种情况\n",
    "## 将同源anchor对转换为bin，bin-tags， loop数目格式\n",
    "def HBin_number(df):\n",
    "    df = df[['chrom_At', 'start_At', 'end_At', 'anchor_number_At', 'chrom_Dt', 'start_Dt', 'end_Dt', 'anchor_number_Dt']].copy()\n",
    "    df_at = df[['chrom_At', 'start_At', 'end_At', 'anchor_number_Dt']].copy()\n",
    "    df_dt = df[['chrom_Dt', 'start_Dt', 'end_Dt', 'anchor_number_At']].copy()\n",
    "    ## 将一个bin包含多个对于的bin所包含的loop 数目进行求和\n",
    "    df_at_sum = df_at.groupby(by=['chrom_At', 'start_At', 'end_At'], as_index=False)['anchor_number_Dt'].sum()\n",
    "    df_dt_sum = df_dt.groupby(by=['chrom_Dt', 'start_Dt', 'end_Dt'], as_index=False)['anchor_number_At'].sum()\n",
    "    df_at_sum.columns = ['chrom', 'start', 'end', 'number']\n",
    "    df_dt_sum.columns = ['chrom', 'start', 'end', 'number']\n",
    "    df_return = pd.concat([df_at_sum, df_dt_sum], axis=0)\n",
    "    return df_return\n",
    "## 分析同源anchor 所包含的loop数目\n",
    "def HBin_LN(df_Hbin, df_bin_number):\n",
    "    df_Hbin, df_bin_number = df_Hbin.copy(), df_bin_number.copy()\n",
    "    df_bin_number.columns = ['chrom_At', 'start_At', 'end_At', 'anchor_number_At']\n",
    "    df_a = pd.merge(df_Hbin, df_bin_number, how='left', on=['chrom_At', 'start_At', 'end_At'])\n",
    "    df_bin_number.columns = ['chrom_Dt', 'start_Dt', 'end_Dt', 'anchor_number_Dt']\n",
    "    df_ab = pd.merge(df_a, df_bin_number, how='left', on=['chrom_Dt', 'start_Dt', 'end_Dt'])\n",
    "    df_ab.fillna(0, inplace=True)\n",
    "    df_return = HBin_number(df_ab)\n",
    "    return df_return\n",
    "## 给loop anchor 打上 对应的同源anchor的数目\n",
    "def loop_anchor_number(df_loop, df_bin_number):\n",
    "    df_loop, df_bin_number = df_loop.copy(), df_bin_number.copy()\n",
    "    df_bin_number.columns = ['chrom1', 'start1', 'end1', 'number1_pair']\n",
    "    df_a = pd.merge(df_loop, df_bin_number, how='left', on=['chrom1', 'start1', 'end1'])\n",
    "    df_bin_number.columns = ['chrom2', 'start2', 'end2', 'number2_pair']\n",
    "    df_ab = pd.merge(df_a, df_bin_number, how='left', on=['chrom2', 'start2', 'end2'])\n",
    "    df_ab.fillna(0, inplace=True)\n",
    "    return df_ab\n",
    "\n",
    "# ## 分析同源保守的loops数目\n",
    "# 1、整理所有同源tags 对\n",
    "def Hbin_tags_pair(df_Hbin_pair):\n",
    "    df_Hbin_pair = df_Hbin_pair.copy()\n",
    "    df_Hbin_pair['pair_tags'] = df_Hbin_pair['At_homo_tags'].map(str) + \"-\" + df_Hbin_pair['Dt_homo_tags'].map(str)\n",
    "    df_a = df_Hbin_pair[['chrom_At', 'start_At', 'end_At', 'pair_tags']].copy()\n",
    "    df_b = df_Hbin_pair[['chrom_Dt', 'start_Dt', 'end_Dt', 'pair_tags']].copy()\n",
    "    df_a.columns = ['chrom', 'start', 'end', 'pair_tags']\n",
    "    df_b.columns = ['chrom', 'start', 'end', 'pair_tags']\n",
    "    df_return = pd.concat([df_a, df_b], axis=0)\n",
    "    df_return.drop_duplicates(keep='first', inplace=True)\n",
    "    #print(df_return.head())\n",
    "    return df_return\n",
    "# 2、给所有的loop anchor 打上 同源tags 对标签\n",
    "def loop_Hbin_tags(df_loop_Hbin_pairs, df_Hbin_tags_pair):\n",
    "    df_loop_Hbin_pairs, df_Hbin_tags_pair = df_loop_Hbin_pairs.copy(), df_Hbin_tags_pair.copy()\n",
    "    df_Hbin_tags_pair.columns = ['chrom1', 'start1', 'end1', 'pair_tags1']\n",
    "    df_a = pd.merge(df_loop_Hbin_pairs, df_Hbin_tags_pair, how='left', on=['chrom1', 'start1', 'end1'])\n",
    "    df_Hbin_tags_pair.columns = ['chrom2', 'start2', 'end2', 'pair_tags2']\n",
    "    df_ab = pd.merge(df_a, df_Hbin_tags_pair, how='left', on=['chrom2', 'start2', 'end2'])\n",
    "    df_ab['pair_tags'] = df_ab['pair_tags1'].map(str) + \"-\" + df_ab['pair_tags2'].map(str)\n",
    "    ## 对标签进行排序\n",
    "    for index, row in df_ab.iterrows():\n",
    "        sorted_tags = \"-\".join(sorted(row['pair_tags'].split(\"-\")))\n",
    "        df_ab.loc[index, 'sorted_pair_tags'] = sorted_tags\n",
    "    df_return = df_ab[['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'sorted_pair_tags']].copy()\n",
    "    df_return.drop_duplicates(keep='first', inplace=True)\n",
    "    #print(df_return.head())\n",
    "    return df_return \n",
    "# 3、计算亚基因组同源loops\n",
    "def Homo_loops(df):\n",
    "    df = df.copy()\n",
    "    df['number'] = [1]*df.shape[0]\n",
    "    df_tags = df.groupby(by=['sorted_pair_tags'], as_index=False)['number'].sum()\n",
    "    df_conserved_loop_tags = df_tags.query(\"number>1\")['sorted_pair_tags'].copy()\n",
    "    df_conserved_loop = pd.merge(df, df_conserved_loop_tags, how='inner', on=['sorted_pair_tags'])\n",
    "    df_return = df_conserved_loop.copy()\n",
    "    return df_return\n",
    "\n",
    "## 1、有非同源anchor的互作\n",
    "def NHanchor_loops(df):\n",
    "    df = df.copy()\n",
    "    df_loops = df.query(\"tags1 == 'NHbin' | tags2 == 'NHbin'\").copy()\n",
    "    df_return = df_loops.copy()\n",
    "    return df_return\n",
    "\n",
    "## 2、有同源anchor 但无loop互作\n",
    "def Hanchor_Nloops(df):\n",
    "    df = df.copy()\n",
    "    #df_loops = df.query(\"tags1 != 'NHbin' & tags2 != 'NHbin' &  number1_pair == 0 & number2_pair == 0\" ).copy()\n",
    "    df_loops = df.query(\"tags1 != 'NHbin' & tags2 != 'NHbin' &  (number1_pair == 0 | number2_pair == 0)\" ).copy()\n",
    "    df_return = df_loops.copy()\n",
    "    return df_return\n",
    "\n",
    "## 4、将所有结果整合，并且给动态loop打上标签\n",
    "def all_loops_tags(df_loops, df_NHanchor_loops, df_Hanchor_Nloops, df_conserved_loops):\n",
    "    df_loops, df_NHanchor_loops, df_Hanchor_Nloops, df_conserved_loops = df_loops.copy(), df_NHanchor_loops.copy(), df_Hanchor_Nloops.copy(), df_conserved_loops.copy()\n",
    "    ## 1、NHanchor_loops\n",
    "    df_NHanchor_loops['tags'] = ['NHanchor']*df_NHanchor_loops.shape[0]\n",
    "    df_NHanchor_loops = df_NHanchor_loops[['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tags']].copy()\n",
    "    ## 2、Hanchor_Nloops\n",
    "    df_Hanchor_Nloops['tags'] = ['NHloops']*df_Hanchor_Nloops.shape[0]\n",
    "    df_Hanchor_Nloops = df_Hanchor_Nloops[['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tags']].copy()\n",
    "    ## 3、dynamic loops\n",
    "    df_conserved_loops['tags'] = ['Conserved']*df_conserved_loops.shape[0]\n",
    "    df_conserved_loops = df_conserved_loops[['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tags']].copy()\n",
    "    df_three = pd.concat([df_NHanchor_loops, df_Hanchor_Nloops, df_conserved_loops], axis=0)\n",
    "    df_loop = pd.merge(df_loops, df_three, how='left', on=['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2'])\n",
    "    df_loop.fillna(\"Dynamic\", inplace=True)\n",
    "    df_loop.drop_duplicates(keep='first', inplace=True)\n",
    "    return df_loop\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/loose/J668_subg_loop_compare2.bed\", sep=\"\\t\")\n",
    "df['number'] = [1]*df.shape[0]\n",
    "df_return = df.groupby(by=['tags', 'tissues'], as_index=False)['number'].sum()\n",
    "df_return.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/loose/J668_subg_loop_compare_statics.bed\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## 2、分析保守的同源loops\n",
    "# def homo_loops(df, ):\n",
    "df_anther_loop = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/final_loops/anther_final_loop.bed\", sep=\"\\t\")\n",
    "df_Hbin = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/strict/J668_Hbin_strict.bed\", sep=\"\\t\")\n",
    "df_Hbin_pair = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/strict/J668_Hbin_pairs_strict.bed\", sep=\"\\t\")\n",
    "# 读取每个bin的loop数目\n",
    "df_LN = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/J668_loop_anchor_number.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'tissues', 'number', 'tags'])\n",
    "# 选择anther 的bin loop数目\n",
    "df_anther_LN = df_LN.query(\"tissues=='anther'\")[['chrom', 'start', 'end', 'number']].copy()\n",
    "\n",
    "# 1、加上同源cis 标签\n",
    "df_anther_loop_tags = loop_anchor_tags(df_anther_loop, df_Hbin) ## 加上同源，非同源标签\n",
    "#print(df_anther_loop_tags.shape[0])\n",
    "# #2、加上对应亚基因组 number标签\n",
    "df_Hbin_LN = HBin_LN(df_Hbin_pair, df_anther_LN) ## 计算每个同源bin 所对应的同源bin 包含的loops 数目\n",
    "df_anther_loop_tags = loop_anchor_number(df_anther_loop_tags, df_Hbin_LN) ## 给bin 加上loop数目标签\n",
    "\n",
    "#! 1、确定为序列缺失loops\n",
    "df_NHanchor_loops = NHanchor_loops(df_anther_loop_tags)\n",
    "#! 2、确定为loop anchor 同源但是缺少对应的loops\n",
    "df_Hanchor_Nloops = Hanchor_Nloops(df_anther_loop_tags)\n",
    "#print(df_NHanchor_loops.shape[0])\n",
    "\n",
    "#! 3、确定保守的loops\n",
    "# 1、确定两端都是同源的loop anchor\n",
    "df_Hanchor = df_anther_loop_tags.query(\"tags1=='Hbin' & tags2=='Hbin'\").copy()\n",
    "df_Hbin_pair = Hbin_tags_pair(df_Hbin_pair)\n",
    "df_loop_tags = loop_Hbin_tags(df_Hanchor, df_Hbin_pair)\n",
    "df_conserved_loops = Homo_loops(df_loop_tags)\n",
    "\n",
    "df_NHanchor_loops.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/strict/anther_NHanchor.bed\", sep=\"\\t\", index=False)\n",
    "df_Hanchor_Nloops.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/strict/anther_Nloops.bed\", sep=\"\\t\", index=False)\n",
    "df_conserved_loops.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/strict/anther_conserved.bed\", sep=\"\\t\", index=False)\n",
    "#! 4、确定动态调控的loops, 并给所有loop 打上标签\n",
    "df_result = all_loops_tags(df_anther_loop, df_NHanchor_loops, df_Hanchor_Nloops, df_conserved_loops)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/strict/anther_subg_loop_compare.bed\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 1、确定非同源基因bin 中并不包含同源基因bin\n",
    "import pandas as pd\n",
    "df_HGbin = pd.read_csv(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 对严格的非同源基因对做整理分析\n",
    "tissues = ['anther', 'cotyledon', 'fiber_10DPA', 'fiber_20DPA', 'hypocotyl', 'leaf', 'ovule', 'petal', 'radicle', 'root', 'stem', 'stigma']\n",
    "#tissues = ['anther']\n",
    "df_Hbin = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/strict/J668_Hbin_strict.bed\", sep=\"\\t\") ## \n",
    "df_Hbin_pair = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/strict/J668_Hbin_pairs_strict.bed\", sep=\"\\t\")\n",
    "df_LN = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/J668_loop_anchor_number.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'tissues', 'number', 'tags'])\n",
    "# print(df_Hbin.head())\n",
    "# print(df_Hbin_pair.head())\n",
    "for t in tissues:\n",
    "    input_loop = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/final_loops/{}_final_loop.bed\".format(t)\n",
    "    output_loop = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/strict/{}_subg_loop_compare.bed\".format(t)\n",
    "    df_Tloops = pd.read_csv(input_loop, sep=\"\\t\")\n",
    "    df_T_LN = df_LN.query(\"tissues == @t\")[['chrom', 'start', 'end', 'number']].copy()\n",
    "    print(df_T_LN.head())\n",
    "    # 1、打上同源cis 标签\n",
    "    df_Tloops = loop_anchor_tags(df_Tloops, df_Hbin)\n",
    "    # 2、加上同源bin所包含的loop数目标签\n",
    "    df_Hbin_LN = HBin_LN(df_Hbin_pair, df_T_LN)\n",
    "    df_Tloops = loop_anchor_number(df_Tloops, df_Hbin_LN) ## 给loop anchor打上对于的同源bin所包含的loop数目标签\n",
    "    # 3、序列缺失loops\n",
    "    df_NHanchor_loops = NHanchor_loops(df_Tloops)\n",
    "    # 4、同源，但loops缺失\n",
    "    df_Hanchor_Nloops = Hanchor_Nloops(df_Tloops)\n",
    "    # 5、保守的loops\n",
    "    df_Hanchor = df_Tloops.query(\"tags1=='Hbin' & tags2=='Hbin'\").copy() ## 选出两端都是同源bin的loops\n",
    "    df_Hbin_pair2 = Hbin_tags_pair(df_Hbin_pair) ## 给同源bin打上 自身tags和对应的tags\n",
    "    df_loop_tags = loop_Hbin_tags(df_Hanchor, df_Hbin_pair2) ## 给两端都是同源bin的loop 打上同源bin对的标签\n",
    "    df_conserved_loops = Homo_loops(df_loop_tags) ## 确定保守的loops\n",
    "    # 6、将所有的loop 进行划分\n",
    "    df_result = all_loops_tags(df_Tloops, df_NHanchor_loops, df_Hanchor_Nloops, df_conserved_loops)\n",
    "    df_result.to_csv(output_loop, sep=\"\\t\", index=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 312,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 调整loop， 之前定义NHloops 将loops 两端都视为 非loops anchor， 现在将其中一端视为非loop anchor\n",
    "tissues = ['anther', 'cotyledon', 'fiber_10DPA', 'fiber_20DPA', 'hypocotyl', 'leaf', 'ovule', 'petal', 'radicle', 'root', 'stem', 'stigma']\n",
    "def loop_type_number(df):\n",
    "    df = df.copy()\n",
    "    df['number'] = [1]*df.shape[0]\n",
    "    df_number = df.groupby(by='tags', as_index=False)['number'].sum()\n",
    "    print(df_number)\n",
    "for t in tissues:\n",
    "    df_tmp = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/loose/{}_subg_loop_compare.bed\".format(t), sep=\"\\t\")\n",
    "    df_NHloops = df_tmp.query(\"tags1=='Hbin' & tags2=='Hbin' & (number1_pair==0 | number2_pair==0)\").copy()\n",
    "    df_NHloops['tags'] = ['NHloops']*df_NHloops.shape[0]\n",
    "    df_result = pd.concat([df_tmp, df_NHloops])\n",
    "    df_result.drop_duplicates(keep='last', inplace=True, subset=['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2'])\n",
    "    output_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/loose/{}_subg_loop_compare2.bed\".format(t)\n",
    "    df_result.to_csv(output_f, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 对宽松的非同源基因对做整理分析\n",
    "tissues = ['hypocotyl', 'anther', 'cotyledon', 'fiber_10DPA', 'fiber_20DPA',  'leaf', 'ovule', 'petal', 'radicle', 'root', 'stem', 'stigma']\n",
    "#tissues = [ 'root', 'stem', 'stigma']\n",
    "#tissues = ['anther']\n",
    "df_Hbin = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/loose/J668_Hbin_loose.bed\", sep=\"\\t\") ## \n",
    "df_Hbin_pair = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/homo_nogene/loose/J668_Hbin_pairs_loose.bed\", sep=\"\\t\")\n",
    "df_LN = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/J668_loop_anchor_number.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'tissues', 'number', 'tags'])\n",
    "\n",
    "for t in tissues:\n",
    "    input_loop = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/final_loops/{}_final_loop.bed\".format(t)\n",
    "    output_loop = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/loose/{}_subg_loop_compare.bed\".format(t)\n",
    "    df_Tloops = pd.read_csv(input_loop, sep=\"\\t\")\n",
    "    df_T_LN = df_LN.query(\"tissues == @t\")[['chrom', 'start', 'end', 'number']].copy()\n",
    "    #print(df_T_LN.head())\n",
    "    # 1、打上同源cis 标签\n",
    "    df_Tloops = loop_anchor_tags(df_Tloops, df_Hbin)\n",
    "    #print(df_Tloops)\n",
    "    # # 2、加上同源bin所包含的loop数目标签\n",
    "    df_Hbin_LN = HBin_LN(df_Hbin_pair, df_T_LN)\n",
    "    df_Tloops = loop_anchor_number(df_Tloops, df_Hbin_LN) ## 给loop anchor打上对于的同源bin所包含的loop数目标签\n",
    "    print(df_Tloops)\n",
    "    # # 3、序列缺失loops\n",
    "    df_NHanchor_loops = NHanchor_loops(df_Tloops)\n",
    "    # # 4、同源，但loops缺失\n",
    "    df_Hanchor_Nloops = Hanchor_Nloops(df_Tloops)\n",
    "    # # 5、保守的loops\n",
    "    df_Hanchor = df_Tloops.query(\"tags1=='Hbin' & tags2=='Hbin'\").copy() ## 选出两端都是同源bin的loops\n",
    "    df_Hbin_pair2 = Hbin_tags_pair(df_Hbin_pair) ## 给同源bin打上 自身tags和对应的tags\n",
    "    df_loop_tags = loop_Hbin_tags(df_Hanchor, df_Hbin_pair2) ## 给两端都是同源bin的loop 打上同源bin对的标签\n",
    "    df_conserved_loops = Homo_loops(df_loop_tags) ## 确定保守的loops\n",
    "    # # 6、将所有的loop 进行划分\n",
    "    df_result = all_loops_tags(df_Tloops, df_NHanchor_loops, df_Hanchor_Nloops, df_conserved_loops)\n",
    "    #print(df_result)\n",
    "    df_result.to_csv(output_loop, sep=\"\\t\", index=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将cis 标签加入loop 的tags\n",
    "#import pandas as pd\n",
    "def loop_add_cis(df_loop, df_cis):\n",
    "    df_loop, df_cis = df_loop.copy(), df_cis.copy()\n",
    "    #df_loop[['type1', 'type2']] = df_loop['loop_type'].str.split(\"-\", expand = True)\n",
    "    df_cis = df_cis[['chrom', 'start', 'end']].copy()\n",
    "    df_cis['cis'] = ['Cis']*df_cis.shape[0]\n",
    "    df_cis.columns = ['chrom1', 'start1', 'end1', 'Cis1']\n",
    "    df_a = pd.merge(df_loop, df_cis, how='left', on=['chrom1', 'start1', 'end1'])\n",
    "    df_cis.columns = ['chrom2', 'start2', 'end2', 'Cis2']\n",
    "    df_ab = pd.merge(df_a, df_cis, how='left', on=['chrom2', 'start2', 'end2'])\n",
    "    df_ab['Cis1'] = df_ab['Cis1'].fillna(df_ab['tags1'])\n",
    "    df_ab['Cis2'] = df_ab['Cis2'].fillna(df_ab['tags2'])\n",
    "    df_ab['loop_type'] = df_ab['Cis1'] + '-' + df_ab['Cis2']\n",
    "    # old_tags = ['Cis-G', 'N-Cis']\n",
    "    # new_tags = ['G-Cis', 'Cis-N']\n",
    "    # df_ab['loop_type'].replace(old_tags, new_tags)\n",
    "    df_return = df_ab[['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2',  'tissues', 'tags', 'loop_type']].copy()\n",
    "    return df_return\n",
    "\n",
    "df_loop = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type.bed\", sep=\"\\t\")\n",
    "df_cis = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_cis.bed\", sep=\"\\t\")\n",
    "df_result = loop_add_cis(df_loop, df_cis)\n",
    "#print(df_result.query(\"chrom1=='Ghir_A01' & start2==2310000 & end2==2313000 & chrom2=='Ghir_A01' & start1==2115000 & end1==2118000\"))\n",
    "#df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type_all2.csv\", index=False)\n",
    "\n",
    "\n",
    "\n",
    "## 将NG, Cis-G, 基因放在前面去\n",
    "def NG_CisG(df):\n",
    "    df_NG_CisG = df.query(\"loop_type=='N-G' | loop_type=='Cis-G' | loop_type=='N-Cis'\").copy()\n",
    "    df_other = df.query(\"~(loop_type=='N-G' | loop_type=='Cis-G' | loop_type=='N-Cis')\").copy()\n",
    "    df_NG_CisG.columns = ['chrom2', 'start2', 'end2', 'chrom1', 'start1', 'end1', 'tissues', 'tags', 'loop_type'].copy()\n",
    "    old_tags, new_tags = ['N-G', 'Cis-G', 'N-Cis'], ['G-N', 'G-Cis', 'Cis-N']\n",
    "    df_NG_CisG['loop_type'].replace(old_tags, new_tags, inplace=True)\n",
    "    df_return = pd.concat([df_other, df_NG_CisG], axis=0)\n",
    "    return df_return\n",
    "df_result = NG_CisG(df_result)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type_cis.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 统计各不同类型loop所包含G-G，G-Cis，G-N的比例\n",
    "import pandas as pd\n",
    "tissues = ['anther', 'cotyledon', 'fiber_10DPA', 'fiber_20DPA', 'hypocotyl', 'leaf', 'ovule', 'petal', 'radicle', 'root', 'stem', 'stigma']\n",
    "df_all = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/loose/{}_subg_loop_compare.bed\".format(t)\n",
    "    df_tmp = pd.read_csv(input_f, sep=\"\\t\")\n",
    "    df_tmp = df_tmp[['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tags']].copy()\n",
    "    df_tmp.columns = ['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'type']\n",
    "    df_tmp['tissues'] = [t]*df_tmp.shape[0]\n",
    "    df_all = pd.concat([df_all, df_tmp], axis=0)\n",
    "df_loop_type = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type_all2.csv\")\n",
    "df_loop = pd.merge(df_all, df_loop_type, how='inner', on=['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tissues'])\n",
    "\n",
    "## 将与基因相关的loop拉出来\n",
    "df_loop_G = df_loop.query(\"loop_type.str.contains('G')\").copy()\n",
    "df_loop_G['number'] = [1]*df_loop_G.shape[0]\n",
    "\n",
    "df_Total_number = df_loop_G.groupby(by=['loop_type', 'tissues'], as_index=False)['number'].sum()\n",
    "df_number = df_loop_G.groupby(by=['loop_type', 'tissues', 'type'], as_index=False)['number'].sum()\n",
    "df_result = pd.merge(df_Total_number, df_number, how='inner', on=['loop_type', 'tissues'])\n",
    "df_result['ratio'] = df_result['number_y']/df_result['number_x']\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/loose/J668_loop_mutip_type_statics.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 313,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将所有不同类型的loop合并为一个文件\n",
    "import pandas as pd\n",
    "tissues = ['anther', 'cotyledon', 'fiber_10DPA', 'fiber_20DPA', 'hypocotyl', 'leaf', 'ovule', 'petal', 'radicle', 'root', 'stem', 'stigma']\n",
    "df_result = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    df_tmp = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/loose/{}_subg_loop_compare2.bed\".format(t), sep=\"\\t\")\n",
    "    df_tmp['tissues'] = [t]*df_tmp.shape[0]\n",
    "    df_result = pd.concat([df_result, df_tmp], axis=0)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/loose/J668_subg_loop_compare2.bed\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "317751 1829710\n",
      "169937\n",
      "3166\n",
      "3862\n",
      "2833\n",
      "hypcotyl 数目： 3862\n",
      "hypcotyl 去重后的数目： 2833\n",
      "          GeneId_At        GeneId_Dt tissues  tags_At   tags_Dt\n",
      "1   Ghir_A03G006670  Ghir_D03G012160  anther  NHloops       NaN\n",
      "4   Ghir_A05G006440  Ghir_D05G006510  anther      NaN   NHloops\n",
      "7   Ghir_A09G014910  Ghir_D09G014170  anther  NHloops       NaN\n",
      "9   Ghir_A01G018960  Ghir_D01G020400  anther  NHloops  NHanchor\n",
      "11  Ghir_A07G022270  Ghir_D07G022350  anther      NaN   NHloops\n",
      "        tissues  NHloop\n",
      "0        anther    3020\n",
      "1     cotyledon    3128\n",
      "2   fiber_10DPA    3142\n",
      "3   fiber_20DPA    3026\n",
      "4     hypocotyl    2833\n",
      "5          leaf    2865\n",
      "6         ovule    2484\n",
      "7         petal    3595\n",
      "8       radicle    2643\n",
      "9          root    2862\n",
      "10         stem    2990\n",
      "11       stigma    3400\n"
     ]
    }
   ],
   "source": [
    "## 分析前三种类型的loops 与同源基因bias表达相关的数目\n",
    "import pandas as pd\n",
    "# 1、找到同源bias 基因bin\n",
    "def BiasG_bin(df_bias, df_GBin):\n",
    "    df_bias, df_GBin = df_bias.copy(), df_GBin.copy()\n",
    "    df_result = pd.merge(df_bias, df_GBin, how='inner', on=['GeneId'])\n",
    "    #print(df_result.shape[0])\n",
    "    return df_result\n",
    "# 2、分析各loops对于Bias 基因的影响\n",
    "def loop_bias(df_loop, df_bias):\n",
    "    df_loop, df_bias = df_loop.copy(), df_bias.copy()\n",
    "    df_bias.columns = ['GeneId1', 'TPM1', 'Bias1', 'tissues', 'chrom1', 'start1', 'end1']\n",
    "    df_a = pd.merge(df_loop, df_bias, how='left', on=['chrom1', 'start1', 'end1', 'tissues'])\n",
    "    #print(df_a.shape[0])\n",
    "    df_bias.columns = ['GeneId2', 'TPM2', 'Bias2', 'tissues', 'chrom2', 'start2', 'end2']\n",
    "    df_ab = pd.merge(df_a, df_bias, how='left', on=['chrom2', 'start2', 'end2', 'tissues'])\n",
    "    #print(df_ab.shape[0])\n",
    "    df_return = df_ab.dropna(subset=['Bias1', 'Bias2'], how='all')\n",
    "    return df_return\n",
    "\n",
    "# 统计数目\n",
    "def stat_number(df):\n",
    "    def tissues_number(df, tags):\n",
    "        df = df.copy()\n",
    "        df['number'] = [1]*df.shape[0]\n",
    "        df_return = df.groupby(by=['tissues'], as_index=False)['number'].sum()\n",
    "        df_return.columns = ['tissues', tags]\n",
    "        return df_return\n",
    "    df = df.copy()\n",
    "    df_NHanchor = df.query(\"tags_At=='NHanchor' | tags_Dt=='NHanchor'\").copy()\n",
    "    df_NHloop = df.query(\"(tags_At=='NHloops' | tags_Dt=='NHloops')\").copy()\n",
    "    df_dynamic = df.query(\"tags_At=='Dynamic' | tags_Dt=='Dynamic'\").copy()\n",
    "    df_a = df_NHloop.query(\"tissues=='hypocotyl'\")\n",
    "    print(\"hypcotyl 数目：\", df_a.shape[0])\n",
    "    \n",
    "    df_NHanchor.drop_duplicates(keep='first', subset=['GeneId_At', 'GeneId_Dt', 'tissues'], inplace=True)\n",
    "    df_NHloop.drop_duplicates(keep='first', subset=['GeneId_At', 'GeneId_Dt', 'tissues'], inplace=True)\n",
    "    df_dynamic.drop_duplicates(keep='first', subset=['GeneId_At', 'GeneId_Dt', 'tissues'], inplace=True)\n",
    "    df_a = df_NHloop.query(\"tissues=='hypocotyl'\")\n",
    "    print(\"hypcotyl 去重后的数目：\", df_a.shape[0])\n",
    "    print(df_NHloop.head())\n",
    "    df_NHanchor = tissues_number(df_NHanchor, 'NHanchor')\n",
    "    df_NHloop = tissues_number(df_NHloop, 'NHloop')\n",
    "    print(df_NHloop)\n",
    "    df_dynamic = tissues_number(df_dynamic , 'Dynamic')\n",
    "    df_a = pd.merge(df_NHanchor, df_NHloop, how='inner', on=['tissues'])\n",
    "    df_return = pd.merge(df_a, df_dynamic, how='inner', on=['tissues'])\n",
    "    return df_return\n",
    "\n",
    "# 3、统计分析各类loop所影响的同源基因bias表达数目\n",
    "def BiasG_stat(df_loop_Bias, df_bias_pairs):\n",
    "    #df_loop_Bias['number'] = [1]*df_loop_Bias.shape[0]\n",
    "    df_loop_Bias, df_bias_pairs = df_loop_Bias.copy(), df_bias_pairs[['GeneId_At', 'GeneId_Dt', 'tissues']].copy()\n",
    "    ## 与loop相关的bias表达基因\n",
    "    print(df_loop_Bias.shape[0])\n",
    "    df_a = df_loop_Bias[['tissues', 'GeneId1', 'tags']].copy()\n",
    "    df_b = df_loop_Bias[['tissues', 'GeneId2', 'tags']].copy()\n",
    "    df_a.columns = ['tissues', 'GeneId', 'tags']\n",
    "    df_b.columns = ['tissues', 'GeneId', 'tags']\n",
    "    df_ab = pd.concat([df_a, df_b], axis=0)\n",
    "    df_ab.drop_duplicates(keep='first', inplace=True)\n",
    "    # df_hypocotyl = df_ab.query(\"tissues=='hypocotyl' & tags=='NHloops'\").copy()\n",
    "    # print(df_hypocotyl.shape[0])\n",
    "    ## bias 表达同源基因对\n",
    "    df_loop_bias = df_ab.dropna()\n",
    "    df_hypocotyl = df_loop_bias.query(\"tissues=='hypocotyl' & tags=='NHloops'\").copy()\n",
    "    print(df_hypocotyl.shape[0])\n",
    "    \n",
    "    df_loop_bias.columns = ['tissues', 'GeneId_At', 'tags_At']\n",
    "    df_a = pd.merge(df_bias_pairs, df_loop_bias, how='left', on=['GeneId_At', 'tissues'])\n",
    "    df_loop_bias.columns = ['tissues', 'GeneId_Dt', 'tags_Dt']\n",
    "    df_ab = pd.merge(df_a, df_loop_bias, how='left', on=['GeneId_Dt', 'tissues'])\n",
    "    df_ab.dropna(subset=['tags_At', 'tags_Dt'], how='all', inplace=True)\n",
    "    #print(df_ab.head())\n",
    "    df_hypocotyl = df_ab.query(\"(tags_At=='NHloops' | tags_Dt=='NHloops') & tissues=='hypocotyl'\").copy()\n",
    "    print(df_hypocotyl.shape[0])\n",
    "    df_hypocotyl.drop_duplicates(keep='first', subset=['GeneId_At', 'GeneId_Dt', 'tissues'], inplace=True)\n",
    "    print(df_hypocotyl.shape[0])\n",
    "    ## 统计各类型loop 所影响的同源基因对数目\n",
    "    df_result = stat_number(df_ab)\n",
    "    return df_result\n",
    "df_Bias = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/all_tissue_bias_gene_long.bed\", sep=\"\\t\")\n",
    "df_Bias_pair = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/all_tissue_bias_gene_pair.bed\", sep=\"\\t\")\n",
    "df_GBin = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/3kb_bin_gene.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'GeneId'])\n",
    "df_loop = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/loose/J668_subg_loop_compare2.bed\", sep=\"\\t\")\n",
    "df_BiasG_bin = BiasG_bin(df_Bias, df_GBin)\n",
    "print(df_BiasG_bin.shape[0], df_loop.shape[0])\n",
    "df_loop_Bias = loop_bias(df_loop, df_BiasG_bin)\n",
    "\n",
    "df_LBias_stat = BiasG_stat(df_loop_Bias, df_Bias_pair)\n",
    "#print(df_LBias_stat)\n",
    "df_LBias_stat.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/test/hg/J668_loop_type_Bias.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 329,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "427\n"
     ]
    }
   ],
   "source": [
    "## 找到hypocotyl 锚点的差异\n",
    "import pandas as pd\n",
    "df_hypocotyl_NHanchor = df_loop_Bias.query(\"tags=='NHanchor' & tissues=='hypocotyl'\").copy()\n",
    "# ## 加上cis的标签，选择有cis差异tags\n",
    "df_cis = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/hypocotyl_Cis.bed\", sep=\"\\t\")\n",
    "df_HGPair_position = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/all_homo_gene_pairs_position.txt\", sep=\"\\t\")\n",
    "df_cis['cis'] =['Cis']*df_cis.shape[0]\n",
    "df_cis.columns = ['chrom1', 'start1', 'end1', 'cis1']\n",
    "df_a = pd.merge(df_hypocotyl_NHanchor, df_cis, how='left', on=['chrom1', 'start1', 'end1'])\n",
    "df_cis.columns = ['chrom2', 'start2', 'end2', 'cis2']\n",
    "df_ab = pd.merge(df_a, df_cis, how='left', on=['chrom2', 'start2', 'end2'])\n",
    "df_NHanchor_cis = df_ab.query(\"cis1=='Cis' | cis2=='Cis'\")\n",
    "## 得到与NHanchor loop 相关的同源bias 表达基因\n",
    "df_a = df_NHanchor_cis[['GeneId1']].copy()\n",
    "df_b = df_NHanchor_cis[['GeneId2']].copy()\n",
    "df_a.columns = ['GeneId']\n",
    "df_b.columns = ['GeneId']\n",
    "df_ab = pd.concat([df_a, df_b], axis=0)\n",
    "df_geneid = df_ab.dropna()\n",
    "print(df_geneid.shape[0])\n",
    "## 找到bias基因所对应的同源基因\n",
    "df_geneid.columns = ['GeneId_At']\n",
    "df_a = pd.merge(df_HGPair_position, df_geneid, how='inner', on=['GeneId_At'])\n",
    "df_geneid.columns = ['GeneId_Dt']\n",
    "df_b = pd.merge(df_HGPair_position, df_geneid, how='inner', on=['GeneId_Dt'])\n",
    "df_ab = pd.concat([df_a, df_b], axis=0)\n",
    "df_ab.drop_duplicates(keep='first', inplace=True)\n",
    "## 得到基因的位置以及标签\n",
    "df_ab['number'] = range(df_ab.shape[0])\n",
    "df_at = df_ab[['chr_at', 'start_at', 'end_at', 'GeneId_At', 'number']].copy()\n",
    "df_dt = df_ab[['chr_dt', 'start_dt', 'end_dt', 'GeneId_Dt', 'number']].copy()\n",
    "df_at.columns = ['chrom', 'start', 'end', 'GeneId', 'number']\n",
    "df_dt.columns = ['chrom', 'start', 'end', 'GeneId', 'number']\n",
    "df_all = pd.concat([df_at, df_dt], axis=0)\n",
    "df_all.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/loose/NHanchor/hypocotyl_NHanchor_Hgene_position.bed\", sep=\"\\t\", index=False, header=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2994\n",
      "2670\n"
     ]
    }
   ],
   "source": [
    "# ## 确定hypocotyl 中NHloops与 同源基因Bias 表达有关\n",
    "# import pandas as pd\n",
    "df_loop_Bias = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/loose/J668_loop_Bias.bed\", sep=\"\\t\")\n",
    "df_HG_pairs = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/all_homo_gene_pairs_position.txt\", sep=\"\\t\")\n",
    "def HG_pairs(df_loop_Bias, df_HG):\n",
    "    df_hypocotyl = df_loop_Bias.query(\"tissues=='hypocotyl' & tags=='NHloops'\").copy()\n",
    "    df_a = df_hypocotyl[['GeneId1']].copy()\n",
    "    df_b = df_hypocotyl[['GeneId2']].copy()\n",
    "    df_a.columns = ['GeneId']\n",
    "    df_b.columns = ['GeneId']\n",
    "    df_ab = pd.concat([df_a, df_b], axis=0)\n",
    "    df_ab.drop_duplicates(keep='first', inplace=True)\n",
    "    print(df_ab.shape[0])\n",
    "    \n",
    "    df_ab.columns = ['GeneId_At']\n",
    "    df_a = pd.merge(df_HG, df_ab, how='inner', on=['GeneId_At'])\n",
    "    df_ab.columns = ['GeneId_Dt']\n",
    "    df_b = pd.merge(df_HG, df_ab, how='inner', on=['GeneId_Dt'])\n",
    "    df_ab = pd.concat([df_a, df_b], axis=0)\n",
    "    df_ab.drop_duplicates(keep='first', inplace=True)\n",
    "    print(df_ab.shape[0])\n",
    "    df_ab['number'] = range(df_ab.shape[0])\n",
    "    df_a = df_ab[['chr_at', 'start_at', 'end_at', 'GeneId_At', 'number']].copy()\n",
    "    df_b = df_ab[['chr_dt', 'start_dt', 'end_dt', 'GeneId_Dt', 'number']].copy()\n",
    "    df_a.columns = ['chrom', 'start', 'end', 'GeneId', 'number']\n",
    "    df_b.columns = ['chrom', 'start', 'end', 'GeneId', 'number']\n",
    "    df_ab = pd.concat([df_a, df_b], axis=0)\n",
    "    return df_ab\n",
    "df_hypocotyl_NHloops_Bias = HG_pairs(df_loop_Bias, df_HG_pairs)\n",
    "df_hypocotyl_NHloops_Bias.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/loose/NHloops/hypocotyl_NHanchor_Hgene_position.bed\", sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2085\n",
      "1495\n"
     ]
    }
   ],
   "source": [
    "## 分析Dynamic 所能影响的同源bias 表达基因\n",
    "import pandas as pd\n",
    "df_loop_Bias = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/loose/J668_loop_Bias.bed\", sep=\"\\t\")\n",
    "df_HG_pairs = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/all_homo_gene_pairs_position.txt\", sep=\"\\t\")\n",
    "def HG_pairs(df_loop_Bias, df_HG):\n",
    "    df_hypocotyl = df_loop_Bias.query(\"tissues=='hypocotyl' & tags=='Dynamic'\").copy()\n",
    "    df_a = df_hypocotyl[['GeneId1']].copy()\n",
    "    df_b = df_hypocotyl[['GeneId2']].copy()\n",
    "    df_a.columns = ['GeneId']\n",
    "    df_b.columns = ['GeneId']\n",
    "    df_ab = pd.concat([df_a, df_b], axis=0)\n",
    "    df_ab.drop_duplicates(keep='first', inplace=True)\n",
    "    print(df_ab.shape[0])\n",
    "    df_ab.columns = ['GeneId_At']\n",
    "    df_a = pd.merge(df_HG, df_ab, how='inner', on=['GeneId_At'])\n",
    "    df_ab.columns = ['GeneId_Dt']\n",
    "    df_b = pd.merge(df_HG, df_ab, how='inner', on=['GeneId_Dt'])\n",
    "    df_ab = pd.concat([df_a, df_b], axis=0)\n",
    "    df_ab.drop_duplicates(keep='first', inplace=True)\n",
    "    print(df_ab.shape[0])\n",
    "    df_ab['number'] = range(df_ab.shape[0])\n",
    "    df_a = df_ab[['chr_at', 'start_at', 'end_at', 'GeneId_At', 'number']].copy()\n",
    "    df_b = df_ab[['chr_dt', 'start_dt', 'end_dt', 'GeneId_Dt', 'number']].copy()\n",
    "    df_a.columns = ['chrom', 'start', 'end', 'GeneId', 'number']\n",
    "    df_b.columns = ['chrom', 'start', 'end', 'GeneId', 'number']\n",
    "    df_ab = pd.concat([df_a, df_b], axis=0)\n",
    "    return df_ab\n",
    "df_hypocotyl_NHloops_Bias = HG_pairs(df_loop_Bias, df_HG_pairs)\n",
    "df_hypocotyl_NHloops_Bias.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/loose/Dynamic/hypocotyl_Dynamic_Hgene_position.bed\", sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_loop_Bias = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/loose/J668_loop_Bias.bed\", sep=\"\\t\")\n",
    "df_hypocotyl_dynamic = df_loop_Bias.query(\"tags=='Dynamic' & tissues=='hypocotyl'\").copy()\n",
    "df_hypocotyl_dynamic.dropna(subset=['GeneId1', 'GeneId2'], how=\"any\", inplace=True)\n",
    "\n",
    "## 找出现两次的基因其所对应的loops\n",
    "df_GeneId1 = df_hypocotyl_dynamic[['GeneId1']].copy()\n",
    "df_GeneId2 = df_hypocotyl_dynamic[['GeneId2']].copy()\n",
    "df_GeneId1.columns = ['GeneId']\n",
    "df_GeneId2.columns = ['GeneId']\n",
    "df_GeneId = pd.concat([df_GeneId1, df_GeneId2], axis=0)\n",
    "df_GeneId['number'] = [1]*df_GeneId.shape[0]\n",
    "df_GeneId_number = df_GeneId.groupby(by=['GeneId'], as_index=False)['number'].sum()\n",
    "df_ab = df_GeneId_number.query(\"number>1\")[['GeneId']].copy()\n",
    "\n",
    "\n",
    "## 找同源基因以及其位置\n",
    "df_HG_pairs = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/all_homo_gene_pairs_position.txt\", sep=\"\\t\")\n",
    "df_ab.columns = ['GeneId_At']\n",
    "df_a = pd.merge(df_HG_pairs, df_ab, how='inner', on=['GeneId_At'])\n",
    "df_ab.columns = ['GeneId_Dt']\n",
    "df_b = pd.merge(df_HG_pairs, df_ab, how='inner', on=['GeneId_Dt'])\n",
    "df_all = pd.concat([df_a, df_b], axis=0)\n",
    "df_all.drop_duplicates(keep='first', inplace=True)\n",
    "df_all['number'] = range(df_all.shape[0])\n",
    "\n",
    "df_at = df_all[['chr_at', 'start_at', 'end_at', 'GeneId_At', 'number']].copy()\n",
    "df_dt = df_all[['chr_dt', 'start_dt', 'end_dt', 'GeneId_Dt', 'number']].copy()\n",
    "df_at.columns = ['chrom', 'start', 'end', 'GeneId', 'number']\n",
    "df_dt.columns = ['chrom', 'start', 'end', 'GeneId', 'number']\n",
    "df = pd.concat([df_at, df_dt], axis=0)\n",
    "df.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/loose/Dynamic/hypocotyl_hbias_anchor.bed\", sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           chrom     start       end           GeneId\n",
      "616501  Ghir_A01   2640000   2643000  Ghir_A01G002890\n",
      "616635  Ghir_A01   9534000   9537000  Ghir_A01G006270\n",
      "616656  Ghir_A01  10170000  10173000  Ghir_A01G006360\n",
      "616660  Ghir_A01  10353000  10356000  Ghir_A01G006440\n",
      "616671  Ghir_A01  10872000  10875000  Ghir_A01G006680\n",
      "...          ...       ...       ...              ...\n",
      "756004  Ghir_D03   8178000   8181000  Ghir_D03G005180\n",
      "756037  Ghir_A05  28830000  28833000  Ghir_A05G027060\n",
      "756044  Ghir_D10  62301000  62304000  Ghir_D10G023600\n",
      "756046  Ghir_D03  50325000  50328000  Ghir_D03G017330\n",
      "756062  Ghir_D12  37581000  37584000  Ghir_D12G010360\n",
      "\n",
      "[3285 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "## 找到hypocotyl 锚点的差异\n",
    "import pandas as pd\n",
    "df_hypocotyl_NHanchor = df_loop_Bias.query(\"tags=='NHanchor' & tissues=='hypocotyl'\").copy()\n",
    "#print(df_hypocotyl_NHanchor)\n",
    "df_a = df_hypocotyl_NHanchor[['chrom1', 'start1', 'end1', 'GeneId1']].copy()\n",
    "df_b = df_hypocotyl_NHanchor[['chrom2', 'start2', 'end2', 'GeneId2']].copy()\n",
    "df_a.columns = ['chrom', 'start', 'end', 'GeneId']\n",
    "df_b.columns = ['chrom', 'start', 'end', 'GeneId']\n",
    "df_ab = pd.concat([df_a, df_b], axis=0)\n",
    "df_need = df_ab.query(\"GeneId != 'Noelement'\").copy()\n",
    "df_need.drop_duplicates(keep='first', inplace=True)\n",
    "print(df_need)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      chrom1    start1      end1    chrom2    start2      end2  \\\n",
      "46  Ghir_A03  15750000  15753000  Ghir_A03  16188000  16191000   \n",
      "48  Ghir_A03  15750000  15753000  Ghir_A03  16188000  16191000   \n",
      "\n",
      "                       sorted_pair_tags  number  \n",
      "46  AtG17302-AtG17307-DtG17302-DtG17307       1  \n",
      "48  AtG17302-AtG17308-DtG17302-DtG17308       1  \n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/strict/anther_conserved.bed\", sep=\"\\t\")\n",
    "df_need = df.query(\"chrom1=='Ghir_A03' & start1==15750000 & end1 == 15753000 & chrom2 == 'Ghir_A03' & start2 == 16188000 & end2 == 16191000\")\n",
    "print(df_need)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     chrom1  start1    end1    chrom2  start2    end2  tags1  tags2  \\\n",
      "0  Ghir_A01  267000  270000  Ghir_A01  327000  330000   Hbin   Hbin   \n",
      "1  Ghir_A01  369000  372000  Ghir_A01  459000  462000  NHbin   Hbin   \n",
      "2  Ghir_A01  399000  402000  Ghir_A01  435000  438000   Hbin   Hbin   \n",
      "3  Ghir_A01  450000  453000  Ghir_A01  498000  501000  NHbin  NHbin   \n",
      "4  Ghir_A01  642000  645000  Ghir_A01  741000  744000  NHbin  NHbin   \n",
      "\n",
      "   number1_pair  number2_pair      tags  \n",
      "0           0.0           3.0   Dynamic  \n",
      "1           0.0           4.0  NHanchor  \n",
      "2           0.0           0.0   NHloops  \n",
      "3           0.0           0.0  NHanchor  \n",
      "4           0.0           0.0  NHanchor  \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df_loop = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/subg_loop_compare/loose/hypocotyl_subg_loop_compare.bed\", sep=\"\\t\")\n",
    "df_loop_type = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type.bed\", sep=\"\\t\")\n",
    "df_hypocotyl_type = df_loop_type.query(\"tissues=='hypocotyl'\")[['chrom1', 'start1', 'end1', 'tags1', 'chrom2', 'start2', 'end2', 'tags2']].copy()\n",
    "print(df_loop.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   A  B  C\n",
      "0  1  1  1\n",
      "1  1  1  1\n",
      "2  1  1  1\n",
      "3  3  6  3\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   A  B  C\n",
       "1  1  1  1\n",
       "2  1  1  1"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.DataFrame({'A': [1, 1, 1, 3], 'B': [1, 1,1,  6], 'C': [1, 1, 1, 3]})\n",
    "print(df)\n",
    "df[df.duplicated()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11832\n"
     ]
    }
   ],
   "source": [
    "## 合并At_vs_Dt 和Dt_vs_At\n",
    "import pandas as pd\n",
    "df_a = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/blastx/Ghir_HCis_Dt_vs_At.bed\", sep=\"\\t\")\n",
    "df_b = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/blastx/Ghir_HCis_Dt_vs_At.bed\", sep=\"\\t\")\n",
    "df_a.columns = ['chrom2', 'start2', 'end2', 'cis2_tags', 'chrom1', 'start1', 'end1', 'cis1_tags']\n",
    "df = pd.concat([df_a, df_b], axis=0)\n",
    "df.drop_duplicates(keep='first', inplace=True)\n",
    "print(df.shape[0])\n",
    "df.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/blastx/J668_Hcis_blastx.bed\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析liftover 所鉴定出的同源cis\n",
    "import pandas as pd\n",
    "bn_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/liftover/At_Cis_Dt_liftover_25.bed\"\n",
    "cis_f1 = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_Cis_At.bed\"\n",
    "cis_f2 = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_Cis_Dt.bed\"\n",
    "df_bn = pd.read_csv(bn_f, sep=\"\\t\", names =['chrom', 'start', 'end', 'bn_tags'])\n",
    "df_cis1 = pd.read_csv(cis_f1, sep=\"\\t\", names=['chrom', 'start', 'end', 'cis1'])\n",
    "df_cis2 = pd.read_csv(cis_f2, sep=\"\\t\", names=['chrom', 'start', 'end', 'cis2'])\n",
    "\n",
    "df_result = pd.DataFrame()\n",
    "for index, row in df_bn.iterrows():\n",
    "    C, S, E, T = row['chrom'], row['start'], row['end'], row['bn_tags']\n",
    "    df_tmp = df_cis2.query(\"chrom==@C & end>=@S & start<=@E\").copy()\n",
    "    df_tmp['cis1'] = [T]*df_tmp.shape[0]\n",
    "    df_result = pd.concat([df_result, df_tmp], axis=0)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/liftover/At_vs_Dt_homo.txt\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析liftover 所鉴定出的同源cis\n",
    "import pandas as pd\n",
    "bn_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/liftover/Dt_Cis_At_liftover_25.bed\"\n",
    "cis_f2 = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_Cis_At.bed\"\n",
    "cis_f1 = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_Cis_Dt.bed\"\n",
    "df_bn = pd.read_csv(bn_f, sep=\"\\t\", names =['chrom', 'start', 'end', 'bn_tags'])\n",
    "df_cis1 = pd.read_csv(cis_f1, sep=\"\\t\", names=['chrom', 'start', 'end', 'cis1'])\n",
    "df_cis2 = pd.read_csv(cis_f2, sep=\"\\t\", names=['chrom', 'start', 'end', 'cis2'])\n",
    "df_result = pd.DataFrame()\n",
    "for index, row in df_bn.iterrows():\n",
    "    C, S, E, T = row['chrom'], row['start'], row['end'], row['bn_tags']\n",
    "    df_tmp = df_cis2.query(\"chrom==@C & end>=@S & start<=@E\").copy()\n",
    "    df_tmp['cis1'] = [T]*df_tmp.shape[0]\n",
    "    df_result = pd.concat([df_result, df_tmp], axis=0)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/liftover/Dt_vs_At_homo.txt\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "# bnmapper 结果合并\n",
    "df1 = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/liftover/At_vs_Dt_homo.txt\", sep=\"\\t\")\n",
    "df2 = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/liftover/Dt_vs_At_homo.txt\", sep=\"\\t\")\n",
    "df1.columns = ['chrom', 'start', 'end', 'cis1', 'cis2']\n",
    "df = pd.concat([df1, df2], axis=0)[['cis1', 'cis2']].copy()\n",
    "df.drop_duplicates(keep='first', subset=['cis1', 'cis2'], inplace=True)\n",
    "\n",
    "df_cis1 = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_Cis_Dt.bed\", sep=\"\\t\", names=['chrom1', 'start1', 'end1','cis1'])\n",
    "df_cis2 = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_Cis_At.bed\", sep=\"\\t\", names=['chrom2', 'start2', 'end2','cis2'])\n",
    "\n",
    "df_a = pd.merge( df, df_cis1, how='left', on=['cis1'])\n",
    "df_ab = pd.merge(df_a, df_cis2, how='left', on=['cis2'])\n",
    "df_ab.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/liftover/J668_cis_hg_liftover.bed\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将liftover， bnmapper 和blastx结果合并\n",
    "df_blastx = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/blastx/J668_Hcis_blastx.bed\", sep=\"\\t\")\n",
    "df_bnmapper = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/bnmapper/J668_cis_hg_bnmapper.bed\", sep=\"\\t\")\n",
    "df_liftover = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/liftover/J668_cis_hg_liftover.bed\", sep=\"\\t\")\n",
    "df_all = pd.concat([df_blastx, df_bnmapper, df_liftover], axis=0)\n",
    "df_all.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/J668_Hcis-element.bed\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 对于hypocotyl 两个亚组bias基因做例子，看看是否具有什么规律\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/all_tissue_bias_gene_pair.bed\", sep=\"\\t\")\n",
    "df_hypocotyl = df.query(\"tissues=='hypocotyl'\").copy()\n",
    "df_a = df_hypocotyl[['GeneId_At']].copy()\n",
    "df_b = df_hypocotyl[['GeneId_Dt']].copy()\n",
    "df_a.columns = ['GeneId']\n",
    "df_b.columns = ['GeneId']\n",
    "df_a['tissues'] = range(df_a.shape[0])\n",
    "df_b['tissues'] = range(df_b.shape[0])\n",
    "df = pd.concat([df_a, df_b], axis=0)\n",
    "df_gene_position = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/Gh_gene_position.bed\", sep=\"\\t\")\n",
    "df_result = pd.merge(df, df_gene_position, how='inner', on=['GeneId'])\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/hypocotyl_bias_gene_position.bed\", sep=\"\\t\", columns=['chrs', 'start', 'end', 'GeneId', 'tissues'], index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "70048\n",
      "69782\n",
      "69782.0\n"
     ]
    }
   ],
   "source": [
    "### 给bias基因打上loop anchor 数目标签\n",
    "import pandas as pd\n",
    "df_bias = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/all_tissue_bias_gene_pair.bed\", sep=\"\\t\")\n",
    "\n",
    "df_anchor = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/step2/J668_gene_loop_anchor_number_final.bed\", sep=\"\\t\")\n",
    "df_position = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/all_homo_gene_pairs_position.txt\", sep=\"\\t\")\n",
    "df_bias_position = pd.merge(df_bias, df_position, how='inner', on=['GeneId_At', 'GeneId_Dt'])\n",
    "df_anchor = df_anchor[['GeneId', 'tissues', 'number']].copy()\n",
    "df_anchor.columns = ['GeneId_At', 'tissues', 'number_At']\n",
    "\n",
    "df_a = pd.merge(df_bias_position, df_anchor, how='left', on=['GeneId_At', 'tissues'])\n",
    "df_anchor.columns = ['GeneId_Dt', 'tissues', 'number_Dt']\n",
    "df_ab = pd.merge(df_a, df_anchor, how='left', on=['GeneId_Dt', 'tissues'])\n",
    "\n",
    "## 去除scaffold\n",
    "print(df_ab.shape[0])\n",
    "df_ab = df_ab.query(\"(~chr_at.str.contains('Scaff')) & (~chr_dt.str.contains('Scaff'))\", engine='python')\n",
    "print(df_ab.shape[0])\n",
    "\n",
    "\n",
    "## 将成对bias 转换为长的数据格式\n",
    "def pair_To_long(df):\n",
    "    df = df.copy()\n",
    "    df_At = df[['GeneId_At', 'TPM_At', 'sig', 'tissues', 'chr_at', 'start_at', 'end_at', 'number_At']].copy()\n",
    "    df_Dt = df[['GeneId_Dt', 'TPM_Dt', 'sig', 'tissues', 'chr_dt', 'start_dt', 'end_dt', 'number_Dt']].copy()\n",
    "    df_At.columns = ['GeneId', 'TPM', 'sig', 'tissues', 'chr', 'start', 'end', 'number']\n",
    "    df_Dt.columns = ['GeneId', 'TPM', 'sig', 'tissues', 'chr', 'start', 'end', 'number']\n",
    "    df_return = pd.concat([df_At, df_Dt], axis=0)\n",
    "    return df_return\n",
    "\n",
    "## 分析基因位于染色体什么位置\n",
    "def gene_chr_position(df):\n",
    "    Ghir_chr = ['Ghir_A01', 'Ghir_A02', 'Ghir_A03', 'Ghir_A04', 'Ghir_A05', 'Ghir_A06', 'Ghir_A07', 'Ghir_A08', 'Ghir_A09', 'Ghir_A10', 'Ghir_A11', 'Ghir_A12', 'Ghir_A13',\n",
    "                'Ghir_D01', 'Ghir_D02', 'Ghir_D03', 'Ghir_D04', 'Ghir_D05', 'Ghir_D06', 'Ghir_D07', 'Ghir_D08', 'Ghir_D09', 'Ghir_D10', 'Ghir_D11', 'Ghir_D12', 'Ghir_D13']\n",
    "    Ghir_size = [117757855, 108092100, 113059412, 85149810, 109419610, 124056332, 97782242, 122376713, 82102766, 114851205, 123212122, 107672493, 108376418,\n",
    "                63209094, 69838496, 52700144, 56430500, 62933208, 66869448, 59257364, 69039221, 52819285, 68005838, 72942580, 62694377, 63343036]\n",
    "    df_result = pd.DataFrame()\n",
    "    for i in range(26):\n",
    "        chrs, size = Ghir_chr[i], Ghir_size[i]\n",
    "        df_tmp = df.query(\"chr==@chrs\").copy()\n",
    "        df_tmp['start_ratio'] = df_tmp['start'].apply(lambda x: x/size)\n",
    "        df_tmp['end_ratio'] = df_tmp['end'].apply(lambda x: (size-x)/size)\n",
    "        df_tmp['position_ratio'] = df_tmp[['start_ratio', 'end_ratio']].min(axis=1)\n",
    "        df_result = pd.concat([df_result, df_tmp], axis=0)\n",
    "    return df_result\n",
    "df = pair_To_long(df_ab)\n",
    "df_result = gene_chr_position(df)\n",
    "print(df_result.shape[0]/2)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_bias/gene_position/J668_bias_gene_position_long.bed\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将bias 基因转换为成对基因格式\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_bias/gene_position/J668_bias_gene_position_long.bed\", sep=\"\\t\")\n",
    "df_pair = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/J668_HG_pairs.csv\")\n",
    "\n",
    "def pair(df, df_pair):\n",
    "    df, df_pair = df.copy(), df_pair.copy()\n",
    "    df.drop_duplicates(keep='first', subset=['GeneId','tissues'], inplace=True)\n",
    "    #print(df.shape[0]/2)\n",
    "    Ghir_chr = ['Ghir_A01', 'Ghir_A02', 'Ghir_A03', 'Ghir_A04', 'Ghir_A05', 'Ghir_A06', 'Ghir_A07', 'Ghir_A08', 'Ghir_A09', 'Ghir_A10', 'Ghir_A11', 'Ghir_A12', 'Ghir_A13',\n",
    "                'Ghir_D01', 'Ghir_D02', 'Ghir_D03', 'Ghir_D04', 'Ghir_D05', 'Ghir_D06', 'Ghir_D07', 'Ghir_D08', 'Ghir_D09', 'Ghir_D10', 'Ghir_D11', 'Ghir_D12', 'Ghir_D13']\n",
    "    tissues = ['anther', 'cotyledon', 'fiber_10DPA', 'fiber_20DPA', 'ovule', 'hypocotyl', 'leaf', 'petal', 'radicle', 'root', 'stem', 'stigma']\n",
    "    df_return = pd.DataFrame()\n",
    "    number = 0 \n",
    "    for t in tissues:\n",
    "        df_a = df.query(\"chr.str.contains('Ghir_A') & tissues == @t\", engine='python')[['GeneId', 'TPM', 'sig', 'tissues', 'number', 'position_ratio']].copy()\n",
    "        df_d = df.query(\"chr.str.contains('Ghir_D') & tissues == @t\", engine='python')[['GeneId', 'TPM', 'sig', 'tissues', 'number', 'position_ratio']].copy()\n",
    "        #print(df_a.shape[0], df_d.shape[0])\n",
    "        number += df_a.shape[0]\n",
    "        df_a.columns = ['GeneId_At', 'TPM_At', 'sig', 'tissues', 'number_At', 'position_At']\n",
    "        df_d.columns = ['GeneId_Dt', 'TPM_Dt', 'sig', 'tissues', 'number_Dt', 'position_Dt']\n",
    "        df_At = pd.merge(df_pair, df_a, how='inner', on=['GeneId_At'])\n",
    "        df_AtDt = pd.merge(df_At, df_d, how='inner', on=['GeneId_Dt', 'sig', 'tissues'])\n",
    "        #print(df_At.shape[0], df_AtDt.shape[0])\n",
    "        df_return = pd.concat([df_return , df_AtDt], axis=0)\n",
    "    return df_return\n",
    "\n",
    "df_result = pair(df, df_pair)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_bias/gene_position/J668_bias_gene_position_pair.bed\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          GeneId_At        GeneId_Dt     TPM_At    sig tissues  number_At  \\\n",
      "0   Ghir_A01G000190  Ghir_D01G000200   1.050841  BiasD  anther          0   \n",
      "1   Ghir_A01G000230  Ghir_D01G000240  69.419666  BiasA  anther          0   \n",
      "5   Ghir_A01G000330  Ghir_D01G000340   4.044175  BiasA  anther          0   \n",
      "8   Ghir_A01G000540  Ghir_D01G000520   4.033785  BiasA  anther          0   \n",
      "12  Ghir_A01G000850  Ghir_D01G000800   0.465625  BiasD  anther          0   \n",
      "\n",
      "    position_At     TPM_Dt  number_Dt  position_Dt    tags  \n",
      "0      0.001200  13.540072          0     0.001897  noloop  \n",
      "1      0.001404  11.728125          0     0.002329  noloop  \n",
      "5      0.001912   0.633772          0     0.003351  noloop  \n",
      "8      0.003567   0.042982          0     0.005783  noloop  \n",
      "12     0.005632   7.436080          0     0.009164  noloop  \n"
     ]
    }
   ],
   "source": [
    "## 将成对的同源bias 基因划分为由loop影响的基因对，不由loop影响的基因对\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_bias/gene_position/J668_bias_gene_position_pair.bed\", sep=\"\\t\")\n",
    "df_noloop = df.query(\"number_At == 0 & number_Dt == 0\").copy()\n",
    "df_loop = df.query(\"number_At != 0 | number_Dt != 0\").copy()\n",
    "df_noloop['tags'] = ['noloop']*df_noloop.shape[0]\n",
    "df_loop['tags'] = ['loop']*df_loop.shape[0]\n",
    "\n",
    "df_all = pd.concat([df_noloop, df_loop], axis=0)\n",
    "print(df_all.head())\n",
    "#df_anther = df_all.query(\"tissues=='anther'\")\n",
    "\n",
    "df_a = df_all[['position_At', 'tags', 'tissues']].copy()\n",
    "df_b = df_all[['position_Dt', 'tags', 'tissues']].copy()\n",
    "df_a.columns = ['position', 'tags', 'tissues']\n",
    "df_b.columns = ['position', 'tags', 'tissues']\n",
    "df = pd.concat([df_a, df_b], axis=0)\n",
    "df.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_bias/gene_position/J668_anther_bias_gene_chr_position.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12430 8086 25679 18171 64366\n"
     ]
    }
   ],
   "source": [
    "## 分析非loop 引起的bias, 多loop引起的bias 和同loop不同互作所引起的bias 表达\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_bias/gene_position/J668_bias_gene_position_pair.bed\", sep=\"\\t\")\n",
    "df_noloop = df.query(\"number_At==0 & number_Dt==0\").copy()\n",
    "df_sameloop = df.query(\"number_At==number_Dt & number_At!=0\").copy()\n",
    "df_zore_other = df.query(\"(number_At==0 & number_Dt!=0) | (number_Dt==0 & number_At!=0)\").copy()\n",
    "df_diffloop = pd.concat([df, df_noloop, df_sameloop, df_zore_other], axis=0)\n",
    "df_diffloop.drop_duplicates(keep=False, inplace=True)\n",
    "print(df_noloop.shape[0], df_sameloop.shape[0], df_zore_other.shape[0], df_diffloop.shape[0], df.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             GeneId_At    sig tissues        GeneId_Dt     TPM_At     TPM_Dt  \\\n",
      "6      Ghir_A09G014910  BiasA  anther  Ghir_D09G014170  77.521673   0.726142   \n",
      "7      Ghir_A01G018960  BiasD  anther  Ghir_D01G020400   1.449371  10.457469   \n",
      "8      Ghir_A01G018960  BiasD  anther  Ghir_D01G020400   1.449371  10.457469   \n",
      "12     Ghir_A12G005210  BiasD  anther  Ghir_D12G005160   0.309670  51.338024   \n",
      "23     Ghir_A08G015650  BiasD  anther  Ghir_D08G016460   4.106777  59.236539   \n",
      "...                ...    ...     ...              ...        ...        ...   \n",
      "70884  Ghir_A05G002550  BiasD  stigma  Ghir_D05G002690   0.214057   1.558473   \n",
      "70891  Ghir_A02G008650  BiasD  stigma  Ghir_D02G008960   0.237562   0.539087   \n",
      "70910  Ghir_A05G022790  BiasD  stigma  Ghir_D05G022710   0.241080   0.691860   \n",
      "70920  Ghir_A02G002580  BiasD  stigma  Ghir_D02G002980   0.720662   1.467550   \n",
      "70921  Ghir_A06G002600  BiasA  stigma  Ghir_D06G002460   0.852241   0.421026   \n",
      "\n",
      "      tags_At tags_Dt  \n",
      "6         G-N    Ncis  \n",
      "7         G-N    Ncis  \n",
      "8         G-N    Ncis  \n",
      "12        G-N    Ncis  \n",
      "23        G-N    Ncis  \n",
      "...       ...     ...  \n",
      "70884    Ncis     G-N  \n",
      "70891     G-N     G-N  \n",
      "70910    Ncis     G-N  \n",
      "70920    Ncis     G-N  \n",
      "70921     G-N    Ncis  \n",
      "\n",
      "[7703 rows x 8 columns]\n"
     ]
    }
   ],
   "source": [
    "## 分析非同源cis调控元件调控差异所引同源基因bias表达\n",
    "\n",
    "# 1、确定与非同源cis调控元件相关的loops\n",
    "import pandas as pd\n",
    "\n",
    "## 将成对的同源cis转换为单列同源cis\n",
    "def pairsTosingle(df):\n",
    "    df_a = df[['chrom1', 'start1', 'end1']].copy()\n",
    "    df_b = df[['chrom2', 'start2', 'end2']].copy()\n",
    "    df_a.columns = ['chrom', 'start', 'end']\n",
    "    df_b.columns = ['chrom', 'start', 'end']\n",
    "    df_return = pd.concat([df_a, df_b], axis=0)\n",
    "    return df_return\n",
    "\n",
    "\n",
    "## 过滤与同源cis相关的cis元件\n",
    "def data_filter(df_cis, df_hcis):\n",
    "    df_NHcis = pd.concat([df_cis, df_hcis], axis=0)\n",
    "    df_NHcis.drop_duplicates(keep=False, inplace=True)\n",
    "    return df_NHcis\n",
    "\n",
    "## 找与非同源cis相关的loops\n",
    "def NHcis_loop(df_loop, df_NHcis):\n",
    "    df_loop, df_NHcis = df_loop.copy(), df_NHcis.copy()\n",
    "    df_G_loop = df_loop.query(\"tags.str.contains('G')\", engine='python').copy()\n",
    "    df_NHcis.columns = ['chrom1', 'start1', 'end1']\n",
    "    df_a = pd.merge(df_G_loop, df_NHcis, how='inner', on=['chrom1', 'start1', 'end1'])\n",
    "    df_NHcis.columns = ['chrom2', 'start2', 'end2']\n",
    "    df_b = pd.merge(df_G_loop, df_NHcis, how='inner', on=['chrom2', 'start2', 'end2'])\n",
    "    df_return = pd.concat([df_a, df_b], axis=0)\n",
    "    df_return.drop_duplicates(keep='first', inplace=True)\n",
    "    return df_return\n",
    "\n",
    "## 找到与同源bias表达基因相关的loops anchor\n",
    "def biasG_anchor():\n",
    "    df_bias_G = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/all_tissue_bias_gene_long.bed\", sep=\"\\t\")\n",
    "    df_gene_anchor = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/3kb_bin_gene.bed\", sep=\"\\t\", names=['chrom1', 'start1', 'end1', 'GeneId'])\n",
    "    df_return = pd.merge(df_bias_G, df_gene_anchor, how='inner', on=['GeneId'])\n",
    "    return df_return\n",
    "\n",
    "## 调整G-N loop格式, 将基因放在前面\n",
    "def data_format(df):\n",
    "    df_a = df.query(\"tags1=='G' & tags2=='N'\")[['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tissues']].copy()\n",
    "    df_b = df.query(\"tags1=='N' & tags2=='G'\")[['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tissues']].copy()\n",
    "    df_b.columns = ['chrom2', 'start2', 'end2', 'chrom1', 'start1', 'end1', 'tissues']\n",
    "    df_return = pd.concat([df_a, df_b], axis=0)\n",
    "    return df_return\n",
    "\n",
    "## 分析哪些bias 基因受到非同源cis的调控\n",
    "def biasG_NHcis(df_gene, df_loop):\n",
    "    df_gene, df_loop = df_gene.copy(), df_loop.copy()\n",
    "    df_return = pd.merge(df_gene, df_loop, how='inner', on=['chrom1', 'start1', 'end1', 'tissues'])\n",
    "    return df_return \n",
    "\n",
    "## 将受到非同源cis调控元件所影响的基因切换为pairs形式\n",
    "def HG_pairs(df):\n",
    "    df = df[['GeneId', 'tissues', 'tags']].copy()\n",
    "    df.columns = ['GeneId_At', 'tissues', 'tags_At']\n",
    "    df_pairs = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/all_tissue_bias_gene_pair.bed\", sep=\"\\t\")\n",
    "    df_a = pd.merge(df_pairs, df, how='left', on=['GeneId_At', 'tissues'])\n",
    "    df.columns = ['GeneId_Dt', 'tissues', 'tags_Dt']\n",
    "    df_return = pd.merge(df_a, df, how='left', on=['GeneId_Dt', 'tissues'])\n",
    "    df_return.fillna('Ncis', inplace=True)\n",
    "    df_return = df_return.query(\"tags_At=='G-N' | tags_Dt=='G-N'\")\n",
    "    print(df_return)\n",
    "    return df_return\n",
    "df_cis = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/cis_gene_homo/J668_cis_hg_filter.bed\", sep=\"\\t\")\n",
    "df_hcis_pairs = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/homo_cis/J668_Hcis-element.bed\", sep=\"\\t\")\n",
    "df_loop = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type.bed\", sep=\"\\t\")\n",
    "df_hcis = pairsTosingle(df_hcis_pairs)\n",
    "df_NHcis = data_filter(df_cis, df_hcis)\n",
    "df_NHcis_loop = NHcis_loop(df_loop, df_NHcis)\n",
    "#print(df_NHcis_loop.head())\n",
    "df_biasG_anchor = biasG_anchor()\n",
    "#print(df_biasG_anchor.head())\n",
    "df_biasG_NHcis = biasG_NHcis(df_biasG_anchor, df_NHcis_loop)\n",
    "df_biasG_NHcis_pairs = HG_pairs(df_biasG_NHcis)\n",
    "df_biasG_NHcis_pairs.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_bias/nHcis_bias/J668_NHcis_biasG_pairs.bed\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5244\n"
     ]
    }
   ],
   "source": [
    "## 找到非同源cis元件所引起的差异表达基因位置\n",
    "import pandas as pd\n",
    "df_biasG = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_bias/nHcis_bias/J668_NHcis_biasG_pairs.bed\", sep=\"\\t\")\n",
    "df_Gposition = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/Gh_gene_position.bed\", sep=\"\\t\")\n",
    "def gene_position(df_biasG, df_Gposition):\n",
    "    df_biasG, df_Gposition = df_biasG.copy(), df_Gposition.copy()\n",
    "    df_a, df_d = df_biasG[['GeneId_At', 'tissues']].copy(), df_biasG[['GeneId_Dt', 'tissues']].copy()\n",
    "    df_a.columns = ['GeneId', 'tissues']\n",
    "    df_d.columns = ['GeneId', 'tissues']\n",
    "    df_a['number'] = range(df_a.shape[0])\n",
    "    df_d['number'] = range(df_d.shape[0])\n",
    "    df = pd.concat([df_a, df_d], axis=0)\n",
    "    df_return = pd.merge(df, df_Gposition, how='inner', on=['GeneId'])\n",
    "    return df_return \n",
    "df_genePosition = gene_position(df_biasG, df_Gposition)\n",
    "print(df_genePosition.shape[0])\n",
    "df_genePosition.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_bias/nHcis_bias/J668_NHcis_biasG_position.bed\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_bias/nHcis_bias/J668_NHcis_biasG_position.bed\", sep=\"\\t\")\n",
    "df_hypocotyl = df.query(\"tissues == 'hypocotyl'\").copy()\n",
    "df_hypocotyl.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_bias/nHcis_bias/J668_NHcis_biasG_hypocotyl.bed\", sep=\"\\t\", columns=['chrs', 'start', 'end', 'GeneId', 'number'], header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             GeneId_At        GeneId_Dt     TPM_At    sig    tissues  \\\n",
      "26248  Ghir_A01G000150  Ghir_D01G000160  22.753054  BiasA  hypocotyl   \n",
      "26252  Ghir_A01G000380  Ghir_D01G000380   0.430021  BiasD  hypocotyl   \n",
      "26253  Ghir_A01G000390  Ghir_D01G000390   1.807266  BiasD  hypocotyl   \n",
      "26254  Ghir_A01G000510  Ghir_D01G000500   0.714468  BiasD  hypocotyl   \n",
      "26255  Ghir_A01G000670  Ghir_D01G000640   1.193409  BiasD  hypocotyl   \n",
      "...                ...              ...        ...    ...        ...   \n",
      "31338  Ghir_A13G023720  Ghir_D13G024440  10.333641  BiasA  hypocotyl   \n",
      "31341  Ghir_A13G023760  Ghir_D13G024490  12.220609  BiasD  hypocotyl   \n",
      "31344  Ghir_A13G024020  Ghir_D13G024760  15.563376  BiasD  hypocotyl   \n",
      "31346  Ghir_A13G024330  Ghir_D13G025060  31.335834  BiasA  hypocotyl   \n",
      "31352  Ghir_A13G024610  Ghir_D13G025380   0.083009  BiasD  hypocotyl   \n",
      "\n",
      "       number_At  position_At     TPM_Dt  number_Dt  position_Dt  \n",
      "26248          1     0.000985   8.449550          0     0.001565  \n",
      "26252          0     0.002104   1.570866          1     0.003856  \n",
      "26253          1     0.002238   4.954495          0     0.003994  \n",
      "26254          1     0.003108   3.362148          0     0.005528  \n",
      "26255          0     0.004273   3.376060          1     0.007010  \n",
      "...          ...          ...        ...        ...          ...  \n",
      "31338          0     0.008698   3.036388          2     0.013381  \n",
      "31341          0     0.008500  26.968364          2     0.013009  \n",
      "31344          0     0.006696  34.291681          1     0.009931  \n",
      "31346          0     0.004562   8.207193          1     0.006737  \n",
      "31352          1     0.001813   0.815918          0     0.003142  \n",
      "\n",
      "[2052 rows x 10 columns]\n"
     ]
    }
   ],
   "source": [
    "# 分析loop 有无对于同源基因差异表达影响\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_bias/gene_position/J668_bias_gene_position_pair.bed\", sep=\"\\t\")\n",
    "df_zore_other = df.query(\"(number_At==0 & number_Dt!=0) | (number_Dt==0 & number_At!=0)\").copy()\n",
    "df_hypocotyl = df_zore_other.query(\"tissues=='hypocotyl'\").copy()\n",
    "def gene_position(df_biasG, df_Gposition):\n",
    "    df_biasG, df_Gposition = df_biasG.copy(), df_Gposition.copy()\n",
    "    df_a, df_d = df_biasG[['GeneId_At', 'tissues']].copy(), df_biasG[['GeneId_Dt', 'tissues']].copy()\n",
    "    df_a.columns = ['GeneId', 'tissues']\n",
    "    df_d.columns = ['GeneId', 'tissues']\n",
    "    df_a['number'] = range(df_a.shape[0])\n",
    "    df_d['number'] = range(df_d.shape[0])\n",
    "    df = pd.concat([df_a, df_d], axis=0)\n",
    "    df_return = pd.merge(df, df_Gposition, how='inner', on=['GeneId'])\n",
    "    return df_return \n",
    "print(df_hypocotyl)\n",
    "df_Gposition = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/Gh_gene_position.bed\", sep=\"\\t\")\n",
    "df_hypocotyl = gene_position(df_hypocotyl, df_Gposition)\n",
    "df_hypocotyl.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_bias/loop_yn_bias/J688_hypocotyl_loop_ys.bed\", sep=\"\\t\", columns=['chrs', 'start', 'end', 'GeneId', 'number'], header=False, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2764\n"
     ]
    }
   ],
   "source": [
    "# 分析loop 调控方式的差异对于基因表达的影响\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_bias/gene_position/J668_bias_gene_position_pair.bed\", sep=\"\\t\")\n",
    "df_diff = df.query(\"number_At>0 & number_Dt>0 & number_At != number_Dt\").copy()\n",
    "df_hypocotyl = df_diff.query(\"tissues=='hypocotyl'\").copy()\n",
    "def gene_position(df_biasG, df_Gposition):\n",
    "    df_biasG, df_Gposition = df_biasG.copy(), df_Gposition.copy()\n",
    "    df_a, df_d = df_biasG[['GeneId_At', 'tissues']].copy(), df_biasG[['GeneId_Dt', 'tissues']].copy()\n",
    "    df_a.columns = ['GeneId', 'tissues']\n",
    "    df_d.columns = ['GeneId', 'tissues']\n",
    "    df_a['number'] = range(df_a.shape[0])\n",
    "    df_d['number'] = range(df_d.shape[0])\n",
    "    df = pd.concat([df_a, df_d], axis=0)\n",
    "    df_return = pd.merge(df, df_Gposition, how='inner', on=['GeneId'])\n",
    "    return df_return \n",
    "df_hypocotyl = gene_position(df_hypocotyl, df_Gposition)\n",
    "df_hypocotyl.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_bias/loop_diff_bias/J668_hypocotyl_loop_diff.bed\", sep=\"\\t\", columns=['chrs', 'start', 'end', 'GeneId', 'number'], header=False, index=False)\n",
    "print(df_hypocotyl.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析同源非bias基因受到相同loop 调控情况\n",
    "import pandas as pd\n",
    "# 1、整理同源基因对G-G 互作loops\n",
    "def G_G_loop(df_loop, df_gene_anchor):\n",
    "    df_loop, df_gene_anchor = df_loop.copy(), df_gene_anchor.copy()\n",
    "    df_gene_anchor.columns = ['chrom1', 'start1', 'end1', 'GeneId_a']\n",
    "    df_a = pd.merge(df_loop, df_gene_anchor, how='inner', on=['chrom1', 'start1', 'end1'])\n",
    "    df_gene_anchor.columns = ['chrom2', 'start2', 'end2', 'GeneId_b']\n",
    "    df_return = pd.merge(df_a, df_gene_anchor, how='inner', on=['chrom2', 'start2', 'end2'])\n",
    "    #print(df_return)\n",
    "    return df_return\n",
    "# 2、确定loop anchor 位置仅有1\n",
    "def one_loop(df_HGG, df_anchor):\n",
    "    df_HGG, df_anchor = df_HGG, df_anchor\n",
    "    df_anchor = df_anchor[['GeneId', 'tissues', 'number']].copy()\n",
    "    df_anchor.columns = ['GeneId_a', 'tissues', 'number_a']\n",
    "    df_a = pd.merge(df_HGG, df_anchor, how='inner', on=['GeneId_a', 'tissues'])\n",
    "    df_anchor.columns = ['GeneId_b', 'tissues', 'number_b']\n",
    "    df_return = pd.merge(df_a, df_anchor, how='inner', on=['GeneId_b', 'tissues'])\n",
    "    df_return = df_return.query(\"number_a==1 & number_b==1\").copy()\n",
    "    return df_return \n",
    "# 3、找到no bias 表达基因\n",
    "def no_bias_gene(df_gene, df_bias_gene):\n",
    "    tissues = ['anther', 'cotyledon', 'fiber_10DPA', 'fiber_20DPA', 'ovule', 'hypocotyl', 'leaf', 'petal', 'radicle', 'root', 'stem', 'stigma']\n",
    "    df_all = pd.DataFrame()\n",
    "    for t in tissues:\n",
    "        df_tmp = df_gene[['GeneId']].copy()\n",
    "        df_tmp['tissues'] = [t]*df_tmp.shape[0]\n",
    "        df_all = pd.concat([df_all, df_tmp], axis=0)\n",
    "    df_return = pd.concat([df_all, df_bias_gene], axis=0)\n",
    "    #print(df_return.shape[0])\n",
    "    df_return.drop_duplicates(keep=False, inplace=True, subset=['GeneId', 'tissues'])\n",
    "    df_return.drop(['TPM', 'Bias'], inplace=True, axis=1)\n",
    "    return(df_return)\n",
    "# 3、确定loop 两端的基因均没有发生bias表达\n",
    "def no_bias(df_GG, df_no_bias):\n",
    "    df_GG, df_no_bias = df_GG.copy(), df_no_bias.copy()\n",
    "    df_no_bias.columns = ['GeneId_a', 'tissues']\n",
    "    df_a = pd.merge(df_GG, df_no_bias, how='inner', on=['GeneId_a', 'tissues'])\n",
    "    df_no_bias.columns = ['GeneId_b', 'tissues']\n",
    "    df_return = pd.merge(df_a, df_no_bias, how='inner', on=['GeneId_b', 'tissues'])\n",
    "    return(df_return)\n",
    "\n",
    "# 4.1 确定重复出现的同源基因对loops\n",
    "def HG_loop_pair(df):\n",
    "    df = df.copy()\n",
    "    df.reset_index(drop=True, inplace=True)\n",
    "    for index, row in df.iterrows():\n",
    "        a, b, c, d = row['GeneId_a'], row['GeneId_a_pair'], row['GeneId_b'], row['GeneId_b_pair']\n",
    "        tags = \"-\".join(sorted([a, b, c, d]))\n",
    "        #print(tags)\n",
    "        df.loc[index, 'tags'] = tags\n",
    "    df_no_tags_duplicated = df.drop_duplicates(keep=False, subset=['tissues', 'tags'])\n",
    "    df_return = pd.concat([df, df_no_tags_duplicated], axis=0)\n",
    "    df_return.drop_duplicates(keep=False, inplace=True)\n",
    "    #print(df_return)\n",
    "    return df_return\n",
    "# 4、找到对应的同源基因对和同源基因对互作的loops\n",
    "def HG_pairs(df_loop, df_HG_pair):\n",
    "    df_At = df_loop.query(\"chrom1.str.contains('Ghir_A')\", engine='python').copy()\n",
    "    df_Dt = df_loop.query(\"chrom2.str.contains('Ghir_D')\", engine='python').copy()\n",
    "    \n",
    "    ## At \n",
    "    df_HG_pair.columns = ['GeneId_a', 'GeneId_a_pair']\n",
    "    df_At_a = pd.merge(df_At, df_HG_pair, how='left', on=['GeneId_a'])\n",
    "    df_HG_pair.columns = ['GeneId_b', 'GeneId_b_pair']\n",
    "    df_At_ab = pd.merge(df_At_a, df_HG_pair, how='left', on=['GeneId_b'])\n",
    "    \n",
    "    ## Dt\n",
    "    df_HG_pair.columns = ['GeneId_a_pair', 'GeneId_a']\n",
    "    df_Dt_a = pd.merge(df_Dt, df_HG_pair, how='left', on=['GeneId_a'])\n",
    "    df_HG_pair.columns = ['GeneId_b_pair', 'GeneId_b']\n",
    "    df_Dt_ab = pd.merge(df_Dt_a, df_HG_pair, how='left', on=['GeneId_b'])\n",
    "    \n",
    "    df_At_Dt = pd.concat([df_At_ab, df_Dt_ab], axis=0)\n",
    "    df_return = HG_loop_pair(df_At_Dt)\n",
    "    df_return.sort_values(by=['tags'], inplace=True)\n",
    "    df_hypocotyl = df_return.query(\"tissues=='hypocotyl'\").copy()\n",
    "    df_hypocotyl.sort_values(by=['tags'], inplace=True)\n",
    "    print(df_hypocotyl)\n",
    "    return df_return\n",
    "\n",
    "## 数据格式整理\n",
    "def data_format(df):\n",
    "    df_a = df[['chrom1', 'start1', 'end1', 'GeneId_a', 'tissues']].copy()\n",
    "    df_b = df[['chrom2', 'start2', 'end2', 'GeneId_b', 'tissues']].copy()\n",
    "    df_a.columns = ['chrom', 'start', 'end', 'GeneId', 'tissues']\n",
    "    df_b.columns = ['chrom', 'start', 'end', 'GeneId', 'tissues']\n",
    "    df_a['tags'] = range(df_a.shape[0])\n",
    "    df_b['tags'] = range(df_b.shape[0])\n",
    "    df_return = pd.concat([df_a, df_b], axis=0)\n",
    "    return df_return \n",
    "\n",
    "df_loop = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/final_loops/J668_all_loop.csv\")\n",
    "df_gene_anchor = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/3kb_bin_gene.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'GeneId'])\n",
    "df_gene_anchor_number = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/anchor/step2/J668_gene_loop_anchor_number_final.bed\", sep=\"\\t\")\n",
    "df_GG = G_G_loop(df_loop, df_gene_anchor)\n",
    "df_GG_loopone = one_loop(df_GG, df_gene_anchor_number) # 得到loop anchor为1 的G-G互作loops\n",
    "\n",
    "## 得到组织非bias表达的同源基因\n",
    "df_hg = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/all_homo_gene_position.txt\", sep=\"\\t\")\n",
    "df_biasG = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/all_tissue_bias_gene_long.bed\", sep=\"\\t\")\n",
    "df_no_biasG = no_bias_gene(df_hg, df_biasG)\n",
    "df_GG_no_bias = no_bias(df_GG_loopone, df_no_biasG)\n",
    "\n",
    "## 得到两亚组保守的互作loops （anchor等于1的）\n",
    "df_pairs = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/Ghir_homo_gene_pairs.txt\", sep=\"\\t\")\n",
    "df_HG_loop_pairs = HG_pairs(df_GG_no_bias, df_pairs)\n",
    "\n",
    "## 数据格式整理\n",
    "df_data = data_format(df_HG_loop_pairs)\n",
    "df_hypocotyl = df_data.query(\"tissues == 'hypocotyl'\")\n",
    "df_hypocotyl.drop_duplicates(keep='first', subset=['tags'], inplace=True)\n",
    "df_hypocotyl.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_bias/loop_conserved_bias/J668_hypocotyl_HG_loop_conserved.bed\", sep=\"\\t\", index=False, header=False, columns=['chrom', 'start', 'end', 'GeneId', 'tags'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析G-G互作的基因是否存在不平衡的表达\n",
    "import pandas as pd\n",
    "# 1、将G-G锚点互作loops 转换为gene-gene 互作loops\n",
    "def G_G(df_loop, df_gene):\n",
    "    df_GG_loop = df_loop.query(\"tags=='G-G'\").copy()\n",
    "    df_gene.columns = ['chrom1', 'start1', 'end1', 'GeneId1']\n",
    "    df_a = pd.merge(df_GG_loop, df_gene, how='inner', on=['chrom1', 'start1', 'end1'])\n",
    "    df_gene.columns = ['chrom2', 'start2', 'end2', 'GeneId2']\n",
    "    df_return = pd.merge(df_a, df_gene, how='inner', on=['chrom2', 'start2', 'end2'])\n",
    "    return df_return\n",
    "# 2、对每个基因添加表达量\n",
    "def G_G_TPM(df_loop, df_TPM):\n",
    "    df_loop, df_TPM = df_loop.copy(), df_TPM.copy()\n",
    "    df_TPM.columns = ['GeneId1', 'tissues', 'TPM1']\n",
    "    df_a = pd.merge(df_loop, df_TPM, how='inner', on=['GeneId1', 'tissues'])\n",
    "    df_TPM.columns = ['GeneId2', 'tissues', 'TPM2']\n",
    "    df_return = pd.merge(df_a, df_TPM, how='inner', on=['GeneId2', 'tissues'])\n",
    "    return df_return\n",
    "# 3、筛选基因，loop两端基因表达差异\n",
    "def loop_nobalance_gene(df_loop):\n",
    "    df_loop = df_loop.copy()\n",
    "    df_loop['ratio'] = (df_loop['TPM1']+0.001)/(df_loop['TPM2']+0.001)\n",
    "    df_loop['sigs'] = df_loop['ratio'].apply(lambda x: 'Blance' if 0.5<x<2 else 'Nblance')\n",
    "    df_loop['number'] = [1]*df_loop.shape[0]\n",
    "    df_loop2 = df_loop.copy()\n",
    "    df_loop2.drop_duplicates(keep='first', subset=['GeneId1', 'GeneId2', 'tissues'], inplace=True)\n",
    "    df_return = df_loop2.groupby(by=['tissues', 'sigs'], as_index=False)['number'].sum()\n",
    "    return df_loop, df_return\n",
    "\n",
    "df_loop = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type.bed\", sep=\"\\t\")\n",
    "df_loop = df_loop[['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tissues', 'tags']].copy()\n",
    "df_gene_anchor = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/3kb_bin_gene.bed\", sep=\"\\t\", names=['chrom1', 'start1', 'end1', 'GeneId'])\n",
    "df_gene_TPM = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data/J668_all_Tissue_mean_TPM_long.csv\")\n",
    "\n",
    "df_GG = G_G(df_loop, df_gene_anchor)\n",
    "df_GG_TPM = G_G_TPM(df_GG, df_gene_TPM)\n",
    "df_no_balance_GG, df_no_balance_GG_sta = loop_nobalance_gene(df_GG_TPM)\n",
    "df_no_balance_GG.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_GG/J688_GG_loop.csv\", index=False)\n",
    "df_no_balance_GG_sta.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_GG/statistics/J668_GG_loop_statistics.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析cis调控bias元件是否bias 哪个亚基因组则受到的调控越多\n",
    "import pandas as pd\n",
    "# 1、锚点前后校正\n",
    "# 将确定真正的cis和基因锚点\n",
    "df_anchor_tags = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type.bed\", sep=\"\\t\")\n",
    "df_anchor_tags = df_anchor_tags[['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tags1', 'tags2', 'tissues']].copy()\n",
    "df_loop_type = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type_all2.csv\")\n",
    "## 锚点和loop类型合并\n",
    "df_tmp = pd.merge(df_anchor_tags, df_loop_type, how='inner', on=['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tissues'])\n",
    "## 由于之前将N-G和Cis-G 这类loop的标签设置为了G-N， G-Cis, 因此调整anthor1 和anchor2的位置使最终结合实际与G-N,G-Cis相一致\n",
    "df_tmp_NG = df_tmp.query(\"tags1=='N' & tags2=='G'\")\n",
    "df_tmp_NG.columns = ['chrom2', 'start2', 'end2', 'chrom1', 'start1', 'end1', 'tags2', 'tags1', 'tissues', 'length', 'subg', 'tags', 'loop_type']\n",
    "df_tmp_not_NG = df_tmp.query(\"~(tags1=='N' & tags2=='G')\")\n",
    "df_result = pd.concat([df_tmp_not_NG, df_tmp_NG], axis=0)\n",
    "df_result.to_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type_all3.csv', index=False)\n",
    "## 找到G-Cis 以及G-N-Cis\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     chrom1   start1     end1    chrom2   start2     end2  length subg  \\\n",
      "0  Ghir_A01   546000   549000  Ghir_A01   576000   579000   30000   At   \n",
      "1  Ghir_A01  1059000  1062000  Ghir_A01  1092000  1095000   33000   At   \n",
      "2  Ghir_A01  1248000  1251000  Ghir_A01  1563000  1566000  315000   At   \n",
      "3  Ghir_A01  1248000  1251000  Ghir_A01  1563000  1566000  315000   At   \n",
      "4  Ghir_A01  1323000  1326000  Ghir_A01  1341000  1344000   18000   At   \n",
      "\n",
      "  tissues   tags loop_type          GeneId1          GeneId2        Cis1  \\\n",
      "0  anther  short       G-N       NotElement  Ghir_A01G000730  NotElement   \n",
      "1  anther  short       G-N       NotElement  Ghir_A01G001330  NotElement   \n",
      "2  anther   long       G-N  Ghir_A01G001400       NotElement  NotElement   \n",
      "3  anther   long       G-N  Ghir_A01G001410       NotElement  NotElement   \n",
      "4  anther  short       G-N       NotElement  Ghir_A01G001540  NotElement   \n",
      "\n",
      "         Cis2  \n",
      "0  NotElement  \n",
      "1  NotElement  \n",
      "2  NotElement  \n",
      "3  NotElement  \n",
      "4  NotElement  \n",
      "            GeneId        TPM   Bias tissues\n",
      "0  Ghir_A01G000190   1.050841  BiasD  anther\n",
      "1  Ghir_A01G000230  69.419666  BiasA  anther\n",
      "2  Ghir_A01G000240   1.207827  BiasA  anther\n",
      "3  Ghir_A01G000250   0.768050  BiasA  anther\n",
      "4  Ghir_A01G000290   4.855359  BiasA  anther\n"
     ]
    }
   ],
   "source": [
    "## 分析bias基因有多少是受到G-Cis和G-G调控\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/final_loops/J668_all_loop_gene_cis.csv\")\n",
    "df_HBG = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/all_tissue_bias_gene_long.bed\", sep=\"\\t\")\n",
    "df_HBG = df_HBG[['GeneId', 'Bias', 'tissues']].copy()\n",
    "df_HBG.columns = ['Bias']\n",
    "\n",
    "df_a = pd.merge(df, df_HBG, how='inner', on=[''])\n",
    "print(df_HBG.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析确定有多少基因受到loop数目正调控影响\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/J668_Gene_anchor.bed\", sep=\"\\t\", names =['GeneId', 'number', 'tissues'])\n",
    "df_HBG = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/all_tissue_bias_gene_long.bed\", sep=\"\\t\")\n",
    "df_all = pd.merge(df, df_HBG, how='inner', on=['GeneId', 'tissues'])[['GeneId', 'tissues', 'number', 'Bias']].copy() # bias 基因loop调控差异\n",
    "df_HG_pair = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/J668_HG_pairs.csv\")\n",
    "## 确定受到loops 正调控的同源bias 基因\n",
    "def loop_positive_gene(df_bias_loop, df_HG_pair):\n",
    "    df_bias_loop, df_HG_pair = df_bias_loop.copy(), df_HG_pair.copy()\n",
    "    df_bias_loop.columns = ['GeneId_At', 'tissues', 'number_At', 'Bias']\n",
    "    df_a = pd.merge(df_bias_loop, df_HG_pair, how='right', on=['GeneId_At'])\n",
    "    df_bias_loop.columns = ['GeneId_Dt', 'tissues', 'number_Dt', 'Bias']\n",
    "    df_ab = pd.merge(df_a, df_bias_loop, how='left', on=['GeneId_Dt','tissues'])\n",
    "    df_ab['Bias_x'].fillna(df_ab['Bias_y'], inplace=True)\n",
    "    df_ab.drop('Bias_y', inplace=True, axis=1)\n",
    "    df_ab.rename(columns={'Bias_x':'Bias'}, inplace=True)\n",
    "    df_ab.fillna(0, inplace=True)\n",
    "    df_return = df_ab.query(\"(Bias=='BiasA' & number_At>number_Dt) | (Bias=='BiasD' & number_Dt>number_At)\").copy()\n",
    "    return df_return\n",
    "df_bias = loop_positive_gene(df_all, df_HG_pair)\n",
    "df_bias.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/J668_loop_positive_hg_bias_pair.csv\", index=False)\n",
    "# 统计受loop正调控的基因数目\n",
    "def sta_number(df):\n",
    "    df = df.copy()\n",
    "    df['number'] = [1]*df.shape[0]\n",
    "    df_result = df.groupby(by=['tissues'], as_index=False)['number'].sum()\n",
    "    return df_result\n",
    "df_bias_sta = sta_number(df_bias)\n",
    "df_bias_sta.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/J668_loop_positive_hg_bias_sta.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty DataFrame\n",
      "Columns: [GeneId2, tissues, chrom2, start2, end2, chrom1, start1, end1, tags, loop_type]\n",
      "Index: []\n",
      "39387 9130 144259 29065\n"
     ]
    }
   ],
   "source": [
    "## 分析这其中有多少基因比例是与G-G, G-Cis，G-N-G， G-N-Cis 有关\n",
    "import pandas as pd\n",
    "df_loop_bias = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/J668_loop_positive_hg_bias_pair.csv\")\n",
    "df_gene_bin = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/3kb_bin_gene.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'GeneId'])\n",
    "df_loop = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type_cis.csv\")\n",
    "# 得到bias基因id\n",
    "def get_bias_gene(df):\n",
    "    df_geneAt = df[['GeneId_At', 'tissues']].copy()\n",
    "    df_geneDt = df[['GeneId_Dt', 'tissues']].copy()\n",
    "    df_geneAt.columns = ['GeneId', 'tissues']\n",
    "    df_geneDt.columns = ['GeneId', 'tissues']\n",
    "    df_return = pd.concat([df_geneAt, df_geneDt], axis=0)\n",
    "    return df_return\n",
    "# 得到bias基因对于的loop anchor\n",
    "def get_loop_anchor(df_gene, df_gene_bin):\n",
    "    df_gene, df_gene_bin = df_gene.copy(), df_gene_bin.copy()\n",
    "    df_return = pd.merge(df_gene, df_gene_bin, how='inner', on=['GeneId'])\n",
    "    return df_return\n",
    "# 分析有多少基因是能和G-G, G-Cis互作\n",
    "def GG_GCis(df_gene_bin, df_loop):\n",
    "    df_gene_bin, df_loop = df_gene_bin.copy(), df_loop.copy()\n",
    "    df_gene_bin.columns = ['GeneId1', 'tissues', 'chrom1', 'start1', 'end1']\n",
    "    df_a = pd.merge(df_gene_bin, df_loop, how='inner', on=['chrom1', 'start1', 'end1', 'tissues'])\n",
    "    df_gene_bin.columns = ['GeneId2', 'tissues', 'chrom2', 'start2', 'end2']\n",
    "    df_b = pd.merge(df_gene_bin, df_loop, how='inner', on=['chrom2', 'start2', 'end2', 'tissues'])\n",
    "    df_return = pd.concat([df_a, df_b], axis=0)\n",
    "    return df_return\n",
    "\n",
    "\n",
    "def GNCis(df_GN, df_CisN):\n",
    "    df_GN, df_CisN = df_GN.copy(), df_CisN.copy()\n",
    "    df_GN.drop_duplicates(subset=['chrom2', 'start2', 'end2', 'tissues'], inplace=True, keep='first')\n",
    "    df_all = pd.concat([df_GN, df_CisN], axis=0)\n",
    "    df_return = df_all[df_all.duplicated(['chrom2', 'start2', 'end2', 'tissues'], keep=False)]\n",
    "    #df_return = df_tmp.drop_duplicates(keep='last', subset=['chrom2', 'start2', 'end2', 'tissues'])\n",
    "    #print(df_return.shape[0])\n",
    "    return df_return\n",
    "\n",
    "# 分析确定GNG和GNCis\n",
    "def GNG_GNCis(df, df_loop):\n",
    "    df, df_loop = df.copy(), df_loop.copy()\n",
    "    df_CisN = df_loop.query(\"loop_type=='Cis-N'\").copy()\n",
    "    df_GN_all = df_loop.query(\"loop_type=='G-N'\").copy()\n",
    "    df_GN = df.query(\"loop_type == 'G-N'\").copy()\n",
    "    df_GCis = df.query(\"loop_type == 'G-Cis'\").copy()\n",
    "    df_GG = df.query(\"loop_type == 'G-G'\").copy()\n",
    "    df_GNG = GNCis(df_GN, df_GN_all)\n",
    "    df_GNCis = GNCis(df_GN, df_CisN)\n",
    "    print(df_GG.shape[0], df_GCis.shape[0], df_GNG.shape[0], df_GNCis.shape[0])\n",
    "    df_return = pd.concat([df_GG, df_GCis, df_GNG, df_GNCis], axis=0)\n",
    "    return df_return\n",
    "\n",
    "df_gene = get_bias_gene(df_loop_bias)\n",
    "df_bias_gene_bin = get_loop_anchor(df_gene, df_gene_bin)\n",
    "df_GG_GCis = GG_GCis(df_bias_gene_bin, df_loop)\n",
    "df_result = GNG_GNCis(df_GG_GCis, df_loop)\n",
    "\n",
    "### 得到所有正loop数目调控的bias基因所相关的GG, G-Cis, GNG, GNCis 相关的loops\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/J668_positive_loop_gene_GG_GN_GNG_GNCis.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11114\n"
     ]
    }
   ],
   "source": [
    "## 分析这其中有多少基因比例是与G-G, G-Cis，G-N-G， G-N-Cis 有关 同源基因相关的所有都仅包含同源基因\n",
    "import pandas as pd\n",
    "df_gene_bin = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/3kb_bin_gene.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'GeneId'])\n",
    "df_loop = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type_cis.csv\")\n",
    "# 得到bias基因id\n",
    "def get_bias_gene(df):\n",
    "    df_geneAt = df[['GeneId_At', 'tissues']].copy()\n",
    "    df_geneDt = df[['GeneId_Dt', 'tissues']].copy()\n",
    "    df_geneAt.columns = ['GeneId', 'tissues']\n",
    "    df_geneDt.columns = ['GeneId', 'tissues']\n",
    "    df_return = pd.concat([df_geneAt, df_geneDt], axis=0)\n",
    "    return df_return\n",
    "# 得到bias基因对于的loop anchor\n",
    "def get_loop_anchor(df_gene, df_gene_bin):\n",
    "    df_gene, df_gene_bin = df_gene.copy(), df_gene_bin.copy()\n",
    "    df_return = pd.merge(df_gene, df_gene_bin, how='inner', on=['GeneId'])\n",
    "    return df_return\n",
    "\n",
    "# 分析有多少基因是能和G-G, G-Cis互作\n",
    "def GG_GCis(df_gene_bin, df_loop):\n",
    "    df_gene_bin, df_loop = df_gene_bin.copy(), df_loop.copy()\n",
    "    df_gene_bin.columns = ['GeneId1', 'tissues', 'chrom1', 'start1', 'end1']\n",
    "    df_a = pd.merge(df_gene_bin, df_loop, how='inner', on=['chrom1', 'start1', 'end1', 'tissues'])\n",
    "    df_gene_bin.columns = ['GeneId2', 'tissues', 'chrom2', 'start2', 'end2']\n",
    "    df_GG = pd.merge(df_gene_bin, df_a, how='inner', on=['chrom2', 'start2', 'end2', 'tissues'])\n",
    "    df_GG.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/J668_HG_GG_loop.csv\", index=False)\n",
    "    df_GCis = df_a.query(\"loop_type=='G-Cis'\").copy()\n",
    "    df_return = pd.concat([df_GG, df_GCis], axis=0)\n",
    "    df_return['GeneId2'].fillna('Cis', inplace=True)\n",
    "    return df_return\n",
    "\n",
    "def GNG(df_gene_bin, df_loop):\n",
    "    df_gene_bin, df_loop = df_gene_bin.copy(), df_loop.copy()\n",
    "    df_gene_bin.columns = ['GeneId1', 'tissues', 'chrom1', 'start1', 'end1']\n",
    "    df_a = pd.merge(df_gene_bin, df_loop, how='inner', on=['chrom1', 'start1', 'end1', 'tissues'])\n",
    "    df_GN = df_a.query(\"loop_type=='G-N'\").copy()\n",
    "    df_GNG = df_GN[df_GN.duplicated(subset=['chrom2', 'start2', 'end2', 'tissues'], keep=False)].copy()\n",
    "    df_GNG['GeneId2']=['NotElement']*df_GNG.shape[0]\n",
    "    return df_GNG\n",
    "\n",
    "def GNCis1(df_GN, df_CisN):\n",
    "    df_GN, df_CisN = df_GN.copy(), df_CisN.copy()\n",
    "    df_GN.drop_duplicates(subset=['chrom2', 'start2', 'end2', 'tissues'], inplace=True, keep='first')\n",
    "    df_all = pd.concat([df_GN, df_CisN], axis=0)\n",
    "    df_return = df_all[df_all.duplicated(['chrom2', 'start2', 'end2', 'tissues'], keep=False)]\n",
    "    return df_return\n",
    "\n",
    "# 分析确定GNG和GNCis\n",
    "def GNCis(df, df_loop):\n",
    "    df, df_loop = df.copy(), df_loop.copy()  \n",
    "    df.columns = ['GeneId1', 'tissues', 'chrom1', 'start1', 'end1']\n",
    "    df_GN = pd.merge(df, df_loop, how='inner', on=['chrom1', 'start1', 'end1', 'tissues']).query(\"loop_type=='G-N'\").copy()\n",
    "    df_GN_Nanchor = df_GN[['chrom2', 'start2', 'end2', 'tissues']].copy()\n",
    "    df_GN_Nanchor.drop_duplicates(keep='first', inplace=True)\n",
    "    print(df_GN_Nanchor.shape[0])\n",
    "    ## 找到与GN的N端重合的CisN loop\n",
    "    df_CisN = df_loop.query(\"loop_type=='Cis-N'\").copy()\n",
    "    df_Cis_N = pd.merge(df_CisN, df_GN_Nanchor, how='inner', on=['chrom2', 'start2', 'end2', 'tissues'])\n",
    "    ## 重叠的N锚点\n",
    "    df_overlap_Nanchor = df_Cis_N[['chrom2', 'start2', 'end2', 'tissues']].copy()\n",
    "    df_overlap_Nanchor.drop_duplicates(keep='first', inplace=True)\n",
    "    ## GN loop\n",
    "    df_GN = pd.merge(df_GN, df_overlap_Nanchor, how='inner', on=['chrom2', 'start2', 'end2', 'tissues'])\n",
    "    \n",
    "    df_GNCis = pd.concat([df_GN, df_Cis_N], axis=0)\n",
    "    df_GNCis['GeneId1'].fillna(\"Cis\", inplace=True)\n",
    "    df_GNCis['GeneId2'] = ['NotElement']*df_GNCis.shape[0]\n",
    "    return df_GNCis\n",
    "df_loop_bias = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/J668_loop_positive_hg_bias_pair.csv\")\n",
    "\n",
    "df_loop_bias = df_loop_bias.query(\"Bias=='BiasD'\")\n",
    "\n",
    "\n",
    "df_gene = get_bias_gene(df_loop_bias)\n",
    "df_bias_gene_bin = get_loop_anchor(df_gene, df_gene_bin)\n",
    "df_GG_GCis = GG_GCis(df_bias_gene_bin, df_loop)\n",
    "df_GNG = GNG(df_bias_gene_bin, df_loop)\n",
    "df_GNCis = GNCis(df_bias_gene_bin, df_loop)\n",
    "df_GNG.sort_values(by=['tissues', 'chrom2', 'start2', 'end2'], inplace=True)\n",
    "df_GNCis.sort_values(by=['tissues', 'chrom2', 'start2', 'end2'], inplace=True)\n",
    "df_result = pd.concat([df_GG_GCis, df_GNG, df_GNCis], axis=0)\n",
    "\n",
    "# #edge at\n",
    "# df_result['tags'] = ['BiasA']*df_result.shape[0]\n",
    "# df_edge = df_result[['GeneId1', 'GeneId2', 'tags', 'tissues']].copy()\n",
    "# df_edge.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/BiasAt_edge.csv\", index=False)\n",
    "\n",
    "#edge dt\n",
    "df_result['tags'] = ['BiasD']*df_result.shape[0]\n",
    "df_edge = df_result[['GeneId1', 'GeneId2', 'tags', 'tissues']].copy()\n",
    "df_edge.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/BiasDt_edge.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 448,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\pc\\AppData\\Local\\Temp\\ipykernel_13004\\2114696152.py:3: DtypeWarning: Columns (0,10) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/J668_positive_loop_gene_GG_GN_GNG_GNCis.csv\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    GeneId1    tissues    chrom1    start1      end1    chrom2    start2  \\\n",
      "494     NaN  hypocotyl  Ghir_A09  69753000  69756000  Ghir_A09  69816000   \n",
      "\n",
      "         end2 tags loop_type          GeneId2  \n",
      "494  69819000  G-G       G-G  Ghir_A09G014510  \n",
      "                Source      Target\n",
      "48548  Ghir_A01G002510  NotElement\n",
      "48577  Ghir_A01G003170  NotElement\n",
      "48597  Ghir_A01G003820  NotElement\n",
      "48615  Ghir_A01G004740  NotElement\n",
      "48635  Ghir_A01G004910  NotElement\n",
      "...                ...         ...\n",
      "70551  Ghir_D13G020130  NotElement\n",
      "70556  Ghir_D13G020390  NotElement\n",
      "70559  Ghir_D13G020400  NotElement\n",
      "70578  Ghir_D13G022400  NotElement\n",
      "70614  Ghir_D13G024260  NotElement\n",
      "\n",
      "[1236 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "## 将所有的互作整理为边格式，  hypocotyl\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/J668_positive_loop_gene_GG_GN_GNG_GNCis.csv\")\n",
    "df_HG_GG = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/J668_HG_GG_loop.csv\")\n",
    "T = 'hypocotyl'\n",
    "## 整理GG 互作, 去掉同源GG \n",
    "def GG(df, df_gene_bin):\n",
    "    df_GG = df.query(\"loop_type=='G-G'\").copy()\n",
    "    #print(df_GG.query(\"GeneId2=='Ghir_A09G014510'\"))\n",
    "    \n",
    "    ## 同源基因\n",
    "    df_GG_bin = df_GG[['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2']].copy()\n",
    "    df_gene_bin.columns = ['GeneId1', 'chrom1', 'start1', 'end1']\n",
    "    df_a = pd.merge(df_GG_bin, df_gene_bin, how='inner', on=['chrom1', 'start1', 'end1'])\n",
    "    df_gene_bin.columns = ['GeneId2', 'chrom2', 'start2', 'end2']\n",
    "    df_ab = pd.merge(df_a, df_gene_bin, how='inner', on=['chrom2', 'start2', 'end2'])\n",
    "    #print(df_ab.query(\"GeneId2.str.contains('Ghir_A09G014510')\", engine='python'))\n",
    "    \n",
    "    df_GG_gene = df_ab[['GeneId1', 'GeneId2']].copy()\n",
    "    \n",
    "    # 非同源基因\n",
    "    df_a_none = df_GG[df_GG['GeneId1'].isnull()]\n",
    "    df_a_not_none = df_GG.dropna(subset=['GeneId1'])\n",
    "    \n",
    "    df_a = pd.concat([df_a_not_none, df_ab], axis=0, ignore_index=True)\n",
    "    df_a.drop_duplicates(keep=False, inplace=True, subset=['GeneId1', 'chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tissues'])\n",
    "    #print(df_a.query(\"GeneId2=='Ghir_A09G014510'\"))\n",
    "    df_b_none = df_a[df_a['GeneId2'].isnull()]\n",
    "    df_b_not_none = df_a.dropna(subset=['GeneId2'])\n",
    "    df_a_b = pd.concat([df_b_not_none, df_ab], axis=0, ignore_index=True)\n",
    "    df_a_b.drop_duplicates(keep=False, inplace=True, subset=['GeneId2', 'chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'tissues'])\n",
    "    df_NHGG = pd.concat([df_a_none, df_b_none, df_a_b], axis=0, ignore_index=True)\n",
    "\n",
    "    print(df_NHGG.query(\"GeneId2=='Ghir_A09G014510'\"))\n",
    "    \n",
    "    df_G1 = df_NHGG.query(\"GeneId2.isnull()\").copy()\n",
    "    df_G2 = df_NHGG.query(\"GeneId1.isnull()\").copy()\n",
    "    df_G1 = df_G1[['GeneId1']].copy()\n",
    "    df_G2 = df_G2[['GeneId2']].copy()\n",
    "    #print(df_G2.query(\"GeneId2=='Ghir_A09G014510'\"))\n",
    "    df_G1.columns = ['Source']\n",
    "    df_G2.columns = ['Source']\n",
    "    df_G1G2 = pd.concat([df_G1, df_G2], axis=0)\n",
    "    df_G1G2['Target'] = ['Gene']*df_G1G2.shape[0]\n",
    "    df_NHG = df_G1G2.drop_duplicates(keep='first')\n",
    "    \n",
    "    #print(df_GG_gene.query(\"GeneId1==GeneId2\"))\n",
    "    #df_GG_gene = df_GG_gene.query(\"GeneId1!=GeneId2\").copy()\n",
    "    df_GG_gene.drop_duplicates(keep='first', inplace=True)\n",
    "    \n",
    "    df_GG_gene.columns = ['Source', 'Target']\n",
    "    #print(df_GG_gene)\n",
    "    ## 总结所有GG\n",
    "    df_return = pd.concat([df_NHG, df_GG_gene], axis=0)\n",
    "    return df_return\n",
    "# def HG_HG(df):\n",
    "#     df_HG = df[['GeneId1', 'GeneId2']].copy()\n",
    "#     df_HG.columns = ['Source', 'Target']\n",
    "#     df_HG = df_HG.query(\"Source != Target\").copy()\n",
    "#     df_HG.drop_duplicates(keep='first', inplace=True)\n",
    "#     return df_HG\n",
    "def GCis(df):\n",
    "    df_GCis = df.query(\"loop_type=='G-Cis'\").copy()\n",
    "    df_return = df_GCis[['GeneId1']].copy()\n",
    "    df_return.columns = ['Source']\n",
    "    df_return['Target'] = ['Cis']*df_return.shape[0]\n",
    "    df_return.drop_duplicates(keep='first', inplace=True)\n",
    "    return df_return\n",
    "def GN(df):\n",
    "    df_GN = df.query(\"loop_type=='G-N'\").copy()\n",
    "    df_GN = df_GN.query(\"~GeneId1.isnull()\")\n",
    "    df_return = df_GN[['GeneId1']].copy()\n",
    "    df_return.columns = ['Source']\n",
    "    df_return['Target'] = ['NotElement']*df_return.shape[0]\n",
    "    df_return.drop_duplicates(keep='first', inplace=True)\n",
    "    #print(df_return)\n",
    "    return df_return\n",
    "\n",
    "df_T = df.query(\"tissues==@T\")\n",
    "df_HG_T = df_HG_GG.query(\"tissues==@T\")\n",
    "\n",
    "df_gbin = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/3kb_bin_gene.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'GeneId'])\n",
    "df_bias = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/J668_loop_positive_hg_bias_pair.csv\")\n",
    "df_Bias_T = df_bias.query(\"tissues=='hypocotyl'\").copy()\n",
    "df_Bias_G = pd.DataFrame(pd.concat([df_Bias_T['GeneId_At'], df_Bias_T['GeneId_Dt']], axis=0, ignore_index=True), columns=['GeneId'])\n",
    "df_Bias_G = df_Bias_G.query(\"GeneId.str.contains('Ghir')\", engine='python').copy()\n",
    "df_Bias_G.drop_duplicates(keep='first', inplace=True)\n",
    "df_Bias_bin = pd.merge(df_Bias_G, df_gbin, how='inner', on=['GeneId'])\n",
    "\n",
    "\n",
    "df_GG = GG(df_T, df_Bias_bin)\n",
    "df_GCis = GCis(df_T)\n",
    "df_GN = GN(df_T)\n",
    "print(df_GN)\n",
    "df_result = pd.concat([df_GG, df_GCis, df_GN], axis=0)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# ## 确定哪些基因是bias a 哪些基因是bias dt\n",
    "# df_bias = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/all_tissue_bias_gene_long.bed\", sep=\"\\t\")\n",
    "# df_bias_T = df_bias.query(\"tissues==@T\")[['GeneId', 'Bias']].copy()\n",
    "# df_bias_T.columns = ['Source', 'Bias']\n",
    "# df_result = pd.merge(df_all, df_bias_T, how='inner', on=['Source'])\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/hypocotyl_network2/hypocotyl_HBG_interaction.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 436,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          GeneId_At        GeneId_Dt   Bias\n",
      "2   Ghir_A01G000150  Ghir_D01G000160  BiasA\n",
      "16  Ghir_A01G000380  Ghir_D01G000380  BiasD\n",
      "27  Ghir_A01G000670  Ghir_D01G000640  BiasD\n",
      "58  Ghir_A01G001190  Ghir_D01G001190  BiasD\n",
      "62  Ghir_A01G001380  Ghir_D01G001380  BiasD\n"
     ]
    }
   ],
   "source": [
    "## 构建网络模型\n",
    "## 得到所有的hypocotyl基因\n",
    "def get_gene(df):\n",
    "    df_geneid = pd.DataFrame(pd.concat([df['Source'], df['Target']], axis=0, ignore_index=True), columns=['GeneId'])\n",
    "    df_geneid = df_geneid.query(\"GeneId.str.contains('Ghir')\").drop_duplicates(keep='first')\n",
    "    return df_geneid\n",
    "## 得到同源基因对\n",
    "def get_hg(df1, df2):\n",
    "    df1, df2 = df1.copy(), df2.copy()\n",
    "    df2.columns = ['GeneId_At']\n",
    "    df_a = pd.merge(df1, df2, how='inner', on=['GeneId_At'])\n",
    "    df2.columns = ['GeneId_Dt']\n",
    "    df_b = pd.merge(df1, df2, how='inner', on=['GeneId_Dt'])\n",
    "    df_return = pd.concat([df_a, df_b], axis=0)\n",
    "    df_return.drop_duplicates(keep='first', inplace=True)\n",
    "    return df_return\n",
    "## 构建网络\n",
    "# 每个基因都与cis，notelement, 非同源基因 有关, \n",
    "def cis_notelement_gene(df):\n",
    "    df_cis, df_notelement, df_gene = df.copy(), df.copy(), df.copy()\n",
    "    df_cis.columns = ['source']\n",
    "    df_notelement.columns = ['source']\n",
    "    df_gene.columns = ['source']\n",
    "    df_cis['target'] = ['Cis']*df_cis.shape[0]\n",
    "    df_notelement['target'] = ['NotElement']*df_notelement.shape[0]\n",
    "    df_gene['target'] = ['Gene']*df_gene.shape[0]\n",
    "    df_result = pd.concat([df_cis, df_notelement, df_gene], axis=0, ignore_index=True)\n",
    "    return df_result\n",
    "# 构建基因和基因间的调控网络\n",
    "def GG(df):\n",
    "    df_source = df.sample(n=3000, replace=True)\n",
    "    df_target = df.sample(n=3000, replace=True)\n",
    "    df_source.columns = ['source']\n",
    "    df_target.columns = ['target']\n",
    "    df_source.reset_index(drop=True, inplace=True)\n",
    "    df_target.reset_index(drop=True, inplace=True)\n",
    "    df_return = pd.concat([df_source, df_target], axis=1)\n",
    "    df_return = df_return.query(\"source != target\").copy()\n",
    "    return df_return\n",
    "\n",
    "def build_net(df):\n",
    "    # Bias A\n",
    "    df_BiasA = df.query(\"Bias=='BiasA'\")[['GeneId_At']].copy()\n",
    "    df_BiasA_GG = GG(df_BiasA)\n",
    "    df_BiasA_cis_notelement = cis_notelement_gene(df_BiasA)\n",
    "    df_BiasA_net = pd.concat([df_BiasA_GG, df_BiasA_cis_notelement])\n",
    "    \n",
    "    # Bias D\n",
    "    df_BiasD = df.query(\"Bias=='BiasD'\")[['GeneId_At']].copy()\n",
    "    df_BiasD_GG = GG(df_BiasD)\n",
    "    df_BiasD_cis_notelement = cis_notelement_gene(df_BiasD)\n",
    "    df_BiasD_net = pd.concat([df_BiasD_GG, df_BiasD_cis_notelement])\n",
    "    \n",
    "    ## 统一节点\n",
    "    df_BiasA.columns = ['GeneId']\n",
    "    df_BiasA['tags'] = ['BiasAt']*df_BiasA.shape[0]\n",
    "    df_BiasD.columns = ['GeneId']\n",
    "    df_BiasD['tags'] = ['BiasDt']*df_BiasD.shape[0]\n",
    "    df_node = pd.concat([df_BiasA, df_BiasD], axis=0, ignore_index=True)\n",
    "    df_node.reset_index(drop=True, inplace=True)\n",
    "    df_node.loc[df_node.shape[0], :] = ['Cis', 'Cis']\n",
    "    df_node.loc[df_node.shape[0], :] = ['NotElement', 'NotElement']\n",
    "    \n",
    "    df_result = pd.concat([df_BiasA_net, df_BiasD_net], axis=0, ignore_index=True)\n",
    "    #print(df_result)\n",
    "    df_result.loc[df_result.shape[0], :] = ['Cis', 'NotElement']\n",
    "    df_result.loc[df_result.shape[0], :] = ['Gene', 'NotElement']\n",
    "    return df_result, df_node\n",
    "\n",
    "\n",
    "# tissue\n",
    "T = \"hypocotyl\"\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/J668_loop_positive_hg_bias_pair.csv\")\n",
    "df_T = df.query(\"tissues==@T\")[['GeneId_At', 'GeneId_Dt', 'Bias']].copy()\n",
    "print(df_T.head())\n",
    "df_hypocotyl = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/hypocotyl_network2/hypocotyl_HBG_interaction.csv\")\n",
    "df_gene = get_gene(df_hypocotyl)\n",
    "df_hg = get_hg(df_T, df_gene)\n",
    "df_hg.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/hypocotyl_network2/hypocotyl_hg_bias.csv\", index=False)\n",
    "df_net, df_node = build_net(df_hg)\n",
    "df_net.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/hypocotyl_network2/hypocotyl_net.csv\", index=False)\n",
    "df_node.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/hypocotyl_network2/hypocotyl_node.csv\", index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 441,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/hypocotyl_network2/hypocotyl_hg_bias.csv\")\n",
    "df_a = df[['GeneId_At', 'Bias']].copy()\n",
    "df_b = df[['GeneId_Dt', 'Bias']].copy()\n",
    "df_a.columns = ['GeneId', 'Bias']\n",
    "df_b.columns = ['GeneId', 'Bias']\n",
    "df_ab = pd.concat([df_a, df_b], axis=0)\n",
    "df_ab.columns = ['Source', 'Bias']\n",
    "df_edge = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/hypocotyl_network2/hypocotyl_HBG_interaction.csv\")\n",
    "df_edge = pd.merge(df_edge, df_ab, how='left', on=['Source'])\n",
    "df_edge.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/hypocotyl_network2/hypocotyl_HBG_interaction.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 449,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "               Source      Target        Type    Id Lable timeset  Weight\n",
      "0     Ghir_A01G000150        Gene  Undirected     0  None    None       1\n",
      "1     Ghir_A01G002510        Gene  Undirected     1  None    None       1\n",
      "2     Ghir_A01G002750        Gene  Undirected     2  None    None       1\n",
      "3     Ghir_A01G002770        Gene  Undirected     3  None    None       1\n",
      "4     Ghir_A01G003820        Gene  Undirected     4  None    None       1\n",
      "...               ...         ...         ...   ...   ...     ...     ...\n",
      "1870  Ghir_A13G016020  NotElement  Undirected  1870  None    None       1\n",
      "1871  Ghir_A13G016700  NotElement  Undirected  1871  None    None       1\n",
      "1872  Ghir_A13G017300  NotElement  Undirected  1872  None    None       1\n",
      "1873  Ghir_A13G021290  NotElement  Undirected  1873  None    None       1\n",
      "1874  Ghir_A13G024350  NotElement  Undirected  1874  None    None       1\n",
      "\n",
      "[1875 rows x 7 columns]\n",
      "               Source      Target        Type    Id Lable timeset  Weight\n",
      "0     Ghir_A01G000380        Gene  Undirected     0  None    None       1\n",
      "1     Ghir_A01G000670        Gene  Undirected     1  None    None       1\n",
      "2     Ghir_A01G001380        Gene  Undirected     2  None    None       1\n",
      "3     Ghir_A01G001440        Gene  Undirected     3  None    None       1\n",
      "4     Ghir_A01G001780        Gene  Undirected     4  None    None       1\n",
      "...               ...         ...         ...   ...   ...     ...     ...\n",
      "2032  Ghir_A13G019390  NotElement  Undirected  2032  None    None       1\n",
      "2033  Ghir_A13G019650  NotElement  Undirected  2033  None    None       1\n",
      "2034  Ghir_A13G019660  NotElement  Undirected  2034  None    None       1\n",
      "2035  Ghir_A13G021560  NotElement  Undirected  2035  None    None       1\n",
      "2036  Ghir_A13G023530  NotElement  Undirected  2036  None    None       1\n",
      "\n",
      "[2037 rows x 7 columns]\n"
     ]
    }
   ],
   "source": [
    "## 生成边\n",
    "## 构建边\n",
    "# At subgenome \n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/hypocotyl_network2/hypocotyl_HBG_interaction.csv\")\n",
    "def build_edge(df, tags, At_gene, Dt_gene):\n",
    "    df_tmp = df.query(\"Source.str.contains(@tags) | Target.str.contains(@tags)\")[['Source', 'Target']].copy()\n",
    "    if tags == 'Ghir_D':\n",
    "        df_tmp.replace(Dt_gene, At_gene, inplace=True)\n",
    "    df_tmp = df_tmp.query(\"Source != Target\").copy()\n",
    "    df_tmp.reset_index(inplace=True, drop=True)    \n",
    "    df_tmp['Type'] = ['Undirected']*df_tmp.shape[0]\n",
    "    df_tmp['Id'] = range(df_tmp.shape[0])\n",
    "    df_tmp['Lable'] = [None]*df_tmp.shape[0]\n",
    "    df_tmp['timeset'] = [None]*df_tmp.shape[0]\n",
    "    df_tmp['Weight'] = [1]*df_tmp.shape[0]\n",
    "    print(df_tmp)\n",
    "    df_tmp.loc[df_tmp.shape[0], :] = ['Cis', 'NotElement', 'Undirected', df_tmp.shape[0], None, None, 1]\n",
    "    df_tmp.loc[df_tmp.shape[0], :] = ['Gene', 'NotElement', 'Undirected', df_tmp.shape[0], None, None, 1]\n",
    "    \n",
    "    \n",
    "    return df_tmp\n",
    "df_hg_pair = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/J668_HG_pairs.csv\")\n",
    "At_gene = df_hg_pair['GeneId_At'].to_list()\n",
    "Dt_gene = df_hg_pair['GeneId_Dt'].to_list()\n",
    "df_At_edge = build_edge(df, 'Ghir_A', At_gene, Dt_gene)\n",
    "df_At_edge.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/hypocotyl_network2/At_gene_edge.csv\", index=False)\n",
    "df_Dt_edge = build_edge(df, 'Ghir_D', At_gene, Dt_gene)\n",
    "df_Dt_edge.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/hypocotyl_network2/Dt_gene_edge.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_at = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/BiasAt_edge.csv\")\n",
    "df_dt = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/BiasDt_edge.csv\")\n",
    "df_atdt = pd.concat([df_at, df_dt], axis=0)\n",
    "df_atdt.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/BiasAtDt_edge.csv\", index=False)\n",
    "df_hypocotyl = df_atdt.query(\"tissues=='hypocotyl'\").copy()\n",
    "df_hypocotyl.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/BiasAtDt_edge_hypocotyl.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              GeneId2    tissues    chrom2     start2       end2  \\\n",
      "15    Ghir_A01G003820  hypocotyl  Ghir_A01    4092000    4095000   \n",
      "71    Ghir_A01G015380  hypocotyl  Ghir_A01   99771000   99774000   \n",
      "115   Ghir_A02G008600  hypocotyl  Ghir_A02   20580000   20583000   \n",
      "133   Ghir_A02G011070  hypocotyl  Ghir_A02   48615000   48618000   \n",
      "150   Ghir_A02G017130  hypocotyl  Ghir_A02  105585000  105588000   \n",
      "...               ...        ...       ...        ...        ...   \n",
      "4089              NaN  hypocotyl  Ghir_D13   34227000   34230000   \n",
      "4048              NaN  hypocotyl  Ghir_D13   34227000   34230000   \n",
      "4139              NaN  hypocotyl  Ghir_D13   58143000   58146000   \n",
      "4140              NaN  hypocotyl  Ghir_D13   58143000   58146000   \n",
      "1180              NaN  hypocotyl  Ghir_D13   58143000   58146000   \n",
      "\n",
      "              GeneId1    chrom1     start1       end1 tags loop_type  \n",
      "15    Ghir_A01G003820  Ghir_A01    4065000    4068000  G-G       G-G  \n",
      "71    Ghir_A01G015380  Ghir_A01   99765000   99768000  G-G       G-G  \n",
      "115   Ghir_A02G008580  Ghir_A02   20040000   20043000  G-G       G-G  \n",
      "133   Ghir_A02G011060  Ghir_A02   48456000   48459000  G-G       G-G  \n",
      "150   Ghir_A02G017110  Ghir_A02  105537000  105540000  G-G       G-G  \n",
      "...               ...       ...        ...        ...  ...       ...  \n",
      "4089  Ghir_D13G011440  Ghir_D13   34275000   34278000  N-G       G-N  \n",
      "4048              NaN  Ghir_D13   34278000   34281000  N-N     Cis-N  \n",
      "4139  Ghir_D13G020390  Ghir_D13   58113000   58116000  G-N       G-N  \n",
      "4140  Ghir_D13G020400  Ghir_D13   58122000   58125000  G-N       G-N  \n",
      "1180              NaN  Ghir_D13   58023000   58026000  N-N     Cis-N  \n",
      "\n",
      "[1857 rows x 11 columns]\n",
      "               GeneId\n",
      "15    Ghir_A01G003820\n",
      "71    Ghir_A01G015380\n",
      "115   Ghir_A02G008580\n",
      "133   Ghir_A02G011060\n",
      "150   Ghir_A02G017110\n",
      "...               ...\n",
      "2752  Ghir_D13G004640\n",
      "2762  Ghir_D13G005720\n",
      "2763  Ghir_D13G007780\n",
      "2778  Ghir_D11G029180\n",
      "2791  Ghir_D13G014490\n",
      "\n",
      "[1121 rows x 1 columns]\n"
     ]
    }
   ],
   "source": [
    "df_hypocotyl = df_result.query(\"tissues=='hypocotyl'\").copy()\n",
    "print(df_hypocotyl)\n",
    "df_hypocotyl_geneid = pd.DataFrame(pd.concat([df_hypocotyl['GeneId1'], df_hypocotyl['GeneId2']], axis=0), columns=['GeneId'])\n",
    "df_hypocotyl_geneid.dropna(inplace=True)\n",
    "df_hypocotyl_geneid.drop_duplicates(keep='first', inplace=True)\n",
    "print(df_hypocotyl_geneid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 375,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31406\n",
      "             GeneId_At        GeneId_Dt    tissues\n",
      "2      Ghir_A01G000150  Ghir_D01G000160  hypocotyl\n",
      "54     Ghir_A01G002510  Ghir_D01G002420  hypocotyl\n",
      "66     Ghir_A01G002750  Ghir_D01G002710  hypocotyl\n",
      "68     Ghir_A01G002770  Ghir_D01G002740  hypocotyl\n",
      "75     Ghir_A01G002890  Ghir_D01G002850  hypocotyl\n",
      "...                ...              ...        ...\n",
      "15812  Ghir_A13G023370  Ghir_D13G024010  hypocotyl\n",
      "15823  Ghir_A13G023530  Ghir_D13G024260  hypocotyl\n",
      "15827  Ghir_A13G023560  Ghir_D13G024290  hypocotyl\n",
      "15831  Ghir_A13G023760  Ghir_D13G024490  hypocotyl\n",
      "15839  Ghir_A13G024020  Ghir_D13G024760  hypocotyl\n",
      "\n",
      "[1820 rows x 3 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\pc\\AppData\\Local\\Temp\\ipykernel_13004\\3119203944.py:4: DtypeWarning: Columns (0,10) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/J668_positive_loop_gene_GG_GN_GNG_GNCis.csv\")\n"
     ]
    }
   ],
   "source": [
    "## 分析各组织的cis基因有多少是能够受到GG, G-Cis，GNCis， GNG调控的\n",
    "import pandas as pd\n",
    "df_loop_positive_hg = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/J668_loop_positive_hg_bias_pair.csv\")\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/J668_positive_loop_gene_GG_GN_GNG_GNCis.csv\")\n",
    "\n",
    "df_a = df[['GeneId1', 'tissues']].copy()\n",
    "df_b = df[['GeneId2', 'tissues']].copy()\n",
    "df_a.columns = ['GeneId', 'tissues']\n",
    "df_b.columns = ['GeneId', 'tissues']\n",
    "df_all = pd.concat([df_a, df_b], axis=0)\n",
    "df_all.dropna(inplace=True)\n",
    "df_all.drop_duplicates(keep='first', inplace=True)\n",
    "print(df_all.shape[0])\n",
    "\n",
    "\n",
    "df_HG = df_loop_positive_hg[['GeneId_At', 'GeneId_Dt', 'tissues']].copy()\n",
    "\n",
    "df_all.columns = ['GeneId_At', 'tissues']\n",
    "df_a = pd.merge(df_HG, df_all, how='inner', on=['GeneId_At', 'tissues'])\n",
    "df_all.columns = ['GeneId_Dt', 'tissues']\n",
    "df_b = pd.merge(df_HG, df_all, how='inner', on=['GeneId_Dt', 'tissues'])\n",
    "df_ab = pd.concat([df_a, df_b], axis=0)\n",
    "df_ab.drop_duplicates(keep='first', inplace=True)\n",
    "print(df_ab.query(\"tissues=='hypocotyl'\"))\n",
    "\n",
    "df_ab.query(\"tissues=='hypocotyl'\").to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/a.csv\", index=False)\n",
    "\n",
    "\n",
    "# df_all['tags'] = ['cis']*df_all.shape[0]\n",
    "# df_all.columns = ['GeneId_At', 'tissues', 'tags_At']\n",
    "# df_all.drop_duplicates(keep='first', inplace=True)\n",
    "# df_a = pd.merge(df_loop_positive_hg, df_all, how='left', on=['GeneId_At', 'tissues'])\n",
    "# print(df_a.shape[0])\n",
    "# df_all.columns = ['GeneId_Dt', 'tissues', 'tags_Dt']\n",
    "# df_b = pd.merge(df_loop_positive_hg, df_all, how='left', on=['GeneId_Dt', 'tissues'])\n",
    "\n",
    "# df_ab = pd.merge(df_a, df_b, how='left', on=['GeneId_At', 'GeneId_Dt', 'tissues', 'number_At', 'number_Dt', 'Bias'])\n",
    "# print(df_ab.head())\n",
    "# ## 确定哪些基因受到了G和Cis的调控\n",
    "# df_result = df_ab.dropna(subset=['tags_At', 'tags_Dt'], how='all')\n",
    "# print(df_result)\n",
    "# df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/J668_loop_positive_hg_bias_pair_G_Cis.csv\", index=False)\n",
    "# df = df_result['tissues'].value_counts()\n",
    "# print(df)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 434,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           GeneId_At    tissues  number_At   Bias        GeneId_Dt  number_Dt\n",
      "0    Ghir_A01G002890  hypocotyl        2.0  BiasA  Ghir_D01G002850        0.0\n",
      "1    Ghir_A01G003170  hypocotyl        2.0  BiasA  Ghir_D01G003140        0.0\n",
      "2    Ghir_A01G003820  hypocotyl        6.0  BiasD  Ghir_D01G003860       10.0\n",
      "3    Ghir_A01G005590  hypocotyl        3.0  BiasA  Ghir_D01G005830        0.0\n",
      "4    Ghir_A01G005970  hypocotyl        2.0  BiasA  Ghir_D01G006270        1.0\n",
      "..               ...        ...        ...    ...              ...        ...\n",
      "555  Ghir_A13G019660  hypocotyl        0.0  BiasD  Ghir_D13G020400        2.0\n",
      "556  Ghir_A13G019870  hypocotyl        0.0  BiasD  Ghir_D13G020650        1.0\n",
      "557  Ghir_A13G019960  hypocotyl        0.0  BiasD  Ghir_D13G020740        2.0\n",
      "558  Ghir_A13G023240  hypocotyl        0.0  BiasD  Ghir_D13G023860        1.0\n",
      "559  Ghir_A13G023530  hypocotyl        0.0  BiasD  Ghir_D13G024260        5.0\n",
      "\n",
      "[892 rows x 6 columns]\n",
      "399 493\n"
     ]
    }
   ],
   "source": [
    "## 构建loop 数目与bias的调控网络\n",
    "import pandas as pd\n",
    "#确定hypocotyl bias受loop数目正调控影响\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/J668_loop_positive_hg_bias_pair.csv\")\n",
    "df_hypocotyl = df.query(\"tissues=='hypocotyl'\").copy()\n",
    "df_hg = pd.read_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/BiasAtDt_edge_hypocotyl.csv')\n",
    "## 得到所有的hypocotyl基因\n",
    "def get_gene(df):\n",
    "    df_geneid = pd.DataFrame(pd.concat([df['GeneId1'], df['GeneId2']], axis=0, ignore_index=True), columns=['GeneId'])\n",
    "    df_geneid = df_geneid.query(\"GeneId.str.contains('Ghir')\").drop_duplicates(keep='first')\n",
    "    return df_geneid\n",
    "## 得到同源基因对\n",
    "def get_hg(df1, df2):\n",
    "    df1, df2 = df1.copy(), df2.copy()\n",
    "    df2.columns = ['GeneId_At']\n",
    "    df_a = pd.merge(df1, df2, how='inner', on=['GeneId_At'])\n",
    "    df2.columns = ['GeneId_Dt']\n",
    "    df_b = pd.merge(df1, df2, how='inner', on=['GeneId_Dt'])\n",
    "    df_return = pd.concat([df_a, df_b], axis=0)\n",
    "    df_return.drop_duplicates(keep='first', inplace=True)\n",
    "    return df_return\n",
    "## 构建网络\n",
    "# 每个基因都与cis，notelement 有关, \n",
    "def cis_notelement(df):\n",
    "    df_cis, df_notelement = df.copy(), df.copy()\n",
    "    df_cis.columns = ['source']\n",
    "    df_notelement.columns = ['source']\n",
    "    df_cis['target'] = ['Cis']*df_cis.shape[0]\n",
    "    df_notelement['target'] = ['NotElement']*df_notelement.shape[0]\n",
    "    df_result = pd.concat([df_cis, df_notelement], axis=0, ignore_index=True)\n",
    "    return df_result\n",
    "# 构建基因和基因间的调控网络\n",
    "def GG(df):\n",
    "    df_source = df.sample(n=3000, replace=True)\n",
    "    df_target = df.sample(n=3000, replace=True)\n",
    "    df_source.columns = ['source']\n",
    "    df_target.columns = ['target']\n",
    "    df_source.reset_index(drop=True, inplace=True)\n",
    "    df_target.reset_index(drop=True, inplace=True)\n",
    "    df_return = pd.concat([df_source, df_target], axis=1)\n",
    "    df_return = df_return.query(\"source != target\").copy()\n",
    "    return df_return\n",
    "\n",
    "def build_net(df):\n",
    "    # Bias A\n",
    "    df_BiasA = df.query(\"Bias=='BiasA'\")[['GeneId_At']].copy()\n",
    "    df_BiasA_GG = GG(df_BiasA)\n",
    "    df_BiasA_cis_notelement = cis_notelement(df_BiasA)\n",
    "    df_BiasA_net = pd.concat([df_BiasA_GG, df_BiasA_cis_notelement])\n",
    "    \n",
    "    # Bias D\n",
    "    df_BiasD = df.query(\"Bias=='BiasD'\")[['GeneId_At']].copy()\n",
    "    df_BiasD_GG = GG(df_BiasD)\n",
    "    df_BiasD_cis_notelement = cis_notelement(df_BiasD)\n",
    "    df_BiasD_net = pd.concat([df_BiasD_GG, df_BiasD_cis_notelement])\n",
    "    \n",
    "    ## 统一节点\n",
    "    df_BiasA.columns = ['GeneId']\n",
    "    df_BiasA['tags'] = ['BiasAt']*df_BiasA.shape[0]\n",
    "    df_BiasD.columns = ['GeneId']\n",
    "    df_BiasD['tags'] = ['BiasDt']*df_BiasD.shape[0]\n",
    "    df_node = pd.concat([df_BiasA, df_BiasD], axis=0, ignore_index=True)\n",
    "    df_node.reset_index(drop=True, inplace=True)\n",
    "    df_node.loc[df_node.shape[0], :] = ['Cis', 'Cis']\n",
    "    df_node.loc[df_node.shape[0], :] = ['NotElement', 'NotElement']\n",
    "    \n",
    "    df_result = pd.concat([df_BiasA_net, df_BiasD_net], axis=0, ignore_index=True)\n",
    "    #print(df_result)\n",
    "    df_result.loc[df_result.shape[0], :] = ['Cis', 'NotElement']\n",
    "    return df_result, df_node\n",
    "\n",
    "df_hypocotyl_hg = get_gene(df_hg) # 一部分同源基因是包含了同源基因间的GG互作和GNG互作，选出来\n",
    "df_hypocotyl_hg = get_hg(df_hypocotyl, df_hypocotyl_hg) ## 将这些同源基因所包含的同源基因对选出来\n",
    "print(df_hypocotyl_hg.query(\"Bias=='BiasA'\").shape[0], df_hypocotyl_hg.query(\"Bias=='BiasD'\").shape[0])\n",
    "df_hypo_net, df_node = build_net(df_hypocotyl_hg) # 构建网络\n",
    "df_hypo_net.drop_duplicates(keep='first', inplace=True)\n",
    "df_hypo_net.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/hypocotyl_network/J668_hypocotyl_network.csv\", index=False)\n",
    "df_node.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/hypocotyl_network/J668_hypocotyl_node.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "713\n"
     ]
    }
   ],
   "source": [
    "## 构建边\n",
    "# At subgenome \n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/BiasAtDt_edge_hypocotyl.csv\")\n",
    "def build_edge(df, tags, At_gene, Dt_gene):\n",
    "    df_tmp = df.query(\"GeneId1.str.contains(@tags) | GeneId2.str.contains(@tags)\")[['GeneId1', 'GeneId2']].copy()\n",
    "    if tags == 'Ghir_D':\n",
    "        df_tmp.replace(Dt_gene, At_gene, inplace=True)\n",
    "    df_tmp.columns = ['Source', 'Target']\n",
    "    df_tmp = df_tmp.query(\"Source != Target\").copy()\n",
    "    df_tmp.reset_index(inplace=True, drop=True)    \n",
    "    df_tmp['Type'] = ['Undirected']*df_tmp.shape[0]\n",
    "    df_tmp['Id'] = range(df_tmp.shape[0])\n",
    "    df_tmp['Lable'] = [None]*df_tmp.shape[0]\n",
    "    df_tmp['timeset'] = [None]*df_tmp.shape[0]\n",
    "    df_tmp['Weight'] = [1]*df_tmp.shape[0]\n",
    "    print(df_tmp.shape[0])\n",
    "    df_tmp.loc[df_tmp.shape[0], :] = ['Cis', 'NotElement', 'Undirected', df_tmp.shape[0], None, None, 1]\n",
    "    return df_tmp\n",
    "df_hg_pair = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/J668_HG_pairs.csv\")\n",
    "At_gene = df_hg_pair['GeneId_At'].to_list()\n",
    "Dt_gene = df_hg_pair['GeneId_Dt'].to_list()\n",
    "# df_At_edge = build_edge(df, 'Ghir_A', At_gene, Dt_gene)\n",
    "# df_At_edge.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/hypocotyl_network/At_gene_edge.csv\", index=False)\n",
    "df_Dt_edge = build_edge(df, 'Ghir_D', At_gene, Dt_gene)\n",
    "df_Dt_edge.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type_bias/hypocotyl_network/Dt_gene_edge.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty DataFrame\n",
      "Columns: [chrom1, start1, end1, chrom2, start2, end2, tags1, tags2, tissues, tags]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "## 整理保守loops\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/loop_type/J668_loop_type.bed\", sep=\"\\t\")\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.3 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "a0f669ea42a4dcea88d385f8f9b4c3d9b966e24234758831d1dbfdac925619a2"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
