{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将TAD 合并为一个文件\n",
    "import pandas as pd\n",
    "tissues = ['anthers', 'cotyledon', 'hypocotyl', 'leaf', 'petal', 'root', 'stem', 'stigma']\n",
    "df = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD/number/{}_choosed.bed\".format(t)\n",
    "    df_tmp = pd.read_csv(input_f, sep=\"\\t\")\n",
    "    df_tmp['Tissue'] = [t]*df_tmp.shape[0]\n",
    "    df = pd.concat([df, df_tmp], axis=0)\n",
    "df.columns = ['Chrs', 'Start', 'End', 'Order', 'Tissue']\n",
    "df['Size'] = (df['End'] - df['Start'])/1000\n",
    "df = df.query(\"Size<=1000\")\n",
    "df.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD/number/all_tissues_TAD.bed\", sep=\"\\t\", index=False, columns=['Chrs', 'Start', 'End', 'Tissue'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 统计亚组中TAD 的数目\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD/number/all_tissues_TAD.bed\", sep=\"\\t\")\n",
    "df['Subgenome'] = df['Chrs'].apply(lambda x: 'At' if \"Ghir_A\" in str(x) else \"Dt\")\n",
    "df['Number'] = [1]*df.shape[0]\n",
    "df_result = df.groupby(by=[\"Tissue\", 'Subgenome'], as_index=False)['Number'].sum()\n",
    "old_tissues = ['anthers', 'cotyledon', 'hypocotyl', 'leaf', 'petal', 'root', 'stem', 'stigma']\n",
    "new_tissues = ['Anther', 'Cotyledon', 'Hypocotyl', 'Leaf', 'Petal', 'Root', 'Stem', 'Stigma']\n",
    "df_result['Tissue'].replace(old_tissues, new_tissues, inplace=True)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD/number/all_tissues_TAD_number.bed\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 分析亚组TAD 的大小\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD/number/all_tissues_TAD.bed\", sep=\"\\t\")\n",
    "df['Subgenome'] = df['Chrs'].apply(lambda x: 'At' if \"Ghir_A\" in str(x) else \"Dt\")\n",
    "df['Size'] = (df['End'] - df['Start'])/1000\n",
    "df['TAD'] = df['Chrs'].map(str) + \"-\" + df['Start'].map(str) + \"-\" + df['End'].map(str)\n",
    "df = df.query(\"Size<=1000\")\n",
    "old_tissues = ['anthers', 'cotyledon', 'hypocotyl', 'leaf', 'petal', 'root', 'stem', 'stigma']\n",
    "new_tissues = ['Anther', 'Cotyledon', 'Hypocotyl', 'Leaf', 'Petal', 'Root', 'Stem', 'Stigma']\n",
    "df['Tissue'].replace(old_tissues, new_tissues, inplace=True)\n",
    "df.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD/TAD_size/all_tissues_TAD_size.bed\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 统计不同类型的TAD 数目\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/J668_all_Tissues_TAD_type.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'number', 'tags'])\n",
    "df['sigs'] = df['number'].apply(lambda x: 'Structural TAD' if x==0 else (\"Isolated TAD\" if x==1 else \"Enrich TAD\"))\n",
    "df['numbers'] = [1]*df.shape[0]\n",
    "df_result = df.groupby(by=['tags', 'sigs'], as_index=False)['numbers'].sum()\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/J668_all_Tissues_TAD_type_number.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/J668_all_Tissues_TAD_type.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'number', 'tags'])\n",
    "df['size'] = df['end'] - df['start']\n",
    "df['type'] = df['number'].apply(lambda x: 'Structural TAD' if x==0 else (\"Isolated TAD\" if x==1 else \"Enrich TAD\"))\n",
    "df.to_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/J668_all_Tissues_TAD_type.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将不同类型的TAD 区分开来\n",
    "tissues = ['anther', 'cotyledon', 'hypocotyl', 'leaf', 'petal', 'root', 'stem', 'stigma', 'fiber_20DPA', 'fiber_10DPA', 'ovule', 'radicle']\n",
    "types= ['Enrich TAD', 'Isolated TAD']\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/J668_all_Tissues_TAD_type.csv\")\n",
    "for t in tissues:\n",
    "    for ty in types:\n",
    "        if ty == \"Enrich TAD\":\n",
    "            z = \"Erich_TAD\"\n",
    "        if ty == 'Isolated TAD':\n",
    "            z = 'Isolated_TAD'\n",
    "        output_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/data/{0}_{1}.bed\".format(t, z)\n",
    "        df_tmp = df.query(\"tags==@t & type==@ty\").copy()\n",
    "        df_tmp.to_csv(output_f, sep=\"\\t\", columns=['chrom', 'start', 'end'], header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data/J668_all_Tissue_mean_TPM.csv\")\n",
    "df_long = pd.melt(df, id_vars=['GeneId'])\n",
    "df_long.columns = ['GeneId', 'tissues', 'TPM']\n",
    "df_long.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data/J668_all_Tissue_mean_TPM_long.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      tissues ex_tag   TAD_tag  number\n",
      "0      anther    EXP     Erich   38889\n",
      "1      anther    EXP  Isolated     331\n",
      "2      anther  NoEXP     Erich   28585\n",
      "3      anther  NoEXP  Isolated     360\n",
      "4   cotyledon    EXP     Erich   41951\n",
      "5   cotyledon    EXP  Isolated     302\n",
      "6   cotyledon  NoEXP     Erich   25584\n",
      "7   cotyledon  NoEXP  Isolated     302\n",
      "8   hypocotyl    EXP     Erich   45596\n",
      "9   hypocotyl    EXP  Isolated     331\n",
      "10  hypocotyl  NoEXP     Erich   22145\n",
      "11  hypocotyl  NoEXP  Isolated     278\n",
      "12       leaf    EXP     Erich   44135\n",
      "13       leaf    EXP  Isolated     297\n",
      "14       leaf  NoEXP     Erich   23340\n",
      "15       leaf  NoEXP  Isolated     318\n",
      "16       root    EXP     Erich   46018\n",
      "17       root    EXP  Isolated     317\n",
      "18       root  NoEXP     Erich   21394\n",
      "19       root  NoEXP  Isolated     282\n",
      "20     stigma    EXP     Erich   40689\n",
      "21     stigma    EXP  Isolated     322\n",
      "22     stigma  NoEXP     Erich   26795\n",
      "23     stigma  NoEXP  Isolated     343\n"
     ]
    }
   ],
   "source": [
    "## 对于基因打上标签, 分析处于\n",
    "import pandas as pd\n",
    "df_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data/J668_all_Tissue_mean_TPM_long.csv\")\n",
    "df_gene['ex_tag'] = df_gene['TPM'].apply(lambda x: \"EXP\" if x>=0.5 else \"NoEXP\")\n",
    "df_tad_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD/type/data/gene_TAD_type.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'GeneId', 'tissues', 'TAD_tag'])\n",
    "df_gene = df_gene[['GeneId', 'tissues', 'ex_tag']].copy()\n",
    "df_tad_gene = df_tad_gene[['GeneId', 'tissues', 'TAD_tag']]\n",
    "df_tad_gene.drop_duplicates(keep='first', subset=['GeneId', 'tissues'])\n",
    "df_result = pd.merge(df_gene, df_tad_gene, how='inner', on=['GeneId', 'tissues'])\n",
    "df_result['number'] = [1]*df_result.shape[0]\n",
    "df_return = df_result.groupby(by=['tissues', 'ex_tag', 'TAD_tag'], as_index=False)['number'].sum()\n",
    "print(df_return)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "               GeneId tissues   TAD_tag tags  number\n",
      "0     Ghir_A01G000240  anther     Erich  TSG       1\n",
      "1     Ghir_A01G000290  anther     Erich  TSG       1\n",
      "2     Ghir_A01G000320  anther     Erich  TSG       1\n",
      "3     Ghir_A01G000350  anther     Erich  TSG       1\n",
      "4     Ghir_A01G000470  anther     Erich  TSG       1\n",
      "...               ...     ...       ...  ...     ...\n",
      "7814  Ghir_D13G025630  stigma     Erich  TSG       1\n",
      "7815  Ghir_A02G008660  stigma  Isolated  TSG       1\n",
      "7816  Ghir_A03G012990  stigma  Isolated  TSG       1\n",
      "7817  Ghir_A10G016630  stigma  Isolated  TSG       1\n",
      "7818  Ghir_D02G012910  stigma  Isolated  TSG       1\n",
      "\n",
      "[7819 rows x 5 columns]\n",
      "      tissues   TAD_tag tags  number\n",
      "0      anther     Erich  TSG    2873\n",
      "1      anther  Isolated  TSG      37\n",
      "2   cotyledon     Erich  TSG     828\n",
      "3   cotyledon  Isolated  TSG       8\n",
      "4   hypocotyl     Erich  TSG     682\n",
      "5   hypocotyl  Isolated  TSG       4\n",
      "6        leaf     Erich  TSG     906\n",
      "7        leaf  Isolated  TSG       8\n",
      "8        root     Erich  TSG    1762\n",
      "9        root  Isolated  TSG      10\n",
      "10     stigma     Erich  TSG     697\n",
      "11     stigma  Isolated  TSG       4\n"
     ]
    }
   ],
   "source": [
    "## 分析组织特异性表达基因是否会偏向分布于Isolated TAD 中\n",
    "import pandas as pd\n",
    "df_tad_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD/type/data/gene_TAD_type.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'GeneId', 'tissues', 'TAD_tag'])\n",
    "df_tad_gene = df_tad_gene[['GeneId', 'tissues', 'TAD_tag']]\n",
    "df_TSG = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/TAU/J668_all_Tissue_TAU_geneid.csv\")\n",
    "#df_TSG = df_TSG[['GeneId']].copy()\n",
    "df_TSG.drop_duplicates(keep='first', inplace=True, subset=['GeneId'])\n",
    "df_TSG['tags'] = ['TSG']*df_TSG.shape[0]\n",
    "df_result = pd.merge(df_tad_gene, df_TSG, how='inner', on=['GeneId', 'tissues'])\n",
    "df_result['number'] = [1]*df_result.shape[0]\n",
    "print(df_result)\n",
    "df_return = df_result.groupby(by=['tissues', 'TAD_tag', 'tags'], as_index=False)['number'].sum()\n",
    "print(df_return)\n",
    "#print(df_gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      tissues   TAD_tag  number\n",
      "0      anther     Erich   67474\n",
      "1      anther  Isolated     691\n",
      "2   cotyledon     Erich   67535\n",
      "3   cotyledon  Isolated     604\n",
      "4   hypocotyl     Erich   67741\n",
      "5   hypocotyl  Isolated     609\n",
      "6        leaf     Erich   67475\n",
      "7        leaf  Isolated     615\n",
      "8        root     Erich   67412\n",
      "9        root  Isolated     599\n",
      "10     stigma     Erich   67484\n",
      "11     stigma  Isolated     665\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df_tad_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD/type/data/gene_TAD_type.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'GeneId', 'tissues', 'TAD_tag'])\n",
    "df_tad_gene = df_tad_gene[['GeneId', 'tissues', 'TAD_tag']]\n",
    "df_tad_gene['number'] = [1]*df_tad_gene.shape[0]\n",
    "df_result = df_tad_gene.groupby(by=['tissues', 'TAD_tag'], as_index=False)['number'].sum()\n",
    "print(df_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                 GeneId    tissues TAD_tag       TAU\n",
      "0       Ghir_A01G000010     anther   Erich  0.730716\n",
      "1       Ghir_A01G000010  cotyledon   Erich  0.730716\n",
      "2       Ghir_A01G000010  hypocotyl   Erich  0.730716\n",
      "3       Ghir_A01G000010       leaf   Erich  0.730716\n",
      "4       Ghir_A01G000010       root   Erich  0.730716\n",
      "...                 ...        ...     ...       ...\n",
      "408899  Ghir_D13G007790       root   Erich  0.850958\n",
      "408900  Ghir_D13G007800       root   Erich  0.836236\n",
      "408901  Ghir_D13G007810       root   Erich  0.999919\n",
      "408902  Ghir_D13G007820       root   Erich  0.448970\n",
      "408903  Ghir_D06G011040     stigma   Erich  0.247399\n",
      "\n",
      "[408904 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "## 分析Isolated gene 是否具备有更高的组织特异化程度\n",
    "import pandas as pd\n",
    "df_tad_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD/type/data/gene_TAD_type.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'GeneId', 'tissues', 'TAD_tag'])\n",
    "df_tad_gene = df_tad_gene[['GeneId', 'tissues', 'TAD_tag']]\n",
    "\n",
    "df_tau = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/TAU/J668_all_tissue_TAU.csv\")\n",
    "\n",
    "df_tau = df_tau[['GeneId', 'TAU']].copy()\n",
    "df_result = pd.merge(df_tad_gene, df_tau, how='inner', on=['GeneId'])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:xlabel='tissues', ylabel='TAU'>"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAEGCAYAAABo25JHAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAAsTAAALEwEAmpwYAAAig0lEQVR4nO3deXxU9fX/8ddJSNiXIogWjFLBql9XjLjUFTegKiCiqFVLVUQt4tLWtWpr3erSUlwo+qPW1oKtoqWKosWtVpRNXMAtRZYEwQVRIGxJzu+Pe4OTZCaZIXOTkPt+Ph48mLn3cz9zPnMn855778y95u6IiEh85TR2ASIi0rgUBCIiMacgEBGJOQWBiEjMKQhERGKuRWMXkKkuXbr4Lrvs0thliIhsU+bOnfuFu3dNNm+bC4JddtmFOXPmNHYZIiLbFDNbkmqedg2JiMScgkBEJOYUBCIiMacgEBGJOQWBiEjMRRYEZjbRzD4zs/dSzDcz+4OZFZnZO2bWJ6paREQktSi3CB4G+tcyfwDQO/w3EnggwlpERCSFyH5H4O6vmtkutTQZBDziwXmw3zCzTma2o7t/mo3HHzduHEVFRTWml5SUANC9e/cq03v16sXo0aO3uu9U/Wbad7o0vq0bX5TPWyaieo7jOL7m8tpM1XdDjK8xf1DWHViWcL84nFYjCMxsJMFWAwUFBfV60PXr19dr+YbuN1MaX9Pqt6nU0JzH1xTGBtv2+CzKC9OEWwRPu/teSeY9A9zm7q+F92cAv3D3ubX1WVhY6PX5ZfGYMWMAGDt27Fb30ZD9NpU6NL7oRVlDcx5fUxhblHVkq18zm+vuhcnmNea3hoqBnRLu9wCWN1ItIiKx1ZhBMBU4J/z20MHA19k6PiAiIumL7BiBmU0CjgK6mFkxcCOQB+Du44FpwECgCCgFRkRVi4iIpBblt4bOqGO+A5dE9fgiIpIe/bJYRCTmFAQiIjGnIBARiTkFgYhIzCkIRERiTkEgIhJzCgIRkZhTEIiIxJyCQEQk5hQEIiIxpyAQEYk5BYGISMwpCEREYk5BICIScwoCEZGYUxCIiMScgkBEJOYUBCIiMacgEBGJOQWBiEjMKQhERGJOQSAiEnMKAhGRmFMQiIjEnIJARCTmFAQiIjGnIBARiTkFgYhIzCkIRERiTkEgIhJzCgIRkZiLNAjMrL+ZfWhmRWZ2dZL5Hc3sX2b2tpktMLMRUdYjIiI1tYiqYzPLBe4DjgOKgdlmNtXdFyY0uwRY6O4nmVlX4EMze9TdN2XyWOPGjaOoqCittpXtxowZk1b7Xr16MXr06EzKybqoxldSUgJA9+7ds9ovZPa8RTW+5rzuQONrCM19fJUiCwKgL1Dk7osAzGwyMAhIDAIH2puZAe2AVUBZpg9UVFTE/Pfep7xN5zrb5mxyAOYuWlln29y1n1FUVJTWCyHKF0Fk41vzJW1blLOx7NM62+ZvDjYeNy6ZU2fbpWtz62yTKIrx5ZauyqiGqES27jS+BtHcx1cpyiDoDixLuF8MHFStzb3AVGA50B443d0rqndkZiOBkQAFBQVJH6y8TWfW7z6w/lUnaDfvL/jGNWm9+UX5RgnRja+g3Sau7fNNVvu9dV6HjJfJ9vhafzAta33VVxTrTuNrOM19fBBtEFiSaV7t/gnAfKAfsCvwgpn9x92rvDO5+wRgAkBhYWH1PiJV0K68SbxRiohEJcogKAZ2Srjfg+CTf6IRwO3u7kCRmX0C7A7MirAuiYmSkpK0d9Vti/t3m/v4pOFEGQSzgd5m1hMoAYYDZ1ZrsxQ4BviPmXUDvg8sirAmiZH169fz8YK3KGhXXmfbqHftRaG5jy/doFPI1V9kQeDuZWb2U2A6kAtMdPcFZjYqnD8euBl42MzeJdiVdJW7fxFVTRI/zX3XXnMeX7pBty2GXFMT5RYB7j4NmFZt2viE28uB46OsQUS2XdkOuqYScjkbvqGoaE2T2eKJNAhERKQmq9iMb9zYJL6RCAoCEZFG0ZR26+lcQyIiMacgEBGJOQWBiEjMKQhERGJOQSAiEnMKAhGRmFMQiIjEnIJARCTmFAQiIjGnIBARiTkFgYhIzCkIRERiTkEgIhJzCgIRkZhTEIiIxJyCQEQk5hQEIiIxpyAQEYk5BYGISMwpCEREYk5BICIScwoCEZGYUxCIiMScgkBEJOYUBCIiMacgEBGJOQWBiEjMtWjsAkRE0lGR357S/S+gvMNOgG1VHz86Mgdrkc/777+fVvvTTjuNwWXleF7rrXq8VOyH15Fnztf5FVnt96wjjNVr1rF582by8vLSXi7SIDCz/sBYIBd4yN1vT9LmKOD3QB7whbsfGWVNIrJtKt3/Ajr33JdObfMx27ogWFGai+W3oaCgIK32S5cupXTjZipaddyqx0slp3QVrXMr2KFNeVb7/XRdDl3LjeLiYnr27Jn2cpEFgZnlAvcBxwHFwGwzm+ruCxPadALuB/q7+1Iz2z6qekRk21beYad6hUAcmBmtW7emtLQ0o+WiPEbQFyhy90XuvgmYDAyq1uZMYIq7LwVw988irEdEtmmmEEjD1jxHUQZBd2BZwv3icFqi3YDvmNnLZjbXzM5J1pGZjTSzOWY25/PPP4+oXBGReIryGEGyWPIkj38AcAzQGphpZm+4+0dVFnKfAEwAKCwsrN6HiEhavvrqK8768VDy8vJYsWIFubm5dO3aFYDp06fTo0cP7r33Xi688MItyxxz9JG0bdcBgPLyco4d8EMuuvRK8lu2TPoYJcuW8tbc2Zw4eGj0A8qSKLcIioGdEu73AJYnafOcu69z9y+AV4F9I6xJRGLsO9/5Ds8++yzz589n1KhRXH755cyfP5/58+fzxBNPcPDBBzNp0qQay/3psSd56oVXmfyv6RQvXcKNV1+R8jFKipcx7aknohxG1kUZBLOB3mbW08zygeHA1Gpt/gkcbmYtzKwNcBCQ3ve6RESyaNKkSdx9990UFxdTUlKStE3btu248da7ePH5Z1m9+qukbX53+83Mnf0Gp/Q/ij8/NJ6SZUs5e+iJnDqwH6cO7Mdbc2YBUFFRwaXX3Mz+Rw9iyDkXM+jsi5jy9PORja82kQWBu5cBPwWmE7y5/93dF5jZKDMbFbZ5H3gOeAeYRfAV0/eiqklEJJlly5axYsUK+vbty2mnncZjjz2Wsm279u3pvlMBSz9ZlHT+5Vf/kgMOPJgpz73MueePonOXLjz06OM8Pu1F7r7vQW678VoAnpv+AkuKlzN3xpM8cNeveXPu25GMLR2R/o7A3acB06pNG1/t/p3AnVHWISJSm8mTJ3PaaacBMHz4cM477zyuuCL17h/P4Ehl2eYybvnl1Xyw8D1ycnNYsigIkNlz53LKiceTk5PDDtt34chDD6zXGOpDvywWkdibNGkSK1eu5NFHHwVg+fLlfPzxx7RMckB43dq1LC9eys49d02r70ceGs92XbsyZfrLVFRU0Kd3DyCzMImazjUkIrH24Ycfsm7dOkpKSli8eDGLFy/mmmuuYfLkyTXarlu3lpuv+wX9jh9Ax06dkvbXtm071q1bu+X+mjXf0HX7buTk5PCvKX+nvDz4NfGBhX146pkXqKioYOXnX/DqzNmRjC8d2iIQkVibNGkSQ4YMqTJt6NChDB8+nHPPPReAEacPwd2pqKjg2P4DGXXplSn7222PPcnNbcGQE45i8LDhnHHOCC678CdMf2YqfQ/5Aa3btAFgwAnH89as1+nTbzC9v7cLB+6/Dx06tItuoLVQEIhILN10000p5+2zzz4sXLiQpUuXMuOlVzI611BeXh4TJ0+pMu3J51/Zcvvyq38JpavIycnh9ht+Tru2bfhy1WoOO3E4e+2+W8bjyAYFgYhIIxly7sV8/fUaNm3ezDVjRrHD9l0apQ4FgYjIVvjog4VcfdnFVabl57dk8tTpaffxwuMPZ7mqraMgEBHZCrvtvidTnnu5scvIipRBYGbjqHpuIAe+AF5y99eiLkxERBpGbVsEc5JM6wzcaWaPufvvoylJREQaUsogcPc/J5tuZuOB1wmuKiYiItu4jI8RuPt6XRxCROJo71260Xv3PbbcH3DSEC64ZEyNduPuvp3CvodwyOHJr7z75JQpfLDwXSbccW1ktWYioyAwsxbA2QSnjxYRaTRX3HI/n61ek9Ey5RWA5dCiRW6Nedt36cy999R+2rOWrVrVeYC4vLyc0VdenVFdja22g8VrqHkhmfXAK8CFNZcQEWk4n61ew6LvHpu9Dle+UnebFI47tA9DTjuT1199iTN/fD6vvTyDI485nhN+eDLvvv0Wt990HaWlpeTn5zNxUvBjs5UrP+Oksy5k0eJlDBpwDLden/rXylGr7RhB+4YsRESkqdu4YQOn9D9qy/0LLh7DgJOD01O0bNmSv055BoDXXp4BwKZNm/jZxRdw1/0Psve++7N2zRpatmoFwML3P2DOC/+gZX4+ex9xIheNOJOduu/YsAMKZbpraFfgDGC4u+8VTUkiIk1TbbuGBpw0uMa0xYuK6LL99uy97/5AcC2DSj849GA6dgju77Hbriwt+bTRgqDOs4+a2Y5mdrmZzQIWALkEYSAiIqHKk8klcndSfbkmPz9/y+3cnBzKysoiq60uKYPAzC4wsxcJjglsB5wPfOruv3L3dxuqQBGRbVXPXXvz+cqVvPv2W0BwLYPGfMNPpbZdQ/cBM4Ez3X0OgJk1oUspiIg0rOrHCA47sh9XXHNDyvb5+fncdf+D3HrDNWzYsIFWrVrx0N8eb4BKM1NbEPQAhgL3mFk34O9AXoNUJSJSh+07tYfl/85ombq+PlqXdxevTDr9hdfnVbl/6z33brm99777M+mfz1WZP+SUU2g9bDAQXKTmyUfur/Oxo1RbEDzn7n2AB8ysBzAc+MzM3geedPem8UsIoKSkhNzSr2n9wbS6G2eivIyVpY1/ETeNL3O5pV+y0bzRP7pEte40PrjnuotTz0xhRWkult+GgoKCtNpv3rwZKsrJ2fB1xo9VO2dzRZa7rIfa3gW2HOFw92J3v8vdDwAGARsjr0xERBpEbVsEXc3sihTzMvs5X8S6d+/Oio0tWL/7wKz2227eX+jWZlNW+9waGl/mWn8wjXYVawh+A9l4olp3Gl/DyMvLY3MFGV2hLB05pavIy2k6h1xrC4JcoB0JWwYJms4IRESkXmoLgk/d/dcNVomIiDSKtI4RiIhI81VbEBzTYFWIiGwDCnffeauWO+7QPny16sta29zxhwkZ9/vIY09x2XW3bFVNiWo76dyqevcuIhKRm269i29WZ/Y2Ve4GZuTm1nzr69ilG7fefW+SpRrGb8c9yFWXjmyUx9bF60Vkm/TN6lVctduirPV3x//Sb/v5yhVceckFrF27hvKycm645bcccNAhPPPPKTx47+9xd47odxxXXlvzV8ejzz+HFcVL2bxpA2Mu+BHn/2gY19/6O9Zv2Ejf44ayx/d78ed77+BvT/yL+yc+yqZNmzlw/334w23Xk5uby58fe5I7xz3EDt260vt7O9My4ZxFW0tBUJuKcpasyeXWeR2y2u2SNbm0LSnJap8i0nCe+ecUfnDk0Vw4+grKy8vZsH49n61YwT23/Zp/PPNvOnTsxAU/GsaM6dM45oSqX629+a6xdM53bHMpp5w6jCEDj+M3117OA3/6G7NeeAKADz7+H49PfY6XnvoLeXl5XHrNzUya8jTHHHEoN991HzOf+zsd27fn+GEj2G+vPZKVmBEFgYhIhvbadz+u/9kYyjaX0e+EAezxf3vz5uv/oe/BP6Dzdl0AOHHwUOa8ObNGEDw68UFmPDsVM1i+fAVFnyxhu86dqrR56bU3eevdhfxg4HAA1m/YSNcunZn91jsccciBdN0uOB3GsJP78/GiJfUej4KgNjm57Nx+I9f2+Sar3d46rwMtu3fPap8i0nAKDzqUR/7xL1558QWuuewSRlx4SZVrDaQya+Z/mfnaK0ya/Bid27Xk7HPOZsPGmidqcHfOGnYyv7nm8irTpz43I+Vpresj0iAws/7AWIIfpz3k7renaHcg8AZwurs3vVPzNVfa9SWyVZYXL2P7HXZk2Jlns750HQvfe4fzLxrNbTddx1ervqRDx05M++eTnDXi/CrLrV3zDR06dqJ169YU/a+IWfPe2TIvLy+PzZs3k5eXx9GHHcypI0Zz6QXnsH2X7Vj11desWbeOA/ffhytvuJ0vV62mQ/u2PPH08+yz5/frPZ7IgsDMcglOZX0cwcXuZ5vZVHdfmKTdHcD0qGoREcmmWTP/y5/+eB8t8lrQpk1bbvvdfXTttgOXXXU9I04fgrtz+NHH0u/4AVWWO+zIfjz214cZdPJJ9PpeT/r22WfLvPPOOpXCY09hv7335M/33sFNvxjNiWeMpMIryGuRx+9vuY6DDtiX66+8mCNPPosdunVl/733oLy8/mevi3KLoC9Q5O6LAMxsMsEJ6xZWazcaeAI4MMJaJJlmvutr48aNLNmgLZ7mqkOnztzxUWbL1PX10brM+SDYHz942HAGDxteY/6Jg4dy4uChNaYnnqb6j488Rk7pKlrnVrBDm/It02+57gpuue7b07sNGzSAYYOqBgnAuacP4dzTh9RZayaiDILuwLKE+8XAQYkNzKw7MAToRy1BYGYjgZFA2qePFZHm7aZrf5bxMpmehjouogyCdE5W93vgKncvr+0AiLtPACYAFBYW6oR3kpaWLVuyU976ZrvFI5ItUQZBMbBTwv0ewPJqbQqByWEIdAEGmlmZuz8VYV0izUJz3/VVfXw/OjKHFaU1ryyWUZ/lkLt5czbKa1aiDILZQG8z6wmUEFzh7MzEBu7es/K2mT0MPK0QEJFk3B13j+Trk82Je+Y7TSILAncvM7OfEnwbKBeY6O4LzGxUOH98VI8tEgfNfddX9fGt3biEVt6JTm3ytzoMVpTmYnnN99Lr7s6G9etp1apVRstF+jsCd58GTKs2LWkAuPuPo6xFRLZtbd56kFVcwOcddmJrz5L/9aYcrEU+69atS6v9qlWr2FRWjudl9xyctmkdeeZ8k5/dCxev3misXrOO/v37Z7ScflksItuEnE1raPfmPfXq4755HWi5cyFjx45Nq/2YMWOYu2hl9i8TO+dh2uSUsXP78robZ2DJmlzafmd7TjrppIyWq+16BCIiEgPaIhARaWhN7Mec2iIQEYk5BYGISMwpCEREYk5BICIScwoCEZGYUxCIiMScgkBEJOYUBCIiMacgEBGJOQWBiEjMKQhERGJOQSAiEnMKAhGRmFMQiIjEnIJARCTmFAQiIjGnIBARiTkFgYhIzCkIRERiTkEgIhJzCgIRkZhTEIiIxJyCQEQk5hQEIiIxpyAQEYk5BYGISMwpCEREYi7SIDCz/mb2oZkVmdnVSeafZWbvhP9eN7N9o6xHRERqiiwIzCwXuA8YAOwJnGFme1Zr9glwpLvvA9wMTIiqHhERSa5FhH33BYrcfRGAmU0GBgELKxu4++sJ7d8Aemztg+WWrqL1B9PqbJez4RsAKlp1qLvT8rKtLSfrNL5AuuPLLV0FrfKyUlt9RbHuNL6G09z/9iDaIOgOLEu4XwwcVEv784Bnk80ws5HASICCgoIa83v16pV2UUVFa4Jlvtctzbab0u47Khrft9IfXzdKSkqgbPXWF5YFUa07ja9hNPe/vUpRBoElmeZJG5odTRAEhyWb7+4TCHcbFRYW1uhj9OjRaRc1ZswYAMaOHZtW241L5qTdd1Q0vm9lPr5Pt7qubIhqbJXtNb5oNfe/vUpRBkExsFPC/R7A8uqNzGwf4CFggLt/GWE9IiKSRJTfGpoN9DaznmaWDwwHpiY2MLMCYApwtrt/FGEtIiKSQmRbBO5eZmY/BaYDucBEd19gZqPC+eOBG4DtgPvNDKDM3QujqklERGqKctcQ7j4NmFZt2viE2+cD50dZg4iI1E6/LBYRiTkFgYhIzCkIRERiTkEgIhJzCgIRkZhTEIiIxJyCQEQk5hQEIiIxpyAQEYk5BYGISMwpCEREYk5BICIScwoCEZGYUxCIiMScgkBEJOYUBCIiMacgEBGJOQWBiEjMKQhERGJOQSAiEnMKAhGRmFMQiIjEnIJARCTmFAQiIjGnIBARiTkFgYhIzCkIRERiTkEgIhJzCgIRkZhTEIiIxJyCQEQk5iINAjPrb2YfmlmRmV2dZL6Z2R/C+e+YWZ8o6xERkZoiCwIzywXuAwYAewJnmNme1ZoNAHqH/0YCD0RVj4iIJNciwr77AkXuvgjAzCYDg4CFCW0GAY+4uwNvmFknM9vR3T+t74OPGzeOoqKiGtMrp40ZM6bK9F69ejF69Oga7ZeuzeXWeR2qTFtZmsOGcku7lla5Trc2FVX67J320slpfFs3vvqOrbLPKMaXamyg8VVqzq9NaLzxRRkE3YFlCfeLgYPSaNMdqBIEZjaSYIuBgoKCehXVunXrtNv26tUr6fTckhJy1q9Pu5/c1q1p2b37lvu9a+m7vjS+byWrob5jg+jGl8nYSFGDxlet/Tb02oTGG58FH8azz8yGASe4+/nh/bOBvu4+OqHNM8Bt7v5aeH8G8At3n5uq38LCQp8zZ04kNYuINFdmNtfdC5PNi/JgcTGwU8L9HsDyrWgjIiIRijIIZgO9zaynmeUDw4Gp1dpMBc4Jvz10MPB1No4PiIhI+iI7RuDuZWb2U2A6kAtMdPcFZjYqnD8emAYMBIqAUmBEVPWIiEhyUR4sxt2nEbzZJ04bn3DbgUuirEFERGqnXxaLiMScgkBEJOYUBCIiMacgEBGJuch+UBYVM/scWNKAD9kF+KIBH6+haXzbruY8NtD4sm1nd++abMY2FwQNzczmpPo1XnOg8W27mvPYQONrSNo1JCIScwoCEZGYUxDUbUJjFxAxjW/b1ZzHBhpfg9ExAhGRmNMWgYhIzCkIRERiTkFQjZkNTry2spm9bGZN4itedaleey3tMhqTmR1lZk/Xr7qUfe9iZu9F0XemwkulXpxGu8iej2qPs7Yeyw4zs/fN7KVs1tQUhK+ZMxu7jmTM7DIza5Nwf5qZdWrEktKiIKhpMFDnm2k6zCw3G/1kYDBZqj2mOgF1BsE24jzgYnc/urELqUt4PZJM3ot2AZpkEACXAVuCwN0HuvvqRqsmTbEIAjN7yszmmtmC8PrHmNlaM7vFzN42szfMrJuZHQqcDNxpZvPNbNewi2FmNsvMPjKzw8Plc83sTjObbWbvmNmF4fSjzOwlM/sb8G6W6j8nfIy3zewvZrazmc0Ip80ws4JktZvZvIQ+eptZjUuAmtnxZjbTzOaZ2T/MrF04vb+ZfWBmrwGnJLTvHD6f74TP2z7h9JvMbGK4tbHIzC7NYIi5ZvZguH6eN7P/S1W7mS02szvC9THLzHqF02s8J+H0bmb2ZPjcvR0+T5jZFWb2XvjvsvChbgd2DZ+/O8PnelBCHY+a2ckZjCtrzOznCa+1XyVMT/bavgE4DBhvZnc2Rr11CT/Vv29m9wPzgP8Xrot3zez0sI2F66HKdIL1dHi4ni5vxDG0NbNnwtfVe2Z2I/Bd4CULt8TC12uX8PYvw7+pF8xskpn9LJz+spn9zsxeDZ+TA81sipl9bGa/SXi8Gus6a9y92f8DOof/twbeA7YDHDgpnP5b4Prw9sPAqQnLvgzcHd4eCPw7vD0yYZmWwBygJ3AUsA7omaXa/w/4EOhSORbgX8C54f2fAE+lqP0lYL/w9q3A6IQxFRL8xP1VoG04/SrgBqAVsIzgWtgG/B14OmwzDrgxvN0PmB/evgl4PXwuugBfAnlpjG8XoCyhzr8DP6ql9sXAdeHtcxLqSvWcPAZcFt7OBToCBxCEdFugHbAA2D+s5b2E2o5M6Kcj8AnBNTyOqnzciF+3a8P/jyf4qqERfHh7Gjgi1Ws7cR039t9eHeu9AjgYGAq8EK6fbsBSYMdapjfI85/GGIYCDybc7xi+PrskTFsc/j0UAvPD9dQe+Bj4WcK6uiO8PYbgcr07hn9LxQnrNOm6zsa/WGwRAJea2dvAGwTXSO4NbCL4gwKYS/DCTGVKknbHE1xmcz7wJkG49A7nzXL3T7JUez/gcXf/AsDdVwGHAH8L5/+F4NNfMg8BIyzYRXV6wjKVDibYlfTfcBznAjsDuwOfuPvHHrzy/pqwzGHhY+LuLwLbmVnHcN4z7r4xrPUzgj/edHzi7vPD25XPcW21T0r4/5DwdqrnpB/wQFhvubt/Hc570t3XuftagvV7ePWi3P0VoJeZbQ+cATzh7mVpjimbjg//vUXw6Xl3vn2tJXttbyuWuPsbBOtjUrh+VgKvAAfWMr2peBc4NtxCPTx8baVyGPBPd1/v7msIPrgkqryM77vAAnf/1N03Aov49rruka3rSK9Q1hSY2VHAscAh7l5qZi8TfOLdHL7JAZRT+3OxMUk7I/iUOj3J463LQulbuiTYeqlNqvlPADcCLwJz3f3LJH2/4O5nVJlotl8tfVotj78xYVpdz2mi6su1pvbaPcXtZDUlk2wMqfwFOIvgmts/yWC5bDLgNnf/Y5WJqV/b24rKv5NU6yOT9dTg3P0jMzuAYE/BbWb2fC3N6xpL5d9ABVX/HiqAFlGv6zhsEXQEvgqfvN0JPgXXZg3BpltdpgMXmVkegJntZmZt61dqUjOA08xsu/BxOhPsghkezj8LeC28XaV2d98Q1vkA8Kckfb8B/CBhP3sbM9sN+ADoad8eI0kMilfDx6x8I/rC3b+p3xBrqqP20xP+nxneTvWczAAuCuvNNbMO4RgGh+NtCwwB/kPydf8wwQFA3H1Bfce1laYDP7Fvj990D7dSMn1tN1WvAqeH66crcAQwq5bp6f6NRsrMvguUuvtfgbuAPqSu7TXgJDNrFa7HH2b4cJGu62a/RQA8B4wys3cI9rW/UUf7ycCDFhzsPLWWdg8R7MKYZ2YGfE7wrZ2scvcFZnYL8IqZlRPsHrgUmGhmPw8fd0Sy2t39f8CjBAd7a3xacffPzezHwCQzaxlOvj78pDMSeMbMviB4Ee8Vzr8J+FP4fJYS7E6KSqraW5rZmwQfZCpDKtVzMgaYYGbnEWxtXOTuM83sYYI3FYCH3P0tADP7rwVfZ33W3X/u7ivN7H3gqUhGmAZ3f97M9gBmBi811hIcR8n0td1UPUmwa+9tgi25X7j7CjNLNf1LoCzcTfKwu/+ukerem+DLGRXAZoIPHIcAz5rZp57wjS13n21mUwnGsoTgmGJtu5Kqi3Rd6xQTzVz4zYSO7v7Lxq4lU8lqN7PFBAdBG+Q87hZ8J/xdoE8d+4BFamVm7dx9bfiaehUY6e7z6lquIcRhiyC2wk9UuxIcMN2mNIXazexYYCJwj0JAsmCCBT/4bAX8uamEAGiLQEQk9uJwsFhERGqhIBARiTkFgYhIzCkIRKh65lEz+66ZPd7YNYk0FB0sFiE4CRrB+Wv2qqutSHOjLQKRQOKZR/8R/qgMC86EOiuc/o4FZ0KtftbJyrNlJp5psjA8DUDlWSonWnD20LcsPKNpsr4bZ+gSd/odgUjgamAvd9+vcusgnD4KGOvuj5pZPsGZMAcCy939hwAJJ91L5TrgRXf/iQUXKZllZv9O0bdIg9MWgUjtZgLXmtlVwM7uvp7MzjoJwZlDr7bgDK8vE/ygqCBF3yINTkEgUgt3/xvBBX/WA9PNrJ+7f8S31zS4zYILwUBwXYXKv6nEM0MaMNTd9wv/Fbj7+8n6bogxiVSnIBAJJD1rpJl9D1jk7n8gOGf8PinOOgnBRUgOCG8PTehmOjA6PDkhZrZ/qr6zPSiRdCgIRIDwegeVZx5NvLzj6cB74W6d3YFHCM46OSucdh1QeTnBXwFjzew/BGc6rXQzkAe8E/Z/cy19izQ4fX1URCTmtEUgIhJzCgIRkZhTEIiIxJyCQEQk5hQEIiIxpyAQEYk5BYGISMz9f4wzw6GaahJLAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "sns.boxplot(x=df_result['tissues'], y=df_result['TAU'], hue=df_result['TAD_tag'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\pc\\AppData\\Local\\Temp\\ipykernel_29592\\4185820667.py:11: FutureWarning: As the xlwt package is no longer maintained, the xlwt engine will be removed in a future version of pandas. This is the only engine in pandas that supports writing in the xls format. Install openpyxl and write to an xlsx file instead. You can set the option io.excel.xls.writer to 'xlwt' to silence this warning. While this option is deprecated and will also raise a warning, it can be globally set and the warning suppressed.\n",
      "  df_result.to_excel(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/mutate/SML_mutate_isolation_TAD.xls\", index=False)\n"
     ]
    }
   ],
   "source": [
    "# 筛选突变体库中的基因是否有位于孤岛基因 (孙梦玲)\n",
    "import pandas as pd\n",
    "df_type_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/data/gene_TAD_type.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'GeneId', 'tissue', 'tags'])\n",
    "df_Isolate_geneid = df_type_gene.query(\"tags=='Isolated'\")[['GeneId']]\n",
    "df_Isolate_geneid.drop_duplicates(keep='first', inplace=True)\n",
    "df_mutate = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/mutants/Mutate_gene.txt\", sep=\"\\t\")\n",
    "df_result = pd.merge(df_Isolate_geneid, df_mutate, how='inner', on=['GeneId'])\n",
    "df_geneid = df_result[['GeneId']].copy()\n",
    "df_gene_TPM = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data/J668_all_Tissue_mean_TPM.csv\")\n",
    "df_result= pd.merge(df_geneid, df_gene_TPM, how='left', on=['GeneId'])\n",
    "df_result.to_excel(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/mutate/SML_mutate_isolation_TAD.xls\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "            GeneId     anther      ovule       root    radicle       leaf  \\\n",
      "0  Ghir_A03G011140   0.799356  29.210959   2.568919  10.288888   6.496146   \n",
      "1  Ghir_A08G010820   0.513364  12.432453   0.726312   0.407916   2.900222   \n",
      "2  Ghir_A11G021210   9.987258   5.531542  45.103682  13.422412  12.830680   \n",
      "3  Ghir_A12G007670  18.389610  13.351511   3.523124   0.823820  65.902990   \n",
      "4  Ghir_D11G019890   0.010051   0.000000   0.000000   0.013704   0.028747   \n",
      "5  Ghir_D11G023270   3.149479   2.403123  46.567595  11.965083  11.037861   \n",
      "6  Ghir_A02G011470  12.342137   8.668066   9.072853   8.614013   4.424294   \n",
      "7  Ghir_A08G008890   0.018365  14.070554   4.155479   1.869885   2.081487   \n",
      "\n",
      "   hypocotyl  cotyledon       stem  fiber_10DPA  fiber_20DPA      petal  \\\n",
      "0  88.615934   5.616829  39.395852     4.118404     3.789097   5.334714   \n",
      "1   2.526649   4.777033   2.173620    22.874353     8.499842   7.479017   \n",
      "2  25.511529  11.533034  28.264296     1.789693     1.662166  64.803886   \n",
      "3  17.387577  46.337480  23.217922    10.253961     6.989545  15.510372   \n",
      "4   0.014874   0.000000   0.000000     0.000000     0.000000   0.000000   \n",
      "5  17.679435  14.270499  24.690523     0.742800     1.432173  72.603335   \n",
      "6   7.362339   5.897661   8.060838    13.876722    15.218599  10.228297   \n",
      "7   3.244210   0.220845   1.706808     0.053123     0.107170   0.011386   \n",
      "\n",
      "      stigma  \n",
      "0  10.701518  \n",
      "1  21.722255  \n",
      "2  19.479849  \n",
      "3  50.989784  \n",
      "4   0.000000  \n",
      "5  15.267159  \n",
      "6  21.244363  \n",
      "7   0.012818  \n"
     ]
    }
   ],
   "source": [
    "## 筛选突变体库中是否有孤岛基因 (张肖)\n",
    "import pandas as pd\n",
    "df_type_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/data/gene_TAD_type.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'GeneId', 'tissue', 'tags'])\n",
    "df_Isolate_geneid = df_type_gene.query(\"tags=='Isolated'\")[['GeneId']]\n",
    "df_Isolate_geneid.drop_duplicates(keep='first', inplace=True)\n",
    "df_mutate = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/mutants/zhangxiao_TF.csv\")\n",
    "df_mutate_id = df_mutate[['GeneId']].copy()\n",
    "df_result = pd.merge(df_Isolate_geneid, df_mutate_id, how='inner', on=['GeneId'])\n",
    "df_gene_TPM = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data/J668_all_Tissue_mean_TPM.csv\")\n",
    "df_result = pd.merge(df_result, df_gene_TPM, how='inner', on=['GeneId'])\n",
    "print(df_result)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/mutate/zhangxiao_mutate_islation_TAD.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 分析不同类型TAD所包含基因其在不同组织中可表达的分布\n",
    "import pandas as pd\n",
    "df_type_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/data/gene_TAD_type.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'GeneId', 'tissue', 'tags'])\n",
    "#df_hypocotyl = df_type_gene.query(\"tissue=='hypocotyl'\")[['GeneId', 'tags']].copy()\n",
    "df_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data/J668_all_Tissue_mean_TPM.csv\", index_col=0)\n",
    "df_gene[df_gene>=0.5] = 1\n",
    "df_gene[df_gene<0.5] = 0\n",
    "df_gene['sum'] = df_gene.sum(axis=1)\n",
    "df_gene['number'] = [1]*df_gene.shape[0]\n",
    "df_gene.reset_index(inplace=True)\n",
    "df_allTissues_gene = pd.merge(df_type_gene, df_gene, how='left', on=['GeneId'])\n",
    "df_allTissues_statistics = df_allTissues_gene.groupby(by=['tissue', 'tags', 'sum'], as_index=False)['number'].sum()\n",
    "df_allTissues_statistics['sigs'] = df_allTissues_statistics['sum'].apply(lambda x: \"Silence gene\" if x==0 else (\"Housekeeping gene\" if x==12 else \"Other gene\"))\n",
    "df_result = df_allTissues_statistics.groupby(by=['tissue', 'tags', 'sigs'], as_index=False)['number'].sum()\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/J668_allTissue_TEG_statistics.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 确定组织功能性TAD区间\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "def TAD_region(df):\n",
    "    df[['chrom', 'start', 'end']] = df['tad_tags'].str.split(\"-\", expand =True)\n",
    "    df_return = df[['chrom', 'start', 'end']].copy()\n",
    "    df_return.drop_duplicates(keep='first', inplace=True)\n",
    "    return df_return\n",
    "tissues = ['anther', 'cotyledon', 'hypocotyl', 'leaf', 'petal', 'root', 'stem', 'stigma', 'fiber_10DPA', 'fiber_20DPA', 'ovule', 'radicle']\n",
    "for t in tissues:\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/{}_TAD_TSG_number.csv\".format(t)\n",
    "    output_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TSG_TAD_region/{}_TAD_TSG_region.bed\".format(t)\n",
    "    df_tmp = pd.read_csv(input_f)\n",
    "    df_True = df_tmp[df_tmp['signif_tags']].copy()\n",
    "    df_result = TAD_region(df_True)\n",
    "    df_result.to_csv(output_f, sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Nutures\n",
    "## 20221123\n",
    "## 分析所包含的矩阵互作\n",
    "\n",
    "import pandas as pd\n",
    "import sys\n",
    "import argparse\n",
    "def main(argv):\n",
    "    parser = argparse.ArgumentParser(description=\"\")\n",
    "    parser.add_argument('-tad', \"--TADfile\", required=True, help=\"TAD indefified by TADLib\")\n",
    "    parser.add_argument('-mtx', \"--matrixfile\", required=True, help=\"Intrachromosomal interaction matrix file generated by HiC-Pro\")\n",
    "    parser.add_argument('-bed', \"--bedfile\", required=True, help=\"Bedfile of Intrachromosomal interaction matrix\")\n",
    "    parser.add_argument('-gene', \"--genefile\", required=True, help=\"The bin file contain gene\")\n",
    "    parser.add_argument('-o', \"--outputfile1\", required=True, help=\"output file that intra-TAD interaction strength\")\n",
    "    args = parser.parse_args()\n",
    "    tad_f = args.TADfile\n",
    "    mtx_f = args.matrixfile\n",
    "    bed_f = args.bedfile\n",
    "    gene_f = args.genefile\n",
    "    out_f = args.outputfile\n",
    "    df_tad = pd.read_csv(tad_f, sep='\\t', names=['chr', 's', 'e'])\n",
    "    df_mtx = pd.read_csv(mtx_f, sep='\\t', names=['bin1', 'bin2', 'freq'])\n",
    "    df_bed = pd.read_csv(bed_f, sep='\\t', names=['chr', 's', 'e', 'bin_id'])\n",
    "    df_gene = pd.read_csv(gene_f, sep=\"\\t\", names=['chr', 's', 'e', 'GeneId'])\n",
    "    ## get TAD bin interaction\n",
    "    df_TAD_interaction = pd.DataFrame()\n",
    "    for index, row in df_tad.iterrows():\n",
    "        c, s, e= row['chr'], row['s'], row['e']\n",
    "        tags = \"{0}-{1}-{2}\".format(c, s, e)\n",
    "        df_bin_id = df_bed.query(\"(chr == @c) & (s>=@s) & (e<=@e)\").copy()\n",
    "        df_bin_id.reset_index(drop=True, inplace=True)\n",
    "        bin_id_s = df_bin_id.loc[0, 'bin_id']\n",
    "        bin_id_e = df_bin_id.loc[df_bin_id.shape[0]-1, 'bin_id']\n",
    "        df_need_matrix = df_mtx.query(\"(bin1>=@bin_id_s & bin1<=@bin_id_e) & (bin2>=@bin_id_s & bin2<=@bin_id_e)\").copy()\n",
    "        df_need_matrix['tags'] = [tags]*df_need_matrix.shape[0]\n",
    "        df_TAD_interaction = pd.concat([df_TAD_interaction, df_need_matrix], axis=0)\n",
    "    ##  get gene bin interaction \n",
    "    df_gene_bin = pd.merge(df_gene, df_bed, how='inner', on=['chr', 's', 'e'])[['bin_id']].copy()\n",
    "    df_gene_bin.columns = ['bin1']\n",
    "    df_gene_bin_interaction1 = pd.merge(df_gene_bin, df_TAD_interaction, how='inner', on=['bin1'])\n",
    "    df_gene_bin.columns = ['bin2']\n",
    "    df_gene_bin_interaction2 = pd.merge(df_gene_bin, df_TAD_interaction, how='inner', on=['bin2'])\n",
    "    df_gene_bin_interaction = pd.concat([df_gene_bin_interaction1, df_gene_bin_interaction2], axis=0)\n",
    "    df_gene_bin_interaction.drop_duplicates(keep='first', on=['GeneId'], inplace=True)\n",
    "    \n",
    "    ## analysis the gene bin interaction \n",
    "    df_gene = df_gene_bin_interaction.groupby(by=['tags'], as_index=False)['freq'].mean()\n",
    "    df_TAD = df_TAD_interaction.groupby(by=['tags'], as_index=False)['freq'].mean()\n",
    "    df_gene.columns = ['tags', 'gene_mean_reads']\n",
    "    df_TAD.columns = ['tags', 'tad_mean_reads']\n",
    "    df_result = pd.merge(df_gene, df_TAD, how='inner', on=['tags'])\n",
    "    df_result['ratio'] = df_result['gene_mean_reads']/df_result['tad_mean_reads']\n",
    "    df_result.to_csv(out_f, sep=\"\\t\", index=False)\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    main(sys.argv[1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import pandas as pd\n",
    "tissues = ['anther', 'leaf', 'root', 'stem', 'stigma', 'cotyledon', 'hypocotyl', 'petal', 'fiber_10DPA', 'radicle', 'fiber_20DPA', 'ovule']\n",
    "df = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    inputf = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/{}_TAD_TSG_number.csv\".format(t)\n",
    "    outputf = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/{}_TAD_TSG.csv\".format(t)\n",
    "    df_tmp = pd.read_csv(inputf)\n",
    "    #print(df_tmp.shape[0])\n",
    "    df_tmp = df_tmp[df_tmp['signif_tags']]\n",
    "    df_tmp.drop_duplicates(keep='first', subset=['tad_tags'], inplace=True)\n",
    "    df_tmp[['chrom', 'start', 'end']] = df_tmp['tad_tags'].str.split(\"-\", expand = True)\n",
    "    df_tmp['tissues'] = [t]*df_tmp.shape[0]\n",
    "    df_tmp.to_csv(outputf, columns=['chrom', 'start', 'end'], index=False)\n",
    "    df = pd.concat([df, df_tmp], axis=0)\n",
    "df.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/J668_TSG_TAD.bed\", sep=\"\\t\", index=False, header=False, columns=['chrom', 'start', 'end', 'tissues'])\n",
    "    #"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析组织特异性基因其subcompartment占比是否会有显著性差异\n",
    "import pandas as pd\n",
    "def get_subcompartment_region(df_TAD, df_subcpt, tags1, tags2):\n",
    "    df_TAD_subcpt = pd.DataFrame()\n",
    "    for index, row in df_TAD.iterrows():\n",
    "        C, S, E,  = row['chrom'], row['start'], row['end']\n",
    "        T = \"{0}-{1}-{2}\".format(C, S, E)\n",
    "        df_tmp = df_subcpt.query(\"chrom==@C & start<=@E & end>=@S\").copy()\n",
    "        df_tmp['tad_tags'] = [T]*df_tmp.shape[0]\n",
    "        df_TAD_subcpt = pd.concat([df_TAD_subcpt, df_tmp], axis=0)\n",
    "    df_TAD_subcpt['number'] = [1]*df_TAD_subcpt.shape[0]\n",
    "    df_TAD_subcpt['tags1'] = [tags1]*df_TAD_subcpt.shape[0]\n",
    "    df_TAD_subcpt['tags2'] = [tags2]*df_TAD_subcpt.shape[0]\n",
    "    return df_TAD_subcpt\n",
    "tissues = ['anther', 'leaf', 'root', 'stem', 'stigma', 'cotyledon', 'hypocotyl', 'petal', 'ovule', 'fiber_10DPA', 'fiber_20DPA', 'radicle']\n",
    "df_result = pd.DataFrame()\n",
    "for t1 in tissues:\n",
    "    tad_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/{}_TAD_TSG.csv\".format(t1)\n",
    "    df_TAD = pd.read_csv(tad_f)\n",
    "    for t2 in tissues:\n",
    "        subcpt_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/compartment/CALDER2/subcompartment/{}_subcompartment_new.bed\".format(t2)\n",
    "        df_subcpt = pd.read_csv(subcpt_f, sep=\"\\t\", names=['chrom', 'start', 'end', 'value', 'subcpt', 'cpt'])\n",
    "        df_subcpt = get_subcompartment_region(df_TAD, df_subcpt, t1, t2)\n",
    "        df_result = pd.concat([df_result, df_subcpt], axis=0)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_subcpt/TSG_TAD_subcpt.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "anther leaf\n",
      "158 67\n",
      "anther root\n",
      "151 84\n",
      "anther stem\n",
      "164 67\n",
      "anther stigma\n",
      "150 67\n",
      "anther cotyledon\n",
      "151 68\n",
      "anther hypocotyl\n",
      "139 84\n",
      "anther petal\n",
      "112 98\n",
      "anther fiber_10DPA\n",
      "168 64\n",
      "anther fiber_20DPA\n",
      "182 49\n",
      "anther ovule\n",
      "169 72\n",
      "anther radicle\n",
      "91 135\n",
      "leaf anther\n",
      "57 47\n",
      "leaf root\n",
      "62 33\n",
      "leaf stem\n",
      "76 29\n",
      "leaf stigma\n",
      "63 38\n",
      "leaf cotyledon\n",
      "63 33\n",
      "leaf hypocotyl\n",
      "57 48\n",
      "leaf petal\n",
      "46 46\n",
      "leaf fiber_10DPA\n",
      "70 37\n",
      "leaf fiber_20DPA\n",
      "77 27\n",
      "leaf ovule\n",
      "87 27\n",
      "leaf radicle\n",
      "53 52\n",
      "root anther\n",
      "133 92\n",
      "root leaf\n",
      "172 53\n",
      "root stem\n",
      "164 54\n",
      "root stigma\n",
      "157 64\n",
      "root cotyledon\n",
      "134 96\n",
      "root hypocotyl\n",
      "124 86\n",
      "root petal\n",
      "148 82\n",
      "root fiber_10DPA\n",
      "147 69\n",
      "root fiber_20DPA\n",
      "166 58\n",
      "root ovule\n",
      "184 47\n",
      "root radicle\n",
      "88 132\n",
      "stem anther\n",
      "32 48\n",
      "stem leaf\n",
      "51 31\n",
      "stem root\n",
      "37 43\n",
      "stem stigma\n",
      "38 36\n",
      "stem cotyledon\n",
      "43 38\n",
      "stem hypocotyl\n",
      "20 49\n",
      "stem petal\n",
      "33 44\n",
      "stem fiber_10DPA\n",
      "33 42\n",
      "stem fiber_20DPA\n",
      "46 34\n",
      "stem ovule\n",
      "58 34\n",
      "stem radicle\n",
      "23 52\n",
      "stigma anther\n",
      "30 27\n",
      "stigma leaf\n",
      "38 24\n",
      "stigma root\n",
      "38 25\n",
      "stigma stem\n",
      "44 20\n",
      "stigma cotyledon\n",
      "40 25\n",
      "stigma hypocotyl\n",
      "35 29\n",
      "stigma petal\n",
      "30 27\n",
      "stigma fiber_10DPA\n",
      "34 25\n",
      "stigma fiber_20DPA\n",
      "42 18\n",
      "stigma ovule\n",
      "50 18\n",
      "stigma radicle\n",
      "31 31\n",
      "cotyledon anther\n",
      "65 42\n",
      "cotyledon leaf\n",
      "78 41\n",
      "cotyledon root\n",
      "69 47\n",
      "cotyledon stem\n",
      "75 37\n",
      "cotyledon stigma\n",
      "78 33\n",
      "cotyledon hypocotyl\n",
      "57 50\n",
      "cotyledon petal\n",
      "68 37\n",
      "cotyledon fiber_10DPA\n",
      "71 39\n",
      "cotyledon fiber_20DPA\n",
      "89 25\n",
      "cotyledon ovule\n",
      "90 32\n",
      "cotyledon radicle\n",
      "57 49\n",
      "hypocotyl anther\n",
      "63 31\n",
      "hypocotyl leaf\n",
      "85 21\n",
      "hypocotyl root\n",
      "67 27\n",
      "hypocotyl stem\n",
      "66 23\n",
      "hypocotyl stigma\n",
      "58 38\n",
      "hypocotyl cotyledon\n",
      "73 28\n",
      "hypocotyl petal\n",
      "70 30\n",
      "hypocotyl fiber_10DPA\n",
      "59 26\n",
      "hypocotyl fiber_20DPA\n",
      "74 20\n",
      "hypocotyl ovule\n",
      "97 24\n",
      "hypocotyl radicle\n",
      "42 43\n",
      "petal anther\n",
      "28 11\n",
      "petal leaf\n",
      "31 6\n",
      "petal root\n",
      "30 8\n",
      "petal stem\n",
      "33 5\n",
      "petal stigma\n",
      "33 4\n",
      "petal cotyledon\n",
      "27 8\n",
      "petal hypocotyl\n",
      "24 12\n",
      "petal fiber_10DPA\n",
      "31 7\n",
      "petal fiber_20DPA\n",
      "31 7\n",
      "petal ovule\n",
      "40 7\n",
      "petal radicle\n",
      "21 16\n",
      "fiber_10DPA anther\n",
      "8 26\n",
      "fiber_10DPA leaf\n",
      "11 23\n",
      "fiber_10DPA root\n",
      "12 21\n",
      "fiber_10DPA stem\n",
      "16 16\n",
      "fiber_10DPA stigma\n",
      "12 21\n",
      "fiber_10DPA cotyledon\n",
      "10 21\n",
      "fiber_10DPA hypocotyl\n",
      "7 26\n",
      "fiber_10DPA petal\n",
      "7 28\n",
      "fiber_10DPA fiber_20DPA\n",
      "14 12\n",
      "fiber_10DPA ovule\n",
      "18 22\n",
      "fiber_10DPA radicle\n",
      "9 25\n",
      "fiber_20DPA anther\n",
      "9 32\n",
      "fiber_20DPA leaf\n",
      "22 22\n",
      "fiber_20DPA root\n",
      "17 26\n",
      "fiber_20DPA stem\n",
      "19 19\n",
      "fiber_20DPA stigma\n",
      "11 30\n",
      "fiber_20DPA cotyledon\n",
      "18 21\n",
      "fiber_20DPA hypocotyl\n",
      "11 30\n",
      "fiber_20DPA petal\n",
      "12 31\n",
      "fiber_20DPA fiber_10DPA\n",
      "13 25\n",
      "fiber_20DPA ovule\n",
      "26 17\n",
      "fiber_20DPA radicle\n",
      "7 36\n",
      "ovule anther\n",
      "219 308\n",
      "ovule leaf\n",
      "247 278\n",
      "ovule root\n",
      "205 302\n",
      "ovule stem\n",
      "268 240\n",
      "ovule stigma\n",
      "244 283\n",
      "ovule cotyledon\n",
      "255 284\n",
      "ovule hypocotyl\n",
      "197 331\n",
      "ovule petal\n",
      "206 327\n",
      "ovule fiber_10DPA\n",
      "221 292\n",
      "ovule fiber_20DPA\n",
      "263 259\n",
      "ovule radicle\n",
      "145 382\n",
      "radicle anther\n",
      "150 50\n",
      "radicle leaf\n",
      "174 47\n",
      "radicle root\n",
      "128 67\n",
      "radicle stem\n",
      "143 55\n",
      "radicle stigma\n",
      "147 51\n",
      "radicle cotyledon\n",
      "150 54\n",
      "radicle hypocotyl\n",
      "120 59\n",
      "radicle petal\n",
      "147 52\n",
      "radicle fiber_10DPA\n",
      "152 44\n",
      "radicle fiber_20DPA\n",
      "176 39\n",
      "radicle ovule\n",
      "188 37\n"
     ]
    }
   ],
   "source": [
    "## 分析有多少TAD 是存在subcpt活性差异\n",
    "import pandas as pd\n",
    "tissues = ['anther', 'leaf', 'root', 'stem', 'stigma', 'cotyledon', 'hypocotyl', 'petal', 'fiber_10DPA', 'fiber_20DPA', 'ovule', 'radicle']\n",
    "old_tags = ['A1', 'A2', 'A3', 'A4', 'B4', 'B3', 'B2', 'B1']\n",
    "new_tags = [8, 7, 6, 5, 4, 3, 2, 1]\n",
    "#old_tags = ['A', 'B']\n",
    "#new_tags = [2, 1]\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_subcpt/TSG_TAD_subcpt.csv\")\n",
    "df['subcpt'].replace(old_tags, new_tags, inplace=True)\n",
    "df_result = pd.DataFrame()\n",
    "for t1 in tissues:\n",
    "    df_t = pd.DataFrame()\n",
    "    for t2 in tissues:\n",
    "        df_tmp = df.query(\"tags1==@t1 & tags2==@t2\")[['chrom', 'start', 'end', 'subcpt', 'tad_tags', 'tags1']].copy()\n",
    "        df_tmp.columns = ['chrom', 'start', 'end', t2, 'tad_tags', 'tags']\n",
    "        if df_t.shape[0] == 0:\n",
    "            df_t = df_tmp.copy()\n",
    "        else:\n",
    "            df_t = pd.merge(df_t, df_tmp, how='inner', on=['chrom', 'start', 'end', 'tad_tags', 'tags'])\n",
    "    df_result = pd.concat([df_result, df_t], axis=0)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_subcpt/TSG_TAD_subcpt_wide.csv\", index=False)\n",
    "\n",
    "## 统计分析活性TAD 差异\n",
    "def bias_TAD(df, tags):\n",
    "    #print(df)\n",
    "    df['number'] = [1]*df.shape[0]\n",
    "    df_active = df[df[tags]>0]\n",
    "    df_deactive = df[df[tags]<0]\n",
    "    #print(df_active)\n",
    "    df_active = df_active.groupby(by=['tad_tags'], as_index=False)['number'].sum()\n",
    "    df_deactive = df_deactive.groupby(by=['tad_tags'], as_index=False)['number'].sum()\n",
    "    df = df.groupby(by=['tad_tags'], as_index=False)['number'].sum()\n",
    "    \n",
    "    df.columns = ['tad_tags', 'total_number']\n",
    "    df_active.columns = ['tad_tags', 'active_number']\n",
    "    df_deactive.columns = ['tad_tags', 'deactive_number']\n",
    "    df_activate = pd.merge(df, df_active, how='left', on=['tad_tags'])\n",
    "    df_deactivate = pd.merge(df, df_deactive, how='left', on=['tad_tags'])\n",
    "    \n",
    "    df_activate.fillna(0, inplace=True)\n",
    "    df_deactivate.fillna(0, inplace=True)\n",
    "    \n",
    "    df_activate['activate_ratio'] = df_activate['active_number']/df_activate['total_number']\n",
    "    df_deactivate['deactivate_ratio'] = df_deactivate['deactive_number']/df_deactivate['total_number']\n",
    "    df_return = pd.merge(df_activate, df_deactivate, how='inner', on=['tad_tags', 'total_number'])\n",
    "    df_return['ratio'] = (df_return['activate_ratio']+1)/(df_return['deactivate_ratio']+1)\n",
    "    #df_return.columns = ['tad_tags', tags]\n",
    "    print(df_return.query(\"ratio>1\").shape[0], df_return.query(\"ratio<1\").shape[0])\n",
    "    return df_return\n",
    "def statistic_biasTAD(df, tags):\n",
    "    tissues = ['anther', 'leaf', 'root', 'stem', 'stigma', 'cotyledon', 'hypocotyl', 'petal', 'fiber_10DPA', 'fiber_20DPA', 'ovule', 'radicle']\n",
    "    df_return = pd.DataFrame()\n",
    "    for t in tissues:\n",
    "        if t != tags:\n",
    "            print(tags, t)\n",
    "            tad_vs_tad = \"{0}-VS-{1}\".format(tags, t)\n",
    "            df[tad_vs_tad] = df[tags] - df[t]\n",
    "            df_tmp = df[['tad_tags', tad_vs_tad]].copy()\n",
    "            df_tmp = bias_TAD(df_tmp, tad_vs_tad)\n",
    "            df_tmp['tags1'] = [tags]*df_tmp.shape[0]\n",
    "            df_tmp['tags2'] = [t]*df_tmp.shape[0]\n",
    "            if df_return.shape[0]==0:\n",
    "                df_return = df_tmp.copy()\n",
    "            else:\n",
    "                df_return = pd.concat([df_return, df_tmp], axis=0)\n",
    "    return df_return\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_subcpt/TSG_TAD_subcpt_wide.csv\")\n",
    "df_result = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    df_t = df.query(\"tags==@t\").copy()\n",
    "    #print(df_t.drop_duplicates(keep='first', subset=['tad_tags']).shape[0])\n",
    "    df_tmp = statistic_biasTAD(df_t, t)\n",
    "    df_result = pd.concat([df_result, df_tmp], axis=0)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_subcpt/TSG_TAD_subcpt_activate_deactivate_ratio.csv\", index=False)\n",
    "\n",
    "## 对上述结果进行统计\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_subcpt/TSG_TAD_subcpt_activate_deactivate_ratio.csv\")\n",
    "df['tags'] = df['ratio'].apply(lambda x: 'deactive' if x>1 else ('active' if x<1 else 'stable'))  ## a TAD 区间相比与b 来说变得不活跃了设定为deactivate 如果是变得活跃了设定为activate\n",
    "df['number'] = [1]*df.shape[0]\n",
    "df_result = df.groupby(by=[ 'tags1', 'tags2', 'tags'], as_index=False)['number'].sum()\n",
    "\n",
    "\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_subcpt/TSG_TAD_subcpt_activate_deactivate_ratio_statistics.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tags2                      tad_tags   tags1   anther cotyledon fiber_10DPA  \\\n",
      "5      Ghir_A01-104980000-105320000  anther  bstable  deactive    deactive   \n",
      "31       Ghir_A01-14940000-15240000  anther  bstable   bstable     bstable   \n",
      "42       Ghir_A01-23680000-23880000  anther  bstable  deactive    deactive   \n",
      "69       Ghir_A01-59180000-59460000  anther  bstable  deactive    deactive   \n",
      "75       Ghir_A01-69920000-71000000  anther  bstable   bstable     bstable   \n",
      "...                             ...     ...      ...       ...         ...   \n",
      "2532     Ghir_D13-40100000-40300000  anther  bstable  deactive    deactive   \n",
      "2547     Ghir_D13-50040000-50300000  anther  bstable   bstable     bstable   \n",
      "2557     Ghir_D13-55440000-55660000  anther  bstable   bstable     bstable   \n",
      "2564     Ghir_D13-58060000-58380000  anther  bstable   bstable     bstable   \n",
      "2572     Ghir_D13-60900000-61340000  anther  bstable   bstable     bstable   \n",
      "\n",
      "tags2 fiber_20DPA hypocotyl      leaf     ovule     petal   radicle      root  \\\n",
      "5        deactive  deactive    active  deactive  deactive    active   bstable   \n",
      "31        bstable    active    active   bstable    active    active    active   \n",
      "42       deactive   bstable  deactive  deactive  deactive  deactive  deactive   \n",
      "69       deactive  deactive  deactive  deactive  deactive  deactive  deactive   \n",
      "75        bstable    active   bstable   bstable   bstable    active   bstable   \n",
      "...           ...       ...       ...       ...       ...       ...       ...   \n",
      "2532     deactive  deactive    active  deactive  deactive    active  deactive   \n",
      "2547      bstable   bstable  deactive  deactive   bstable    active   bstable   \n",
      "2557      bstable   bstable   bstable   bstable   bstable   bstable   bstable   \n",
      "2564      bstable   bstable   bstable   bstable   bstable   bstable   bstable   \n",
      "2572      bstable   bstable  deactive  deactive   bstable   bstable  deactive   \n",
      "\n",
      "tags2      stem    stigma  \n",
      "5        active  deactive  \n",
      "31      bstable   bstable  \n",
      "42     deactive  deactive  \n",
      "69     deactive  deactive  \n",
      "75      bstable   bstable  \n",
      "...         ...       ...  \n",
      "2532   deactive  deactive  \n",
      "2547   deactive   bstable  \n",
      "2557    bstable   bstable  \n",
      "2564    bstable   bstable  \n",
      "2572    bstable   bstable  \n",
      "\n",
      "[307 rows x 14 columns]\n"
     ]
    }
   ],
   "source": [
    "## 整理TAD 包含的subcompartment的变化\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_subcpt/TSG_TAD_subcpt_activate_deactivate_ratio.csv\")\n",
    "df['tags'] = df['ratio'].apply(lambda x: 'deactive' if x>1 else ('active' if x<1 else 'bstable')) ## >1 值对照组的活性高，实验组中变得不活跃\n",
    "## 数据调整为宽数据格式\n",
    "df_wide = pd.pivot(df, index=['tad_tags', 'tags1'], columns='tags2', values='tags')\n",
    "df_wide.reset_index(inplace=True)\n",
    "df_wide.fillna('bstable', inplace=True)\n",
    "## anther\n",
    "df_wide_anther = df_wide.query(\"tags1=='anther'\").copy()\n",
    "print(df_wide_anther)\n",
    "tissues = ['anther', 'ovule', 'root', 'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df_wide_anther = df_wide_anther.sort_values(by=tissues)\n",
    "df_wide_anther['x_lab'] = range(df_wide_anther.shape[0])\n",
    "df_anther_long = pd.melt(df_wide_anther, id_vars=['x_lab', 'tags1', 'tad_tags'])\n",
    "df_anther_long.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_subcpt/anther_TSG_TAD_subcpt.csv\", index=False)\n",
    "\n",
    "# radicle\n",
    "# df_wide_anther = df_wide.query(\"tags1=='radicle'\").copy()\n",
    "# print(df_wide_anther)\n",
    "# tissues = ['radicle', 'anther', 'ovule', 'root',  'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "# df_wide_anther = df_wide_anther.sort_values(by=tissues)\n",
    "# df_wide_anther['x_lab'] = range(df_wide_anther.shape[0])\n",
    "# df_anther_long = pd.melt(df_wide_anther, id_vars=['x_lab', 'tags1', 'tad_tags'])\n",
    "# df_anther_long.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_subcpt/radicle_TSG_TAD_subcpt.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 将所有大于比例大于1.5的TAD 选出来\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_subcpt/TSG_TAD_subcpt_activate_deactivate_ratio.csv\")\n",
    "df_choose = df.query(\"ratio>=1.5\").copy()\n",
    "df_choose.drop_duplicates(keep='first', subset=['tad_tags'], inplace=True)\n",
    "df_choose[['chr', 'start', 'end']] = df_choose['tad_tags'].str.split(\"-\", expand=True)\n",
    "df_choose.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_subcpt/example/TSG_TAD_subcpt.bed\", sep=\"\\t\", index=False, header=False, columns=['chr', 'start', 'end', 'tags1'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        tags  number\n",
      "0     anther     307\n",
      "1  cotyledon     154\n",
      "2  hypocotyl     165\n",
      "3       leaf     154\n",
      "4      petal      53\n",
      "5       root     327\n",
      "6       stem     127\n",
      "7     stigma      91\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df =  pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_subcpt/TSG_TAD_subcpt_wide.csv\")\n",
    "df['number'] = [1]*df.shape[0]\n",
    "df.drop_duplicates(keep='first', subset=['tad_tags', 'tags'], inplace=True)\n",
    "df_tmp = df.groupby(by=['tags'], as_index=False)['number'].sum()\n",
    "print(df_tmp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 拉取包含基因的bin\n",
    "import pandas as pd\n",
    "df_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_subcpt/all_gene_bin_10Kb.bed\", sep=\"\\t\", names=['chrom', 'start', 'end'])\n",
    "df_subcpt = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_subcpt/TSG_TAD_subcpt.csv\")\n",
    "df_gene['start'] = df_gene['start']+1\n",
    "df_gene['end'] = df_gene['end']\n",
    "print(df_gene)\n",
    "df_result = pd.merge(df_subcpt, df_gene, how='inner', on=['chrom', 'start', 'end'])\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_subcpt/TSG_gene_TAD_subcpt.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_subcpt/TSG_gene_TAD_subcpt.csv\")\n",
    "df_result = df.groupby(by=['cpt', 'tags1', 'tags2'], as_index=False)['number'].sum()\n",
    "df_result.to_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_subcpt/TSG_gene_TAD_subcpt_statistics.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_subcpt/TSG_TAD_subcpt.csv\")\n",
    "df_result = df.groupby(by=['cpt', 'tags1', 'tags2'], as_index=False)['number'].sum()\n",
    "df_result.to_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_subcpt/TSG_TAD_subcpt_statistics.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "94288\n",
      "98738\n",
      "94350\n",
      "98774\n",
      "106790\n",
      "95032\n",
      "108520\n",
      "112332\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "tissues = ['anther', 'cotyledon', 'hypocotyl','leaf', 'petal', 'root', 'stem', 'stigma']\n",
    "## 分析转换区间所能包含的\n",
    "for t2 in tissues:\n",
    "    subcpt_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/compartment/CALDER2/subcompartment/{}_subcompartment_new.bed\".format(t2)\n",
    "    df = pd.read_csv(subcpt_f, sep=\"\\t\",names=['chrom', 'start', 'end', 'value', 'subcpt', 'cpt'])\n",
    "    df_a = df.query(\"cpt=='A'\").copy()\n",
    "    df_b = df.query(\"cpt=='B'\").copy()\n",
    "    print(df_b.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析TSG 所包含的subcompartment\n",
    "import pandas as pd\n",
    "df_anther_TSG_TAD = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/anther_TAD_TSG.csv\")\n",
    "df_dynamic_subcpt = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/compartment/CALDER2/subcompartment/J668_tissues_subcpt_wide.bed\", sep=\"\\t\")\n",
    "df_TSG_TAD_subcpt = pd.DataFrame() \n",
    "## 将所有位于TSG_TAD的动态subcpt都选出来\n",
    "for index, row in df_anther_TSG_TAD.iterrows():\n",
    "    C, S, E = row['chrom'], row['start'], row['end']\n",
    "    df_tmp = df_dynamic_subcpt.query(\"chr==@C & start<= @E & end>=@S\").copy()\n",
    "    df_TSG_TAD_subcpt = pd.concat([df_TSG_TAD_subcpt, df_tmp], axis=0)\n",
    "## 将结果整理为ggplot可识别格式\n",
    "#df_TSG_TAD_subcpt_A02 = df_TSG_TAD_subcpt.query(\"chr.str.contains('A01')\", engine='python').copy()\n",
    "df_TSG_TAD_subcpt_A02 = df_TSG_TAD_subcpt.copy()\n",
    "#print(df_TSG_TAD_subcpt_A02)\n",
    "# 排序\n",
    "list_sorted = ['A1', 'A2', 'A3', 'A4', 'B4', 'B3', 'B2', 'B1']\n",
    "df_TSG_TAD_subcpt_A02['anthers'] = df_TSG_TAD_subcpt_A02['anthers'].astype('category').cat.set_categories(list_sorted)\n",
    "#df_TSG_TAD_subcpt_A02['ovules'] = df_TSG_TAD_subcpt_A02['ovules'].astype('category').cat.set_categories(list_sorted)\n",
    "#df_TSG_TAD_subcpt_A02['root'] = df_TSG_TAD_subcpt_A02['root'].astype('category').cat.set_categories(list_sorted)\n",
    "\n",
    "df_TSG_TAD_subcpt_A02 = df_TSG_TAD_subcpt_A02.sort_values(by=['anthers'], ascending=True)\n",
    "df_TSG_TAD_subcpt_A02['x_lab'] = range(df_TSG_TAD_subcpt_A02.shape[0])\n",
    "#df_tmp = df_TSG_TAD_subcpt_A02[['x_lab', ]]\n",
    "df_long = pd.melt(df_TSG_TAD_subcpt_A02, id_vars=['x_lab', 'chr', 'start', 'end'])\n",
    "df_long.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_subcpt/anther_TSG_TAD_subcpt.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 选出所有的孤岛基因\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/data/gene_TAD_type.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'GeneId', 'tissue', 'type'])\n",
    "tissues = ['anther', 'cotyledon', 'hypocotyl','leaf', 'petal', 'root', 'stem', 'stigma', 'fiber_10DPA', 'fiber_20DPA', 'ovule', 'radicle']\n",
    "for t in tissues:\n",
    "    output_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/Rawdata/{}_isolated_gene.bed\".format(t)\n",
    "    df_tmp = df.query(\"tissue==@t & type=='Isolated'\")[['GeneId']].copy()\n",
    "    df_tmp.to_csv(output_f, index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析孤岛TAD 中所包含的同源基因，分析其是否具有明显差异\n",
    "## 1、确定同源基因对的孤岛TAD\n",
    "import pandas as pd\n",
    "def isolated_Hg_gene(df_tissues_isolated_gene):\n",
    "    df_hg = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/homo_gene2/Ghir_homo_gene_pairs.txt\", sep=\"\\t\")\n",
    "    df_Isolate_geneid = df_tissues_isolated_gene.copy()\n",
    "    df_Isolate_geneid['tags'] = ['Isolated']*df_Isolate_geneid.shape[0]\n",
    "    df_Isolate_geneid.columns = ['GeneId_At', 'At_tags']\n",
    "    df_a = pd.merge(df_hg, df_Isolate_geneid, how='left', on=['GeneId_At'])\n",
    "    df_Isolate_geneid.columns = ['GeneId_Dt', 'Dt_tags']\n",
    "    df_ad = pd.merge(df_a, df_Isolate_geneid, how='left', on=['GeneId_Dt'])\n",
    "    df_return = df_ad.query(\"At_tags=='Isolated' | Dt_tags=='Isolated'\").copy()\n",
    "    df_return.fillna(\"NIsolated\", inplace=True)\n",
    "    return df_return\n",
    "def ITADHG_bias(df_ITADHG, T):\n",
    "    df_bias = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/homo_gene2/gene_bias/all_tissue_bias_gene_pair.bed\", sep=\"\\t\")\n",
    "    df_bias_tag = df_bias.query(\"tags==@T\")[['GeneId_At', 'GeneId_Dt', 'sig']].copy()\n",
    "    df_return = pd.merge(df_ITADHG, df_bias_tag, how='left', on=['GeneId_At', 'GeneId_Dt'])\n",
    "    df_return.fillna(\"BiasN\", inplace=True)\n",
    "    return df_return\n",
    "    \n",
    "tissues = ['anther', 'cotyledon', 'fiber_10DPA', 'fiber_20DPA', 'hypocotyl', 'leaf', 'petal', 'radicle', 'root', 'stem', 'stigma', 'ovule']\n",
    "for t in tissues:\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/Rawdata/{}_isolated_gene.bed\".format(t)\n",
    "    output_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/Rawdata/{}_isolated_Hg.csv\".format(t)\n",
    "    df_Isolated_gene = pd.read_csv(input_f, names=['geneid'])\n",
    "    df_Isolated_HG = isolated_Hg_gene(df_Isolated_gene) ## 分析确定同源孤岛基因\n",
    "    df_Isolated_HG_bias = ITADHG_bias(df_Isolated_HG, t) ## 确定同源基因对之间是否发生了bias\n",
    "    df_Isolated_HG_bias.to_csv(output_f, index=False)\n",
    "\n",
    "## 2、分析孤岛同源基因对间的差异\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1621\n",
      "        tissues   type\n",
      "0        anther  516.0\n",
      "1     cotyledon  429.0\n",
      "2   fiber_10DPA  570.0\n",
      "3   fiber_20DPA  599.0\n",
      "4     hypocotyl  434.0\n",
      "5          leaf  440.0\n",
      "6         ovule  544.0\n",
      "7         petal  506.0\n",
      "8       radicle  475.0\n",
      "9          root  424.0\n",
      "10         stem  430.0\n",
      "11       stigma  490.0\n"
     ]
    }
   ],
   "source": [
    "## 统计分析各组织间动态孤岛基因数目\n",
    "import pandas as pd\n",
    "tissues = ['anther', 'ovule', 'root', 'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df_result = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/Rawdata/{}_isolated_gene.bed\".format(t)\n",
    "    df_tmp = pd.read_csv(input_f, sep=\"\\t\", names=['GeneId'])\n",
    "    df_tmp[t] = [1]*df_tmp.shape[0]\n",
    "    if df_result.shape[0] == 0:\n",
    "        df_result = df_tmp.copy()\n",
    "    else:\n",
    "        df_result = pd.merge(df_result, df_tmp, how='outer', on=['GeneId'])\n",
    "df_result.fillna(0 , inplace=True)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/J668_isolated_TAD_gene.csv\", index=False)\n",
    "df_result['allsum'] = df_result[tissues].sum(axis=1)\n",
    "print(df_result.shape[0])\n",
    "df_stable = df_result.query(\"allsum==12\").copy()\n",
    "df_stable.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/stable_ITAD_gene/J668_Multip_tissues_Stable_ITAD_gene.csv\", index=False)\n",
    "df_dynamic = df_result.query(\"allsum<12\")[tissues].copy()\n",
    "df_dynamic.reset_index(inplace=True, drop=True)\n",
    "df_dynamic.reset_index(inplace=True)\n",
    "df_dynamic.sort_values(by=tissues, inplace=True)\n",
    "df_dynamic_long = pd.melt(df_dynamic, id_vars='index')\n",
    "df_dynamic_long.columns = ['index', 'tissues', 'type']\n",
    "print(df_dynamic_long.groupby(by=['tissues'],as_index=False)['type'].sum())\n",
    "df_dynamic_long.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/J668_dynamic_isolated_TAD_gene.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析孤岛基因在各组织的两个亚基因组同源基因中的数目\n",
    "import pandas as pd\n",
    "df_enrich_isolated = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/data/gene_TAD_type.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'GeneId', 'tissue', 'type'])\n",
    "df_isolated = df_enrich_isolated.query(\"type=='Isolated'\")[['GeneId', 'tissue', 'type']].copy()\n",
    "df_isolated_At = df_isolated.query(\"GeneId.str.contains('Ghir_A')\", engine='python').copy()\n",
    "df_isolated_Dt = df_isolated.query(\"GeneId.str.contains('Ghir_D')\", engine='python').copy()\n",
    "#print(df_isolated_At.shape[0], df_isolated_Dt.shape[0])\n",
    "df_hg_pair = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/J668_HG_pairs.csv\")\n",
    "###  Bias At\n",
    "df_isolated_At.columns = ['GeneId_At', 'tissue', 'type_At']\n",
    "df_isolated_At = pd.merge(df_hg_pair, df_isolated_At, how='inner', on=['GeneId_At'])\n",
    "#print(df_isolated_At)\n",
    "#print(df_isolated_At.shape[0])\n",
    "### Bias Dt\n",
    "df_isolated_Dt.columns = ['GeneId_Dt', 'tissue', 'type_Dt']\n",
    "df_isolated_Dt = pd.merge(df_hg_pair, df_isolated_Dt, how='inner', on=['GeneId_Dt'])\n",
    "df_isolated_AtDt = pd.merge(df_isolated_At, df_isolated_Dt, how='outer', on=['GeneId_At', 'GeneId_Dt', 'tissue'])\n",
    "df_isolated_AtDt.fillna(\"NIsolated\", inplace=True)\n",
    "df_isolated_AtDt.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/HG/J668_Isolated_TAD_HG_pair.csv\", index=False)\n",
    "df_isolated_AtDt['Tags'] = df_isolated_AtDt['type_At'].map(str) + \"-\" + df_isolated_AtDt['type_Dt'].map(str)\n",
    "old_tags = ['Isolated-Isolated', 'Isolated-NIsolated', 'NIsolated-Isolated']\n",
    "new_tags = ['AtIso-DtIso', 'AtIso-DtNIso', 'AtNIso-DtNIso']\n",
    "df_isolated_AtDt['Tags'].replace(old_tags, new_tags, inplace=True)\n",
    "df_isolated_AtDt['number'] = [1]*df_isolated_AtDt.shape[0]\n",
    "df_statistics_result = df_isolated_AtDt.groupby(by=['tissue', 'Tags'], as_index=False)['number'].sum()\n",
    "df_statistics_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/HG/J668_HIsoGene_statistics.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 根据所确定的孤岛基因来确定所有的孤岛TAD 结构\n",
    "import pandas as pd\n",
    "tissues = ['cotyledon', 'fiber_10DPA', 'fiber_20DPA', 'hypocotyl', 'leaf', 'petal', 'radicle', 'root', 'stem', 'stigma', 'ovule']\n",
    "df_anther = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/Rawdata/anther_isolated_gene.bed\", names=['GeneId'])\n",
    "df_tmp = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/Rawdata/{}_isolated_gene.bed\".format(t)\n",
    "    df_tissue = pd.read_csv(input_f, names=['GeneId'])\n",
    "    if df_tmp.shape[0] == 0:\n",
    "        df_tmp = pd.merge(df_anther, df_tissue, how='inner', on=['GeneId'])\n",
    "    else:\n",
    "        df_tmp = pd.merge(df_tmp, df_tissue, how='inner', on=['GeneId'])\n",
    "df_tmp.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/Rawdata/J668_Multip_tissues_Stable_ITAD_gene.bed\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "40\n"
     ]
    }
   ],
   "source": [
    "## 分析这些基因的TAU值的大小\n",
    "import pandas as pd\n",
    "df_stable = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/Rawdata/J668_Multip_tissues_Stable_ITAD_gene.bed\")\n",
    "df_TAU = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/TAU/J668_all_tissue_TAU.csv\")\n",
    "df_TAU = df_TAU[['GeneId', 'TAU']].copy()\n",
    "df = pd.merge(df_stable, df_TAU, how='inner', on=['GeneId'])\n",
    "print(df.query(\"TAU>=0.85\").shape[0])\n",
    "df.to_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/Rawdata/J668_Multip_tissues_Stable_ITAD_gene_TAU.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 确定基因位置\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/stable_ITAD_gene/J668_Multip_tissues_Stable_ITAD_gene.bed\")\n",
    "df_position = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/Gh_gene_position.bed\", sep=\"\\t\")\n",
    "df_result = pd.merge(df, df_position, how='inner', on=['GeneId'])\n",
    "df_result.to_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/stable_ITAD_gene/J668_Multip_tissues_Stable_ITAD_gene.bed', sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将每个组织TAD分为五等份\n",
    "import pandas as pd\n",
    "\n",
    "## 每个TAD划分五等份\n",
    "def TAD_Quintiles(df):\n",
    "    df['TAD_size'] = df['end'] - df['start']\n",
    "    df['length'] = df['TAD_size']/5\n",
    "    tags = ['one', 'two', 'three', 'four', 'five']\n",
    "    df_return = ['df1', 'df2', 'df3', 'df4', 'df5']\n",
    "    for i in range(5):\n",
    "        df[tags[i]] = df['start'] + df['length']*(i+1)\n",
    "        df[tags[i]] = df[tags[i]].map(int)\n",
    "        if i==0:\n",
    "            df_return[i] = df[['chrom', 'start', tags[i]]].copy()\n",
    "        else:\n",
    "            df_return[i] = df[['chrom', tags[i-1], tags[i]]].copy()\n",
    "    return df_return\n",
    "## 数据格式标准化\n",
    "def data_format(input_f):\n",
    "    df = pd.read_csv(input_f, sep=\"\\t\")\n",
    "    df.columns = ['chrom', 'start', 'end', 'order']\n",
    "    df_return = df[['chrom', 'start', 'end']].copy()\n",
    "    return df_return\n",
    "\n",
    "tissues = ['anther', 'cotyledon', 'fiber_10DPA', 'fiber_20DPA', 'ovule', 'hypocotyl', 'leaf', 'petal', 'radicle', 'root', 'stem', 'stigma']\n",
    "for t in tissues:\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_data/{0}_choosed.bed\".format(t)\n",
    "    df = data_format(input_f)\n",
    "    df_TAD_quintiles = TAD_Quintiles(df)\n",
    "    # 输出文件\n",
    "    for i in range(5):\n",
    "        output_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/{0}_TAD{1}.bed\".format(t, i)\n",
    "        df_TAD_quintiles[i].to_csv(output_f, sep=\"\\t\", index=False, header=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析确定五等份TAD所包含的基因表达量\n",
    "import pandas \n",
    "tissues = ['anther', 'cotyledon', 'fiber_10DPA', 'fiber_20DPA', 'ovule', 'hypocotyl', 'leaf', 'petal', 'radicle', 'root', 'stem', 'stigma']\n",
    "df_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data/J668_all_Tissue_mean_TPM_long.csv\")\n",
    "df_result = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    df_tissues_gene = df_gene.query(\"tissues==@t\").copy()\n",
    "    TAD_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_gene/{}_TAD_quintiles_gene.bed\".format(t)\n",
    "    df_TAD = pd.read_csv(TAD_f, sep=\"\\t\", names=['GeneId', 'level'])\n",
    "    df_tissues_tmp = pd.merge(df_TAD, df_tissues_gene, how='left', on=['GeneId'])\n",
    "    df_result = pd.concat([df_result, df_tissues_tmp], axis=0)\n",
    "old_level = [0, 1, 2, 3, 4]\n",
    "new_level = ['one', 'two', 'three', 'four', 'five']\n",
    "df_result['level'].replace(old_level, new_level, inplace=True)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_gene/J668_TAD_quintiles_gene_TPM.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析有哪些差异表达基因是由于TAD结构变化差异所导致\n",
    "import pandas as pd\n",
    "tissues = ['anther', 'ovule', 'root', 'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df_result = pd.DataFrame()\n",
    "for i in range(11):\n",
    "    one_tissues = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_gene/{}_TAD_quintiles_gene.bed\".format(tissues[i])\n",
    "    df_one = pd.read_csv(one_tissues, sep=\"\\t\", names=['GeneId', 'one'])\n",
    "    df_one.drop_duplicates(keep='first', subset=['GeneId'], inplace=True)\n",
    "    for j in range(i+1, 12):\n",
    "        two_tissues = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_gene/{}_TAD_quintiles_gene.bed\".format(tissues[j])\n",
    "        df_two = pd.read_csv(two_tissues, sep=\"\\t\", names=['GeneId', 'two'])\n",
    "        df_two.drop_duplicates(keep='first', subset=['GeneId'], inplace=True)\n",
    "        DEG_f = \"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/{0}_{1}_DEG.csv\".format(tissues[i], tissues[j])\n",
    "        df_DEG = pd.read_csv(DEG_f)\n",
    "        df_a = pd.merge(df_DEG, df_one, how='left', on=['GeneId'])\n",
    "        df_ab = pd.merge(df_a, df_two, how='left', on=['GeneId'])\n",
    "        Tags = \"{0}_{1}\".format(tissues[i], tissues[j])\n",
    "        df_ab['tags'] = [Tags]*df_ab.shape[0]\n",
    "        df_result = pd.concat([df_result, df_ab], axis=0)\n",
    "old_tags = [0.0, 1.0, 2.0, 3.0, 4.0]\n",
    "new_tags = [3, 2, 1, 2, 3]\n",
    "df_result['one'].replace(old_tags, new_tags, inplace=True)\n",
    "df_result['two'].replace(old_tags, new_tags, inplace=True)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_gene/DEG/J668_allTissues_DEG_TADquintiles.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析确定哪些差异表达基因与基因在TAD上的相对位置差异有关\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_gene/DEG/J668_allTissues_DEG_TADquintiles.csv\")\n",
    "df_need = df.query(\"(sig=='Down' & two< one) | (sig=='Up' & two>one)\").copy()\n",
    "#df_need = df.query(\"sig!='Not' & two != one\").copy()\n",
    "df_need.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_gene/DEG/J668_allTissues_DEG_TADquintiles_need.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 统计差异表达基因与基因在TAD位置上差异的关系\n",
    "import pandas as pd\n",
    "df_TAD_quintiles = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_gene/DEG/J668_allTissues_DEG_TADquintiles.csv\")\n",
    "df_TAD_quintiles.dropna(inplace=True)\n",
    "df_TAD_quintiles['number'] = [1]*df_TAD_quintiles.shape[0]\n",
    "df_DEG = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/J668_mutipTissues_DEG.csv\")\n",
    "df_DEG['number'] = [1]*df_DEG.shape[0]\n",
    "df_TAD_quintiles_DEG = df_TAD_quintiles.query(\"sig!='Not' & two != one\").copy()\n",
    "df_TAD_quintiles_DEG_need = df_TAD_quintiles.query(\"(sig=='Down' & two< one) | (sig=='Up' & two>one)\").copy()\n",
    "df_DEG = df_DEG.query(\"sig!='Not'\").copy()\n",
    "df_all = df_DEG.groupby(by=['tags'], as_index=False)['number'].sum()\n",
    "df_a = df_TAD_quintiles_DEG.groupby(by=['tags'], as_index=False)['number'].sum()\n",
    "df_b = df_TAD_quintiles_DEG_need.groupby(by=['tags'], as_index=False)['number'].sum()\n",
    "df_all_a = pd.merge(df_all, df_a, how='inner', on=['tags'])\n",
    "df_all_ab = pd.merge(df_all_a, df_b, how='inner', on=['tags'])\n",
    "df_all_ab.columns = ['tags', 'All_DEG', 'TAD_DEG', 'TAD_S_DEG']\n",
    "df_all_ab.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_gene/DEG/J668_DEG_gene_number.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析cotyledon 与fiber_20DPA 之间联系\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_gene/DEG/J668_allTissues_DEG_TADquintiles_need.csv\")\n",
    "df_cotyledon_fiber_20DPA = df.query(\"tags=='cotyledon_fiber_20DPA'\").copy()\n",
    "df_cotyledon = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_gene/cotyledon_TAD_gene.csv\")\n",
    "df_fiber_20DPA = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_gene/fiber_20DPA_TAD_gene.csv\")\n",
    "df_cotyledon = df_cotyledon[['GeneId', 'tags']].copy()\n",
    "df_fiber_20DPA = df_fiber_20DPA[['GeneId', 'tags']].copy()\n",
    "df_cotyledon.columns = ['GeneId', 'cotyledon']\n",
    "df_fiber_20DPA.columns = ['GeneId', 'fiber_20DPA']\n",
    "df_cotyledon_fiber_20DPA = pd.merge(df_cotyledon_fiber_20DPA, df_cotyledon, how='left', on=['GeneId'])\n",
    "df_cotyledon_fiber_20DPA = pd.merge(df_cotyledon_fiber_20DPA, df_fiber_20DPA, how='left', on=['GeneId'])\n",
    "df_cotyledon_fiber_20DPA.sort_values(by=['logFC'], inplace=True)\n",
    "df_cotyledon_fiber_20DPA.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_gene/DEG/J668_cotyledon_fiber_20DPA.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析确定TAD所包含的基因\n",
    "import pandas as pd\n",
    "df_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/RNA_seq/Gh_gene_position.bed\", sep=\"\\t\")\n",
    "tissues = ['cotyledon', 'fiber_20DPA', 'anther', 'ovule', 'root', 'radicle', 'leaf', 'hypocotyl', 'stem', 'fiber_10DPA', 'petal', 'stigma']\n",
    "for t in tissues:\n",
    "    TAD_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_data/{}_choosed.bed\".format(t)\n",
    "    output_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_gene/{}_TAD_gene.csv\".format(t)\n",
    "    df_tad = pd.read_csv(TAD_f, sep=\"\\t\")\n",
    "    df_tad.columns = ['chrs', 'start', 'end', 'order']\n",
    "    df_out = pd.DataFrame()\n",
    "    for index, row in df_tad.iterrows():\n",
    "        chrom, start, end = row['chrs'], row['start'], row['end']\n",
    "        tags = \"{0}-{1}-{2}\".format(chrom, start, end)\n",
    "        df_tmp = df_gene.query(\"chrs==@chrom & start<=@end & end >= @start\").copy()\n",
    "        df_tmp['tags'] = [tags]*df_tmp.shape[0]\n",
    "        df_out = pd.concat([df_out, df_tmp], axis=0)\n",
    "    df_out.drop_duplicates(keep='first', inplace=True)\n",
    "    df_out.to_csv(output_f, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将TAD 转换为juicerbox 可识别的格式\n",
    "import pandas as pd\n",
    "tissues = ['cotyledon', 'fiber_20DPA', 'anther', 'ovule', 'root', 'radicle', 'leaf', 'hypocotyl', 'stem', 'fiber_10DPA', 'petal', 'stigma']\n",
    "old_chr = ['Ghir_A01', 'Ghir_A02', 'Ghir_A03', 'Ghir_A04', 'Ghir_A05', 'Ghir_A06', 'Ghir_A07', 'Ghir_A08', 'Ghir_A09', 'Ghir_A10', 'Ghir_A11', 'Ghir_A12', 'Ghir_A13', \n",
    "           'Ghir_D01', 'Ghir_D02', 'Ghir_D03', 'Ghir_D04', 'Ghir_D05', 'Ghir_D06', 'Ghir_D07', 'Ghir_D08', 'Ghir_D09', 'Ghir_D10', 'Ghir_D11', 'Ghir_D12', 'Ghir_D13']\n",
    "new_chr = ['GHIR_A01', 'GHIR_A02', 'GHIR_A03', 'GHIR_A04', 'GHIR_A05', 'GHIR_A06', 'GHIR_A07', 'GHIR_A08', 'GHIR_A09', 'GHIR_A10', 'GHIR_A11', 'GHIR_A12', 'GHIR_A13', \n",
    "           'GHIR_D01', 'GHIR_D02', 'GHIR_D03', 'GHIR_D04', 'GHIR_D05', 'GHIR_D06', 'GHIR_D07', 'GHIR_D08', 'GHIR_D09', 'GHIR_D10', 'GHIR_D11', 'GHIR_D12', 'GHIR_D13']\n",
    "for t in tissues:\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_data/{}_choosed.bed\".format(t)\n",
    "    output_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_data/{}_choosed.bedpe\".format(t)\n",
    "    df = pd.read_csv(input_f, sep=\"\\t\")\n",
    "    df['chr'].replace(old_chr, new_chr, inplace=True)\n",
    "    df['chr1'] = df['chr']\n",
    "    df['s1'] = df['s']\n",
    "    df['e1'] = df['e']\n",
    "    df['a'] = ['.']*df.shape[0]\n",
    "    df['b'] = ['.']*df.shape[0]\n",
    "    df['c'] = ['.']*df.shape[0]\n",
    "    df['d'] = ['.']*df.shape[0]\n",
    "    df['h'] = ['255,255,0']*df.shape[0]\n",
    "    df.to_csv(output_f, sep=\"\\t\", index=False, columns=['chr', 's', 'e', 'chr1', 's1', 'e1', 'a', 'b', 'c', 'd', 'h'], header=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 合并五等份TAD\n",
    "tissues = ['cotyledon', 'fiber_20DPA', 'anther', 'ovule', 'root', 'radicle', 'leaf', 'hypocotyl', 'stem', 'fiber_10DPA', 'petal', 'stigma']\n",
    "input_dir = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region\"\n",
    "for t in tissues:\n",
    "    for i in range(5):\n",
    "        input_f = \"{}/{}_TAD{}.bed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析亚基因组bias表达基因分别具有TAD位置差异的影响\n",
    "import pandas as pd\n",
    "tissues = ['anther', 'cotyledon', 'fiber_20DPA', 'ovule', 'root', 'radicle', 'leaf', 'hypocotyl', 'stem', 'fiber_10DPA', 'petal', 'stigma']\n",
    "df_hg = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data/J668_HG_pairs.csv\")\n",
    "df_TAD_quintiles_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_gene/J668_TAD_quintiles_gene_TPM.csv\")\n",
    "def hg_bias_data_format(df):\n",
    "    df_tmp = df[['GeneId', 'sig']].copy()\n",
    "    df_tmp.columns = ['GeneId_At', 'sig']\n",
    "    df_return = pd.merge(df_tmp, df_hg, how='inner', on=['GeneId_At'])\n",
    "    old_tags = ['Down', 'Up']\n",
    "    new_tags = ['Bias_At', 'Bias_Dt']\n",
    "    df_return['sig'].replace(old_tags, new_tags, inplace=True)\n",
    "    return df_return \n",
    "def hg_format(df1,df2, tags):\n",
    "    df2 = df2[['GeneId', 'level']].copy()\n",
    "    df2.columns = ['GeneId_At', 'level_At']\n",
    "    df_a = pd.merge(df1, df2, how='left', on=['GeneId_At'])\n",
    "    df2.columns = ['GeneId_Dt', 'level_Dt']\n",
    "    df_ab = pd.merge(df_a, df2, how='left', on=['GeneId_Dt'])\n",
    "    df_return = df_ab.dropna().copy()\n",
    "    df_return['tags'] = [tags]*df_return.shape[0]\n",
    "    df_return.drop_duplicates(keep='first', subset=['GeneId_At', 'GeneId_Dt'], inplace=True)\n",
    "    return df_return\n",
    "df_result = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    hg_bias_f = \"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/hg/{}_DEG.bed\".format(t)\n",
    "    df_tissues_hg = pd.read_csv(hg_bias_f, sep=\"\\t\")\n",
    "    df_tissues_hg = hg_bias_data_format(df_tissues_hg) ## 同源基因对bias基因\n",
    "    df_tissues_TADG = df_TAD_quintiles_gene.query(\"tissues==@t\").copy()\n",
    "    df_tmp = hg_format(df_tissues_hg, df_tissues_TADG, t)\n",
    "    df_result = pd.concat([df_result, df_tmp], axis=0)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_hg_bias_gene/J668_TAD_quintiles_hg_bias_gene.csv\", index=False)    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           tags  number\n",
      "0        anther    4340\n",
      "1     cotyledon    3776\n",
      "2   fiber_10DPA    4056\n",
      "3   fiber_20DPA    3722\n",
      "4     hypocotyl    3605\n",
      "5          leaf    3758\n",
      "6         ovule    3367\n",
      "7         petal    4509\n",
      "8       radicle    3226\n",
      "9          root    3560\n",
      "10         stem    3737\n",
      "11       stigma    4614\n"
     ]
    }
   ],
   "source": [
    "## 分析有多少同源bias基因受到TAD结构差异的影响\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_hg_bias_gene/J668_TAD_quintiles_hg_bias_gene.csv\")\n",
    "old_tags = ['one', 'two', 'three', 'four', 'five']\n",
    "new_tags = [0, 1, 2, 1, 0]\n",
    "df['level_At'].replace(old_tags, new_tags, inplace=True)\n",
    "df['level_Dt'].replace(old_tags, new_tags, inplace=True)\n",
    "df_diff = df.query(\"level_At != level_Dt\").copy()\n",
    "df_diff['number'] = [1]*df_diff.shape[0]\n",
    "df_TAD = df_diff.groupby(by=['tags'], as_index=False)['number'].sum()\n",
    "print(df_TAD)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           tags  number\n",
      "0        anther    2219\n",
      "1     cotyledon    1951\n",
      "2   fiber_10DPA    2102\n",
      "3   fiber_20DPA    1925\n",
      "4     hypocotyl    1841\n",
      "5          leaf    1882\n",
      "6         ovule    1724\n",
      "7         petal    2334\n",
      "8       radicle    1658\n",
      "9          root    1818\n",
      "10         stem    1915\n",
      "11       stigma    2376\n"
     ]
    }
   ],
   "source": [
    "## 分析有多少同源bias基因是符合越靠近boundary表达水平越高的\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_hg_bias_gene/J668_TAD_quintiles_hg_bias_gene.csv\")\n",
    "old_tags = ['one', 'two', 'three', 'four', 'five']\n",
    "new_tags = [0, 1, 2, 1, 0]\n",
    "df['level_At'].replace(old_tags, new_tags, inplace=True)\n",
    "df['level_Dt'].replace(old_tags, new_tags, inplace=True)\n",
    "df_diff = df.query(\"(sig=='Bias_At' & level_At<level_Dt) | (sig=='Bias_Dt' & level_Dt<level_At)\").copy()\n",
    "df_diff['number'] = [1]*df_diff.shape[0]\n",
    "df_TAD_true = df_diff.groupby(by=['tags'], as_index=False)['number'].sum()\n",
    "print(df_TAD_true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           tags  number\n",
      "0        anther    8185\n",
      "1     cotyledon    7155\n",
      "2   fiber_10DPA    7858\n",
      "3   fiber_20DPA    7416\n",
      "4     hypocotyl    7019\n",
      "5          leaf    7046\n",
      "6         ovule    6271\n",
      "7         petal    8591\n",
      "8       radicle    6322\n",
      "9          root    6907\n",
      "10         stem    6996\n",
      "11       stigma    8706\n"
     ]
    }
   ],
   "source": [
    "## 将所有数据都整理好，1、各同源差异表达基因对都位于TAD上，2、各同源差异表达基因位于不同的TAD位置，3、各同源差异表达基因位置与表达差异正相关\n",
    "import pandas as pd\n",
    "df = pd.read_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_hg_bias_gene/J668_TAD_quintiles_hg_bias_gene.csv')\n",
    "df['number'] = [1]*df.shape[0]\n",
    "df_number = df.groupby(by=['tags'], as_index=False)['number'].sum()\n",
    "print(df_number)\n",
    "df_number.columns = ['tags', 'all_TAD_diff_hg']\n",
    "df_TAD.columns = ['tags', 'all_diff_TAD_diff_hg']\n",
    "df_TAD_true.columns = ['tags', 'all_diff_TAD_diff_hg_true']\n",
    "df_a = pd.merge(df_number, df_TAD, how='inner', on=['tags'])\n",
    "df_result = pd.merge(df_a, df_TAD_true, how='inner', on=['tags'])\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_hg_bias_gene/J668_TAD_quintiles_statistics.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "            GeneId        color     chrom     start       end\n",
      "0  Ghir_A04G013310   218,131,61  Ghir_A04  82083000  82086000\n",
      "1  Ghir_A04G013310   218,131,61  Ghir_A04  82086000  82089000\n",
      "2  Ghir_A12G010900  233,111,183  Ghir_A12  77670000  77673000\n",
      "3  Ghir_A12G010900  233,111,183  Ghir_A12  77673000  77676000\n",
      "4  Ghir_A12G010900  233,111,183  Ghir_A12  77676000  77679000\n"
     ]
    }
   ],
   "source": [
    "## 确定hypocotyl 组织能够发生bias 表达并且符合越靠近边界其表达水平越高\n",
    "import pandas as pd\n",
    "import random as r\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_hg_bias_gene/J668_TAD_quintiles_hg_bias_gene.csv\")\n",
    "old_tags = ['one', 'two', 'three', 'four', 'five']\n",
    "new_tags = [0, 1, 2, 1, 0]\n",
    "df['level_At'].replace(old_tags, new_tags, inplace=True)\n",
    "df['level_Dt'].replace(old_tags, new_tags, inplace=True)\n",
    "\n",
    "# 确定符合约靠近边界其表达水平越高的同源基因对\n",
    "def TAD_true_bias(df):\n",
    "    df_return = df.query(\"(level_At>level_Dt & sig=='Bias_Dt') | (level_At<level_Dt & sig=='Bias_At')\").copy()\n",
    "    return df_return\n",
    "\n",
    "# 将数据转换为juicer 可识别格式\n",
    "def trans_juicerbox(df):\n",
    "    print(df.head())\n",
    "    df_a = df[['chrom', 'start', 'end', 'color']].copy()\n",
    "    df_a.columns = ['chrom1', 'start1', 'end1', 'color']\n",
    "    df_a['chrom2'], df_a['start2'], df_a['end2'] = df_a['chrom1'], df_a['start1'], df_a['end1']\n",
    "    df_a['a'] = [\".\"]*df_a.shape[0]\n",
    "    df_a['b'] = [\".\"]*df_a.shape[0]\n",
    "    df_a['c'] = [\".\"]*df_a.shape[0]\n",
    "    df_a['d'] = [\".\"]*df_a.shape[0]\n",
    "    return df_a\n",
    "\n",
    "# 分析同源基因在亚基因组上的位置\n",
    "def true_hg_bedpe(df_hg, df_bin):\n",
    "    # 确定基因所在的bin区间\n",
    "    df_hg, df_bin = df_hg.copy(), df_bin.copy()\n",
    "    #print(\",\".join([str(r.randint(0, 255)), str(r.randint(0, 255)), str(r.randint(0, 255))]))\n",
    "    color_list = [\",\".join([str(r.randint(0, 255)), str(r.randint(0, 255)), str(r.randint(0, 255))]) for i in range(df_hg.shape[0])]\n",
    "    df_hg['color'] = color_list\n",
    "    df_a = df_hg[['GeneId_At', 'color']].copy()\n",
    "    df_b = df_hg[['GeneId_Dt', 'color']].copy()\n",
    "    df_a.columns = ['GeneId', 'color']\n",
    "    df_b.columns = ['GeneId', 'color']\n",
    "    df_ab = pd.concat([df_a, df_b], axis=0)\n",
    "    df_ab_bin = pd.merge(df_ab, df_bin, how='inner', on=['GeneId'])\n",
    "    # 将基因区域写称juicerbox 可识别的格式\n",
    "    #print(df_ab_bin)\n",
    "    df_G_bedpe = trans_juicerbox(df_ab_bin)\n",
    "    df_G_bedpe.sort_values(by=['color'], inplace=True)\n",
    "    return df_G_bedpe\n",
    "\n",
    "df_TAD_true = TAD_true_bias(df)\n",
    "df_TAD_true_hypocotyl = df_TAD_true.query(\"tags=='hypocotyl'\")\n",
    "df_G_bin = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/3kb_bin_gene.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'GeneId'])\n",
    "df_result = true_hg_bedpe(df_TAD_true_hypocotyl, df_G_bin)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_hg_bias_gene/J668_hypocotyl_diffTAD_diffHG_True.bedpe\", sep=\"\\t\", index=False, columns=['chrom1', 'start1', 'end1', 'chrom2', 'start2', 'end2', 'a', 'b', 'c', 'd', 'color'])\n",
    "\n",
    "#print(df_TAD_true_hypocotyl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "            GeneId_At        GeneId_Dt    chr_at  start_at    end_at  \\\n",
      "9795  Ghir_A06G008060  Ghir_D06G008290  Ghir_A06  20260598  20265166   \n",
      "9796  Ghir_A06G008070  Ghir_D06G008300  Ghir_A06  20265136  20268037   \n",
      "9797  Ghir_A06G008090  Ghir_D06G008310  Ghir_A06  20429594  20431519   \n",
      "9798  Ghir_A06G008100  Ghir_D06G008320  Ghir_A06  20432621  20435043   \n",
      "9799  Ghir_A06G008120  Ghir_D06G008330  Ghir_A06  20464622  20466555   \n",
      "9800  Ghir_A06G008130  Ghir_D06G008350  Ghir_A06  20493938  20496252   \n",
      "9801  Ghir_A06G008140  Ghir_D06G008360  Ghir_A06  20496258  20497564   \n",
      "9802  Ghir_A06G008150  Ghir_D06G008370  Ghir_A06  20547877  20548299   \n",
      "9803  Ghir_A06G008160  Ghir_D06G008380  Ghir_A06  20549380  20549713   \n",
      "9804  Ghir_A06G008170  Ghir_D06G008390  Ghir_A06  20549787  20550004   \n",
      "9805  Ghir_A06G008180  Ghir_D06G008400  Ghir_A06  20550570  20553752   \n",
      "9806  Ghir_A06G008200  Ghir_D06G008430  Ghir_A06  20608736  20614696   \n",
      "9807  Ghir_A06G008210  Ghir_D06G008440  Ghir_A06  20614899  20617190   \n",
      "9808  Ghir_A06G008220  Ghir_D06G008450  Ghir_A06  20712519  20714210   \n",
      "9809  Ghir_A06G008240  Ghir_D06G008460  Ghir_A06  20976818  20978444   \n",
      "9810  Ghir_A06G008250  Ghir_D06G008470  Ghir_A06  20978995  20980283   \n",
      "9811  Ghir_A06G008260  Ghir_D06G008480  Ghir_A06  20982301  20985225   \n",
      "9812  Ghir_A06G008270  Ghir_D06G008490  Ghir_A06  21035709  21039424   \n",
      "\n",
      "        chr_dt  start_dt    end_dt  \n",
      "9795  Ghir_D06  14472590  14477390  \n",
      "9796  Ghir_D06  14477264  14480211  \n",
      "9797  Ghir_D06  14512049  14514143  \n",
      "9798  Ghir_D06  14518098  14520481  \n",
      "9799  Ghir_D06  14581789  14584697  \n",
      "9800  Ghir_D06  14622117  14625794  \n",
      "9801  Ghir_D06  14625007  14625429  \n",
      "9802  Ghir_D06  14650685  14653187  \n",
      "9803  Ghir_D06  14652455  14652787  \n",
      "9804  Ghir_D06  14652862  14653101  \n",
      "9805  Ghir_D06  14653344  14657133  \n",
      "9806  Ghir_D06  14679133  14685102  \n",
      "9807  Ghir_D06  14685123  14687714  \n",
      "9808  Ghir_D06  14746881  14748589  \n",
      "9809  Ghir_D06  14797519  14800044  \n",
      "9810  Ghir_D06  14800086  14801207  \n",
      "9811  Ghir_D06  14803150  14806085  \n",
      "9812  Ghir_D06  14844961  14848748  \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/all_homo_gene_pairs_position.txt\" , sep=\"\\t\")\n",
    "#at_start, at_end, dt_start, dt_end = 82780000, 83180000, 19300000, 19580000\n",
    "# at_start, at_end, dt_start, dt_end = 82780000, 83180000, 43640000, 44160000\n",
    "# at_chr, dt_chr = 'Ghir_A10', 'Ghir_D10'\n",
    "# df_need = df.query(\"chr_at==@at_chr & chr_dt==@dt_chr & start_at>=@at_start & end_at<=@at_end & start_dt>=@dt_start & end_dt<=@dt_end\").copy()\n",
    "\n",
    "#dt_chr, dt_start, dt_end = 'Ghir_D07', 51180000, 51580000\n",
    "dt_chr, dt_start, dt_end = 'Ghir_D06', 14460000, 14860000\n",
    "\n",
    "df_need = df.query(\"chr_dt==@dt_chr &  start_dt>=@dt_start & end_dt<=@dt_end\").copy()\n",
    "print(df_need)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析组织特异性基因在TAD边界处的距离\n",
    "import pandas as pd\n",
    "tissues = ['anther', 'cotyledon', 'fiber_20DPA', 'ovule', 'root', 'radicle', 'leaf', 'hypocotyl', 'stem', 'fiber_10DPA', 'petal', 'stigma']\n",
    "input_dir = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_gene\"\n",
    "df_result = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    input_f = \"{0}/{1}_TAD_gene.csv\".format(input_dir, t)\n",
    "    df_input = pd.read_csv(input_f)\n",
    "    df_input[['TAD_chr', 'TAD_start', 'TAD_end']] = df_input['tags'].str.split(\"-\", expand=True)\n",
    "    df_input['start_distance'] = df_input['start'].map(int) - df_input['TAD_start'].map(int) ## 基因开始位点与TAD边界间的距离\n",
    "    df_input['end_distance'] = df_input['TAD_end'].map(int) - df_input['end'].map(int) ## 基因终止位点与TAD边界间的距离\n",
    "    df_input['TAD_distance'] = df_input[['start_distance', 'end_distance']].min(axis=1) ## 起始位点与终止位点间距离TAD 边界的最小距离\n",
    "    df_input['TAD_distance2'] = df_input['TAD_distance'].apply(lambda x: 0 if x<0 else x)\n",
    "    df_input['tissues'] = [t]*df_input.shape[0]\n",
    "    df_result = pd.concat([df_result, df_input], axis=0)\n",
    "## 给基因打上亚基因组标签\n",
    "def subg_tags(df):\n",
    "    df['subg'] = df['chrs'].apply(lambda x: 'At' if 'Ghir_A' in str(x) else 'Dt')\n",
    "    return df \n",
    "\n",
    "## 计算距离\n",
    "def distance(df, tags):\n",
    "    #df_tmp = df[['chrs', 'start', 'end', 'GeneId', 'TAD_start', 'TAD_end']].copy()\n",
    "    df_tmp = df.copy()\n",
    "    if (tags=='start'):\n",
    "        df_tmp['TSS_site'] = df_tmp['start']\n",
    "    else:\n",
    "        df_tmp['TSS_site'] = df_tmp['end']\n",
    "    df_tmp['TSS_TAD_start'] = df_tmp['TSS_site'] - df_tmp['TAD_start'].map(int)\n",
    "    df_tmp['TSS_TAD_end'] = df_tmp['TAD_end'].map(int) - df_tmp['TSS_site']\n",
    "    df_tmp['TSS_TAD_distance'] = df_tmp[['TSS_TAD_end', 'TSS_TAD_start']].min(axis=1)\n",
    "    df_tmp['TSS_TAD_distance2'] = df_tmp['TSS_TAD_distance'].apply(lambda x: 0 if x<0 else x)\n",
    "    return df_tmp\n",
    "    \n",
    "## 转录起始位点距离TAD边界的具体\n",
    "def TSS_distance(df):\n",
    "    df_gene = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/all_gene_sorted.bed6\", sep=\"\\t\", names = ['chrs', 'start', 'end', 'GeneId', 'a', 'directory'])\n",
    "    df_gene = df_gene[['GeneId', 'directory']].copy()\n",
    "    df_tmp = pd.merge(df, df_gene, how='left', on=['GeneId'])\n",
    "    df_tmp_a = df_tmp.query(\"directory=='+'\").copy()\n",
    "    df_tmp_b = df_tmp.query(\"directory=='-'\").copy()\n",
    "    df_tmp_a = distance(df_tmp_a, 'start')\n",
    "    df_tmp_b = distance(df_tmp_b, 'end')\n",
    "    df_tmp = pd.concat([df_tmp_a, df_tmp_b], axis=0)\n",
    "    df_tmp['TSS_TAD_distance_ratio'] =  (df_tmp['TSS_TAD_distance2']/(df_tmp['TAD_end'].map(int) - df_tmp['TAD_start'].map(int)))*100\n",
    "    return df_tmp\n",
    "## 组织特异性表达基因打上标签\n",
    "def TSG_file(df):\n",
    "    df_TSG = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/TAU/J668_all_Tissue_TAU_geneid.csv\")\n",
    "    df_TSG_TAD = pd.merge(df, df_TSG, how='inner', on=['GeneId', 'tissues'])\n",
    "    df_TSG_TAD['TSG_tags'] = ['TSG']*df_TSG_TAD.shape[0] \n",
    "    \n",
    "    df_return = pd.concat([df_TSG_TAD, df], axis=0)\n",
    "    df_return.drop_duplicates(keep='first', inplace=True, subset=['GeneId', 'tissues'])\n",
    "    df_return.fillna('No_TSG', inplace=True)\n",
    "    df_return = subg_tags(df_return)\n",
    "    df_return['TAD_distance_ratio'] = (df_return['TAD_distance2']/(df_return['TAD_end'].map(int) - df_return['TAD_start'].map(int)))*100\n",
    "    df_return = TSS_distance(df_return)\n",
    "    return df_return\n",
    "df_result = TSG_file(df_result)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_gene/J668_all_tissues_gene_TAD_distances.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将特定组织中所有其它组织的TSG选出来，分析其位置\n",
    "import pandas as pd\n",
    "df_TAU = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/TAU/J668_all_Tissue_TAU_geneid.csv\")\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_gene/J668_all_tissues_gene_TAD_distances.csv\")\n",
    "df_fiber_20DPA = df_TAU.query(\"tissues=='fiber_20DPA'\")[['GeneId']].copy()\n",
    "df_fiber_20DPA = pd.merge(df, df_fiber_20DPA, how='inner', on='GeneId')\n",
    "df_fiber_20DPA.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_gene/J668_TSG_fiber_20DPA_TAD_distances.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 划分TAD boundary和TAD center\n",
    "import pandas as pd\n",
    "tissues = ['anther', 'cotyledon', 'fiber_10DPA', 'fiber_20DPA', 'hypocotyl', 'leaf', 'ovule', 'petal', 'radicle', 'root', 'stem', 'stigma']\n",
    "for t in tissues:\n",
    "    boundary_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/boundary_interior/{}_boundary.bed\".format(t)\n",
    "    interior_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/boundary_interior/{}_interior.bed\".format(t)\n",
    "    df_boundary = pd.DataFrame()\n",
    "    df_interior = pd.DataFrame()\n",
    "    for i in range(5):\n",
    "        input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region/{0}_TAD{1}.bed\".format(t, i)\n",
    "        df_tmp = pd.read_csv(input_f, sep=\"\\t\", names=['chrom', 'start', 'end'])\n",
    "        if i == 0 or i == 4:\n",
    "            df_boundary = pd.concat([df_boundary, df_tmp], axis=0)\n",
    "        else:\n",
    "            df_interior = pd.concat([df_interior, df_tmp], axis=0)\n",
    "    df_boundary.to_csv(boundary_f, sep=\"\\t\", index=False)\n",
    "    df_interior.to_csv(interior_f, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 确定TAD 边界位点\n",
    "import pandas as pd\n",
    "tissues = ['anther', 'cotyledon', 'fiber_10DPA', 'fiber_20DPA', 'hypocotyl', 'leaf', 'ovule', 'petal', 'radicle', 'root', 'stem', 'stigma']\n",
    "def TAD_boundary(df):\n",
    "    df_a = df[['chr', 's']].copy()\n",
    "    df_b = df[['chr', 'e']].copy()\n",
    "    df_a.columns = ['chr', 'boundary']\n",
    "    df_b.columns = ['chr', 'boundary']\n",
    "    df = pd.concat([df_a, df_b], axis=0)\n",
    "    df.drop_duplicates(keep='first', inplace=True)\n",
    "    return df\n",
    "for t in tissues:\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_data/{}_choosed.bed\".format(t)\n",
    "    output_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_data/{}_boundary_site.bed\".format(t)\n",
    "    df = pd.read_csv(input_f, sep=\"\\t\")\n",
    "    df_boundary = TAD_boundary(df)\n",
    "    df_boundary.to_csv(output_f, sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##统计确定保守的TAD boundary, 20230524\n",
    "import pandas as pd\n",
    "tissues = ['anther', 'ovule', 'root', 'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "dirs='/cotton/public/public_data/xhhuang/Jin668_AllTissues_3Dgenome/Hi-C/TAD/conserved_dynamic_TAD_boundary'\n",
    "# one\n",
    "def conserved_dynamic_TAD():\n",
    "    df_return = pd.DataFrame(columns=['control', 'treatment', 'conserved_boundary', 'total_boundary'])\n",
    "    for i in range(11):\n",
    "        for j in range(i+1, 12):\n",
    "            T1, T2 = tissues[i], tissues[j]\n",
    "            T1_input = '{0}/{1}_{2}.bed'.format(dirs, T1, T2)\n",
    "            T2_input = '{0}/{1}_{2}.bed'.format(dirs, T2, T1)\n",
    "            df_T1 = pd.read_csv(T1_input, names=['chorm', 'bounrady', 'tags'], sep=\"\\t\")\n",
    "            df_T2 = pd.read_csv(T2_input, names=['chorm', 'bounrady', 'tags'], sep=\"\\t\")\n",
    "            total_boundary = df_T1.shape[0] + df_T2.shape[0]\n",
    "            conserved_boundary = df_T1.query(\"tags>0\").shape[0] + df_T2.query(\"tags>0\").shape[0]\n",
    "            df_return.loc[df_return.shape[0], ] = [T1,T2, conserved_boundary, total_boundary]\n",
    "    df_return['ratio'] = df_return['conserved_boundary']/df_return['total_boundary']\n",
    "    return df_return\n",
    "# two \n",
    "def one_sample_conserved():\n",
    "    for i in tissues:\n",
    "        output_f = \"{0}/{1}_boundary_statistics.csv\".format(dirs, i)\n",
    "        df_output = pd.DataFrame()\n",
    "        tags = []\n",
    "        for j in tissues:\n",
    "            T1, T2 = i, j\n",
    "            if i != j:\n",
    "                tags.append('{0}_{1}'.format(T1, T2))\n",
    "                T1_input = '{0}/{1}.bed'.format(dirs, T1, T2)\n",
    "                df_T1 = pd.read_csv(T1_input, names=['chorm', 'bounrady', '{0}_{1}'.format(T1, T2)], sep=\"\\t\")\n",
    "                if df_output.shape[0] == 0:\n",
    "                    df_output = df_T1\n",
    "                else:\n",
    "                    df_output = pd.concat(df_output, df_T1, how='inner', on=['chrom', 'boundary'])\n",
    "        df_output['total'] = df_output[tags].sum(axis=1)\n",
    "        df_output.to_csv(output_f, index=False)\n",
    "\n",
    "## statistics all conserved boundary\n",
    "def all_conserved_boundary():\n",
    "    df_result = pd.DataFrame()\n",
    "    for t in tissues:\n",
    "        input_f = \"{0}_boundary_statistics.csv\".format(t)\n",
    "        df = pd.read_csv(input_f)\n",
    "        df['tags'] = [t]*df.shape[0]\n",
    "        df_result = pd.concat([df_result, df[['chrom', 'boundary', 'total', 'tags']]], axis=0)\n",
    "    df_result['tags2'] = df_result['total'].apply(lambda x: \"conserverd\" if x==11 else (\"dynamic\" if x>0 else \"special\"))\n",
    "    df_result.to_csv(\"J668_Mutiple_tissues_boundary_conserved_statistics.csv\", index=False)\n",
    "\n",
    "\n",
    "# df_result = conserved_dynamic_TAD()\n",
    "# df_result.to_csv(\"J668_Mutiple_tissues_conserved_dynamic_boundary.bed\", sep=\"\\t\", index=False)\n",
    "\n",
    "#df_result = one_sample_conserved()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           chrom     start       end        is tissues  order       tags\n",
      "345     Ghir_A01   6900000   6920000  0.297314  anther      0  conserved\n",
      "346     Ghir_A01   6920000   6940000  0.314435  anther      1  conserved\n",
      "347     Ghir_A01   6940000   6960000  0.090598  anther      2  conserved\n",
      "348     Ghir_A01   6960000   6980000  0.059191  anther      3  conserved\n",
      "349     Ghir_A01   6980000   7000000 -0.062873  anther      4  conserved\n",
      "...          ...       ...       ...       ...     ...    ...        ...\n",
      "893032  Ghir_D13  51200000  51220000 -0.284395  stigma     15    dynamic\n",
      "893033  Ghir_D13  51220000  51240000 -0.098515  stigma     16    dynamic\n",
      "893034  Ghir_D13  51240000  51260000  0.038385  stigma     17    dynamic\n",
      "893035  Ghir_D13  51260000  51280000  0.285436  stigma     18    dynamic\n",
      "893036  Ghir_D13  51280000  51300000  0.395037  stigma     19    dynamic\n",
      "\n",
      "[1248638 rows x 7 columns]\n"
     ]
    }
   ],
   "source": [
    "## 分析确定保守boundary、动态boundary和special boundary 周围IS 的变化\n",
    "import pandas as pd\n",
    "# 分析各类TAD上下所包含的TAD boundary\n",
    "df_boundary = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/conserved_TAD/J668_Mutiple_tissues_boundary_conserved_statistics.csv\")\n",
    "df_IS = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/conserved_TAD/J668_Multip_tissues_boundary_IS.bed\", sep=\"\\t\")\n",
    "df_boundary.rename(columns={'tags':'tissues'}, inplace=True)\n",
    "\n",
    "df_result = pd.DataFrame()\n",
    "for index, row in df_boundary.iterrows():\n",
    "    C = row['chrom']\n",
    "    S = row['boundary'] - 200000\n",
    "    E = row['boundary'] + 200000\n",
    "    T = row['tags2']\n",
    "    Tiss = row['tissues']\n",
    "    df_tmp = df_IS.query(\"chrom==@C & start>=@S & end<=@E & tissues==@Tiss\").copy()\n",
    "    df_tmp['order'] = range(df_tmp.shape[0])\n",
    "    df_tmp['tags'] = [T]*df_tmp.shape[0]\n",
    "    df_tmp['tissues'] = [Tiss]*df_tmp.shape[0]\n",
    "    df_result = pd.concat([df_result, df_tmp],axis=0)\n",
    "## 统计各类型TAD 边界保守的loops\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/conserved_TAD/J668_conserved_boundary_IS.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将所有的保守boundary的IS取个平均值\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/conserved_TAD/J668_conserved_boundary_IS.csv\")\n",
    "df_result = df.groupby(by=['tags', 'order'], as_index=False)['is'].mean()\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/conserved_TAD/J668_conserved_boundary_IS_mean.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 将数据整理为coorplot 可识别格式\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/conserved_TAD/J668_Mutiple_tissues_conserved_dynamic_boundary.bed\", sep=\"\\t\")\n",
    "df_upper = df[['control', 'treatment', 'ratio']].copy()\n",
    "df_low = df[['treatment', 'control', 'ratio']].copy()\n",
    "df_low.columns = ['control', 'treatment', 'ratio']\n",
    "tissues = ['anther', 'ovule', 'root', 'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "ratios = [0.85, 0.85, 0.85, 0.85, 0.85, 0.85, 0.85, 0.85, 0.85, 0.85, 0.85, 0.85]\n",
    "df_center = pd.DataFrame({'control':tissues, 'treatment':tissues, 'ratio':ratios})\n",
    "df_all = pd.concat([df_upper, df_low, df_center], axis=0)\n",
    "df_wide = df_all.pivot(index='control', columns='treatment', values='ratio')\n",
    "df_wide.to_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/conserved_TAD/J668_dynamic_TAD_coorplot.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#计算所有TAD的数目\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/conserved_TAD/J668_Mutiple_tissues_boundary_conserved_statistics.csv\")\n",
    "df['number'] = [1]*df.shape[0]\n",
    "df_sta = df.groupby(by=['total'], as_index=False)['number'].sum()\n",
    "df_sta.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/conserved_TAD/J668_conserved_TAD_number.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "conserverd    0.564080\n",
      "dynamic       0.427499\n",
      "special       0.008421\n",
      "Name: tags2, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "## 计算保守TAD，动态TAD和孤立TAD的数目\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/conserved_TAD/J668_Mutiple_tissues_boundary_conserved_statistics.csv\")\n",
    "df_v = df['tags2'].value_counts()\n",
    "c = df_v/df.shape[0]\n",
    "print(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 在分析保守boundary时候，对于起始为0的TAD 边界加了一个40Kb，因此需要替换回来\n",
    "import pandas as pd\n",
    "df_boundary = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/conserved_TAD/J668_Mutiple_tissues_boundary_conserved_statistics.csv\")\n",
    "df_boundary['boundary'].replace(40000, 0, inplace=True)\n",
    "df_boundary.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/conserved_TAD/J668_Mutiple_tissues_boundary_conserved_statistics.csv\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "anther\n",
      "ovule\n",
      "root\n",
      "radicle\n",
      "leaf\n",
      "hypocotyl\n",
      "cotyledon\n",
      "stem\n",
      "fiber_10DPA\n",
      "fiber_20DPA\n",
      "petal\n",
      "stigma\n"
     ]
    }
   ],
   "source": [
    "# 根据保守和动态TAD boundary确定保守和动态TAD\n",
    "import pandas as pd\n",
    "df_boundary = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/conserved_TAD/J668_Mutiple_tissues_boundary_conserved_statistics.csv\")\n",
    "tissues = ['anther', 'ovule', 'root', 'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df_result = pd.DataFrame()\n",
    "output_f ='G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/conserved_TAD/J668_conserved_TAD.csv' \n",
    "for t in tissues:\n",
    "    print(t)\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_data/{}_choosed.bed\".format(t)\n",
    "    df = pd.read_csv(input_f, sep=\"\\t\")\n",
    "    for index, row in df.iterrows():\n",
    "        chrom, start, end = row['chr'], row['s'], row['e']\n",
    "        #print(t, chrom, start, end)\n",
    "        # boundary one\n",
    "        df_tmp = df_boundary.query(\"tags==@t & chrom==@chrom & boundary==@start\")\n",
    "        df_tmp.reset_index(drop=True, inplace=True)\n",
    "        tags1 = df_tmp.loc[0, 'tags2']\n",
    "        df.loc[index, 'tags1'] = tags1\n",
    "        # boundary two\n",
    "        df_tmp = df_boundary.query(\"tags==@t & chrom==@chrom & boundary==@end\")\n",
    "        df_tmp.reset_index(drop=True, inplace=True)\n",
    "        tags2 = df_tmp.loc[0, 'tags2']\n",
    "        df.loc[index, 'tags2'] = tags2\n",
    "        ## 确定哪些TAD是保守TAD，哪些是动态TAD，哪些是特有TAD\n",
    "        if tags1=='conserved' and tags2=='conserved':\n",
    "            tags = 'conserved'\n",
    "        elif tags1 == 'special' and tags2=='special':\n",
    "            tags = 'special'\n",
    "        else:\n",
    "            tags = 'dynamic'\n",
    "        df.loc[index, 'tags'] = tags\n",
    "    df['tissues'] = [t]*df.shape[0]\n",
    "    df_result = pd.concat([df_result, df], axis=0)\n",
    "df_result.to_csv(output_f, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析孤岛TAD是否在组织中更加的动态相比其它类型TAD\n",
    "import pandas as pd\n",
    "df_TAD_type = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/J668_all_Tissues_TAD_type.csv\")\n",
    "df_TAD_conserved = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/conserved_TAD/J668_conserved_TAD.csv\")\n",
    "df_TAD_conserved = df_TAD_conserved[['chr', 's', 'e', 'tags', 'tissues']]\n",
    "df_TAD_conserved.columns = ['chrom', 'start', 'end', 'tags', 'tissues' ]\n",
    "df_TAD_type.rename(columns={'tags':'tissues'}, inplace=True)\n",
    "df_TAD = pd.merge(df_TAD_type, df_TAD_conserved, how='inner', on=['chrom', 'start', 'end', 'tissues'])\n",
    "df_TAD['number'] = [1]*df_TAD.shape[0]\n",
    "df_result = df_TAD.groupby(by=['type', 'tags'], as_index=False)['number'].sum()\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/conserved_TAD/J668_conserved_TAD_type_statistics.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "#分析基因位于TAD 边界和TAD center处所包含的差异表达基因数目\n",
    "## 确定所有基因是boundary 还是interior基因\n",
    "tissues = ['anther', 'ovule', 'root', 'radicle', 'leaf', 'hypocotyl', 'cotyledon', 'stem', 'fiber_10DPA', 'fiber_20DPA', 'petal', 'stigma']\n",
    "df_result = pd.DataFrame()\n",
    "for t in tissues:\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_gene/{}_TAD_quintiles_gene.bed\".format(t)\n",
    "    df = pd.read_csv(input_f, sep=\"\\t\", names=['GeneId', 'tags'])\n",
    "    old_tags = [0, 1, 2, 3, 4]\n",
    "    new_tags = ['boundary', 'interior', 'interior', 'interior', 'boundary']\n",
    "    df['tags'].replace(old_tags, new_tags, inplace=True)\n",
    "    df.drop_duplicates(keep='first', inplace=True, subset=['GeneId'])\n",
    "    df.rename(columns={'tags':t}, inplace=True)\n",
    "    if df_result.shape[0] == 0:\n",
    "        df_result = df.copy()\n",
    "    else:\n",
    "        df_result = pd.merge(df_result, df, how='outer', on=['GeneId'])\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_gene/J668_gene_boundary_interior.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 比较任意两个组织分析有多少基因差异表达与所处TAD位置不同相关\n",
    "df_gene_TB = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_gene/J668_gene_boundary_interior.csv\")\n",
    "\n",
    "# 分析确定差异表达基因数目1、基因位于TAD上的差异表达基因数目。 2、在TAD上位置分布具有差异的差异表达基因数目。3、TAD位置分布不同且有符合边界基因表达更高的差异表达基因\n",
    "\n",
    "def ABC(df):\n",
    "    A = df.shape[0]\n",
    "    B = df.query(\"C!=T\").shape[0]\n",
    "    C = df.query(\"(sig=='Down' & C=='boundary' & T=='interior') | (sig=='Up' & C=='interior' & T=='boundary')\").shape[0]\n",
    "    return A, B, C\n",
    "df_result = pd.DataFrame(columns=['control', 'treatment', 'A', 'B', 'C'])\n",
    "for i in range(11):\n",
    "    for j in range(i+1, 12):\n",
    "        C, T = tissues[i], tissues[j]\n",
    "        df_TB_CT = df_gene_TB[['GeneId', C, T]].copy()\n",
    "        df_TB_CT.columns = ['GeneId', 'C', 'T']\n",
    "        input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/DEG/tissue/{0}_{1}_DEG.csv\".format(C, T)\n",
    "        df = pd.read_csv(input_f)\n",
    "        df_up_down = df.query(\"sig!='Not'\").copy()\n",
    "        df_overlap = pd.merge(df_up_down, df_TB_CT, how='inner', on=['GeneId'])\n",
    "        df_overlap.dropna(inplace=True)\n",
    "        A, B, K = ABC(df_overlap)\n",
    "        df_result.loc[df_result.shape[0]] = [C, T, A, B, K]\n",
    "df_result['tags'] = df_result['control'].map(str) + \"_\" + df_result['treatment'].map(str)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_quintiles/TAD_region_gene/DEG/J668_DEG_boundary_interior.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析亚基因组孤岛基因bias表达\n",
    "# 以下胚轴为例子说明孤岛TAD影响同源基因bias表达, At是孤岛基因,Dt是非孤岛基因，同时Dt bias\n",
    "import pandas as pd\n",
    "df = pd.read_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/HG/J668_Isolated_TAD_HG_pair.csv')\n",
    "df_hypocotyl = df.query(\"tissue=='hypocotyl' & type_At=='Isolated' & type_Dt=='NIsolated'\").copy()\n",
    "df_hypocotyl.rename(columns={'tissue':'tissues'}, inplace=True)\n",
    "df_bias = pd.read_csv('G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/all_tissue_bias_gene_pair.bed', sep=\"\\t\")\n",
    "df_bias_hypocotyl = df_bias.query(\"tissues=='hypocotyl'\").copy()\n",
    "df_result = pd.merge(df_hypocotyl, df_bias_hypocotyl, how='inner', on=['GeneId_At', 'GeneId_Dt', 'tissues'])\n",
    "df_hypocotyl_Bias_Dt = df_result.query(\"sig=='BiasD'\").copy()\n",
    "df_hypocotyl_Bias_Dt.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/hypocotyl_bias_Dt/hypocotyl_AtIso_DtNIso_BiasDt.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析这些基因所在的TAD\n",
    "import pandas as pd\n",
    "df_HTAD = pd.read_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/hypocotyl_bias_Dt/hypocotyl_AtIso_DtNIso_BiasDt.csv')\n",
    "df_At = df_HTAD[['GeneId_At']].copy()\n",
    "df_Dt = df_HTAD[['GeneId_Dt']].copy()\n",
    "df_At['number'] = range(df_At.shape[0])\n",
    "df_Dt['number'] = range(df_Dt.shape[0])\n",
    "df_At.columns = ['GeneId', 'number']\n",
    "df_Dt.columns = ['GeneId', 'number']\n",
    "df_gene = pd.concat([df_At, df_Dt], axis=0)\n",
    "df_gene_TAD = pd.read_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_gene/hypocotyl_TAD_gene.csv')\n",
    "df_result = pd.merge(df_gene, df_gene_TAD, how='inner', on=['GeneId'])\n",
    "df_result[['TAD_chr', 'TAD_start', 'TAD_end']] = df_result['tags'].str.split('-', expand=True)\n",
    "df_result.to_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/hypocotyl_bias_Dt/hypocotyl_AtIso_DtNIso_BiasDt_TAD.csv', index=False)\n",
    "df_result.to_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/hypocotyl_bias_Dt/hypocotyl_AtIso_DtNIso_BiasDt_TAD.bedpe', sep=\"\\t\", index=False, columns=['TAD_chr', 'TAD_start', 'TAD_end','TAD_chr', 'TAD_start', 'TAD_end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析亚基因组孤岛基因bias表达\n",
    "# 以下胚轴为例子说明孤岛TAD影响同源基因bias表达, At是孤岛基因,Dt是非孤岛基因，同时Dt bias\n",
    "import pandas as pd\n",
    "df = pd.read_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/HG/J668_Isolated_TAD_HG_pair.csv')\n",
    "df_hypocotyl = df.query(\"tissue=='stigma' & type_At=='Isolated' & type_Dt=='NIsolated'\").copy()\n",
    "df_hypocotyl.rename(columns={'tissue':'tissues'}, inplace=True)\n",
    "df_bias = pd.read_csv('G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/all_tissue_bias_gene_pair.bed', sep=\"\\t\")\n",
    "df_bias_hypocotyl = df_bias.query(\"tissues=='stigma'\").copy()\n",
    "df_result = pd.merge(df_hypocotyl, df_bias_hypocotyl, how='inner', on=['GeneId_At', 'GeneId_Dt', 'tissues'])\n",
    "df_hypocotyl_Bias_Dt = df_result.query(\"sig=='BiasD'\").copy()\n",
    "df_hypocotyl_Bias_Dt.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/stigma_bias_Dt/stigma_AtIso_DtNIso_BiasDt.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析这些基因所在的TAD\n",
    "import pandas as pd\n",
    "df_HTAD = pd.read_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/stigma_bias_Dt/stigma_AtIso_DtNIso_BiasDt.csv')\n",
    "df_At = df_HTAD[['GeneId_At']].copy()\n",
    "df_Dt = df_HTAD[['GeneId_Dt']].copy()\n",
    "df_At['number'] = range(df_At.shape[0])\n",
    "df_Dt['number'] = range(df_Dt.shape[0])\n",
    "df_At.columns = ['GeneId', 'number']\n",
    "df_Dt.columns = ['GeneId', 'number']\n",
    "df_gene = pd.concat([df_At, df_Dt], axis=0)\n",
    "df_gene_TAD = pd.read_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_gene/stigma_TAD_gene.csv')\n",
    "df_result = pd.merge(df_gene, df_gene_TAD, how='inner', on=['GeneId'])\n",
    "df_result[['TAD_chr', 'TAD_start', 'TAD_end']] = df_result['tags'].str.split('-', expand=True)\n",
    "df_result.to_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/stigma_bias_Dt/stigma_AtIso_DtNIso_BiasDt_TAD.csv', index=False)\n",
    "df_result.to_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/stigma_bias_Dt/stigma_AtIso_DtNIso_BiasDt_TAD.bedpe', sep=\"\\t\", index=False, columns=['TAD_chr', 'TAD_start', 'TAD_end','TAD_chr', 'TAD_start', 'TAD_end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析亚基因组孤岛基因bias表达\n",
    "# 以下胚轴为例子说明孤岛TAD影响同源基因bias表达, At是孤岛基因,Dt是非孤岛基因，同时Dt bias\n",
    "import pandas as pd\n",
    "df = pd.read_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/HG/J668_Isolated_TAD_HG_pair.csv')\n",
    "df_hypocotyl = df.query(\"tissue=='ovule' & type_At=='Isolated' & type_Dt=='NIsolated'\").copy()\n",
    "df_hypocotyl.rename(columns={'tissue':'tissues'}, inplace=True)\n",
    "df_bias = pd.read_csv('G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/homo_gene2/gene_bias/all_tissue_bias_gene_pair.bed', sep=\"\\t\")\n",
    "df_bias_hypocotyl = df_bias.query(\"tissues=='ovule'\").copy()\n",
    "df_result = pd.merge(df_hypocotyl, df_bias_hypocotyl, how='inner', on=['GeneId_At', 'GeneId_Dt', 'tissues'])\n",
    "df_hypocotyl_Bias_Dt = df_result.query(\"sig=='BiasD'\").copy()\n",
    "df_hypocotyl_Bias_Dt.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/ovule_bias_Dt/ovule_AtIso_DtNIso_BiasDt.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析这些基因所在的TAD\n",
    "import pandas as pd\n",
    "df_HTAD = pd.read_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/ovule_bias_Dt/ovule_AtIso_DtNIso_BiasDt.csv')\n",
    "df_At = df_HTAD[['GeneId_At']].copy()\n",
    "df_Dt = df_HTAD[['GeneId_Dt']].copy()\n",
    "df_At['number'] = range(df_At.shape[0])\n",
    "df_Dt['number'] = range(df_Dt.shape[0])\n",
    "df_At.columns = ['GeneId', 'number']\n",
    "df_Dt.columns = ['GeneId', 'number']\n",
    "df_gene = pd.concat([df_At, df_Dt], axis=0)\n",
    "df_gene_TAD = pd.read_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/TAD_gene/ovule_TAD_gene.csv')\n",
    "df_result = pd.merge(df_gene, df_gene_TAD, how='inner', on=['GeneId'])\n",
    "df_result[['TAD_chr', 'TAD_start', 'TAD_end']] = df_result['tags'].str.split('-', expand=True)\n",
    "df_result.to_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/ovule_bias_Dt/ovule_AtIso_DtNIso_BiasDt_TAD.csv', index=False)\n",
    "df_result.to_csv('G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/isolated_TAD/ovule_bias_Dt/ovule_AtIso_DtNIso_BiasDt_TAD.bedpe', sep=\"\\t\", index=False, columns=['TAD_chr', 'TAD_start', 'TAD_end','TAD_chr', 'TAD_start', 'TAD_end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 统计各类型TAD的占比\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/J668_all_Tissues_TAD_type.csv\")\n",
    "df['subg'] = df['chrom'].apply(lambda x: 'At' if 'Ghir_A' in x else 'Dt')\n",
    "df_result = df.groupby(by=['tags', 'subg', 'type'], as_index=False)['size'].sum()\n",
    "df_result.columns = ['tissues', 'subg', 'sigs', 'number']\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_type/J668_TAD_type_length.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 统计组织特异高活性TAD占比\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/J668_TSG_TAD.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'tissues'])\n",
    "df['sigs'] = ['active TAD']*df.shape[0]\n",
    "df_TAD = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TAD_data/J668_TAD_allTissues.csv\")\n",
    "df_TAD.columns = ['chrom', 'start', 'end', 'order', 'tissues']\n",
    "df_TAD.drop(\"order\", axis=1, inplace=True)\n",
    "df_result = pd.merge(df_TAD, df, how='left', on=['chrom', 'start', 'end', 'tissues'])\n",
    "df_result.fillna(\"normal TAD\", inplace=True)\n",
    "df_result['subg'] = df_result['chrom'].apply(lambda x : 'At' if 'Ghir_A' in x else 'Dt')\n",
    "df_result['number'] = df_result['end'] - df_result['start']\n",
    "df_sta = pd.DataFrame(df_result.groupby(by=['tissues', 'subg', 'sigs'], axis=False)['number'].sum(), columns=['number'])\n",
    "df_sta.reset_index(inplace=True)\n",
    "df_sta.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/TAD_new/TSG_TAD/J668_TSG_TAD_length_sta.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.3 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "a0f669ea42a4dcea88d385f8f9b4c3d9b966e24234758831d1dbfdac925619a2"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
