{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# analysis micore environment in PETRE\n",
    "import pandas as pd\n",
    "import os\n",
    "import sys\n",
    "import argparse\n",
    "def main(argv):\n",
    "    #! init\n",
    "    parser = argparse.ArgumentParser(description=\"\")\n",
    "    parser.add_argument(\"-g\", \"--gene\", required=True, help=\"The gene file\")\n",
    "    parser.add_argument(\"-c\", \"--chip\", required=True, help=\"The chip file\")\n",
    "    parser.add_argument(\"-o\", \"--output\", required=True, help='The output file')\n",
    "    parser.add_argument(\"-l\", \"--length\", required='True', help=\"The length of region to be extend\")\n",
    "    parser.add_argument(\"-n\", \"--number\", required='True', help=\"The number of splits for the region\")\n",
    "    parser.add_argument(\"-cn\", \"--chrom_name\", required='True', help=\"The number of chromsome\")\n",
    "    args = parser.parse_args()\n",
    "    genef = args.gene\n",
    "    chipf = args.chip\n",
    "    L = int(args.length)\n",
    "    N = int(args.number)\n",
    "    cn = args.chrom_name\n",
    "    outputf = args.output\n",
    "    #! analysis\n",
    "    #* 1、region extend\n",
    "    df_gene = pd.read_csv(genef, sep=\"\\t\")\n",
    "    df_gene = region_extend(df_gene, L)\n",
    "    df_chip = pd.read_csv(chipf, sep=\"\\t\")\n",
    "    df_result = pd.DataFrame()\n",
    "    #* single chrom \n",
    "    #for c in ['Ghir_A01', 'Ghir_A02', 'Ghir_A03', 'Ghir_A04', 'Ghir_A05', 'Ghir_A06', 'Ghir_A07', 'Ghir_A08', 'Ghir_A09', 'Ghir_A10', 'Ghir_A11', 'Ghir_A12', 'Ghir_A13', 'Ghir_D01', 'Ghir_D02', 'Ghir_D03', 'Ghir_D04', 'Ghir_D05', 'Ghir_D06', 'Ghir_D07', 'Ghir_D08', 'Ghir_D09', 'Ghir_D10', 'Ghir_D11', 'Ghir_D12', 'Ghir_D13']:\n",
    "    for c in [cn]:\n",
    "        df_gene_chr = df_gene.query(\"chrom==@c\").copy()\n",
    "        df_chip_chr = df_chip.query(\"chrom==@c\").copy()\n",
    "        print(df_gene_chr)\n",
    "        print(df_chip_chr)\n",
    "        #* 2、region chip modified \n",
    "        df_result = region_chip(df_gene_chr, df_chip_chr, L, N)\n",
    "    df_result.to_csv(outputf)\n",
    "# extend the region\n",
    "def region_extend(df, length):\n",
    "    df['s'] = df['start'] - length\n",
    "    df['e'] = df['end'] + length\n",
    "    df_return = df[['chrom', 's', 'e', 'GeneId']]\n",
    "    df_return.columns = ['chrom', 'start', 'end', 'GeneId']\n",
    "    return df_return\n",
    "# find chip modified on the region\n",
    "def region_chip(df_gene, df_chip, length, number):\n",
    "    for index, row in df_gene.iterrows():\n",
    "        ## add geneid region\n",
    "        C, S, E, G = row['chrom'], row['start'], row['end'], row['GeneId']\n",
    "        average_length = int((E-S)/number)\n",
    "        gene_region_length = E-S\n",
    "        #print(average_length)\n",
    "        for i in range(number):\n",
    "            tags_n = 'E{}'.format(i)\n",
    "            S_tmp = S\n",
    "            E_tmp = S + average_length\n",
    "            df_tmp = df_chip.query(\"chrom==@C & start <= @E_tmp & end>= @S_tmp\").copy()\n",
    "            tags_number = df_tmp.shape[0]\n",
    "            df_gene.loc[index, tags_n] = tags_number\n",
    "            S = E_tmp+1\n",
    "    return df_gene\n",
    "if __name__ == \"__main__\":\n",
    "    main(sys.argv[1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 合并PETERs 周围ATAC, H3K4me3, H3K27ac 信号，并且用umap降维聚类\n",
    "import pandas as pd\n",
    "chroms = ['Ghir_A01', 'Ghir_A02', 'Ghir_A03', 'Ghir_A04', 'Ghir_A05', 'Ghir_A06', 'Ghir_A07', 'Ghir_A08', 'Ghir_A09', 'Ghir_A10', 'Ghir_A11', 'Ghir_A12', 'Ghir_A13', \n",
    "          'Ghir_D01', 'Ghir_D02', 'Ghir_D03', 'Ghir_D04', 'Ghir_D05', 'Ghir_D06', 'Ghir_D07', 'Ghir_D08', 'Ghir_D09', 'Ghir_D10', 'Ghir_D11', 'Ghir_D12', 'Ghir_D13']\n",
    "#tags = ['ATAC', 'H3K27ac', 'H3K4me3']\n",
    "tags = ['H3K27ac', 'H3K4me3']\n",
    "df_result = pd.DataFrame()\n",
    "## 数据合并\n",
    "for t in tags:\n",
    "    df_tmp_tags = pd.DataFrame()\n",
    "    for c in chroms:\n",
    "        #input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/peaks_single/PATREs_peaks/{0}_PETERs_{1}.csv\".format(t, c)\n",
    "        input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/peaks_single/PATREs_peaks/hypocotyl_250bp/{0}_PETERs_{1}.csv\".format(t, c)\n",
    "        df = pd.read_csv(input_f)\n",
    "        df_tmp_tags = pd.concat([df_tmp_tags, df], axis=0)\n",
    "    if df_result.shape[0]==0:\n",
    "        df_result = df_tmp_tags.copy()\n",
    "    else:\n",
    "        df_result = pd.merge(df_result, df_tmp_tags, how='inner', on=['chrom', 'start', 'end', 'GeneId'])\n",
    "#umap 降维\n",
    "def zscore_umap(df):\n",
    "    import scipy.stats as stats\n",
    "    from sklearn.preprocessing import StandardScaler\n",
    "    import umap.umap_ as umap\n",
    "    #z-score\n",
    "    df.drop(['chrom', 'start', 'end'], axis=1, inplace=True)\n",
    "    df.set_index('GeneId', inplace=True)\n",
    "    df = df.loc[~(df==0).all(axis=1), :]\n",
    "    print(df.shape[0])\n",
    "    id_list = df.index.to_list()\n",
    "    df_numpy = df.to_numpy()\n",
    "    #df_zscore = stats.zscore(df_numpy, axis=1)\n",
    "    #df_zscore = pd.DataFrame(df_zscore)\n",
    "    df_zscore = pd.DataFrame(df_numpy)\n",
    "    # umap\n",
    "    reducer = umap.UMAP()\n",
    "    scaled_penguin_data = StandardScaler().fit_transform(df_zscore)\n",
    "    embedding = reducer.fit_transform(scaled_penguin_data)\n",
    "    df_umap = pd.DataFrame(embedding, columns =['umap1', 'umap2'], index=id_list)\n",
    "    return df_umap\n",
    "df_umap = zscore_umap(df_result)\n",
    "df_umap.reset_index(inplace=True)\n",
    "df_umap.rename(columns={'index':'PATRE_id'}, inplace=True)\n",
    "df_umap.to_csv('G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/peaks_single/Hypocotyl_PETRE_peaks_single_250bp.csv', index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 给umap 降维的结果也打上标签\n",
    "import pandas as pd\n",
    "df_umap = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/peaks_single/Hypocotyl_PETRE_peaks_single_250bp.csv\")\n",
    "df_patre = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/peaks_single/hypocotyl_PATRE_tags.csv\", names=['chrom', 'start', 'end', 'PATRE_id', 'tags'])\n",
    "df_result = pd.merge(df_umap, df_patre, how='outer', on=['PATRE_id'])\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/peaks_single/hypcotyl_PETRE_peaks_single_250bp_tags.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 合并PETERs 周围ATAC, H3K4me3, H3K27ac 信号，并且用umap降维聚类\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "def add_tags(df):\n",
    "    df = df.copy()\n",
    "    df[df<=0] = 0\n",
    "    df['total'] = df.select_dtypes(np.number).gt(0.1).sum(axis=1)\n",
    "    df['tags'] = df['total'].apply(lambda x: 'T' if x>=20 else 'F')\n",
    "    df.drop('total', inplace=True, axis=1)\n",
    "    return df\n",
    "\n",
    "def get_tags(df1, df2):\n",
    "    df1 = df1.copy()\n",
    "    df2 = df2.copy()\n",
    "    df1.set_index('PATRE_Id', inplace=True)\n",
    "    df2.set_index('PATRE_Id', inplace=True)\n",
    "    df1 = add_tags(df1)\n",
    "    df2 = add_tags(df2)\n",
    "    df1.reset_index(inplace=True)\n",
    "    df2.reset_index( inplace=True)\n",
    "    print(df1.head())\n",
    "    #print(df1)\n",
    "    df = pd.merge(df1, df2, how='inner', on=['PATRE_Id'])\n",
    "    df['tags'] = df['tags_x'].map(str) + \"-\" + df['tags_y'].map(str)\n",
    "    df_tags = df[['PATRE_Id', 'tags']].copy()\n",
    "    return df_tags\n",
    "\n",
    "\n",
    "#umap 降维\n",
    "def zscore_umap(df):\n",
    "    import scipy.stats as stats\n",
    "    from sklearn.preprocessing import StandardScaler\n",
    "    import umap.umap_ as umap\n",
    "    #z-score\n",
    "    #df.drop(['chrom', 'start', 'end'], axis=1, inplace=True)\n",
    "    df.set_index('PATRE_Id', inplace=True)\n",
    "    df = df.loc[~(df==0).all(axis=1), :]\n",
    "    df.fillna(0, inplace=True)\n",
    "    #df[df<=0] = 0\n",
    "    #print(df.shape[0])\n",
    "    id_list = df.index.to_list()\n",
    "    #print(df.shape)\n",
    "    df_numpy = df.to_numpy()\n",
    "    df_zscore = stats.zscore(df_numpy, axis=1)\n",
    "    df_zscore = pd.DataFrame(df_zscore)\n",
    "    #df_zscore = pd.DataFrame(df_numpy)\n",
    "    df_zscore.fillna(0, inplace=True)\n",
    "    # umap\n",
    "    reducer = umap.UMAP()\n",
    "    scaled_penguin_data = StandardScaler().fit_transform(df_zscore)\n",
    "    embedding = reducer.fit_transform(scaled_penguin_data)\n",
    "    df_umap = pd.DataFrame(embedding, columns =['umap1', 'umap2'], index=id_list)\n",
    "    return df_umap\n",
    "\n",
    "\n",
    "## 数据合并\n",
    "output_f = \"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/THR/J668_PATRE_UMAP.csv\"\n",
    "df_all = pd.DataFrame()\n",
    "for t in ['hypocotyl', 'ovule', 'stigma', 'radicle']:\n",
    "    for i in ['ATAC', 'H3K27ac', 'H3K4me3']:\n",
    "        input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/THR/{0}_{1}_1K_100bp.csv\".format(t, i)\n",
    "        df_tmp = pd.read_csv(input_f)\n",
    "        if df_all.shape[0] == 0:\n",
    "            df_all = df_tmp.copy()\n",
    "        else:\n",
    "            df_all = pd.merge(df_all, df_tmp, how='inner', on=['PATRE_Id'])\n",
    "df_annot = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/THR/J668_PATRE_annotation.bed\", sep=\" \")\n",
    "df_annot['promoter'] = df_annot['annotation'].apply(lambda x: -100 if 'Distal' in str(x) else 100)\n",
    "df_annot = df_annot[['V4', 'promoter']].copy()\n",
    "df_annot.columns = ['PATRE_Id', 'promoter']\n",
    "#print(df_annot)\n",
    "df_result = pd.merge(df_all, df_annot, how='left', on=['PATRE_Id'])\n",
    "#df_result = df_all.copy()\n",
    "#print(df_result.head())\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/THR/a.csv\", index=False)\n",
    "\n",
    "df_H3K27ac = pd.DataFrame()\n",
    "df_H3K4me3 = pd.DataFrame()\n",
    "## H3K27ac\n",
    "for i in ['hypocotyl', 'ovule', 'stigma', 'radicle']:\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/THR/{0}_H3K27ac_1K_100bp.csv\".format(i)\n",
    "    df_tissue = pd.read_csv(input_f)\n",
    "    #print(df_tissue)\n",
    "    if(df_H3K27ac.shape[0]==0):\n",
    "        df_H3K27ac = df_tissue.copy()\n",
    "    else:\n",
    "        #print(df_H3K27ac.head())\n",
    "        #print(df_tissue.head())\n",
    "        df_H3K27ac = pd.merge(df_H3K27ac, df_tissue, how='inner', on=['PATRE_Id'])\n",
    "df_H3K4me3 = pd.DataFrame()\n",
    "for i in ['hypocotyl', 'ovule', 'stigma', 'radicle']:\n",
    "    input_f = \"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/THR/{0}_H3K4me3_1K_100bp.csv\".format(i)\n",
    "    df_tissue = pd.read_csv(input_f)\n",
    "    if(df_H3K4me3.shape[0]==0):\n",
    "        df_H3K4me3 = df_tissue.copy()\n",
    "    else:\n",
    "        df_H3K4me3 = pd.merge(df_H3K4me3, df_tissue, how='inner', on=['PATRE_Id'])\n",
    "\n",
    "df_PATRE_Tag = get_tags(df_H3K4me3, df_H3K27ac)\n",
    "\n",
    "df_umap = zscore_umap(df_result)\n",
    "df_umap.reset_index(inplace=True)\n",
    "df_umap.rename(columns={'index':'PATRE_Id'}, inplace=True)\n",
    "\n",
    "# ## 打上chromHMM 标签\n",
    "df_result = pd.merge(df_umap, df_PATRE_Tag, how='left', on =['PATRE_Id'])\n",
    "old_tags = ['F-F', 'F-T', 'T-F', 'T-T']\n",
    "#new_tags = ['Non-functional elements', 'Candidate enhancers', 'Candidate promoters', 'Candidate promoters']\n",
    "new_tags = ['Non-functional elements', 'Candidate enhancers', 'Candidate promoters', 'Candidate promoters']\n",
    "\n",
    "df_result['tags'].replace(old_tags, new_tags, inplace=True)\n",
    "df_result.to_csv(output_f, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析各个组织活性的cis元件\n",
    "## 根据所鉴定的cis调控元件在其它组织中是否受到loop调控确定是否为组织激活cis调控元件\n",
    "import pandas as pd\n",
    "df_cis = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7824\n",
      "5574\n"
     ]
    }
   ],
   "source": [
    "### 对上一步umap的结果进行校正，用PATRE的注释信息来校正，\n",
    "# 被定义为候选启动子，但是位于远端调控区域，修正为候选增强子\n",
    "# 被定义为候选增强子，但是位于非远端调控区域，修正为候选启动子\n",
    "import pandas as pd\n",
    "df_PATRE = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/THR/J668_PATRE_UMAP.csv\")\n",
    "df_annot = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/THR/J668_PATRE_annotation.bed\", sep=\" \")\n",
    "df_distal = df_annot.query(\"annotation.str.contains('Distal')\", engine='python')[['V4']].copy()\n",
    "df_distal['tags1'] = ['Distal']*df_distal.shape[0]\n",
    "df_distal.columns = ['PATRE_Id', 'tags1']\n",
    "\n",
    "df_distal.fillna(\"Clost\", inplace=True)\n",
    "df_PATRE = pd.merge(df_PATRE, df_distal, how='left', on=['PATRE_Id'])\n",
    "## 更正候选增强子\n",
    "df_enhencer_correct = df_PATRE.query(\"tags=='Candidate promoters' & tags1=='Distal'\").copy()\n",
    "df_enhencer_correct['tags'] = ['Candidate enhancers']*df_enhencer_correct.shape[0]\n",
    "print(df_enhencer_correct.shape[0]) \n",
    "## 更正候选启动子\n",
    "df_promoter_correct = df_PATRE.query(\"tags=='Candidate enhancers' & tags1!='Distal'\").copy()\n",
    "df_promoter_correct['tags'] = ['Candidate promoters']*df_promoter_correct.shape[0]\n",
    "print(df_promoter_correct.shape[0])\n",
    "df_result = pd.concat([df_PATRE, df_enhencer_correct, df_promoter_correct], axis=0)\n",
    "df_result.drop_duplicates(keep='last', subset=['PATRE_Id'], inplace=True)\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/THR/J668_PATRE_UMAP_correct.csv\", columns=['PATRE_Id', 'umap1', 'umap2', 'tags'], index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                      tags  number\n",
      "0      Candidate enhancers   18520\n",
      "1      Candidate promoters   26561\n",
      "2  Non-functional elements    9531\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/THR/J668_PATRE_UMAP_correct.csv\")\n",
    "df['number'] = [1]*df.shape[0]\n",
    "df_result = df.groupby(by=['tags'], as_index=False)['number'].sum()\n",
    "print(df_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析所鉴定的活性启动子所调控的基因与其它基因在表达水平上的差异\n",
    "import pandas as pd\n",
    "df_PATRE = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/THR/J668_PATRE_UMAP_correct.csv\")\n",
    "df_PATRE_promoter = df_PATRE.query(\"tags=='Candidate promoters'\").copy()\n",
    "df_annot = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/THR/J668_PATRE_annotation.bed\", sep=\" \")\n",
    "df_annot = df_annot[['V4', 'geneId']].copy()\n",
    "df_annot.columns = ['PATRE_Id', 'GeneId']\n",
    "df_promoter_annot = pd.merge(df_annot, df_PATRE_promoter, how='inner', on=['PATRE_Id'])[['GeneId', 'tags']].copy()\n",
    "df_promoter_annot.drop_duplicates(keep='first', inplace=True)\n",
    "df_gene_TPM = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data/J668_all_Tissue_mean_TPM_long.csv\")\n",
    "df_PATRE_TPM = pd.merge(df_gene_TPM, df_promoter_annot, how='left', on=['GeneId'])\n",
    "df_PATRE_TPM.fillna(\"no active promoter\", inplace=True)\n",
    "old_tags = ['Candidate promoters', 'no active promoter']\n",
    "new_tags = ['A_P', 'N-A_P']\n",
    "df_PATRE_TPM['tags'].replace(old_tags, new_tags, inplace=True)\n",
    "df_PATRE_TPM.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/THR/J668_PATRE_promoter_gene.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分析活性promoter启动子调控基因中管家基因，沉默基因和动态基因的分布\n",
    "df_gene_type = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/new_RNA_Seq/RNA_Seq_data/J668_gene_type.csv\")\n",
    "df_promoter_tags = df_PATRE_TPM[['GeneId', 'tags']].drop_duplicates(keep='first').copy()\n",
    "df_tmp = pd.merge(df_gene_type, df_promoter_tags, how='inner', on=['GeneId'])\n",
    "df_tmp['number'] = [1]*df_tmp.shape[0]\n",
    "df_result = df_tmp.groupby(by=['sigs', 'tags'], as_index=False)['number'].sum()\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/THR/J668_PATRE_promoter_gene_type.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/hypocotyl/hypocotyl_micro_state.csv\")\n",
    "df_RNA = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/hypocotyl/hypocotyl_peaks_1Kb_RNA_Seq.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'PATRE_Id', 'Reads'])\n",
    "df_RNA.drop(['chrom', 'start', 'end'], inplace=True, axis=1)\n",
    "df_result = pd.merge(df, df_RNA, how='left', on=['PATRE_Id'])\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/hypocotyl/hypocotyl_micro_state.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 将所有组织的chromHMM 合并为一个文件\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"all_chromHMM_segments.bed\", sep=\"\\t\", names=['chr', 'start', 'end','leaf', 'hypocotyl', 'cotyledon', 'stigma', 'radicle'])\n",
    "df_same = df.query(\"leaf==hypocotyl & hypocotyl==cotypldon & cotyledon==stigma & stigma==radicle\").copy()\n",
    "df_dynamic = pd.concat([df, df_same], axis=0)\n",
    "df_dynamic.drop_duplicates(keep=False, inplace=True)\n",
    "df_same['tags'] = df_same['leaf']\n",
    "for index, row in df_dynamic.iterrows():\n",
    "    if 'E2' in row:\n",
    "        t = 'E2'\n",
    "    elif 'E1' in row and 'E4' in row:\n",
    "        t = 'E2'\n",
    "    elif 'E1' in row:\n",
    "        t = 'E1'\n",
    "    elif 'E4' in row:\n",
    "        t = 'E4'\n",
    "    else:\n",
    "        t = 'E3'\n",
    "    df_dynamic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 给PATERs打上标签 \n",
    "# Nutures\n",
    "# 20222203\n",
    "# add chromHMM tags for PATERs\n",
    "import pandas as pd\n",
    "import sys\n",
    "paters = sys.argv[1]\n",
    "chromHMM = sys.argv[2]\n",
    "output = sys.argv[3]\n",
    "Chrom = sys.argv[4]\n",
    "df_paters = pd.read_csv(paters, sep=\"\\t\", names=['chrom', 'start', 'end', 'PATRE_ID'])\n",
    "df_chromHMM = pd.read_csv(chromHMM, sep=\"\\t\", names=['chrom', 'start', 'end', 'tags'])\n",
    "c = Chrom\n",
    "all_chrom = ['Ghir_A01', 'Ghir_A02', 'Ghir_A03', 'Ghir_A04', 'Ghir_A05', 'Ghir_A06', 'Ghir_A07', 'Ghir_A08', 'Ghir_A09', 'Ghir_A10', 'Ghir_A11', 'Ghir_A12', 'Ghir_A13',\n",
    "             'Ghir_D01', 'Ghir_D02', 'Ghir_D03', 'Ghir_D04', 'Ghir_D05', 'Ghir_D06', 'Ghir_D07', 'Ghir_D08', 'Ghir_D09', 'Ghir_D10', 'Ghir_D11', 'Ghir_D12', 'Ghir_D13']\n",
    "df_paters_c = df_paters.query(\"chrom==@c\").copy()\n",
    "df_chromHMM_c = df_chromHMM.query(\"chrom==@c\").copy()\n",
    "for index, row in df_paters_c.iterrows():\n",
    "    C, S, E = row['chrom'], row['start'], row['end']\n",
    "    print(C, S, E)\n",
    "    df_tmp = df_chromHMM_c.query(\"chrom==@C & start<=@E & end>= @S\").copy()\n",
    "    print(df_tmp)\n",
    "    all_tags = \"-\".join(list(set(df_tmp['tags'].tolist())))\n",
    "    print(all_tags)\n",
    "    df_paters_c.loc[index, 'tags'] = all_tags\n",
    "df_paters_c.to_csv(output, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 给ATAC 位点打上PATERs 标签\n",
    "# Nutures\n",
    "# 202221004\n",
    "# add chromHMM tags for ATAC site\n",
    "import pandas as pd\n",
    "import sys\n",
    "atac = sys.argv[1]\n",
    "chromHMM = sys.argv[2]\n",
    "output = sys.argv[3]\n",
    "Chrom = sys.argv[4]\n",
    "df_atac = pd.read_csv(atac, sep=\"\\t\", names=['chrom', 'start', 'end', 'ID'])\n",
    "df_chrom = pd.read_csv(chromHMM, sep=\"\\t\", names=['chrom', 'start', 'end', 'Tags'])\n",
    "df_atac_c = df_atac.query(\"chrom==@Chrom\").copy()\n",
    "df_chrom_c = df_chrom.query(\"chrom==@Chrom\").copy()\n",
    "for index, row in df_atac_c.iterrows():\n",
    "    C, S, E = row['chrom'], row['start'], row['start']\n",
    "    df_tmp = df_chrom_c.query(\"chrom==@C & start<=@E & end>= @S\").copy()\n",
    "    if df_tmp.shape[0] == 0:\n",
    "        all_tags = 'E0'\n",
    "    else:\n",
    "        df_tmp.reset_index(drop=True, inplace=True)\n",
    "        all_tags = df_tmp.loc[0, 'Tags']\n",
    "    df_atac_c.loc[index, 'Tags'] = all_tags\n",
    "df_atac_c.to_csv(output, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## hypocotyl , ovule, radicle , stigma这几种组织的染色质开放区域置合并起来\n",
    "df_hypocotyl = pd.read_csv(\"hypocotyl/hypocotyl_PATRE_around_1K.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'id'])\n",
    "df_radicle = pd.read_csv(\"radicle/radicle_PATRE_around_1K.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'id'])\n",
    "df_ovule = pd.read_csv(\"ovule/ovule_PATRE_around_1K.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'id'])\n",
    "df_stigma = pd.read_csv(\"stigma/radicle_PATRE_around_1K.bed\", sep='\\t', names=['chrom', 'start', 'end', 'id'])\n",
    "\n",
    "df_all = pd.concat([df_hypocotyl, df_radicle, df_ovule, df_stigma], axis=0)\n",
    "df_all.drop_duplicates(keep='first', subset=['chrom', 'start', 'end'], inplace=True)\n",
    "df_all['id'] = ['PATER']*df_all.shape[0]\n",
    "df_all['number'] = range(df_all.shape[0])\n",
    "df_all['id2'] = df_all['id'].map(str) + df_all['number'].map(str)\n",
    "df_all.to_csv(\"J668_PATRE.bed\", sep=\"\\t\", index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 给PATRE打上标签\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/THR/J668_PATRE.bed\", sep=\"\\t\", names=['chrom', 'start', 'end', 'PATRE_Id'])\n",
    "df_tags = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/THR/J668_PATRE_UMAP_correct.csv\")\n",
    "df_result = pd.merge(df, df_tags, how='inner', on=['PATRE_Id'])\n",
    "df_result.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/ATAC/PETRE/THR/J668_PATRE_tags.bed\", sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 分为At和Dt两个Cis\n",
    "import pandas as pd\n",
    "df = pd.read_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_cis.bed\", sep=\"\\t\")\n",
    "df_At = df.query(\"chrom.str.contains('Ghir_A')\", engine='python').copy()\n",
    "df_Dt = df.query(\"chrom.str.contains('Ghir_D')\", engine='python').copy()\n",
    "df_At['tags1'] = ['At']*df_At.shape[0]\n",
    "df_At['tags2'] = range(df_At.shape[0])\n",
    "df_At['tags'] = df_At['tags1'].map(str) + df_At['tags2'].map(str)\n",
    "\n",
    "df_Dt['tags1'] = ['Dt']*df_Dt.shape[0]\n",
    "df_Dt['tags2'] = range(df_Dt.shape[0])\n",
    "df_Dt['tags'] = df_Dt['tags1'].map(str) + df_Dt['tags2'].map(str)\n",
    "\n",
    "df_At.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_Cis_At.bed\", sep=\"\\t\", index=False, columns=['chrom', 'start', 'end', 'tags'], header=False)\n",
    "df_Dt.to_csv(\"G:/Billfish/J668_multip_tissue_3D-genome/HiC/loops/Cis-element/J668_Cis_Dt.bed\", sep=\"\\t\", index=False, columns=['chrom', 'start', 'end', 'tags'], header=False)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.3 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "a0f669ea42a4dcea88d385f8f9b4c3d9b966e24234758831d1dbfdac925619a2"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
